{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython - classe `PairwiseAligner`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La classe `PairwiseAligner` del package `Align` implementa algoritmi di allineamento di due sequenze. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importare il package `Bio`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importare il package `Align`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.Align in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.Align - Code for dealing with sequence alignments.\n",
      "\n",
      "DESCRIPTION\n",
      "    One of the most important things in this module is the MultipleSeqAlignment\n",
      "    class, used in the Bio.AlignIO module.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    AlignInfo\n",
      "    Applications (package)\n",
      "    _aligners\n",
      "    a2m\n",
      "    bed\n",
      "    bigbed\n",
      "    bigmaf\n",
      "    bigpsl\n",
      "    clustal\n",
      "    emboss\n",
      "    exonerate\n",
      "    fasta\n",
      "    hhr\n",
      "    interfaces\n",
      "    maf\n",
      "    mauve\n",
      "    msf\n",
      "    nexus\n",
      "    phylip\n",
      "    psl\n",
      "    sam\n",
      "    stockholm\n",
      "    substitution_matrices (package)\n",
      "    tabular\n",
      "\n",
      "CLASSES\n",
      "    _algorithms.PairwiseAligner(builtins.object)\n",
      "        PairwiseAligner\n",
      "    builtins.object\n",
      "        Alignment\n",
      "            PairwiseAlignment\n",
      "        MultipleSeqAlignment\n",
      "        PairwiseAlignments\n",
      "    \n",
      "    class Alignment(builtins.object)\n",
      "     |  Alignment(sequences, coordinates=None)\n",
      "     |  \n",
      "     |  Represents a sequence alignment.\n",
      "     |  \n",
      "     |  An Alignment object has a `.sequences` attribute storing the sequences\n",
      "     |  (Seq, MutableSeq, SeqRecord, or string objects) that were aligned, as well\n",
      "     |  as a `.coordinates` attribute storing the sequence coordinates defining the\n",
      "     |  alignment as a numpy array.\n",
      "     |  \n",
      "     |  Other commonly used attributes (which may or may not be present) are:\n",
      "     |       - annotations        - A dictionary with annotations describing the\n",
      "     |                              alignment;\n",
      "     |       - column_annotations - A dictionary with annotations describing each\n",
      "     |                              column in the alignment;\n",
      "     |       - score              - The alignment score.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __array__(self, dtype=None)\n",
      "     |  \n",
      "     |  __eq__(self, other)\n",
      "     |      Check if two Alignment objects specify the same alignment.\n",
      "     |  \n",
      "     |  __format__(self, format_spec)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      Wrapper for self.format().\n",
      "     |  \n",
      "     |  __ge__(self, other)\n",
      "     |      Check if self should come after or is equal to other.\n",
      "     |  \n",
      "     |  __getitem__(self, key)\n",
      "     |      Return self[key].\n",
      "     |      \n",
      "     |      Indices of the form\n",
      "     |      \n",
      "     |      self[:, :]\n",
      "     |      \n",
      "     |      return a copy of the Alignment object;\n",
      "     |      \n",
      "     |      self[:, i:]\n",
      "     |      self[:, :j]\n",
      "     |      self[:, i:j]\n",
      "     |      self[:, iterable] (where iterable returns integers)\n",
      "     |      \n",
      "     |      return a new Alignment object spanning the selected columns;\n",
      "     |      \n",
      "     |      self[k, i]\n",
      "     |      self[k, i:]\n",
      "     |      self[k, :j]\n",
      "     |      self[k, i:j]\n",
      "     |      self[k, iterable] (where iterable returns integers)\n",
      "     |      self[k] (equivalent to self[k, :])\n",
      "     |      \n",
      "     |      return a string with the aligned sequence (including gaps) for the\n",
      "     |      selected columns, where k = 0 represents the target and k = 1\n",
      "     |      represents the query sequence; and\n",
      "     |      \n",
      "     |      self[:, i]\n",
      "     |      \n",
      "     |      returns a string with the selected column in the alignment.\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> alignments = aligner.align(\"ACCGGTTT\", \"ACGGGTT\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ACCGG-TTT 8\n",
      "     |                        0 ||-||-||- 9\n",
      "     |      query             0 AC-GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment[0, :]\n",
      "     |      'ACCGG-TTT'\n",
      "     |      >>> alignment[1, :]\n",
      "     |      'AC-GGGTT-'\n",
      "     |      >>> alignment[0]\n",
      "     |      'ACCGG-TTT'\n",
      "     |      >>> alignment[1]\n",
      "     |      'AC-GGGTT-'\n",
      "     |      >>> alignment[0, 1:-2]\n",
      "     |      'CCGG-T'\n",
      "     |      >>> alignment[1, 1:-2]\n",
      "     |      'C-GGGT'\n",
      "     |      >>> alignment[0, (1, 5, 2)]\n",
      "     |      'C-C'\n",
      "     |      >>> alignment[1, ::2]\n",
      "     |      'A-GT-'\n",
      "     |      >>> alignment[1, range(0, 9, 2)]\n",
      "     |      'A-GT-'\n",
      "     |      >>> alignment[:, 0]\n",
      "     |      'AA'\n",
      "     |      >>> alignment[:, 5]\n",
      "     |      '-G'\n",
      "     |      >>> alignment[:, 1:]  # doctest:+ELLIPSIS\n",
      "     |      <Alignment object (2 rows x 8 columns) at 0x...>\n",
      "     |      >>> print(alignment[:, 1:])\n",
      "     |      target            1 CCGG-TTT 8\n",
      "     |                        0 |-||-||- 8\n",
      "     |      query             1 C-GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 2:])\n",
      "     |      target            2 CGG-TTT 8\n",
      "     |                        0 -||-||- 7\n",
      "     |      query             2 -GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 3:])\n",
      "     |      target            3 GG-TTT 8\n",
      "     |                        0 ||-||- 6\n",
      "     |      query             2 GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 3:-1])\n",
      "     |      target            3 GG-TT 7\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             2 GGGTT 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, ::2])\n",
      "     |      target            0 ACGTT 5\n",
      "     |                        0 |-||- 5\n",
      "     |      query             0 A-GT- 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, range(1, 9, 2)])\n",
      "     |      target            0 CG-T 3\n",
      "     |                        0 ||-| 4\n",
      "     |      query             0 CGGT 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, (2, 7, 3)])\n",
      "     |      target            0 CTG 3\n",
      "     |                        0 -|| 3\n",
      "     |      query             0 -TG 2\n",
      "     |      <BLANKLINE>\n",
      "     |  \n",
      "     |  __gt__(self, other)\n",
      "     |      Check if self should come after other.\n",
      "     |  \n",
      "     |  __init__(self, sequences, coordinates=None)\n",
      "     |      Initialize a new Alignment object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - sequences   - A list of the sequences (Seq, MutableSeq, SeqRecord,\n",
      "     |                       or string objects)that were aligned.\n",
      "     |       - coordinates - The sequence coordinates that define the alignment.\n",
      "     |                       If None (the default value), assume that the sequences\n",
      "     |                       align to each other without any gaps.\n",
      "     |  \n",
      "     |  __le__(self, other)\n",
      "     |      Check if self should come before or is equal to other.\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return the number of sequences in the alignment.\n",
      "     |  \n",
      "     |  __lt__(self, other)\n",
      "     |      Check if self should come before other.\n",
      "     |  \n",
      "     |  __ne__(self, other)\n",
      "     |      Check if two Alignment objects have different alignments.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return a representation of the alignment, including its shape.\n",
      "     |      \n",
      "     |      The representation cannot be used with eval() to recreate the object,\n",
      "     |      which is usually possible with simple python objects.  For example:\n",
      "     |      \n",
      "     |      <Alignment object (2 rows x 14 columns) at 0x10403d850>\n",
      "     |      \n",
      "     |      The hex string is the memory address of the object and can be used to\n",
      "     |      distinguish different Alignment objects.  See help(id) for more\n",
      "     |      information.\n",
      "     |      \n",
      "     |      >>> import numpy\n",
      "     |      >>> from Bio.Align import Alignment\n",
      "     |      >>> alignment = Alignment((\"ACCGT\", \"ACGT\"),\n",
      "     |      ...                       coordinates = numpy.array([[0, 2, 3, 5],\n",
      "     |      ...                                                  [0, 2, 2, 4],\n",
      "     |      ...                                                 ]))\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ACCGT 5\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AC-GT 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment  # doctest:+ELLIPSIS\n",
      "     |      <Alignment object (2 rows x 5 columns) at 0x...>\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return a human-readable string representation of the alignment.\n",
      "     |      \n",
      "     |      For sequence alignments, each line has at most 80 columns.\n",
      "     |      The first 10 columns show the (possibly truncated) sequence name,\n",
      "     |      which may be the id attribute of a SeqRecord, or otherwise 'target'\n",
      "     |      or 'query' for pairwise alignments.\n",
      "     |      The next 10 columns show the sequence coordinate, using zero-based\n",
      "     |      counting as usual in Python.\n",
      "     |      The remaining 60 columns shown the sequence, using dashes to represent\n",
      "     |      gaps.\n",
      "     |      At the end of the alignment, the end coordinates are shown on the right\n",
      "     |      of the sequence, again in zero-based coordinates.\n",
      "     |      \n",
      "     |      Pairwise alignments have an additional line between the two sequences\n",
      "     |      showing whether the sequences match ('|') or mismatch ('.'), or if\n",
      "     |      there is a gap ('-').\n",
      "     |      The coordinates shown for this line are the column indices, which can\n",
      "     |      be useful when extracting a subalignment.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      \n",
      "     |      >>> seqA = \"TTAACCCCATTTG\"\n",
      "     |      >>> seqB = \"AAGCCCCTTT\"\n",
      "     |      >>> seqC = \"AAAGGGGCTT\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(seqA, seqB)\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 TTAA-CCCCATTTG 13\n",
      "     |                        0 --||-||||-|||- 14\n",
      "     |      query             0 --AAGCCCC-TTT- 10\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Note that seqC is the reverse complement of seqB. Aligning it to the\n",
      "     |      reverse strand gives the same alignment, but the query coordinates are\n",
      "     |      switched:\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(seqA, seqC, strand=\"-\")\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 TTAA-CCCCATTTG 13\n",
      "     |                        0 --||-||||-|||- 14\n",
      "     |      query            10 --AAGCCCC-TTT-  0\n",
      "     |      <BLANKLINE>\n",
      "     |  \n",
      "     |  format(self, fmt='', *args, **kwargs)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - fmt       - File format. Acceptable values are an empty string to\n",
      "     |                     create a human-readable representation of the alignment,\n",
      "     |                     or any of the alignment file formats supported by\n",
      "     |                     `Bio.Align` (some have not yet been implemented).\n",
      "     |      \n",
      "     |      All other arguments are passed to the format-specific writer functions:\n",
      "     |       - mask      - PSL format only. Specify if repeat regions in the target\n",
      "     |                     sequence are masked and should be reported in the\n",
      "     |                     `repMatches` field of the PSL file instead of in the\n",
      "     |                     `matches` field. Acceptable values are\n",
      "     |                     None   : no masking (default);\n",
      "     |                     \"lower\": masking by lower-case characters;\n",
      "     |                     \"upper\": masking by upper-case characters.\n",
      "     |       - wildcard  - PSL format only. Report alignments to the wildcard\n",
      "     |                     character in the target or query sequence in the\n",
      "     |                     `nCount` field of the PSL file instead of in the\n",
      "     |                     `matches`, `misMatches`, or `repMatches` fields.\n",
      "     |                     Default value is 'N'.\n",
      "     |       - md        - SAM format only. If True, calculate the MD tag from\n",
      "     |                     the alignment and include it in the output. If False\n",
      "     |                     (default), do not include the MD tag in the output.\n",
      "     |  \n",
      "     |  map(self, alignment)\n",
      "     |      Map the alignment to self.target and return the resulting alignment.\n",
      "     |      \n",
      "     |      Here, self.query and alignment.target are the same sequence.\n",
      "     |      \n",
      "     |      A typical example is where self is the pairwise alignment between a\n",
      "     |      chromosome and a transcript, the argument is the pairwise alignment\n",
      "     |      between the transcript and a sequence (e.g., as obtained by RNA-seq),\n",
      "     |      and we want to find the alignment of the sequence to the chromosome:\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = 'local'\n",
      "     |      >>> aligner.open_gap_score = -1\n",
      "     |      >>> aligner.extend_gap_score = 0\n",
      "     |      >>> chromosome = \"AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA\"\n",
      "     |      >>> transcript = \"CCCCCCCGGGGGG\"\n",
      "     |      >>> alignments1 = aligner.align(chromosome, transcript)\n",
      "     |      >>> len(alignments1)\n",
      "     |      1\n",
      "     |      >>> alignment1 = alignments1[0]\n",
      "     |      >>> print(alignment1)\n",
      "     |      target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32\n",
      "     |                        0 |||||||-----------|||||| 24\n",
      "     |      query             0 CCCCCCC-----------GGGGGG 13\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> sequence = \"CCCCGGGG\"\n",
      "     |      >>> alignments2 = aligner.align(transcript, sequence)\n",
      "     |      >>> len(alignments2)\n",
      "     |      1\n",
      "     |      >>> alignment2 = alignments2[0]\n",
      "     |      >>> print(alignment2)\n",
      "     |      target            3 CCCCGGGG 11\n",
      "     |                        0 ||||||||  8\n",
      "     |      query             0 CCCCGGGG  8\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment = alignment1.map(alignment2)\n",
      "     |      >>> print(alignment)\n",
      "     |      target           11 CCCCAAAAAAAAAAAGGGG 30\n",
      "     |                        0 ||||-----------|||| 19\n",
      "     |      query             0 CCCC-----------GGGG  8\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> format(alignment, \"psl\")\n",
      "     |      '8\\t0\\t0\\t0\\t0\\t0\\t1\\t11\\t+\\tquery\\t8\\t0\\t8\\ttarget\\t40\\t11\\t30\\t2\\t4,4,\\t0,4,\\t11,26,\\n'\n",
      "     |      \n",
      "     |      Mapping the alignment does not depend on the sequence contents. If we\n",
      "     |      delete the sequence contents, the same alignment is found in PSL format\n",
      "     |      (though we obviously lose the ability to print the sequence alignment):\n",
      "     |      \n",
      "     |      >>> alignment1.target = Seq(None, len(alignment1.target))\n",
      "     |      >>> alignment1.query = Seq(None, len(alignment1.query))\n",
      "     |      >>> alignment2.target = Seq(None, len(alignment2.target))\n",
      "     |      >>> alignment2.query = Seq(None, len(alignment2.query))\n",
      "     |      >>> alignment = alignment1.map(alignment2)\n",
      "     |      >>> format(alignment, \"psl\")\n",
      "     |      '8\\t0\\t0\\t0\\t0\\t0\\t1\\t11\\t+\\tquery\\t8\\t0\\t8\\ttarget\\t40\\t11\\t30\\t2\\t4,4,\\t0,4,\\t11,26,\\n'\n",
      "     |  \n",
      "     |  sort(self, key=None, reverse=False)\n",
      "     |      Sort the sequences of the alignment in place.\n",
      "     |      \n",
      "     |      By default, this sorts the sequences alphabetically using their id\n",
      "     |      attribute if available, or by their sequence contents otherwise.\n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> aligner.gap_score = -1\n",
      "     |      >>> alignments = aligner.align(\"AATAA\", \"AAGAA\")\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AATAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AAGAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.sort()\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AAGAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AATAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Alternatively, a key function can be supplied that maps each sequence\n",
      "     |      to a sort value.  For example, you could sort on the GC content of each\n",
      "     |      sequence.\n",
      "     |      \n",
      "     |      >>> from Bio.SeqUtils import GC\n",
      "     |      >>> alignment.sort(key=GC)\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AATAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AAGAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      You can reverse the sort order by passing `reverse=True`:\n",
      "     |      \n",
      "     |      >>> alignment.sort(key=GC, reverse=True)\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AAGAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AATAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      The sequences are now sorted by decreasing GC content value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  infer_coordinates(lines, skipped_columns=None) from builtins.type\n",
      "     |      Infer the coordinates from a printed alignment.\n",
      "     |      \n",
      "     |      This method is primarily employed in Biopython's alignment parsers,\n",
      "     |      though it may be useful for other purposes.\n",
      "     |      \n",
      "     |      For an alignment consisting of N sequences, printed as N lines with\n",
      "     |      the same number of columns, where gaps are represented by dashes,\n",
      "     |      this method will calculate the sequence coordinates that define the\n",
      "     |      alignment. The coordinates are returned as a numpy array of integers,\n",
      "     |      and can be used to create an Alignment object.\n",
      "     |      \n",
      "     |      The argument skipped columns should be None (the default) or an empty\n",
      "     |      list. If skipped_columns is a list, then the indices of any columns in\n",
      "     |      the alignment with a gap in all lines are appended to skipped_columns.\n",
      "     |      \n",
      "     |      This is an example for the alignment of three sequences TAGGCATACGTG,\n",
      "     |      AACGTACGT, and ACGCATACTTG, with gaps in the second and third sequence:\n",
      "     |      \n",
      "     |      >>> from Bio.Align import Alignment\n",
      "     |      >>> lines = [\"TAGGCATACGTG\",\n",
      "     |      ...          \"AACG--TACGT-\",\n",
      "     |      ...          \"-ACGCATACTTG\",\n",
      "     |      ...         ]\n",
      "     |      >>> sequences = [line.replace(\"-\", \"\") for line in lines]\n",
      "     |      >>> sequences\n",
      "     |      ['TAGGCATACGTG', 'AACGTACGT', 'ACGCATACTTG']\n",
      "     |      >>> coordinates = Alignment.infer_coordinates(lines)\n",
      "     |      >>> coordinates\n",
      "     |      array([[ 0,  1,  4,  6, 11, 12],\n",
      "     |             [ 0,  1,  4,  4,  9,  9],\n",
      "     |             [ 0,  0,  3,  5, 10, 11]])\n",
      "     |      >>> alignment = Alignment(sequences, coordinates)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  aligned\n",
      "     |      Return the indices of subsequences aligned to each other.\n",
      "     |      \n",
      "     |      This property returns the start and end indices of subsequences\n",
      "     |      in the target and query sequence that were aligned to each other.\n",
      "     |      If the alignment between target (t) and query (q) consists of N\n",
      "     |      chunks, you get two tuples of length N:\n",
      "     |      \n",
      "     |          (((t_start1, t_end1), (t_start2, t_end2), ..., (t_startN, t_endN)),\n",
      "     |           ((q_start1, q_end1), (q_start2, q_end2), ..., (q_startN, q_endN)))\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> alignments = aligner.align(\"GAACT\", \"GAT\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GAACT 5\n",
      "     |                        0 ||--| 5\n",
      "     |      query             0 GA--T 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      array([[[0, 2],\n",
      "     |              [4, 5]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 2],\n",
      "     |              [2, 3]]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GAACT 5\n",
      "     |                        0 |-|-| 5\n",
      "     |      query             0 G-A-T 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      array([[[0, 1],\n",
      "     |              [2, 3],\n",
      "     |              [4, 5]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 1],\n",
      "     |              [1, 2],\n",
      "     |              [2, 3]]])\n",
      "     |      \n",
      "     |      Note that different alignments may have the same subsequences\n",
      "     |      aligned to each other. In particular, this may occur if alignments\n",
      "     |      differ from each other in terms of their gap placement only:\n",
      "     |      \n",
      "     |      >>> aligner.mismatch_score = -10\n",
      "     |      >>> alignments = aligner.align(\"AAACAAA\", \"AAAGAAA\")\n",
      "     |      >>> len(alignments)\n",
      "     |      2\n",
      "     |      >>> print(alignments[0])\n",
      "     |      target            0 AAAC-AAA 7\n",
      "     |                        0 |||--||| 8\n",
      "     |      query             0 AAA-GAAA 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[0].aligned\n",
      "     |      array([[[0, 3],\n",
      "     |              [4, 7]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 3],\n",
      "     |              [4, 7]]])\n",
      "     |      >>> print(alignments[1])\n",
      "     |      target            0 AAA-CAAA 7\n",
      "     |                        0 |||--||| 8\n",
      "     |      query             0 AAAG-AAA 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[1].aligned\n",
      "     |      array([[[0, 3],\n",
      "     |              [4, 7]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 3],\n",
      "     |              [4, 7]]])\n",
      "     |      \n",
      "     |      The property can be used to identify alignments that are identical\n",
      "     |      to each other in terms of their aligned sequences.\n",
      "     |  \n",
      "     |  indices\n",
      "     |      Return the sequence index of each lettter in the alignment.\n",
      "     |      \n",
      "     |      This property returns a 2D numpy array with the sequence index of each\n",
      "     |      letter in the alignment. Gaps are indicated by -1.  The array has the\n",
      "     |      same number of rows and columns as the alignment, as given by\n",
      "     |      `self.shape`.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"AATG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AA-TG 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5],\n",
      "     |             [ 0,  1, -1,  2,  3]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             0 AA-T-G 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5,  6],\n",
      "     |             [ 0,  1, -1,  2, -1,  3]])\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"CATT\", strand=\"-\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             4 AA-TG 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5],\n",
      "     |             [ 3,  2, -1,  1,  0]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             4 AA-T-G 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5,  6],\n",
      "     |             [ 3,  2, -1,  1, -1,  0]])\n",
      "     |  \n",
      "     |  inverse_indices\n",
      "     |      Return the alignment column index for each letter in each sequence.\n",
      "     |      \n",
      "     |      This property returns a list of 1D numpy arrays; the number of arrays\n",
      "     |      is equal to the number of aligned sequences, and the length of each\n",
      "     |      array is equal to the length of the corresponding sequence. For each\n",
      "     |      letter in each sequence, the array contains the corresponding column\n",
      "     |      index in the alignment. Letters not included in the alignment are\n",
      "     |      indicated by -1.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"AATG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AA-TG 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4, -1]), array([0, 1, 3, 4])]\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             0 AA-T-G 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4,  5]), array([0, 1, 3, 5])]\n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"CATT\", strand=\"-\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             4 AA-TG 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4, -1]), array([4, 3, 1, 0])]\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             4 AA-T-G 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4,  5]), array([5, 3, 1, 0])]\n",
      "     |  \n",
      "     |  shape\n",
      "     |      Return the shape of the alignment as a tuple of two integer values.\n",
      "     |      \n",
      "     |      The first integer value is the number of sequences in the alignment as\n",
      "     |      returned by len(alignment), which is always 2 for pairwise alignments.\n",
      "     |      \n",
      "     |      The second integer value is the number of columns in the alignment when\n",
      "     |      it is printed, and is equal to the sum of the number of matches, number\n",
      "     |      of mismatches, and the total length of gaps in the target and query.\n",
      "     |      Sequence sections beyond the aligned segment are not included in the\n",
      "     |      number of columns.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"global\"\n",
      "     |      >>> alignments = aligner.align(\"GACCTG\", \"CGATCG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 -GACCT-G 6\n",
      "     |                        0 -||--|-| 8\n",
      "     |      query             0 CGA--TCG 6\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> len(alignment)\n",
      "     |      2\n",
      "     |      >>> alignment.shape\n",
      "     |      (2, 8)\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      >>> alignments = aligner.align(\"GACCTG\", \"CGATCG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GACCT-G 6\n",
      "     |                        0 ||--|-| 7\n",
      "     |      query             1 GA--TCG 6\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> len(alignment)\n",
      "     |      2\n",
      "     |      >>> alignment.shape\n",
      "     |      (2, 7)\n",
      "     |  \n",
      "     |  substitutions\n",
      "     |      Return an Array with the number of substitutions of letters in the alignment.\n",
      "     |      \n",
      "     |      As an example, consider a sequence alignment of two RNA sequences:\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> target = \"ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTTATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG\"  # human spliceosomal small nuclear RNA U1\n",
      "     |      >>> query = \"ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTCTTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG\"  # sea lamprey spliceosomal small RNA U1\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> aligner.gap_score = -10\n",
      "     |      >>> alignments = aligner.align(target, query)\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTT\n",
      "     |                        0 |||||||||||.||||||||..|||||||||||..|||||||..|||||||||||||||.\n",
      "     |      query             0 ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTC\n",
      "     |      <BLANKLINE>\n",
      "     |      target           60 ATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT\n",
      "     |                       60 .|||||||||||.|||..|.|.||||||||||||||||||||||||||||||||||||||\n",
      "     |      query            60 TTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT\n",
      "     |      <BLANKLINE>\n",
      "     |      target          120 GCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG 164\n",
      "     |                      120 |.||||||||||||||||||||||||||||||||||.|||||.| 164\n",
      "     |      query           120 GTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG 164\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> m = alignment.substitutions\n",
      "     |      >>> print(m)\n",
      "     |           A    C    G    T\n",
      "     |      A 28.0  1.0  2.0  1.0\n",
      "     |      C  0.0 39.0  1.0  2.0\n",
      "     |      G  2.0  0.0 45.0  0.0\n",
      "     |      T  2.0  5.0  1.0 35.0\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Note that the matrix is not symmetric: rows correspond to the target\n",
      "     |      sequence, and columns to the query sequence.  For example, the number\n",
      "     |      of T's in the target sequence that are aligned to a C in the query\n",
      "     |      sequence is\n",
      "     |      \n",
      "     |      >>> m['T', 'C']\n",
      "     |      5.0\n",
      "     |      \n",
      "     |      and the number of C's in the query sequence tat are aligned to a T in\n",
      "     |      the query sequence is\n",
      "     |      \n",
      "     |      >>> m['C', 'T']\n",
      "     |      2.0\n",
      "     |      \n",
      "     |      For some applications (for example, to define a scoring matrix from\n",
      "     |      the substitution matrix), a symmetric matrix may be preferred, which\n",
      "     |      can be calculated as follows:\n",
      "     |      \n",
      "     |      >>> m += m.transpose()\n",
      "     |      >>> m /= 2.0\n",
      "     |      >>> print(m)\n",
      "     |           A    C    G    T\n",
      "     |      A 28.0  0.5  2.0  1.5\n",
      "     |      C  0.5 39.0  0.5  3.5\n",
      "     |      G  2.0  0.5 45.0  0.5\n",
      "     |      T  1.5  3.5  0.5 35.0\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      The matrix is now symmetric, with counts divided equally on both sides\n",
      "     |      of the diagonal:\n",
      "     |      \n",
      "     |      >>> m['C', 'T']\n",
      "     |      3.5\n",
      "     |      >>> m['T', 'C']\n",
      "     |      3.5\n",
      "     |      \n",
      "     |      The total number of substitutions between T's and C's in the alignment\n",
      "     |      is 3.5 + 3.5 = 7.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  path\n",
      "     |      Return the path through the trace matrix.\n",
      "     |  \n",
      "     |  query\n",
      "     |      Return self.sequences[1] for a pairwise alignment.\n",
      "     |  \n",
      "     |  target\n",
      "     |      Return self.sequences[0] for a pairwise alignment.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes defined here:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "    \n",
      "    class MultipleSeqAlignment(builtins.object)\n",
      "     |  MultipleSeqAlignment(records, alphabet=None, annotations=None, column_annotations=None)\n",
      "     |  \n",
      "     |  Represents a classical multiple sequence alignment (MSA).\n",
      "     |  \n",
      "     |  By this we mean a collection of sequences (usually shown as rows) which\n",
      "     |  are all the same length (usually with gap characters for insertions or\n",
      "     |  padding). The data can then be regarded as a matrix of letters, with well\n",
      "     |  defined columns.\n",
      "     |  \n",
      "     |  You would typically create an MSA by loading an alignment file with the\n",
      "     |  AlignIO module:\n",
      "     |  \n",
      "     |  >>> from Bio import AlignIO\n",
      "     |  >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      "     |  >>> print(align)\n",
      "     |  Alignment with 7 rows and 156 columns\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  In some respects you can treat these objects as lists of SeqRecord objects,\n",
      "     |  each representing a row of the alignment. Iterating over an alignment gives\n",
      "     |  the SeqRecord object for each row:\n",
      "     |  \n",
      "     |  >>> len(align)\n",
      "     |  7\n",
      "     |  >>> for record in align:\n",
      "     |  ...     print(\"%s %i\" % (record.id, len(record)))\n",
      "     |  ...\n",
      "     |  gi|6273285|gb|AF191659.1|AF191 156\n",
      "     |  gi|6273284|gb|AF191658.1|AF191 156\n",
      "     |  gi|6273287|gb|AF191661.1|AF191 156\n",
      "     |  gi|6273286|gb|AF191660.1|AF191 156\n",
      "     |  gi|6273290|gb|AF191664.1|AF191 156\n",
      "     |  gi|6273289|gb|AF191663.1|AF191 156\n",
      "     |  gi|6273291|gb|AF191665.1|AF191 156\n",
      "     |  \n",
      "     |  You can also access individual rows as SeqRecord objects via their index:\n",
      "     |  \n",
      "     |  >>> print(align[0].id)\n",
      "     |  gi|6273285|gb|AF191659.1|AF191\n",
      "     |  >>> print(align[-1].id)\n",
      "     |  gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  And extract columns as strings:\n",
      "     |  \n",
      "     |  >>> print(align[:, 1])\n",
      "     |  AAAAAAA\n",
      "     |  \n",
      "     |  Or, take just the first ten columns as a sub-alignment:\n",
      "     |  \n",
      "     |  >>> print(align[:, :10])\n",
      "     |  Alignment with 7 rows and 10 columns\n",
      "     |  TATACATTAA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  Combining this alignment slicing with alignment addition allows you to\n",
      "     |  remove a section of the alignment. For example, taking just the first\n",
      "     |  and last ten columns:\n",
      "     |  \n",
      "     |  >>> print(align[:, :10] + align[:, -10:])\n",
      "     |  Alignment with 7 rows and 20 columns\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAAGTGTACCAGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAAGTATACCAGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  Note - This object does NOT attempt to model the kind of alignments used\n",
      "     |  in next generation sequencing with multiple sequencing reads which are\n",
      "     |  much shorter than the alignment, and where there is usually a consensus or\n",
      "     |  reference sequence with special status.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __add__(self, other)\n",
      "     |      Combine two alignments with the same number of rows by adding them.\n",
      "     |      \n",
      "     |      If you have two multiple sequence alignments (MSAs), there are two ways to think\n",
      "     |      about adding them - by row or by column. Using the extend method adds by row.\n",
      "     |      Using the addition operator adds by column. For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a1 = SeqRecord(Seq(\"AAAAC\"), id=\"Alpha\")\n",
      "     |      >>> b1 = SeqRecord(Seq(\"AAA-C\"), id=\"Beta\")\n",
      "     |      >>> c1 = SeqRecord(Seq(\"AAAAG\"), id=\"Gamma\")\n",
      "     |      >>> a2 = SeqRecord(Seq(\"GT\"), id=\"Alpha\")\n",
      "     |      >>> b2 = SeqRecord(Seq(\"GT\"), id=\"Beta\")\n",
      "     |      >>> c2 = SeqRecord(Seq(\"GT\"), id=\"Gamma\")\n",
      "     |      >>> left = MultipleSeqAlignment([a1, b1, c1],\n",
      "     |      ...                             annotations={\"tool\": \"demo\", \"name\": \"start\"},\n",
      "     |      ...                             column_annotations={\"stats\": \"CCCXC\"})\n",
      "     |      >>> right = MultipleSeqAlignment([a2, b2, c2],\n",
      "     |      ...                             annotations={\"tool\": \"demo\", \"name\": \"end\"},\n",
      "     |      ...                             column_annotations={\"stats\": \"CC\"})\n",
      "     |      \n",
      "     |      Now, let's look at these two alignments:\n",
      "     |      \n",
      "     |      >>> print(left)\n",
      "     |      Alignment with 3 rows and 5 columns\n",
      "     |      AAAAC Alpha\n",
      "     |      AAA-C Beta\n",
      "     |      AAAAG Gamma\n",
      "     |      >>> print(right)\n",
      "     |      Alignment with 3 rows and 2 columns\n",
      "     |      GT Alpha\n",
      "     |      GT Beta\n",
      "     |      GT Gamma\n",
      "     |      \n",
      "     |      And add them:\n",
      "     |      \n",
      "     |      >>> combined = left + right\n",
      "     |      >>> print(combined)\n",
      "     |      Alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      \n",
      "     |      For this to work, both alignments must have the same number of records (here\n",
      "     |      they both have 3 rows):\n",
      "     |      \n",
      "     |      >>> len(left)\n",
      "     |      3\n",
      "     |      >>> len(right)\n",
      "     |      3\n",
      "     |      >>> len(combined)\n",
      "     |      3\n",
      "     |      \n",
      "     |      The individual rows are SeqRecord objects, and these can be added together. Refer\n",
      "     |      to the SeqRecord documentation for details of how the annotation is handled. This\n",
      "     |      example is a special case in that both original alignments shared the same names,\n",
      "     |      meaning when the rows are added they also get the same name.\n",
      "     |      \n",
      "     |      Any common annotations are preserved, but differing annotation is lost. This is\n",
      "     |      the same behaviour used in the SeqRecord annotations and is designed to prevent\n",
      "     |      accidental propagation of inappropriate values:\n",
      "     |      \n",
      "     |      >>> combined.annotations\n",
      "     |      {'tool': 'demo'}\n",
      "     |      \n",
      "     |      Similarly any common per-column-annotations are combined:\n",
      "     |      \n",
      "     |      >>> combined.column_annotations\n",
      "     |      {'stats': 'CCCXCCC'}\n",
      "     |  \n",
      "     |  __format__(self, format_spec)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      The format should be a lower case string supported as an output\n",
      "     |      format by Bio.AlignIO (such as \"fasta\", \"clustal\", \"phylip\",\n",
      "     |      \"stockholm\", etc), which is used to turn the alignment into a\n",
      "     |      string.\n",
      "     |      \n",
      "     |      e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\", description=\"\")\n",
      "     |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\", description=\"\")\n",
      "     |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\", description=\"\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> print(format(align, \"fasta\"))\n",
      "     |      >Alpha\n",
      "     |      ACTGCTAGCTAG\n",
      "     |      >Beta\n",
      "     |      ACT-CTAGCTAG\n",
      "     |      >Gamma\n",
      "     |      ACTGCTAGATAG\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(format(align, \"phylip\"))\n",
      "     |       3 12\n",
      "     |      Alpha      ACTGCTAGCT AG\n",
      "     |      Beta       ACT-CTAGCT AG\n",
      "     |      Gamma      ACTGCTAGAT AG\n",
      "     |      <BLANKLINE>\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |      Access part of the alignment.\n",
      "     |      \n",
      "     |      Depending on the indices, you can get a SeqRecord object\n",
      "     |      (representing a single row), a Seq object (for a single columns),\n",
      "     |      a string (for a single characters) or another alignment\n",
      "     |      (representing some part or all of the alignment).\n",
      "     |      \n",
      "     |      align[r,c] gives a single character as a string\n",
      "     |      align[r] gives a row as a SeqRecord\n",
      "     |      align[r,:] gives a row as a SeqRecord\n",
      "     |      align[:,c] gives a column as a Seq\n",
      "     |      \n",
      "     |      align[:] and align[:,:] give a copy of the alignment\n",
      "     |      \n",
      "     |      Anything else gives a sub alignment, e.g.\n",
      "     |      align[0:2] or align[0:2,:] uses only row 0 and 1\n",
      "     |      align[:,1:3] uses only columns 1 and 2\n",
      "     |      align[0:2,1:3] uses only rows 0 & 1 and only cols 1 & 2\n",
      "     |      \n",
      "     |      We'll use the following example alignment here for illustration:\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      "     |      >>> d = SeqRecord(Seq(\"AAAACGT\"), id=\"Delta\")\n",
      "     |      >>> e = SeqRecord(Seq(\"AAA-GGT\"), id=\"Epsilon\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c, d, e])\n",
      "     |      \n",
      "     |      You can access a row of the alignment as a SeqRecord using an integer\n",
      "     |      index (think of the alignment as a list of SeqRecord objects here):\n",
      "     |      \n",
      "     |      >>> first_record = align[0]\n",
      "     |      >>> print(\"%s %s\" % (first_record.id, first_record.seq))\n",
      "     |      Alpha AAAACGT\n",
      "     |      >>> last_record = align[-1]\n",
      "     |      >>> print(\"%s %s\" % (last_record.id, last_record.seq))\n",
      "     |      Epsilon AAA-GGT\n",
      "     |      \n",
      "     |      You can also access use python's slice notation to create a sub-alignment\n",
      "     |      containing only some of the SeqRecord objects:\n",
      "     |      \n",
      "     |      >>> sub_alignment = align[2:5]\n",
      "     |      >>> print(sub_alignment)\n",
      "     |      Alignment with 3 rows and 7 columns\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAAACGT Delta\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      This includes support for a step, i.e. align[start:end:step], which\n",
      "     |      can be used to select every second sequence:\n",
      "     |      \n",
      "     |      >>> sub_alignment = align[::2]\n",
      "     |      >>> print(sub_alignment)\n",
      "     |      Alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      Or to get a copy of the alignment with the rows in reverse order:\n",
      "     |      \n",
      "     |      >>> rev_alignment = align[::-1]\n",
      "     |      >>> print(rev_alignment)\n",
      "     |      Alignment with 5 rows and 7 columns\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      AAAACGT Delta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAACGT Alpha\n",
      "     |      \n",
      "     |      You can also use two indices to specify both rows and columns. Using simple\n",
      "     |      integers gives you the entry as a single character string. e.g.\n",
      "     |      \n",
      "     |      >>> align[3, 4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      This is equivalent to:\n",
      "     |      \n",
      "     |      >>> align[3][4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      or:\n",
      "     |      \n",
      "     |      >>> align[3].seq[4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      To get a single column (as a string) use this syntax:\n",
      "     |      \n",
      "     |      >>> align[:, 4]\n",
      "     |      'CCGCG'\n",
      "     |      \n",
      "     |      Or, to get part of a column,\n",
      "     |      \n",
      "     |      >>> align[1:3, 4]\n",
      "     |      'CG'\n",
      "     |      \n",
      "     |      However, in general you get a sub-alignment,\n",
      "     |      \n",
      "     |      >>> print(align[1:5, 3:6])\n",
      "     |      Alignment with 4 rows and 3 columns\n",
      "     |      -CG Beta\n",
      "     |      AGG Gamma\n",
      "     |      ACG Delta\n",
      "     |      -GG Epsilon\n",
      "     |      \n",
      "     |      This should all seem familiar to anyone who has used the NumPy\n",
      "     |      array or matrix objects.\n",
      "     |  \n",
      "     |  __init__(self, records, alphabet=None, annotations=None, column_annotations=None)\n",
      "     |      Initialize a new MultipleSeqAlignment object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - records - A list (or iterator) of SeqRecord objects, whose\n",
      "     |                   sequences are all the same length.  This may be an be an\n",
      "     |                   empty list.\n",
      "     |       - alphabet - For backward compatibility only; its value should always\n",
      "     |                    be None.\n",
      "     |       - annotations - Information about the whole alignment (dictionary).\n",
      "     |       - column_annotations - Per column annotation (restricted dictionary).\n",
      "     |                    This holds Python sequences (lists, strings, tuples)\n",
      "     |                    whose length matches the number of columns. A typical\n",
      "     |                    use would be a secondary structure consensus string.\n",
      "     |      \n",
      "     |      You would normally load a MSA from a file using Bio.AlignIO, but you\n",
      "     |      can do this from a list of SeqRecord objects too:\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c],\n",
      "     |      ...                              annotations={\"tool\": \"demo\"},\n",
      "     |      ...                              column_annotations={\"stats\": \"CCCXCCC\"})\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      >>> align.annotations\n",
      "     |      {'tool': 'demo'}\n",
      "     |      >>> align.column_annotations\n",
      "     |      {'stats': 'CCCXCCC'}\n",
      "     |  \n",
      "     |  __iter__(self)\n",
      "     |      Iterate over alignment rows as SeqRecord objects.\n",
      "     |      \n",
      "     |      e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> for record in align:\n",
      "     |      ...    print(record.id)\n",
      "     |      ...    print(record.seq)\n",
      "     |      ...\n",
      "     |      Alpha\n",
      "     |      ACTGCTAGCTAG\n",
      "     |      Beta\n",
      "     |      ACT-CTAGCTAG\n",
      "     |      Gamma\n",
      "     |      ACTGCTAGATAG\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return the number of sequences in the alignment.\n",
      "     |      \n",
      "     |      Use len(alignment) to get the number of sequences (i.e. the number of\n",
      "     |      rows), and alignment.get_alignment_length() to get the length of the\n",
      "     |      longest sequence (i.e. the number of columns).\n",
      "     |      \n",
      "     |      This is easy to remember if you think of the alignment as being like a\n",
      "     |      list of SeqRecord objects.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return a representation of the object for debugging.\n",
      "     |      \n",
      "     |      The representation cannot be used with eval() to recreate the object,\n",
      "     |      which is usually possible with simple python objects.  For example:\n",
      "     |      \n",
      "     |      <Bio.Align.MultipleSeqAlignment instance (2 records of length 14)\n",
      "     |      at a3c184c>\n",
      "     |      \n",
      "     |      The hex string is the memory address of the object, see help(id).\n",
      "     |      This provides a simple way to visually distinguish alignments of\n",
      "     |      the same size.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return a multi-line string summary of the alignment.\n",
      "     |      \n",
      "     |      This output is intended to be readable, but large alignments are\n",
      "     |      shown truncated.  A maximum of 20 rows (sequences) and 50 columns\n",
      "     |      are shown, with the record identifiers.  This should fit nicely on a\n",
      "     |      single screen. e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 3 rows and 12 columns\n",
      "     |      ACTGCTAGCTAG Alpha\n",
      "     |      ACT-CTAGCTAG Beta\n",
      "     |      ACTGCTAGATAG Gamma\n",
      "     |      \n",
      "     |      See also the alignment's format method.\n",
      "     |  \n",
      "     |  append(self, record)\n",
      "     |      Add one more SeqRecord object to the alignment as a new row.\n",
      "     |      \n",
      "     |      This must have the same length as the original alignment (unless this is\n",
      "     |      the first record).\n",
      "     |      \n",
      "     |      >>> from Bio import AlignIO\n",
      "     |      >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 7 rows and 156 columns\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |      >>> len(align)\n",
      "     |      7\n",
      "     |      \n",
      "     |      We'll now construct a dummy record to append as an example:\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> dummy = SeqRecord(Seq(\"N\"*156), id=\"dummy\")\n",
      "     |      \n",
      "     |      Now append this to the alignment,\n",
      "     |      \n",
      "     |      >>> align.append(dummy)\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 8 rows and 156 columns\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |      NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN dummy\n",
      "     |      >>> len(align)\n",
      "     |      8\n",
      "     |  \n",
      "     |  extend(self, records)\n",
      "     |      Add more SeqRecord objects to the alignment as rows.\n",
      "     |      \n",
      "     |      They must all have the same length as the original alignment. For\n",
      "     |      example,\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      "     |      >>> d = SeqRecord(Seq(\"AAAACGT\"), id=\"Delta\")\n",
      "     |      >>> e = SeqRecord(Seq(\"AAA-GGT\"), id=\"Epsilon\")\n",
      "     |      \n",
      "     |      First we create a small alignment (three rows):\n",
      "     |      \n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      \n",
      "     |      Now we can extend this alignment with another two rows:\n",
      "     |      \n",
      "     |      >>> align.extend([d, e])\n",
      "     |      >>> print(align)\n",
      "     |      Alignment with 5 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAAACGT Delta\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      Because the alignment object allows iteration over the rows as\n",
      "     |      SeqRecords, you can use the extend method with a second alignment\n",
      "     |      (provided its sequences have the same length as the original alignment).\n",
      "     |  \n",
      "     |  get_alignment_length(self)\n",
      "     |      Return the maximum length of the alignment.\n",
      "     |      \n",
      "     |      All objects in the alignment should (hopefully) have the same\n",
      "     |      length. This function will go through and find this length\n",
      "     |      by finding the maximum length of sequences in the alignment.\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> align.get_alignment_length()\n",
      "     |      12\n",
      "     |      \n",
      "     |      If you want to know the number of sequences in the alignment,\n",
      "     |      use len(align) instead:\n",
      "     |      \n",
      "     |      >>> len(align)\n",
      "     |      3\n",
      "     |  \n",
      "     |  sort(self, key=None, reverse=False)\n",
      "     |      Sort the rows (SeqRecord objects) of the alignment in place.\n",
      "     |      \n",
      "     |      This sorts the rows alphabetically using the SeqRecord object id by\n",
      "     |      default. The sorting can be controlled by supplying a key function\n",
      "     |      which must map each SeqRecord to a sort value.\n",
      "     |      \n",
      "     |      This is useful if you want to add two alignments which use the same\n",
      "     |      record identifiers, but in a different order. For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align1 = MultipleSeqAlignment([\n",
      "     |      ...              SeqRecord(Seq(\"ACGT\"), id=\"Human\"),\n",
      "     |      ...              SeqRecord(Seq(\"ACGG\"), id=\"Mouse\"),\n",
      "     |      ...              SeqRecord(Seq(\"ACGC\"), id=\"Chicken\"),\n",
      "     |      ...          ])\n",
      "     |      >>> align2 = MultipleSeqAlignment([\n",
      "     |      ...              SeqRecord(Seq(\"CGGT\"), id=\"Mouse\"),\n",
      "     |      ...              SeqRecord(Seq(\"CGTT\"), id=\"Human\"),\n",
      "     |      ...              SeqRecord(Seq(\"CGCT\"), id=\"Chicken\"),\n",
      "     |      ...          ])\n",
      "     |      \n",
      "     |      If you simple try and add these without sorting, you get this:\n",
      "     |      \n",
      "     |      >>> print(align1 + align2)\n",
      "     |      Alignment with 3 rows and 8 columns\n",
      "     |      ACGTCGGT <unknown id>\n",
      "     |      ACGGCGTT <unknown id>\n",
      "     |      ACGCCGCT Chicken\n",
      "     |      \n",
      "     |      Consult the SeqRecord documentation which explains why you get a\n",
      "     |      default value when annotation like the identifier doesn't match up.\n",
      "     |      However, if we sort the alignments first, then add them we get the\n",
      "     |      desired result:\n",
      "     |      \n",
      "     |      >>> align1.sort()\n",
      "     |      >>> align2.sort()\n",
      "     |      >>> print(align1 + align2)\n",
      "     |      Alignment with 3 rows and 8 columns\n",
      "     |      ACGCCGCT Chicken\n",
      "     |      ACGTCGTT Human\n",
      "     |      ACGGCGGT Mouse\n",
      "     |      \n",
      "     |      As an example using a different sort order, you could sort on the\n",
      "     |      GC content of each sequence.\n",
      "     |      \n",
      "     |      >>> from Bio.SeqUtils import gc_fraction\n",
      "     |      >>> print(align1)\n",
      "     |      Alignment with 3 rows and 4 columns\n",
      "     |      ACGC Chicken\n",
      "     |      ACGT Human\n",
      "     |      ACGG Mouse\n",
      "     |      >>> align1.sort(key = lambda record: gc_fraction(record.seq))\n",
      "     |      >>> print(align1)\n",
      "     |      Alignment with 3 rows and 4 columns\n",
      "     |      ACGT Human\n",
      "     |      ACGC Chicken\n",
      "     |      ACGG Mouse\n",
      "     |      \n",
      "     |      There is also a reverse argument, so if you wanted to sort by ID\n",
      "     |      but backwards:\n",
      "     |      \n",
      "     |      >>> align1.sort(reverse=True)\n",
      "     |      >>> print(align1)\n",
      "     |      Alignment with 3 rows and 4 columns\n",
      "     |      ACGG Mouse\n",
      "     |      ACGT Human\n",
      "     |      ACGC Chicken\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  substitutions\n",
      "     |      Return an Array with the number of substitutions of letters in the alignment.\n",
      "     |      \n",
      "     |      As an example, consider a multiple sequence alignment of three DNA sequences:\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> seq1 = SeqRecord(Seq(\"ACGT\"), id=\"seq1\")\n",
      "     |      >>> seq2 = SeqRecord(Seq(\"A--A\"), id=\"seq2\")\n",
      "     |      >>> seq3 = SeqRecord(Seq(\"ACGT\"), id=\"seq3\")\n",
      "     |      >>> seq4 = SeqRecord(Seq(\"TTTC\"), id=\"seq4\")\n",
      "     |      >>> alignment = MultipleSeqAlignment([seq1, seq2, seq3, seq4])\n",
      "     |      >>> print(alignment)\n",
      "     |      Alignment with 4 rows and 4 columns\n",
      "     |      ACGT seq1\n",
      "     |      A--A seq2\n",
      "     |      ACGT seq3\n",
      "     |      TTTC seq4\n",
      "     |      \n",
      "     |      >>> m = alignment.substitutions\n",
      "     |      >>> print(m)\n",
      "     |          A   C   G   T\n",
      "     |      A 3.0 0.5 0.0 2.5\n",
      "     |      C 0.5 1.0 0.0 2.0\n",
      "     |      G 0.0 0.0 1.0 1.0\n",
      "     |      T 2.5 2.0 1.0 1.0\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Note that the matrix is symmetric, with counts divided equally on both\n",
      "     |      sides of the diagonal. For example, the total number of substitutions\n",
      "     |      between A and T in the alignment is 3.5 + 3.5 = 7.\n",
      "     |      \n",
      "     |      Any weights associated with the sequences are taken into account when\n",
      "     |      calculating the substitution matrix.  For example, given the following\n",
      "     |      multiple sequence alignment::\n",
      "     |      \n",
      "     |          GTATC  0.5\n",
      "     |          AT--C  0.8\n",
      "     |          CTGTC  1.0\n",
      "     |      \n",
      "     |      For the first column we have::\n",
      "     |      \n",
      "     |          ('A', 'G') : 0.5 * 0.8 = 0.4\n",
      "     |          ('C', 'G') : 0.5 * 1.0 = 0.5\n",
      "     |          ('A', 'C') : 0.8 * 1.0 = 0.8\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  column_annotations\n",
      "     |      Dictionary of per-letter-annotation for the sequence.\n",
      "    \n",
      "    class PairwiseAligner(_algorithms.PairwiseAligner)\n",
      "     |  PairwiseAligner(scoring=None, **kwargs)\n",
      "     |  \n",
      "     |  Performs pairwise sequence alignment using dynamic programming.\n",
      "     |  \n",
      "     |  This provides functions to get global and local alignments between two\n",
      "     |  sequences.  A global alignment finds the best concordance between all\n",
      "     |  characters in two sequences.  A local alignment finds just the\n",
      "     |  subsequences that align the best.\n",
      "     |  \n",
      "     |  To perform a pairwise sequence alignment, first create a PairwiseAligner\n",
      "     |  object.  This object stores the match and mismatch scores, as well as the\n",
      "     |  gap scores.  Typically, match scores are positive, while mismatch scores\n",
      "     |  and gap scores are negative or zero.  By default, the match score is 1,\n",
      "     |  and the mismatch and gap scores are zero.  Based on the values of the gap\n",
      "     |  scores, a PairwiseAligner object automatically chooses the appropriate\n",
      "     |  alignment algorithm (the Needleman-Wunsch, Smith-Waterman, Gotoh, or\n",
      "     |  Waterman-Smith-Beyer global or local alignment algorithm).\n",
      "     |  \n",
      "     |  Calling the \"score\" method on the aligner with two sequences as arguments\n",
      "     |  will calculate the alignment score between the two sequences.\n",
      "     |  Calling the \"align\" method on the aligner with two sequences as arguments\n",
      "     |  will return a generator yielding the alignments between the two\n",
      "     |  sequences.\n",
      "     |  \n",
      "     |  Some examples:\n",
      "     |  \n",
      "     |  >>> from Bio import Align\n",
      "     |  >>> aligner = Align.PairwiseAligner()\n",
      "     |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      "     |  >>> for alignment in sorted(alignments):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 3.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -|-|| 5\n",
      "     |  query             0 -A-CG 3\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 3.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -||-| 5\n",
      "     |  query             0 -AC-G 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Specify the aligner mode as local to generate local alignments:\n",
      "     |  \n",
      "     |  >>> aligner.mode = 'local'\n",
      "     |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      "     |  >>> for alignment in sorted(alignments):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 3.0:\n",
      "     |  target            1 ACCG 5\n",
      "     |                    0 |-|| 4\n",
      "     |  query             0 A-CG 3\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 3.0:\n",
      "     |  target            1 ACCG 5\n",
      "     |                    0 ||-| 4\n",
      "     |  query             0 AC-G 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Do a global alignment.  Identical characters are given 2 points,\n",
      "     |  1 point is deducted for each non-identical character.\n",
      "     |  \n",
      "     |  >>> aligner.mode = 'global'\n",
      "     |  >>> aligner.match_score = 2\n",
      "     |  >>> aligner.mismatch_score = -1\n",
      "     |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 6.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -||-| 5\n",
      "     |  query             0 -AC-G 3\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 6.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -|-|| 5\n",
      "     |  query             0 -A-CG 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Same as above, except now 0.5 points are deducted when opening a\n",
      "     |  gap, and 0.1 points are deducted when extending it.\n",
      "     |  \n",
      "     |  >>> aligner.open_gap_score = -0.5\n",
      "     |  >>> aligner.extend_gap_score = -0.1\n",
      "     |  >>> aligner.target_end_gap_score = 0.0\n",
      "     |  >>> aligner.query_end_gap_score = 0.0\n",
      "     |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 5.5:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -|-|| 5\n",
      "     |  query             0 -A-CG 3\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 5.5:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -||-| 5\n",
      "     |  query             0 -AC-G 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  The alignment function can also use known matrices already included in\n",
      "     |  Biopython:\n",
      "     |  \n",
      "     |  >>> from Bio.Align import substitution_matrices\n",
      "     |  >>> aligner = Align.PairwiseAligner()\n",
      "     |  >>> aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
      "     |  >>> alignments = aligner.align(\"KEVLA\", \"EVL\")\n",
      "     |  >>> alignments = list(alignments)\n",
      "     |  >>> print(\"Number of alignments: %d\" % len(alignments))\n",
      "     |  Number of alignments: 1\n",
      "     |  >>> alignment = alignments[0]\n",
      "     |  >>> print(\"Score = %.1f\" % alignment.score)\n",
      "     |  Score = 13.0\n",
      "     |  >>> print(alignment)\n",
      "     |  target            0 KEVLA 5\n",
      "     |                    0 -|||- 5\n",
      "     |  query             0 -EVL- 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  You can also set the value of attributes directly during construction\n",
      "     |  of the PairwiseAligner object by providing them as keyword arguments:\n",
      "     |  \n",
      "     |  >>> aligner = Align.PairwiseAligner(mode='global', match_score=2, mismatch_score=-1)\n",
      "     |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 6.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -||-| 5\n",
      "     |  query             0 -AC-G 3\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 6.0:\n",
      "     |  target            0 TACCG 5\n",
      "     |                    0 -|-|| 5\n",
      "     |  query             0 -A-CG 3\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PairwiseAligner\n",
      "     |      _algorithms.PairwiseAligner\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __init__(self, scoring=None, **kwargs)\n",
      "     |      Initialize a new PairwiseAligner as specified by the keyword arguments.\n",
      "     |      \n",
      "     |      If scoring is None, use the default scoring scheme match = 1.0,\n",
      "     |      mismatch = 0.0, gap score = 0.0\n",
      "     |      If scoring is \"blastn\", \"megablast\", or \"blastp\", use the default\n",
      "     |      substitution matrix and gap scores for BLASTN, MEGABLAST, or BLASTP,\n",
      "     |      respectively.\n",
      "     |      \n",
      "     |      Loops over the remaining keyword arguments and sets them as attributes\n",
      "     |      on the object.\n",
      "     |  \n",
      "     |  __setattr__(self, key, value)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  align(self, seqA, seqB, strand='+')\n",
      "     |      Return the alignments of two sequences using PairwiseAligner.\n",
      "     |  \n",
      "     |  score(self, seqA, seqB, strand='+')\n",
      "     |      Return the alignments score of two sequences using PairwiseAligner.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  algorithm\n",
      "     |      alignment algorithm\n",
      "     |  \n",
      "     |  alphabet\n",
      "     |      alphabet\n",
      "     |  \n",
      "     |  end_extend_gap_score\n",
      "     |      end extend gap score\n",
      "     |  \n",
      "     |  end_gap_score\n",
      "     |      end gap score\n",
      "     |  \n",
      "     |  end_open_gap_score\n",
      "     |      end open gap score\n",
      "     |  \n",
      "     |  epsilon\n",
      "     |      roundoff epsilon\n",
      "     |  \n",
      "     |  extend_gap_score\n",
      "     |      extend gap score\n",
      "     |  \n",
      "     |  gap_score\n",
      "     |      gap score\n",
      "     |  \n",
      "     |  internal_extend_gap_score\n",
      "     |      internal extend gap score\n",
      "     |  \n",
      "     |  internal_gap_score\n",
      "     |      internal gap score\n",
      "     |  \n",
      "     |  internal_open_gap_score\n",
      "     |      internal open gap score\n",
      "     |  \n",
      "     |  left_extend_gap_score\n",
      "     |      left extend gap score\n",
      "     |  \n",
      "     |  left_gap_score\n",
      "     |      left gap score\n",
      "     |  \n",
      "     |  left_open_gap_score\n",
      "     |      left open gap score\n",
      "     |  \n",
      "     |  match\n",
      "     |      match score\n",
      "     |  \n",
      "     |  match_score\n",
      "     |      match score\n",
      "     |  \n",
      "     |  mismatch\n",
      "     |      mismatch score\n",
      "     |  \n",
      "     |  mismatch_score\n",
      "     |      mismatch score\n",
      "     |  \n",
      "     |  mode\n",
      "     |      alignment mode ('global' or 'local')\n",
      "     |  \n",
      "     |  open_gap_score\n",
      "     |      internal and end open gap score\n",
      "     |  \n",
      "     |  query_end_extend_gap_score\n",
      "     |      query end extend score\n",
      "     |  \n",
      "     |  query_end_gap_score\n",
      "     |      query end score\n",
      "     |  \n",
      "     |  query_end_open_gap_score\n",
      "     |      query end open score\n",
      "     |  \n",
      "     |  query_extend_gap_score\n",
      "     |      query gap extend score\n",
      "     |  \n",
      "     |  query_gap_score\n",
      "     |      query gap score\n",
      "     |  \n",
      "     |  query_internal_extend_gap_score\n",
      "     |      query internal extend gap score\n",
      "     |  \n",
      "     |  query_internal_gap_score\n",
      "     |      query internal gap score\n",
      "     |  \n",
      "     |  query_internal_open_gap_score\n",
      "     |      query internal open gap score\n",
      "     |  \n",
      "     |  query_left_extend_gap_score\n",
      "     |      query left extend score\n",
      "     |  \n",
      "     |  query_left_gap_score\n",
      "     |      query left score\n",
      "     |  \n",
      "     |  query_left_open_gap_score\n",
      "     |      query left open score\n",
      "     |  \n",
      "     |  query_open_gap_score\n",
      "     |      query gap open score\n",
      "     |  \n",
      "     |  query_right_extend_gap_score\n",
      "     |      query right extend score\n",
      "     |  \n",
      "     |  query_right_gap_score\n",
      "     |      query right score\n",
      "     |  \n",
      "     |  query_right_open_gap_score\n",
      "     |      query right open score\n",
      "     |  \n",
      "     |  right_extend_gap_score\n",
      "     |      right extend gap score\n",
      "     |  \n",
      "     |  right_gap_score\n",
      "     |      right gap score\n",
      "     |  \n",
      "     |  right_open_gap_score\n",
      "     |      right open gap score\n",
      "     |  \n",
      "     |  substitution_matrix\n",
      "     |      substitution_matrix\n",
      "     |  \n",
      "     |  target_end_extend_gap_score\n",
      "     |      target end extend gap score\n",
      "     |  \n",
      "     |  target_end_gap_score\n",
      "     |      target end gap score\n",
      "     |  \n",
      "     |  target_end_open_gap_score\n",
      "     |      target end open gap score\n",
      "     |  \n",
      "     |  target_extend_gap_score\n",
      "     |      target extend gap score\n",
      "     |  \n",
      "     |  target_gap_score\n",
      "     |      target gap score\n",
      "     |  \n",
      "     |  target_internal_extend_gap_score\n",
      "     |      target internal extend gap score\n",
      "     |  \n",
      "     |  target_internal_gap_score\n",
      "     |      target internal gap score\n",
      "     |  \n",
      "     |  target_internal_open_gap_score\n",
      "     |      target internal open gap score\n",
      "     |  \n",
      "     |  target_left_extend_gap_score\n",
      "     |      target left extend score\n",
      "     |  \n",
      "     |  target_left_gap_score\n",
      "     |      target left score\n",
      "     |  \n",
      "     |  target_left_open_gap_score\n",
      "     |      target left open score\n",
      "     |  \n",
      "     |  target_open_gap_score\n",
      "     |      target open gap score\n",
      "     |  \n",
      "     |  target_right_extend_gap_score\n",
      "     |      target right extend score\n",
      "     |  \n",
      "     |  target_right_gap_score\n",
      "     |      target right score\n",
      "     |  \n",
      "     |  target_right_open_gap_score\n",
      "     |      target right open score\n",
      "     |  \n",
      "     |  wildcard\n",
      "     |      wildcard character\n",
      "    \n",
      "    class PairwiseAlignment(Alignment)\n",
      "     |  PairwiseAlignment(target, query, path, score)\n",
      "     |  \n",
      "     |  Represents a pairwise sequence alignment.\n",
      "     |  \n",
      "     |  Internally, the pairwise alignment is stored as the path through\n",
      "     |  the traceback matrix, i.e. a tuple of pairs of indices corresponding\n",
      "     |  to the vertices of the path in the traceback matrix.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PairwiseAlignment\n",
      "     |      Alignment\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, target, query, path, score)\n",
      "     |      Initialize a new PairwiseAlignment object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - target  - The first sequence, as a plain string, without gaps.\n",
      "     |       - query   - The second sequence, as a plain string, without gaps.\n",
      "     |       - path    - The path through the traceback matrix, defining an\n",
      "     |                   alignment.\n",
      "     |       - score   - The alignment score.\n",
      "     |      \n",
      "     |      You would normally obtain a PairwiseAlignment object by iterating\n",
      "     |      over a PairwiseAlignments object.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from Alignment:\n",
      "     |  \n",
      "     |  __array__(self, dtype=None)\n",
      "     |  \n",
      "     |  __eq__(self, other)\n",
      "     |      Check if two Alignment objects specify the same alignment.\n",
      "     |  \n",
      "     |  __format__(self, format_spec)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      Wrapper for self.format().\n",
      "     |  \n",
      "     |  __ge__(self, other)\n",
      "     |      Check if self should come after or is equal to other.\n",
      "     |  \n",
      "     |  __getitem__(self, key)\n",
      "     |      Return self[key].\n",
      "     |      \n",
      "     |      Indices of the form\n",
      "     |      \n",
      "     |      self[:, :]\n",
      "     |      \n",
      "     |      return a copy of the Alignment object;\n",
      "     |      \n",
      "     |      self[:, i:]\n",
      "     |      self[:, :j]\n",
      "     |      self[:, i:j]\n",
      "     |      self[:, iterable] (where iterable returns integers)\n",
      "     |      \n",
      "     |      return a new Alignment object spanning the selected columns;\n",
      "     |      \n",
      "     |      self[k, i]\n",
      "     |      self[k, i:]\n",
      "     |      self[k, :j]\n",
      "     |      self[k, i:j]\n",
      "     |      self[k, iterable] (where iterable returns integers)\n",
      "     |      self[k] (equivalent to self[k, :])\n",
      "     |      \n",
      "     |      return a string with the aligned sequence (including gaps) for the\n",
      "     |      selected columns, where k = 0 represents the target and k = 1\n",
      "     |      represents the query sequence; and\n",
      "     |      \n",
      "     |      self[:, i]\n",
      "     |      \n",
      "     |      returns a string with the selected column in the alignment.\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> alignments = aligner.align(\"ACCGGTTT\", \"ACGGGTT\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ACCGG-TTT 8\n",
      "     |                        0 ||-||-||- 9\n",
      "     |      query             0 AC-GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment[0, :]\n",
      "     |      'ACCGG-TTT'\n",
      "     |      >>> alignment[1, :]\n",
      "     |      'AC-GGGTT-'\n",
      "     |      >>> alignment[0]\n",
      "     |      'ACCGG-TTT'\n",
      "     |      >>> alignment[1]\n",
      "     |      'AC-GGGTT-'\n",
      "     |      >>> alignment[0, 1:-2]\n",
      "     |      'CCGG-T'\n",
      "     |      >>> alignment[1, 1:-2]\n",
      "     |      'C-GGGT'\n",
      "     |      >>> alignment[0, (1, 5, 2)]\n",
      "     |      'C-C'\n",
      "     |      >>> alignment[1, ::2]\n",
      "     |      'A-GT-'\n",
      "     |      >>> alignment[1, range(0, 9, 2)]\n",
      "     |      'A-GT-'\n",
      "     |      >>> alignment[:, 0]\n",
      "     |      'AA'\n",
      "     |      >>> alignment[:, 5]\n",
      "     |      '-G'\n",
      "     |      >>> alignment[:, 1:]  # doctest:+ELLIPSIS\n",
      "     |      <Alignment object (2 rows x 8 columns) at 0x...>\n",
      "     |      >>> print(alignment[:, 1:])\n",
      "     |      target            1 CCGG-TTT 8\n",
      "     |                        0 |-||-||- 8\n",
      "     |      query             1 C-GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 2:])\n",
      "     |      target            2 CGG-TTT 8\n",
      "     |                        0 -||-||- 7\n",
      "     |      query             2 -GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 3:])\n",
      "     |      target            3 GG-TTT 8\n",
      "     |                        0 ||-||- 6\n",
      "     |      query             2 GGGTT- 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, 3:-1])\n",
      "     |      target            3 GG-TT 7\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             2 GGGTT 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, ::2])\n",
      "     |      target            0 ACGTT 5\n",
      "     |                        0 |-||- 5\n",
      "     |      query             0 A-GT- 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, range(1, 9, 2)])\n",
      "     |      target            0 CG-T 3\n",
      "     |                        0 ||-| 4\n",
      "     |      query             0 CGGT 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(alignment[:, (2, 7, 3)])\n",
      "     |      target            0 CTG 3\n",
      "     |                        0 -|| 3\n",
      "     |      query             0 -TG 2\n",
      "     |      <BLANKLINE>\n",
      "     |  \n",
      "     |  __gt__(self, other)\n",
      "     |      Check if self should come after other.\n",
      "     |  \n",
      "     |  __le__(self, other)\n",
      "     |      Check if self should come before or is equal to other.\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return the number of sequences in the alignment.\n",
      "     |  \n",
      "     |  __lt__(self, other)\n",
      "     |      Check if self should come before other.\n",
      "     |  \n",
      "     |  __ne__(self, other)\n",
      "     |      Check if two Alignment objects have different alignments.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return a representation of the alignment, including its shape.\n",
      "     |      \n",
      "     |      The representation cannot be used with eval() to recreate the object,\n",
      "     |      which is usually possible with simple python objects.  For example:\n",
      "     |      \n",
      "     |      <Alignment object (2 rows x 14 columns) at 0x10403d850>\n",
      "     |      \n",
      "     |      The hex string is the memory address of the object and can be used to\n",
      "     |      distinguish different Alignment objects.  See help(id) for more\n",
      "     |      information.\n",
      "     |      \n",
      "     |      >>> import numpy\n",
      "     |      >>> from Bio.Align import Alignment\n",
      "     |      >>> alignment = Alignment((\"ACCGT\", \"ACGT\"),\n",
      "     |      ...                       coordinates = numpy.array([[0, 2, 3, 5],\n",
      "     |      ...                                                  [0, 2, 2, 4],\n",
      "     |      ...                                                 ]))\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ACCGT 5\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AC-GT 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment  # doctest:+ELLIPSIS\n",
      "     |      <Alignment object (2 rows x 5 columns) at 0x...>\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return a human-readable string representation of the alignment.\n",
      "     |      \n",
      "     |      For sequence alignments, each line has at most 80 columns.\n",
      "     |      The first 10 columns show the (possibly truncated) sequence name,\n",
      "     |      which may be the id attribute of a SeqRecord, or otherwise 'target'\n",
      "     |      or 'query' for pairwise alignments.\n",
      "     |      The next 10 columns show the sequence coordinate, using zero-based\n",
      "     |      counting as usual in Python.\n",
      "     |      The remaining 60 columns shown the sequence, using dashes to represent\n",
      "     |      gaps.\n",
      "     |      At the end of the alignment, the end coordinates are shown on the right\n",
      "     |      of the sequence, again in zero-based coordinates.\n",
      "     |      \n",
      "     |      Pairwise alignments have an additional line between the two sequences\n",
      "     |      showing whether the sequences match ('|') or mismatch ('.'), or if\n",
      "     |      there is a gap ('-').\n",
      "     |      The coordinates shown for this line are the column indices, which can\n",
      "     |      be useful when extracting a subalignment.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      \n",
      "     |      >>> seqA = \"TTAACCCCATTTG\"\n",
      "     |      >>> seqB = \"AAGCCCCTTT\"\n",
      "     |      >>> seqC = \"AAAGGGGCTT\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(seqA, seqB)\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 TTAA-CCCCATTTG 13\n",
      "     |                        0 --||-||||-|||- 14\n",
      "     |      query             0 --AAGCCCC-TTT- 10\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Note that seqC is the reverse complement of seqB. Aligning it to the\n",
      "     |      reverse strand gives the same alignment, but the query coordinates are\n",
      "     |      switched:\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(seqA, seqC, strand=\"-\")\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 TTAA-CCCCATTTG 13\n",
      "     |                        0 --||-||||-|||- 14\n",
      "     |      query            10 --AAGCCCC-TTT-  0\n",
      "     |      <BLANKLINE>\n",
      "     |  \n",
      "     |  format(self, fmt='', *args, **kwargs)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - fmt       - File format. Acceptable values are an empty string to\n",
      "     |                     create a human-readable representation of the alignment,\n",
      "     |                     or any of the alignment file formats supported by\n",
      "     |                     `Bio.Align` (some have not yet been implemented).\n",
      "     |      \n",
      "     |      All other arguments are passed to the format-specific writer functions:\n",
      "     |       - mask      - PSL format only. Specify if repeat regions in the target\n",
      "     |                     sequence are masked and should be reported in the\n",
      "     |                     `repMatches` field of the PSL file instead of in the\n",
      "     |                     `matches` field. Acceptable values are\n",
      "     |                     None   : no masking (default);\n",
      "     |                     \"lower\": masking by lower-case characters;\n",
      "     |                     \"upper\": masking by upper-case characters.\n",
      "     |       - wildcard  - PSL format only. Report alignments to the wildcard\n",
      "     |                     character in the target or query sequence in the\n",
      "     |                     `nCount` field of the PSL file instead of in the\n",
      "     |                     `matches`, `misMatches`, or `repMatches` fields.\n",
      "     |                     Default value is 'N'.\n",
      "     |       - md        - SAM format only. If True, calculate the MD tag from\n",
      "     |                     the alignment and include it in the output. If False\n",
      "     |                     (default), do not include the MD tag in the output.\n",
      "     |  \n",
      "     |  map(self, alignment)\n",
      "     |      Map the alignment to self.target and return the resulting alignment.\n",
      "     |      \n",
      "     |      Here, self.query and alignment.target are the same sequence.\n",
      "     |      \n",
      "     |      A typical example is where self is the pairwise alignment between a\n",
      "     |      chromosome and a transcript, the argument is the pairwise alignment\n",
      "     |      between the transcript and a sequence (e.g., as obtained by RNA-seq),\n",
      "     |      and we want to find the alignment of the sequence to the chromosome:\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = 'local'\n",
      "     |      >>> aligner.open_gap_score = -1\n",
      "     |      >>> aligner.extend_gap_score = 0\n",
      "     |      >>> chromosome = \"AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA\"\n",
      "     |      >>> transcript = \"CCCCCCCGGGGGG\"\n",
      "     |      >>> alignments1 = aligner.align(chromosome, transcript)\n",
      "     |      >>> len(alignments1)\n",
      "     |      1\n",
      "     |      >>> alignment1 = alignments1[0]\n",
      "     |      >>> print(alignment1)\n",
      "     |      target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32\n",
      "     |                        0 |||||||-----------|||||| 24\n",
      "     |      query             0 CCCCCCC-----------GGGGGG 13\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> sequence = \"CCCCGGGG\"\n",
      "     |      >>> alignments2 = aligner.align(transcript, sequence)\n",
      "     |      >>> len(alignments2)\n",
      "     |      1\n",
      "     |      >>> alignment2 = alignments2[0]\n",
      "     |      >>> print(alignment2)\n",
      "     |      target            3 CCCCGGGG 11\n",
      "     |                        0 ||||||||  8\n",
      "     |      query             0 CCCCGGGG  8\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment = alignment1.map(alignment2)\n",
      "     |      >>> print(alignment)\n",
      "     |      target           11 CCCCAAAAAAAAAAAGGGG 30\n",
      "     |                        0 ||||-----------|||| 19\n",
      "     |      query             0 CCCC-----------GGGG  8\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> format(alignment, \"psl\")\n",
      "     |      '8\\t0\\t0\\t0\\t0\\t0\\t1\\t11\\t+\\tquery\\t8\\t0\\t8\\ttarget\\t40\\t11\\t30\\t2\\t4,4,\\t0,4,\\t11,26,\\n'\n",
      "     |      \n",
      "     |      Mapping the alignment does not depend on the sequence contents. If we\n",
      "     |      delete the sequence contents, the same alignment is found in PSL format\n",
      "     |      (though we obviously lose the ability to print the sequence alignment):\n",
      "     |      \n",
      "     |      >>> alignment1.target = Seq(None, len(alignment1.target))\n",
      "     |      >>> alignment1.query = Seq(None, len(alignment1.query))\n",
      "     |      >>> alignment2.target = Seq(None, len(alignment2.target))\n",
      "     |      >>> alignment2.query = Seq(None, len(alignment2.query))\n",
      "     |      >>> alignment = alignment1.map(alignment2)\n",
      "     |      >>> format(alignment, \"psl\")\n",
      "     |      '8\\t0\\t0\\t0\\t0\\t0\\t1\\t11\\t+\\tquery\\t8\\t0\\t8\\ttarget\\t40\\t11\\t30\\t2\\t4,4,\\t0,4,\\t11,26,\\n'\n",
      "     |  \n",
      "     |  sort(self, key=None, reverse=False)\n",
      "     |      Sort the sequences of the alignment in place.\n",
      "     |      \n",
      "     |      By default, this sorts the sequences alphabetically using their id\n",
      "     |      attribute if available, or by their sequence contents otherwise.\n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> aligner.gap_score = -1\n",
      "     |      >>> alignments = aligner.align(\"AATAA\", \"AAGAA\")\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AATAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AAGAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.sort()\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AAGAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AATAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Alternatively, a key function can be supplied that maps each sequence\n",
      "     |      to a sort value.  For example, you could sort on the GC content of each\n",
      "     |      sequence.\n",
      "     |      \n",
      "     |      >>> from Bio.SeqUtils import GC\n",
      "     |      >>> alignment.sort(key=GC)\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AATAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AAGAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      You can reverse the sort order by passing `reverse=True`:\n",
      "     |      \n",
      "     |      >>> alignment.sort(key=GC, reverse=True)\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 AAGAA 5\n",
      "     |                        0 ||.|| 5\n",
      "     |      query             0 AATAA 5\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      The sequences are now sorted by decreasing GC content value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from Alignment:\n",
      "     |  \n",
      "     |  infer_coordinates(lines, skipped_columns=None) from builtins.type\n",
      "     |      Infer the coordinates from a printed alignment.\n",
      "     |      \n",
      "     |      This method is primarily employed in Biopython's alignment parsers,\n",
      "     |      though it may be useful for other purposes.\n",
      "     |      \n",
      "     |      For an alignment consisting of N sequences, printed as N lines with\n",
      "     |      the same number of columns, where gaps are represented by dashes,\n",
      "     |      this method will calculate the sequence coordinates that define the\n",
      "     |      alignment. The coordinates are returned as a numpy array of integers,\n",
      "     |      and can be used to create an Alignment object.\n",
      "     |      \n",
      "     |      The argument skipped columns should be None (the default) or an empty\n",
      "     |      list. If skipped_columns is a list, then the indices of any columns in\n",
      "     |      the alignment with a gap in all lines are appended to skipped_columns.\n",
      "     |      \n",
      "     |      This is an example for the alignment of three sequences TAGGCATACGTG,\n",
      "     |      AACGTACGT, and ACGCATACTTG, with gaps in the second and third sequence:\n",
      "     |      \n",
      "     |      >>> from Bio.Align import Alignment\n",
      "     |      >>> lines = [\"TAGGCATACGTG\",\n",
      "     |      ...          \"AACG--TACGT-\",\n",
      "     |      ...          \"-ACGCATACTTG\",\n",
      "     |      ...         ]\n",
      "     |      >>> sequences = [line.replace(\"-\", \"\") for line in lines]\n",
      "     |      >>> sequences\n",
      "     |      ['TAGGCATACGTG', 'AACGTACGT', 'ACGCATACTTG']\n",
      "     |      >>> coordinates = Alignment.infer_coordinates(lines)\n",
      "     |      >>> coordinates\n",
      "     |      array([[ 0,  1,  4,  6, 11, 12],\n",
      "     |             [ 0,  1,  4,  4,  9,  9],\n",
      "     |             [ 0,  0,  3,  5, 10, 11]])\n",
      "     |      >>> alignment = Alignment(sequences, coordinates)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties inherited from Alignment:\n",
      "     |  \n",
      "     |  aligned\n",
      "     |      Return the indices of subsequences aligned to each other.\n",
      "     |      \n",
      "     |      This property returns the start and end indices of subsequences\n",
      "     |      in the target and query sequence that were aligned to each other.\n",
      "     |      If the alignment between target (t) and query (q) consists of N\n",
      "     |      chunks, you get two tuples of length N:\n",
      "     |      \n",
      "     |          (((t_start1, t_end1), (t_start2, t_end2), ..., (t_startN, t_endN)),\n",
      "     |           ((q_start1, q_end1), (q_start2, q_end2), ..., (q_startN, q_endN)))\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> alignments = aligner.align(\"GAACT\", \"GAT\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GAACT 5\n",
      "     |                        0 ||--| 5\n",
      "     |      query             0 GA--T 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      array([[[0, 2],\n",
      "     |              [4, 5]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 2],\n",
      "     |              [2, 3]]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GAACT 5\n",
      "     |                        0 |-|-| 5\n",
      "     |      query             0 G-A-T 3\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      array([[[0, 1],\n",
      "     |              [2, 3],\n",
      "     |              [4, 5]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 1],\n",
      "     |              [1, 2],\n",
      "     |              [2, 3]]])\n",
      "     |      \n",
      "     |      Note that different alignments may have the same subsequences\n",
      "     |      aligned to each other. In particular, this may occur if alignments\n",
      "     |      differ from each other in terms of their gap placement only:\n",
      "     |      \n",
      "     |      >>> aligner.mismatch_score = -10\n",
      "     |      >>> alignments = aligner.align(\"AAACAAA\", \"AAAGAAA\")\n",
      "     |      >>> len(alignments)\n",
      "     |      2\n",
      "     |      >>> print(alignments[0])\n",
      "     |      target            0 AAAC-AAA 7\n",
      "     |                        0 |||--||| 8\n",
      "     |      query             0 AAA-GAAA 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[0].aligned\n",
      "     |      array([[[0, 3],\n",
      "     |              [4, 7]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 3],\n",
      "     |              [4, 7]]])\n",
      "     |      >>> print(alignments[1])\n",
      "     |      target            0 AAA-CAAA 7\n",
      "     |                        0 |||--||| 8\n",
      "     |      query             0 AAAG-AAA 7\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[1].aligned\n",
      "     |      array([[[0, 3],\n",
      "     |              [4, 7]],\n",
      "     |      <BLANKLINE>\n",
      "     |             [[0, 3],\n",
      "     |              [4, 7]]])\n",
      "     |      \n",
      "     |      The property can be used to identify alignments that are identical\n",
      "     |      to each other in terms of their aligned sequences.\n",
      "     |  \n",
      "     |  indices\n",
      "     |      Return the sequence index of each lettter in the alignment.\n",
      "     |      \n",
      "     |      This property returns a 2D numpy array with the sequence index of each\n",
      "     |      letter in the alignment. Gaps are indicated by -1.  The array has the\n",
      "     |      same number of rows and columns as the alignment, as given by\n",
      "     |      `self.shape`.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"AATG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AA-TG 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5],\n",
      "     |             [ 0,  1, -1,  2,  3]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             0 AA-T-G 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5,  6],\n",
      "     |             [ 0,  1, -1,  2, -1,  3]])\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"CATT\", strand=\"-\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             4 AA-TG 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5],\n",
      "     |             [ 3,  2, -1,  1,  0]])\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             4 AA-T-G 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.indices\n",
      "     |      array([[ 1,  2,  3,  4,  5,  6],\n",
      "     |             [ 3,  2, -1,  1, -1,  0]])\n",
      "     |  \n",
      "     |  inverse_indices\n",
      "     |      Return the alignment column index for each letter in each sequence.\n",
      "     |      \n",
      "     |      This property returns a list of 1D numpy arrays; the number of arrays\n",
      "     |      is equal to the number of aligned sequences, and the length of each\n",
      "     |      array is equal to the length of the corresponding sequence. For each\n",
      "     |      letter in each sequence, the array contains the corresponding column\n",
      "     |      index in the alignment. Letters not included in the alignment are\n",
      "     |      indicated by -1.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      \n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"AATG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             0 AA-TG 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4, -1]), array([0, 1, 3, 4])]\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             0 AA-T-G 4\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4,  5]), array([0, 1, 3, 5])]\n",
      "     |      >>> alignments = aligner.align(\"GAACTGG\", \"CATT\", strand=\"-\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTG 6\n",
      "     |                        0 ||-|| 5\n",
      "     |      query             4 AA-TG 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4, -1]), array([4, 3, 1, 0])]\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            1 AACTGG 7\n",
      "     |                        0 ||-|-| 6\n",
      "     |      query             4 AA-T-G 0\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.inverse_indices\n",
      "     |      [array([-1,  0,  1,  2,  3,  4,  5]), array([5, 3, 1, 0])]\n",
      "     |  \n",
      "     |  shape\n",
      "     |      Return the shape of the alignment as a tuple of two integer values.\n",
      "     |      \n",
      "     |      The first integer value is the number of sequences in the alignment as\n",
      "     |      returned by len(alignment), which is always 2 for pairwise alignments.\n",
      "     |      \n",
      "     |      The second integer value is the number of columns in the alignment when\n",
      "     |      it is printed, and is equal to the sum of the number of matches, number\n",
      "     |      of mismatches, and the total length of gaps in the target and query.\n",
      "     |      Sequence sections beyond the aligned segment are not included in the\n",
      "     |      number of columns.\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> aligner.mode = \"global\"\n",
      "     |      >>> alignments = aligner.align(\"GACCTG\", \"CGATCG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 -GACCT-G 6\n",
      "     |                        0 -||--|-| 8\n",
      "     |      query             0 CGA--TCG 6\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> len(alignment)\n",
      "     |      2\n",
      "     |      >>> alignment.shape\n",
      "     |      (2, 8)\n",
      "     |      >>> aligner.mode = \"local\"\n",
      "     |      >>> alignments = aligner.align(\"GACCTG\", \"CGATCG\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 GACCT-G 6\n",
      "     |                        0 ||--|-| 7\n",
      "     |      query             1 GA--TCG 6\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> len(alignment)\n",
      "     |      2\n",
      "     |      >>> alignment.shape\n",
      "     |      (2, 7)\n",
      "     |  \n",
      "     |  substitutions\n",
      "     |      Return an Array with the number of substitutions of letters in the alignment.\n",
      "     |      \n",
      "     |      As an example, consider a sequence alignment of two RNA sequences:\n",
      "     |      \n",
      "     |      >>> from Bio.Align import PairwiseAligner\n",
      "     |      >>> target = \"ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTTATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG\"  # human spliceosomal small nuclear RNA U1\n",
      "     |      >>> query = \"ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTCTTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG\"  # sea lamprey spliceosomal small RNA U1\n",
      "     |      >>> aligner = PairwiseAligner()\n",
      "     |      >>> aligner.gap_score = -10\n",
      "     |      >>> alignments = aligner.align(target, query)\n",
      "     |      >>> len(alignments)\n",
      "     |      1\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      target            0 ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTT\n",
      "     |                        0 |||||||||||.||||||||..|||||||||||..|||||||..|||||||||||||||.\n",
      "     |      query             0 ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTC\n",
      "     |      <BLANKLINE>\n",
      "     |      target           60 ATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT\n",
      "     |                       60 .|||||||||||.|||..|.|.||||||||||||||||||||||||||||||||||||||\n",
      "     |      query            60 TTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT\n",
      "     |      <BLANKLINE>\n",
      "     |      target          120 GCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG 164\n",
      "     |                      120 |.||||||||||||||||||||||||||||||||||.|||||.| 164\n",
      "     |      query           120 GTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG 164\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> m = alignment.substitutions\n",
      "     |      >>> print(m)\n",
      "     |           A    C    G    T\n",
      "     |      A 28.0  1.0  2.0  1.0\n",
      "     |      C  0.0 39.0  1.0  2.0\n",
      "     |      G  2.0  0.0 45.0  0.0\n",
      "     |      T  2.0  5.0  1.0 35.0\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      Note that the matrix is not symmetric: rows correspond to the target\n",
      "     |      sequence, and columns to the query sequence.  For example, the number\n",
      "     |      of T's in the target sequence that are aligned to a C in the query\n",
      "     |      sequence is\n",
      "     |      \n",
      "     |      >>> m['T', 'C']\n",
      "     |      5.0\n",
      "     |      \n",
      "     |      and the number of C's in the query sequence tat are aligned to a T in\n",
      "     |      the query sequence is\n",
      "     |      \n",
      "     |      >>> m['C', 'T']\n",
      "     |      2.0\n",
      "     |      \n",
      "     |      For some applications (for example, to define a scoring matrix from\n",
      "     |      the substitution matrix), a symmetric matrix may be preferred, which\n",
      "     |      can be calculated as follows:\n",
      "     |      \n",
      "     |      >>> m += m.transpose()\n",
      "     |      >>> m /= 2.0\n",
      "     |      >>> print(m)\n",
      "     |           A    C    G    T\n",
      "     |      A 28.0  0.5  2.0  1.5\n",
      "     |      C  0.5 39.0  0.5  3.5\n",
      "     |      G  2.0  0.5 45.0  0.5\n",
      "     |      T  1.5  3.5  0.5 35.0\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      The matrix is now symmetric, with counts divided equally on both sides\n",
      "     |      of the diagonal:\n",
      "     |      \n",
      "     |      >>> m['C', 'T']\n",
      "     |      3.5\n",
      "     |      >>> m['T', 'C']\n",
      "     |      3.5\n",
      "     |      \n",
      "     |      The total number of substitutions between T's and C's in the alignment\n",
      "     |      is 3.5 + 3.5 = 7.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Alignment:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  path\n",
      "     |      Return the path through the trace matrix.\n",
      "     |  \n",
      "     |  query\n",
      "     |      Return self.sequences[1] for a pairwise alignment.\n",
      "     |  \n",
      "     |  target\n",
      "     |      Return self.sequences[0] for a pairwise alignment.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from Alignment:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "    \n",
      "    class PairwiseAlignments(builtins.object)\n",
      "     |  PairwiseAlignments(seqA, seqB, score, paths)\n",
      "     |  \n",
      "     |  Implements an iterator over pairwise alignments returned by the aligner.\n",
      "     |  \n",
      "     |  This class also supports indexing, which is fast for increasing indices,\n",
      "     |  but may be slow for random access of a large number of alignments.\n",
      "     |  \n",
      "     |  Note that pairwise aligners can return an astronomical number of alignments,\n",
      "     |  even for relatively short sequences, if they align poorly to each other. We\n",
      "     |  therefore recommend to first check the number of alignments, accessible as\n",
      "     |  len(alignments), which can be calculated quickly even if the number of\n",
      "     |  alignments is very large.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |  \n",
      "     |  __init__(self, seqA, seqB, score, paths)\n",
      "     |      Initialize a new PairwiseAlignments object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - seqA  - The first sequence, as a plain string, without gaps.\n",
      "     |       - seqB  - The second sequence, as a plain string, without gaps.\n",
      "     |       - score - The alignment score.\n",
      "     |       - paths - An iterator over the paths in the traceback matrix;\n",
      "     |                 each path defines one alignment.\n",
      "     |      \n",
      "     |      You would normally obtain a PairwiseAlignments object by calling\n",
      "     |      aligner.align(seqA, seqB), where aligner is a PairwiseAligner object.\n",
      "     |  \n",
      "     |  __iter__(self)\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return the number of alignments.\n",
      "     |  \n",
      "     |  __next__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "\n",
      "FUNCTIONS\n",
      "    parse(source, fmt)\n",
      "        Parse an alignment file and return an iterator over alignments.\n",
      "        \n",
      "        Arguments:\n",
      "         - source - File or file-like object to read from, or filename as string.\n",
      "         - fmt    - String describing the file format (case-insensitive).\n",
      "        \n",
      "        Typical usage, opening a file to read in, and looping over the aligments:\n",
      "        \n",
      "        >>> from Bio import Align\n",
      "        >>> filename = \"Exonerate/exn_22_m_ner_cigar.exn\"\n",
      "        >>> for alignment in Align.parse(filename, \"exonerate\"):\n",
      "        ...    print(\"Number of sequences in alignment\", len(alignment))\n",
      "        ...    print(\"Alignment score:\", alignment.score)\n",
      "        Number of sequences in alignment 2\n",
      "        Alignment score: 6150\n",
      "        Number of sequences in alignment 2\n",
      "        Alignment score: 502\n",
      "        Number of sequences in alignment 2\n",
      "        Alignment score: 440\n",
      "        \n",
      "        For lazy-loading file formats such as bigMaf, for which the file contents\n",
      "        is read on demand only, ensure that the file remains open while extracting\n",
      "        alignment data.\n",
      "        \n",
      "        You can use the Bio.Align.read(...) function when the file contains only\n",
      "        one alignment.\n",
      "    \n",
      "    read(handle, fmt)\n",
      "        Parse a file containing one alignment, and return it.\n",
      "        \n",
      "        Arguments:\n",
      "         - source - File or file-like object to read from, or filename as string.\n",
      "         - fmt    - String describing the file format (case-insensitive).\n",
      "        \n",
      "        This function is for use parsing alignment files containing exactly one\n",
      "        alignment.  For example, reading a Clustal file:\n",
      "        \n",
      "        >>> from Bio import Align\n",
      "        >>> alignment = Align.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      "        >>> print(\"Alignment shape:\", alignment.shape)\n",
      "        Alignment shape: (7, 156)\n",
      "        >>> for sequence in alignment.sequences:\n",
      "        ...     print(sequence.id, len(sequence))\n",
      "        gi|6273285|gb|AF191659.1|AF191 146\n",
      "        gi|6273284|gb|AF191658.1|AF191 148\n",
      "        gi|6273287|gb|AF191661.1|AF191 146\n",
      "        gi|6273286|gb|AF191660.1|AF191 146\n",
      "        gi|6273290|gb|AF191664.1|AF191 150\n",
      "        gi|6273289|gb|AF191663.1|AF191 150\n",
      "        gi|6273291|gb|AF191665.1|AF191 156\n",
      "        \n",
      "        If the file contains no records, or more than one record, an exception is\n",
      "        raised.  For example:\n",
      "        \n",
      "        >>> from Bio import Align\n",
      "        >>> filename = \"Exonerate/exn_22_m_ner_cigar.exn\"\n",
      "        >>> alignment = Align.read(filename, \"exonerate\")\n",
      "        Traceback (most recent call last):\n",
      "            ...\n",
      "        ValueError: More than one alignment found in file\n",
      "        \n",
      "        Use the Bio.Align.parse function if you want to read a file containing\n",
      "        more than one alignment.\n",
      "    \n",
      "    write(alignments, target, fmt, *args, **kwargs)\n",
      "        Write alignments to a file.\n",
      "        \n",
      "        Arguments:\n",
      "         - alignments - List (or iterator) of Alignment objects, or a single\n",
      "           Alignment.\n",
      "         - target     - File or file-like object to write to, or filename as string.\n",
      "         - fmt        - String describing the file format (case-insensitive).\n",
      "        \n",
      "        Note if providing a file or file-like object, your code should close the\n",
      "        target after calling this function, or call .flush(), to ensure the data\n",
      "        gets flushed to disk.\n",
      "        \n",
      "        Returns the number of alignments written (as an integer).\n",
      "\n",
      "DATA\n",
      "    formats = ('a2m', 'bed', 'bigbed', 'bigmaf', 'bigpsl', 'clustal', 'emb...\n",
      "\n",
      "FILE\n",
      "    /opt/miniconda3/lib/python3.9/site-packages/Bio/Align/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Align)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Allineamento di due sequenze"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importare la classe `PairwiseAligner`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import PairwiseAligner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class PairwiseAligner in module Bio.Align:\n",
      "\n",
      "class PairwiseAligner(_algorithms.PairwiseAligner)\n",
      " |  PairwiseAligner(scoring=None, **kwargs)\n",
      " |  \n",
      " |  Performs pairwise sequence alignment using dynamic programming.\n",
      " |  \n",
      " |  This provides functions to get global and local alignments between two\n",
      " |  sequences.  A global alignment finds the best concordance between all\n",
      " |  characters in two sequences.  A local alignment finds just the\n",
      " |  subsequences that align the best.\n",
      " |  \n",
      " |  To perform a pairwise sequence alignment, first create a PairwiseAligner\n",
      " |  object.  This object stores the match and mismatch scores, as well as the\n",
      " |  gap scores.  Typically, match scores are positive, while mismatch scores\n",
      " |  and gap scores are negative or zero.  By default, the match score is 1,\n",
      " |  and the mismatch and gap scores are zero.  Based on the values of the gap\n",
      " |  scores, a PairwiseAligner object automatically chooses the appropriate\n",
      " |  alignment algorithm (the Needleman-Wunsch, Smith-Waterman, Gotoh, or\n",
      " |  Waterman-Smith-Beyer global or local alignment algorithm).\n",
      " |  \n",
      " |  Calling the \"score\" method on the aligner with two sequences as arguments\n",
      " |  will calculate the alignment score between the two sequences.\n",
      " |  Calling the \"align\" method on the aligner with two sequences as arguments\n",
      " |  will return a generator yielding the alignments between the two\n",
      " |  sequences.\n",
      " |  \n",
      " |  Some examples:\n",
      " |  \n",
      " |  >>> from Bio import Align\n",
      " |  >>> aligner = Align.PairwiseAligner()\n",
      " |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      " |  >>> for alignment in sorted(alignments):\n",
      " |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      " |  ...     print(alignment)\n",
      " |  ...\n",
      " |  Score = 3.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -|-|| 5\n",
      " |  query             0 -A-CG 3\n",
      " |  <BLANKLINE>\n",
      " |  Score = 3.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -||-| 5\n",
      " |  query             0 -AC-G 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  Specify the aligner mode as local to generate local alignments:\n",
      " |  \n",
      " |  >>> aligner.mode = 'local'\n",
      " |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      " |  >>> for alignment in sorted(alignments):\n",
      " |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      " |  ...     print(alignment)\n",
      " |  ...\n",
      " |  Score = 3.0:\n",
      " |  target            1 ACCG 5\n",
      " |                    0 |-|| 4\n",
      " |  query             0 A-CG 3\n",
      " |  <BLANKLINE>\n",
      " |  Score = 3.0:\n",
      " |  target            1 ACCG 5\n",
      " |                    0 ||-| 4\n",
      " |  query             0 AC-G 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  Do a global alignment.  Identical characters are given 2 points,\n",
      " |  1 point is deducted for each non-identical character.\n",
      " |  \n",
      " |  >>> aligner.mode = 'global'\n",
      " |  >>> aligner.match_score = 2\n",
      " |  >>> aligner.mismatch_score = -1\n",
      " |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      " |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      " |  ...     print(alignment)\n",
      " |  ...\n",
      " |  Score = 6.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -||-| 5\n",
      " |  query             0 -AC-G 3\n",
      " |  <BLANKLINE>\n",
      " |  Score = 6.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -|-|| 5\n",
      " |  query             0 -A-CG 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  Same as above, except now 0.5 points are deducted when opening a\n",
      " |  gap, and 0.1 points are deducted when extending it.\n",
      " |  \n",
      " |  >>> aligner.open_gap_score = -0.5\n",
      " |  >>> aligner.extend_gap_score = -0.1\n",
      " |  >>> aligner.target_end_gap_score = 0.0\n",
      " |  >>> aligner.query_end_gap_score = 0.0\n",
      " |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      " |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      " |  ...     print(alignment)\n",
      " |  ...\n",
      " |  Score = 5.5:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -|-|| 5\n",
      " |  query             0 -A-CG 3\n",
      " |  <BLANKLINE>\n",
      " |  Score = 5.5:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -||-| 5\n",
      " |  query             0 -AC-G 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  The alignment function can also use known matrices already included in\n",
      " |  Biopython:\n",
      " |  \n",
      " |  >>> from Bio.Align import substitution_matrices\n",
      " |  >>> aligner = Align.PairwiseAligner()\n",
      " |  >>> aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
      " |  >>> alignments = aligner.align(\"KEVLA\", \"EVL\")\n",
      " |  >>> alignments = list(alignments)\n",
      " |  >>> print(\"Number of alignments: %d\" % len(alignments))\n",
      " |  Number of alignments: 1\n",
      " |  >>> alignment = alignments[0]\n",
      " |  >>> print(\"Score = %.1f\" % alignment.score)\n",
      " |  Score = 13.0\n",
      " |  >>> print(alignment)\n",
      " |  target            0 KEVLA 5\n",
      " |                    0 -|||- 5\n",
      " |  query             0 -EVL- 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also set the value of attributes directly during construction\n",
      " |  of the PairwiseAligner object by providing them as keyword arguments:\n",
      " |  \n",
      " |  >>> aligner = Align.PairwiseAligner(mode='global', match_score=2, mismatch_score=-1)\n",
      " |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      " |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      " |  ...     print(alignment)\n",
      " |  ...\n",
      " |  Score = 6.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -||-| 5\n",
      " |  query             0 -AC-G 3\n",
      " |  <BLANKLINE>\n",
      " |  Score = 6.0:\n",
      " |  target            0 TACCG 5\n",
      " |                    0 -|-|| 5\n",
      " |  query             0 -A-CG 3\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      PairwiseAligner\n",
      " |      _algorithms.PairwiseAligner\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __getstate__(self)\n",
      " |  \n",
      " |  __init__(self, scoring=None, **kwargs)\n",
      " |      Initialize a new PairwiseAligner as specified by the keyword arguments.\n",
      " |      \n",
      " |      If scoring is None, use the default scoring scheme match = 1.0,\n",
      " |      mismatch = 0.0, gap score = 0.0\n",
      " |      If scoring is \"blastn\", \"megablast\", or \"blastp\", use the default\n",
      " |      substitution matrix and gap scores for BLASTN, MEGABLAST, or BLASTP,\n",
      " |      respectively.\n",
      " |      \n",
      " |      Loops over the remaining keyword arguments and sets them as attributes\n",
      " |      on the object.\n",
      " |  \n",
      " |  __setattr__(self, key, value)\n",
      " |      Implement setattr(self, name, value).\n",
      " |  \n",
      " |  __setstate__(self, state)\n",
      " |  \n",
      " |  align(self, seqA, seqB, strand='+')\n",
      " |      Return the alignments of two sequences using PairwiseAligner.\n",
      " |  \n",
      " |  score(self, seqA, seqB, strand='+')\n",
      " |      Return the alignments score of two sequences using PairwiseAligner.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _algorithms.PairwiseAligner:\n",
      " |  \n",
      " |  __repr__(self, /)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __str__(self, /)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from _algorithms.PairwiseAligner:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from _algorithms.PairwiseAligner:\n",
      " |  \n",
      " |  algorithm\n",
      " |      alignment algorithm\n",
      " |  \n",
      " |  alphabet\n",
      " |      alphabet\n",
      " |  \n",
      " |  end_extend_gap_score\n",
      " |      end extend gap score\n",
      " |  \n",
      " |  end_gap_score\n",
      " |      end gap score\n",
      " |  \n",
      " |  end_open_gap_score\n",
      " |      end open gap score\n",
      " |  \n",
      " |  epsilon\n",
      " |      roundoff epsilon\n",
      " |  \n",
      " |  extend_gap_score\n",
      " |      extend gap score\n",
      " |  \n",
      " |  gap_score\n",
      " |      gap score\n",
      " |  \n",
      " |  internal_extend_gap_score\n",
      " |      internal extend gap score\n",
      " |  \n",
      " |  internal_gap_score\n",
      " |      internal gap score\n",
      " |  \n",
      " |  internal_open_gap_score\n",
      " |      internal open gap score\n",
      " |  \n",
      " |  left_extend_gap_score\n",
      " |      left extend gap score\n",
      " |  \n",
      " |  left_gap_score\n",
      " |      left gap score\n",
      " |  \n",
      " |  left_open_gap_score\n",
      " |      left open gap score\n",
      " |  \n",
      " |  match\n",
      " |      match score\n",
      " |  \n",
      " |  match_score\n",
      " |      match score\n",
      " |  \n",
      " |  mismatch\n",
      " |      mismatch score\n",
      " |  \n",
      " |  mismatch_score\n",
      " |      mismatch score\n",
      " |  \n",
      " |  mode\n",
      " |      alignment mode ('global' or 'local')\n",
      " |  \n",
      " |  open_gap_score\n",
      " |      internal and end open gap score\n",
      " |  \n",
      " |  query_end_extend_gap_score\n",
      " |      query end extend score\n",
      " |  \n",
      " |  query_end_gap_score\n",
      " |      query end score\n",
      " |  \n",
      " |  query_end_open_gap_score\n",
      " |      query end open score\n",
      " |  \n",
      " |  query_extend_gap_score\n",
      " |      query gap extend score\n",
      " |  \n",
      " |  query_gap_score\n",
      " |      query gap score\n",
      " |  \n",
      " |  query_internal_extend_gap_score\n",
      " |      query internal extend gap score\n",
      " |  \n",
      " |  query_internal_gap_score\n",
      " |      query internal gap score\n",
      " |  \n",
      " |  query_internal_open_gap_score\n",
      " |      query internal open gap score\n",
      " |  \n",
      " |  query_left_extend_gap_score\n",
      " |      query left extend score\n",
      " |  \n",
      " |  query_left_gap_score\n",
      " |      query left score\n",
      " |  \n",
      " |  query_left_open_gap_score\n",
      " |      query left open score\n",
      " |  \n",
      " |  query_open_gap_score\n",
      " |      query gap open score\n",
      " |  \n",
      " |  query_right_extend_gap_score\n",
      " |      query right extend score\n",
      " |  \n",
      " |  query_right_gap_score\n",
      " |      query right score\n",
      " |  \n",
      " |  query_right_open_gap_score\n",
      " |      query right open score\n",
      " |  \n",
      " |  right_extend_gap_score\n",
      " |      right extend gap score\n",
      " |  \n",
      " |  right_gap_score\n",
      " |      right gap score\n",
      " |  \n",
      " |  right_open_gap_score\n",
      " |      right open gap score\n",
      " |  \n",
      " |  substitution_matrix\n",
      " |      substitution_matrix\n",
      " |  \n",
      " |  target_end_extend_gap_score\n",
      " |      target end extend gap score\n",
      " |  \n",
      " |  target_end_gap_score\n",
      " |      target end gap score\n",
      " |  \n",
      " |  target_end_open_gap_score\n",
      " |      target end open gap score\n",
      " |  \n",
      " |  target_extend_gap_score\n",
      " |      target extend gap score\n",
      " |  \n",
      " |  target_gap_score\n",
      " |      target gap score\n",
      " |  \n",
      " |  target_internal_extend_gap_score\n",
      " |      target internal extend gap score\n",
      " |  \n",
      " |  target_internal_gap_score\n",
      " |      target internal gap score\n",
      " |  \n",
      " |  target_internal_open_gap_score\n",
      " |      target internal open gap score\n",
      " |  \n",
      " |  target_left_extend_gap_score\n",
      " |      target left extend score\n",
      " |  \n",
      " |  target_left_gap_score\n",
      " |      target left score\n",
      " |  \n",
      " |  target_left_open_gap_score\n",
      " |      target left open score\n",
      " |  \n",
      " |  target_open_gap_score\n",
      " |      target open gap score\n",
      " |  \n",
      " |  target_right_extend_gap_score\n",
      " |      target right extend score\n",
      " |  \n",
      " |  target_right_gap_score\n",
      " |      target right score\n",
      " |  \n",
      " |  target_right_open_gap_score\n",
      " |      target right open score\n",
      " |  \n",
      " |  wildcard\n",
      " |      wildcard character\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(PairwiseAligner)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Costruire un oggetto `PairwiseAligner`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner = PairwiseAligner()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ottenere l'algoritmo correntemente impostato all'attributo `algorithm`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Needleman-Wunsch'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assegnare lo schema di punteggio."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner.match_score = 1.0\n",
    "aligner.mismatch_score = -1.0\n",
    "aligner.gap_score = -1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Si considerino le due sequenze:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence1 = 'GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG'\n",
    "sequence2 = 'GAAATAACTGCGAGGGAAATAACTGCGAGGGAAATAACTGCGAGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ottenere lo *score* di allineamento ottimo tramite il metodo `score()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "84.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.score(sequence1, sequence2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determininare tutti gli allineamenti ottimi delle due sequenze."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments = aligner.align(sequence1, sequence2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualizzare uno dopo l'altro gli allineamenti."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***Alignment 1\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||||-|-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-A-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 2\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||||-|-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 3\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||||-|-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 4\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||||-|-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 5\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||||-||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 6\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||||-||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 7\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||||-||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 8\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||||-||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 9\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||-|||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 10\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||-|||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 11\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||-|||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 12\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||-|||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 13\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||-||||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 14\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||-||||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 15\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||-||||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 16\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||-||||-|||||.||||||||-||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-ATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 17\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||||-|-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-A-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 18\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||||-|-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 19\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||||-|-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 20\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||||-|-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 21\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||||-||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 22\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||||-||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 23\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||||-||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 24\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||||-||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 25\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||-|||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 26\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||-|||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 27\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||-|||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 28\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||-|||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 29\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||-||||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 30\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||-||||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 31\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||-||||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 32\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||-||||-|||||.|||||||-|||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 33\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||||-|-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-A-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 34\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||||-|-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 35\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||||-|-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 36\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||||-|-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 37\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||||-||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 38\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||||-||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 39\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||||-||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 40\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||||-||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 41\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||-|||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 42\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||-|||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 43\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||-|||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 44\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||-|||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 45\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||-||||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 46\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||-||||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 47\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||-||||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 48\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||-||||-|||||.||||||-||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 49\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||||-|-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGGG-A-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 50\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||||-|-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 51\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||||-|-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 52\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||||-|-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 53\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||||-||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAGG-GA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 54\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||||-||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 55\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||||-||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 56\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||||-||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 57\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.||||||-|||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGAG-GGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 58\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.||||||-|||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 59\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.||||||-|||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 60\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.||||||-|||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 61\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||||-|-|||||.|||||-||||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGGG-A-AATAACTGCGA-GGGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 62\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||||-||-|||||.|||||-||||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAGG-GA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 63\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.||||||-|||-|||||.|||||-||||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGAG-GGA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n",
      "***Alignment 64\n",
      "target            0 GATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGGGATAATAAG\n",
      "                  0 ||-|||||.|||||-||||-|||||.|||||-||||-|||||.|||||-|||||||||||\n",
      "query             0 GA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGA-AATAACTGCGA-GGGATAATAAG\n",
      "\n",
      "target           60 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG 102\n",
      "                 60 |||||||||||||||||||||||||||||||||||||||||| 102\n",
      "query            54 TGCGAGGGGATAATAAGTGCGAGGGGATAATAAGTGCGAGGG  96\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (i, alignment) in enumerate(alignments):\n",
    "    print('***Alignment', i+1)\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Si considerino ora queste le due sequenze:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence3 = 'AGAAGGGGGGAAGAA'\n",
    "sequence4 = 'TTTTGGGGGGTTTTT'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ottenere lo *score* di allineamento ottimo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-3.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.score(sequence3, sequence4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determininare tutti gli allineamenti ottimi."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments = aligner.align(sequence3, sequence4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stampare uno dopo l'altro gli allineamenti."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***Alignment 1\n",
      "target            0 AGAAGGGGGGAAGAA 15\n",
      "                  0 ....||||||..... 15\n",
      "query             0 TTTTGGGGGGTTTTT 15\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (i, alignment) in enumerate(alignments):\n",
    "    print('***Alignment', i+1)\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Allinearle con l'algoritmo di Smith-Waterman. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner.mode = 'local'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Smith-Waterman'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner.score(sequence3, sequence4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments = aligner.align(sequence3, sequence4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***Alignment 1\n",
      "target            4 GGGGGG 10\n",
      "                  0 ||||||  6\n",
      "query             4 GGGGGG 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (i, alignment) in enumerate(alignments):\n",
    "    print('***Alignment', i+1)\n",
    "    print(alignment)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
