{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Esercizio 4\n",
    "\n",
    "EMBL (https://www.ebi.ac.uk) è una banca di sequenze nucleotidiche sviluppata da EMBL-EBI, che è l'European Bioinformatics Institute dell'European Molecular Biology Laboratory (EMBL),. Ogni sequenza nucleotidica (*entry*) viene memorizzata, con altre informazioni, in un file di puro testo in un formato che prende il nome di `EMBL`.\n",
    "\n",
    "**Il formato `EMBL`** è composto da *record* a struttura variabile che iniziano con un codice a due lettere maiuscole che ne indica il contenuto. I soli *record* che non iniziano con un codice a due lettere sono quelli contenenti la sequenza nucleotidica.\n",
    "Il record `//` segnala la fine dell'*entry*.\n",
    "\n",
    "L'esercizio richiede di leggere il file `M10051.embl` che contiene la sequenza nucleotidica di un mRNA (cioé di un trascritto espresso da un gene). `M10051` è l'identificatore univoco (Accession Number) della sequenza.\n",
    "\n",
    "Si richiede di produrre, in un file in formato `FASTA`, la sequenza nucleotidica della coding sequence (CDS) separata in *record* di 60 basi.\n",
    "\n",
    "Inoltre si richiede di determinare:\n",
    "\n",
    "- le frequenze dei codoni presenti nella CDS, elencate in ordine decrescente\n",
    "- le frequenze degli amminoacidi presenti nella proteina, elencate in ordine decrescente\n",
    "\n",
    "Inoltre, preso in input un file `genetic-code.txt` contenente il codice genetico:\n",
    "- effettuare la validazione della proteina specificata nell'entry `EMBL` rispetto alla sequenza ottenuta per traduzione della sequenza CDS (ottenuta prima)\n",
    "- trovare una CDS \"sinonima\" (cioé che produca la stessa proteina) sostituendo il maggior numero possibile di codoni.\n",
    "\n",
    "***\n",
    "\n",
    "Requisiti:\n",
    "\n",
    "- nell’*header* `FASTA` della sequenza della CDS devono comparire l’identificatore univoco della sequenza, l’organismo, la lunghezza della CDS e se la CDS inizia con lo start codon e finisce con lo stop codon, rispettando il formato seguente: \n",
    "\n",
    "        >M10051-HUM; len = [length]; start = [yes|no]; end = [yes|no]\n",
    "        \n",
    "***\n",
    "\n",
    "## Indicazioni sul formato `EMBL`\n",
    "\n",
    "Fare riferimento all'esercizio 3.\n",
    "\n",
    "## Indicazioni sul file del codice genetico\n",
    "\n",
    "Il file del codice genetico è strutturato in record di campi separati da `,`, di cui il primo è il simbolo di un amminoacido e i restanti campi sono i codoni che codificano tale amminoacido. L'ultimo record fornisce gli stop codon a cui corrisponde il simbolo `s`.\n",
    "\n",
    "    A,gct,gcc,gca,gcg\n",
    "    C,tgt,tgc\n",
    "    D,gat,gac\n",
    "    E,gaa,gag\n",
    "    F,ttt,ttc\n",
    "    G,ggt,ggc,gga,ggg\n",
    "    H,cat,cac\n",
    "    I,att,atc,ata\n",
    "    K,aaa,aag\n",
    "    L,tta,ttg,ctt,ctc,cta,ctg\n",
    "    M,atg\n",
    "    N,aat,aac\n",
    "    P,cct,ccc,cca,ccg\n",
    "    Q,caa,cag\n",
    "    R,cgt,cgc,cga,cgg,aga,agg\n",
    "    S,tct,tcc,tca,tcg,agt,agc\n",
    "    T,act,acc,aca,acg\n",
    "    V,gtt,gtc,gta,gtg\n",
    "    W,tgg\n",
    "    Y,tat,tac\n",
    "    s,tga,taa,tag\n",
    "    \n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parametri in input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "embl_file_name = './M10051.embl'\n",
    "genetic_code_name = './genetic-code.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importare il modulo `re`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Definire la funzione che formatta una sequenza in `FASTA` \n",
    "\n",
    "Fare riferimento all'esercizio 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_fasta(header, sequence, record = 80):    \n",
    "    p = re.compile('\\w{,' + str(record) + '}')\n",
    "    return header + '\\n' + '\\n'.join(re.findall(p, sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Leggere il file del codice genetico"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(genetic_code_name, 'r') as input_file:\n",
    "    genetic_code_rows = input_file.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A,gct,gcc,gca,gcg\\n',\n",
       " 'C,tgt,tgc\\n',\n",
       " 'D,gat,gac\\n',\n",
       " 'E,gaa,gag\\n',\n",
       " 'F,ttt,ttc\\n',\n",
       " 'G,ggt,ggc,gga,ggg\\n',\n",
       " 'H,cat,cac\\n',\n",
       " 'I,att,atc,ata\\n',\n",
       " 'K,aaa,aag\\n',\n",
       " 'L,tta,ttg,ctt,ctc,cta,ctg\\n',\n",
       " 'M,atg\\n',\n",
       " 'N,aat,aac\\n',\n",
       " 'P,cct,ccc,cca,ccg\\n',\n",
       " 'Q,caa,cag\\n',\n",
       " 'R,cgt,cgc,cga,cgg,aga,agg\\n',\n",
       " 'S,tct,tcc,tca,tcg,agt,agc\\n',\n",
       " 'T,act,acc,aca,acg\\n',\n",
       " 'V,gtt,gtc,gta,gtg\\n',\n",
       " 'W,tgg\\n',\n",
       " 'Y,tat,tac\\n',\n",
       " 's,tga,taa,tag']"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_code_rows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Costruire il dizionario del codice genetico"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Costruire il dizionario che contiene il codice genetico:\n",
    "\n",
    "- **chiave**: codone\n",
    "- **valore**: simbolo dell'amminoacido codificato\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Costruire la lista delle tuple *(chiave, valore)*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_genetic_code = [row.rstrip().split(',') for row in genetic_code_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_value_list = []\n",
    "\n",
    "for record_list in split_genetic_code:\n",
    "    key_value_list.extend(list(itertools.product(record_list[1:], record_list[:1])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('gct', 'A'),\n",
       " ('gcc', 'A'),\n",
       " ('gca', 'A'),\n",
       " ('gcg', 'A'),\n",
       " ('tgt', 'C'),\n",
       " ('tgc', 'C'),\n",
       " ('gat', 'D'),\n",
       " ('gac', 'D'),\n",
       " ('gaa', 'E'),\n",
       " ('gag', 'E'),\n",
       " ('ttt', 'F'),\n",
       " ('ttc', 'F'),\n",
       " ('ggt', 'G'),\n",
       " ('ggc', 'G'),\n",
       " ('gga', 'G'),\n",
       " ('ggg', 'G'),\n",
       " ('cat', 'H'),\n",
       " ('cac', 'H'),\n",
       " ('att', 'I'),\n",
       " ('atc', 'I'),\n",
       " ('ata', 'I'),\n",
       " ('aaa', 'K'),\n",
       " ('aag', 'K'),\n",
       " ('tta', 'L'),\n",
       " ('ttg', 'L'),\n",
       " ('ctt', 'L'),\n",
       " ('ctc', 'L'),\n",
       " ('cta', 'L'),\n",
       " ('ctg', 'L'),\n",
       " ('atg', 'M'),\n",
       " ('aat', 'N'),\n",
       " ('aac', 'N'),\n",
       " ('cct', 'P'),\n",
       " ('ccc', 'P'),\n",
       " ('cca', 'P'),\n",
       " ('ccg', 'P'),\n",
       " ('caa', 'Q'),\n",
       " ('cag', 'Q'),\n",
       " ('cgt', 'R'),\n",
       " ('cgc', 'R'),\n",
       " ('cga', 'R'),\n",
       " ('cgg', 'R'),\n",
       " ('aga', 'R'),\n",
       " ('agg', 'R'),\n",
       " ('tct', 'S'),\n",
       " ('tcc', 'S'),\n",
       " ('tca', 'S'),\n",
       " ('tcg', 'S'),\n",
       " ('agt', 'S'),\n",
       " ('agc', 'S'),\n",
       " ('act', 'T'),\n",
       " ('acc', 'T'),\n",
       " ('aca', 'T'),\n",
       " ('acg', 'T'),\n",
       " ('gtt', 'V'),\n",
       " ('gtc', 'V'),\n",
       " ('gta', 'V'),\n",
       " ('gtg', 'V'),\n",
       " ('tgg', 'W'),\n",
       " ('tat', 'Y'),\n",
       " ('tac', 'Y'),\n",
       " ('tga', 's'),\n",
       " ('taa', 's'),\n",
       " ('tag', 's')]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_value_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Costruire il dizionario."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_code_dict = dict(key_value_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gct': 'A',\n",
       " 'gcc': 'A',\n",
       " 'gca': 'A',\n",
       " 'gcg': 'A',\n",
       " 'tgt': 'C',\n",
       " 'tgc': 'C',\n",
       " 'gat': 'D',\n",
       " 'gac': 'D',\n",
       " 'gaa': 'E',\n",
       " 'gag': 'E',\n",
       " 'ttt': 'F',\n",
       " 'ttc': 'F',\n",
       " 'ggt': 'G',\n",
       " 'ggc': 'G',\n",
       " 'gga': 'G',\n",
       " 'ggg': 'G',\n",
       " 'cat': 'H',\n",
       " 'cac': 'H',\n",
       " 'att': 'I',\n",
       " 'atc': 'I',\n",
       " 'ata': 'I',\n",
       " 'aaa': 'K',\n",
       " 'aag': 'K',\n",
       " 'tta': 'L',\n",
       " 'ttg': 'L',\n",
       " 'ctt': 'L',\n",
       " 'ctc': 'L',\n",
       " 'cta': 'L',\n",
       " 'ctg': 'L',\n",
       " 'atg': 'M',\n",
       " 'aat': 'N',\n",
       " 'aac': 'N',\n",
       " 'cct': 'P',\n",
       " 'ccc': 'P',\n",
       " 'cca': 'P',\n",
       " 'ccg': 'P',\n",
       " 'caa': 'Q',\n",
       " 'cag': 'Q',\n",
       " 'cgt': 'R',\n",
       " 'cgc': 'R',\n",
       " 'cga': 'R',\n",
       " 'cgg': 'R',\n",
       " 'aga': 'R',\n",
       " 'agg': 'R',\n",
       " 'tct': 'S',\n",
       " 'tcc': 'S',\n",
       " 'tca': 'S',\n",
       " 'tcg': 'S',\n",
       " 'agt': 'S',\n",
       " 'agc': 'S',\n",
       " 'act': 'T',\n",
       " 'acc': 'T',\n",
       " 'aca': 'T',\n",
       " 'acg': 'T',\n",
       " 'gtt': 'V',\n",
       " 'gtc': 'V',\n",
       " 'gta': 'V',\n",
       " 'gtg': 'V',\n",
       " 'tgg': 'W',\n",
       " 'tat': 'Y',\n",
       " 'tac': 'Y',\n",
       " 'tga': 's',\n",
       " 'taa': 's',\n",
       " 'tag': 's'}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_code_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Leggere il file `EMBL`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fare riferimento all'esercizio 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(embl_file_name, 'r') as embl_file:\n",
    "    embl_str = embl_file.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID   M10051; SV 1; linear; mRNA; STD; HUM; 4723 BP.\n",
      "XX\n",
      "AC   M10051;\n",
      "XX\n",
      "DT   02-JUL-1986 (Rel. 09, Created)\n",
      "DT   14-NOV-2006 (Rel. 89, Last updated, Version 7)\n",
      "XX\n",
      "DE   Human insulin receptor mRNA, complete cds.\n",
      "XX\n",
      "KW   insulin receptor; tyrosine kinase.\n",
      "XX\n",
      "OS   Homo sapiens (human)\n",
      "OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;\n",
      "OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;\n",
      "OC   Homo.\n",
      "XX\n",
      "RN   [1]\n",
      "RP   1-4723\n",
      "RX   DOI; 10.1016/0092-8674(85)90334-4.\n",
      "RX   PUBMED; 2859121.\n",
      "RA   Ebina Y., Ellis L., Jarnagin K., Edery M., Graf L., Clauser E., Ou J.-H.,\n",
      "RA   Masiarz F., Kan Y.W., Goldfine I.D., Roth R.A., Rutter W.J.;\n",
      "RT   \"The human insulin receptor cDNA: the structural basis for\n",
      "RT   hormone-activated transmembrane signalling\";\n",
      "RL   Cell 40(4):747-758(1985).\n",
      "XX\n",
      "DR   MD5; e4e6ebf2e723a500c1dd62385c279351.\n",
      "DR   Ensembl-Gn; ENSG00000171105; homo_sapiens.\n",
      "DR   Ensembl-Tr; ENST00000302850; homo_sapiens.\n",
      "DR   Ensembl-Tr; ENST00000341500; homo_sapiens.\n",
      "DR   EuropePMC; PMC2739203; 19682364.\n",
      "DR   EuropePMC; PMC3164640; 21909271.\n",
      "DR   EuropePMC; PMC452597; 15146055.\n",
      "XX\n",
      "CC   [1] suggests that the insulin receptor may be the cellular homolog\n",
      "CC   of the v-ros transforming (oncogene) protein.  [1] notes\n",
      "CC   similarities between the insulin receptor and several growth factor\n",
      "CC   receptors and oncogenes.  Insulin receptor is a heterodimer\n",
      "CC   consisting of 2 alpha and 2 beta subunits.  Beta-prime may be a\n",
      "CC   cleavage product produced upon binding of insulin.  [1] suggests\n",
      "CC   that translation may begin at the 'atg' start codon at positions\n",
      "CC   79-81 with protein cleavage occurring after position 120 to yield\n",
      "CC   the signal peptide.  [1] gives illustrations of the various domains\n",
      "CC   present in the protein.  A draft entry and sequence for [1] in\n",
      "CC   computer-readable form were kindly provided by K. Jarnagin\n",
      "CC   (30-JUL-1985).\n",
      "XX\n",
      "FH   Key             Location/Qualifiers\n",
      "FH\n",
      "FT   source          1..4723\n",
      "FT                   /organism=\"Homo sapiens\"\n",
      "FT                   /map=\"19p13.3-p13.2\"\n",
      "FT                   /mol_type=\"mRNA\"\n",
      "FT                   /db_xref=\"taxon:9606\"\n",
      "FT   sig_peptide     137..219\n",
      "FT                   /note=\"insulin receptor signal peptide\"\n",
      "FT   CDS             139..4287\n",
      "FT                   /codon_start=1\n",
      "FT                   /gene=\"INSR\"\n",
      "FT                   /note=\"insulin receptor precursor\"\n",
      "FT                   /db_xref=\"GOA:P06213\"\n",
      "FT                   /db_xref=\"H-InvDB:HIT000194074.15\"\n",
      "FT                   /db_xref=\"HGNC:HGNC:6091\"\n",
      "FT                   /db_xref=\"InterPro:IPR000494\"\n",
      "FT                   /db_xref=\"InterPro:IPR000719\"\n",
      "FT                   /db_xref=\"InterPro:IPR001245\"\n",
      "FT                   /db_xref=\"InterPro:IPR002011\"\n",
      "FT                   /db_xref=\"InterPro:IPR003961\"\n",
      "FT                   /db_xref=\"InterPro:IPR006211\"\n",
      "FT                   /db_xref=\"InterPro:IPR006212\"\n",
      "FT                   /db_xref=\"InterPro:IPR008266\"\n",
      "FT                   /db_xref=\"InterPro:IPR009030\"\n",
      "FT                   /db_xref=\"InterPro:IPR011009\"\n",
      "FT                   /db_xref=\"InterPro:IPR013783\"\n",
      "FT                   /db_xref=\"InterPro:IPR016246\"\n",
      "FT                   /db_xref=\"InterPro:IPR017441\"\n",
      "FT                   /db_xref=\"InterPro:IPR020635\"\n",
      "FT                   /db_xref=\"InterPro:IPR032675\"\n",
      "FT                   /db_xref=\"PDB:1GAG\"\n",
      "FT                   /db_xref=\"PDB:1I44\"\n",
      "FT                   /db_xref=\"PDB:1IR3\"\n",
      "FT                   /db_xref=\"PDB:1IRK\"\n",
      "FT                   /db_xref=\"PDB:1P14\"\n",
      "FT                   /db_xref=\"PDB:1RQQ\"\n",
      "FT                   /db_xref=\"PDB:2AUH\"\n",
      "FT                   /db_xref=\"PDB:2B4S\"\n",
      "FT                   /db_xref=\"PDB:2HR7\"\n",
      "FT                   /db_xref=\"PDB:2MFR\"\n",
      "FT                   /db_xref=\"PDB:2Z8C\"\n",
      "FT                   /db_xref=\"PDB:3BU3\"\n",
      "FT                   /db_xref=\"PDB:3BU5\"\n",
      "FT                   /db_xref=\"PDB:3BU6\"\n",
      "FT                   /db_xref=\"PDB:3EKK\"\n",
      "FT                   /db_xref=\"PDB:3EKN\"\n",
      "FT                   /db_xref=\"PDB:3ETA\"\n",
      "FT                   /db_xref=\"PDB:3W11\"\n",
      "FT                   /db_xref=\"PDB:3W12\"\n",
      "FT                   /db_xref=\"PDB:3W13\"\n",
      "FT                   /db_xref=\"PDB:3W14\"\n",
      "FT                   /db_xref=\"PDB:4IBM\"\n",
      "FT                   /db_xref=\"PDB:4OGA\"\n",
      "FT                   /db_xref=\"PDB:4XLV\"\n",
      "FT                   /db_xref=\"PDB:4XSS\"\n",
      "FT                   /db_xref=\"PDB:4XST\"\n",
      "FT                   /db_xref=\"PDB:4ZXB\"\n",
      "FT                   /db_xref=\"PDB:5E1S\"\n",
      "FT                   /db_xref=\"PDB:5HHW\"\n",
      "FT                   /db_xref=\"UniProtKB/Swiss-Prot:P06213\"\n",
      "FT                   /protein_id=\"AAA59174.1\"\n",
      "FT                   /translation=\"MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLT\n",
      "FT                   RLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFP\n",
      "FT                   NLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRI\n",
      "FT                   LDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTI\n",
      "FT                   CKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCV\n",
      "FT                   NFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCH\n",
      "FT                   LLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRR\n",
      "FT                   SYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYN\n",
      "FT                   PKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWE\n",
      "FT                   PYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHP\n",
      "FT                   GWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSS\n",
      "FT                   SQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDS\n",
      "FT                   QKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAE\n",
      "FT                   DPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGL\n",
      "FT                   RHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMW\n",
      "FT                   QEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSL\n",
      "FT                   AGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLFSVVIGSIYLFLRKRQPDGPLG\n",
      "FT                   PLYASSNPEYLSASDVFPCSVYVPDEWEVSREKITLLRELGQGSFGMVYEGNARDIIKG\n",
      "FT                   EAETRVAVKTVNESASLRERIEFLNEASVMKGFTCHHVVRLLGVVSKGQPTLVVMELMA\n",
      "FT                   HGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMAAEIADGMAYLNAKKFVHRDLAARNCM\n",
      "FT                   VAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPVRWMAPESLKDGVFTTSSDMWSFGVV\n",
      "FT                   LWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDNCPERVTDLMRMCWQFNPKMRPTFL\n",
      "FT                   EIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEMEFEDMENVPLDRSSHCQREEAGGR\n",
      "FT                   DGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSNPS\"\n",
      "FT   mat_peptide     220..2424\n",
      "FT                   /gene=\"INSR\"\n",
      "FT                   /note=\"insulin receptor alpha subunit\"\n",
      "FT   mat_peptide     2425..4284\n",
      "FT                   /gene=\"INSR\"\n",
      "FT                   /note=\"insulin receptor beta subunit\"\n",
      "FT   mat_peptide     2425..2469\n",
      "FT                   /partial\n",
      "FT                   /gene=\"INSR\"\n",
      "FT                   /note=\"insulin receptor beta-prime subunit\"\n",
      "XX\n",
      "SQ   Sequence 4723 BP; 1068 A; 1298 C; 1311 G; 1046 T; 0 other;\n",
      "     ggggggctgc gcggccgggt cggtgcgcac acgagaagga cgcgcggccc ccagcgctct        60\n",
      "     tgggggccgc ctcggagcat gacccccgcg ggccagcgcc gcgcgcctga tccgaggaga       120\n",
      "     ccccgcgctc ccgcagccat gggcaccggg ggccggcggg gggcggcggc cgcgccgctg       180\n",
      "     ctggtggcgg tggccgcgct gctactgggc gccgcgggcc acctgtaccc cggagaggtg       240\n",
      "     tgtcccggca tggatatccg gaacaacctc actaggttgc atgagctgga gaattgctct       300\n",
      "     gtcatcgaag gacacttgca gatactcttg atgttcaaaa cgaggcccga agatttccga       360\n",
      "     gacctcagtt tccccaaact catcatgatc actgattact tgctgctctt ccgggtctat       420\n",
      "     gggctcgaga gcctgaagga cctgttcccc aacctcacgg tcatccgggg atcacgactg       480\n",
      "     ttctttaact acgcgctggt catcttcgag atggttcacc tcaaggaact cggcctctac       540\n",
      "     aacctgatga acatcacccg gggttctgtc cgcatcgaga agaacaatga gctctgttac       600\n",
      "     ttggccacta tcgactggtc ccgtatcctg gattccgtgg aggataatca catcgtgttg       660\n",
      "     aacaaagatg acaacgagga gtgtggagac atctgtccgg gtaccgcgaa gggcaagacc       720\n",
      "     aactgccccg ccaccgtcat caacgggcag tttgtcgaac gatgttggac tcatagtcac       780\n",
      "     tgccagaaag tttgcccgac catctgtaag tcacacggct gcaccgccga aggcctctgt       840\n",
      "     tgccacagcg agtgcctggg caactgttct cagcccgacg accccaccaa gtgcgtggcc       900\n",
      "     tgccgcaact tctacctgga cggcaggtgt gtggagacct gcccgccccc gtactaccac       960\n",
      "     ttccaggact ggcgctgtgt gaacttcagc ttctgccagg acctgcacca caaatgcaag      1020\n",
      "     aactcgcgga ggcagggctg ccaccaatac gtcattcaca acaacaagtg catccctgag      1080\n",
      "     tgtccctccg ggtacacgat gaattccagc aacttgctgt gcaccccatg cctgggtccc      1140\n",
      "     tgtcccaagg tgtgccacct cctagaaggc gagaagacca tcgactcggt gacgtctgcc      1200\n",
      "     caggagctcc gaggatgcac cgtcatcaac gggagtctga tcatcaacat tcgaggaggc      1260\n",
      "     aacaatctgg cagctgagct agaagccaac ctcggcctca ttgaagaaat ttcagggtat      1320\n",
      "     ctaaaaatcc gccgatccta cgctctggtg tcactttcct tcttccggaa gttacgtctg      1380\n",
      "     attcgaggag agaccttgga aattgggaac tactccttct atgccttgga caaccagaac      1440\n",
      "     ctaaggcagc tctgggactg gagcaaacac aacctcacca ccactcaggg gaaactcttc      1500\n",
      "     ttccactata accccaaact ctgcttgtca gaaatccaca agatggaaga agtttcagga      1560\n",
      "     accaaggggc gccaggagag aaacgacatt gccctgaaga ccaatgggga caaggcatcc      1620\n",
      "     tgtgaaaatg agttacttaa attttcttac attcggacat cttttgacaa gatcttgctg      1680\n",
      "     agatgggagc cgtactggcc ccccgacttc cgagacctct tggggttcat gctgttctac      1740\n",
      "     aaagaggccc cttatcagaa tgtgacggag ttcgatgggc aggatgcgtg tggttccaac      1800\n",
      "     agttggacgg tggtagacat tgacccaccc ctgaggtcca acgaccccaa atcacagaac      1860\n",
      "     cacccagggt ggctgatgcg gggtctcaag ccctggaccc agtatgccat ctttgtgaag      1920\n",
      "     accctggtca ccttttcgga tgaacgccgg acctatgggg ccaagagtga catcatttat      1980\n",
      "     gtccagacag atgccaccaa cccctctgtg cccctggatc caatctcagt gtctaactca      2040\n",
      "     tcatcccaga ttattctgaa gtggaaacca ccctccgacc ccaatggcaa catcacccac      2100\n",
      "     tacctggttt tctgggagag gcaggcggaa gacagtgagc tgttcgagct ggattattgc      2160\n",
      "     ctcaaagggc tgaagctgcc ctcgaggacc tggtctccac cattcgagtc tgaagattct      2220\n",
      "     cagaagcaca accagagtga gtatgaggat tcggccggcg aatgctgctc ctgtccaaag      2280\n",
      "     acagactctc agatcctgaa ggagctggag gagtcctcgt ttaggaagac gtttgaggat      2340\n",
      "     tacctgcaca acgtggtttt cgtccccaga aaaacctctt caggcactgg tgccgaggac      2400\n",
      "     cctaggccat ctcggaaacg caggtccctt ggcgatgttg ggaatgtgac ggtggccgtg      2460\n",
      "     cccacggtgg cagctttccc caacacttcc tcgaccagcg tgcccacgag tccggaggag      2520\n",
      "     cacaggcctt ttgagaaggt ggtgaacaag gagtcgctgg tcatctccgg cttgcgacac      2580\n",
      "     ttcacgggct atcgcatcga gctgcaggct tgcaaccagg acacccctga ggaacggtgc      2640\n",
      "     agtgtggcag cctacgtcag tgcgaggacc atgcctgaag ccaaggctga tgacattgtt      2700\n",
      "     ggccctgtga cgcatgaaat ctttgagaac aacgtcgtcc acttgatgtg gcaggagccg      2760\n",
      "     aaggagccca atggtctgat cgtgctgtat gaagtgagtt atcggcgata tggtgatgag      2820\n",
      "     gagctgcatc tctgcgtctc ccgcaagcac ttcgctctgg aacggggctg caggctgcgt      2880\n",
      "     gggctgtcac cggggaacta cagcgtgcga atccgggcca cctcccttgc gggcaacggc      2940\n",
      "     tcttggacgg aacccaccta tttctacgtg acagactatt tagacgtccc gtcaaatatt      3000\n",
      "     gcaaaaatta tcatcggccc cctcatcttt gtctttctct tcagtgttgt gattggaagt      3060\n",
      "     atttatctat tcctgagaaa gaggcagcca gatgggccgc tgggaccgct ttacgcttct      3120\n",
      "     tcaaaccctg agtatctcag tgccagtgat gtgtttccat gctctgtgta cgtgccggac      3180\n",
      "     gagtgggagg tgtctcgaga gaagatcacc ctccttcgag agctggggca gggctccttc      3240\n",
      "     ggcatggtgt atgagggcaa tgccagggac atcatcaagg gtgaggcaga gacccgcgtg      3300\n",
      "     gcggtgaaga cggtcaacga gtcagccagt ctccgagagc ggattgagtt cctcaatgag      3360\n",
      "     gcctcggtca tgaagggctt cacctgccat cacgtggtgc gcctcctggg agtggtgtcc      3420\n",
      "     aagggccagc ccacgctggt ggtgatggag ctgatggctc acggagacct gaagagctac      3480\n",
      "     ctccgttctc tgcggccaga ggctgagaat aatcctggcc gccctccccc tacccttcaa      3540\n",
      "     gagatgattc agatggcggc agagattgct gacgggatgg cctacctgaa cgccaagaag      3600\n",
      "     tttgtgcatc gggacctggc agcgagaaac tgcatggtcg cccatgattt tactgtcaaa      3660\n",
      "     attggagact ttggaatgac cagagacatc tatgaaacgg attactaccg gaaagggggc      3720\n",
      "     aagggtctgc tccctgtacg gtggatggca ccggagtccc tgaaggatgg ggtcttcacc      3780\n",
      "     acttcttctg acatgtggtc ctttggcgtg gtcctttggg aaatcaccag cttggcagaa      3840\n",
      "     cagccttacc aaggcctgtc taatgaacag gtgttgaaat ttgtcatgga tggagggtat      3900\n",
      "     ctggatcaac ccgacaactg tccagagaga gtcactgacc tcatgcgcat gtgctggcaa      3960\n",
      "     ttcaacccca agatgaggcc aaccttcctg gagattgtca acctgctcaa ggacgacctg      4020\n",
      "     caccccagct ttccagaggt gtcgttcttc cacagcgagg agaacaaggc tcccgagagt      4080\n",
      "     gaggagctgg agatggagtt tgaggacatg gagaatgtgc ccctggaccg ttcctcgcac      4140\n",
      "     tgtcagaggg aggaggcggg gggccgggat ggagggtcct cgctgggttt caagcggagc      4200\n",
      "     tacgaggaac acatccctta cacacacatg aacggaggca agaaaaacgg gcggattctg      4260\n",
      "     accttgcctc ggtccaatcc ttcctaacag tgcctaccgt ggcgggggcg ggcaggggtt      4320\n",
      "     cccattttcg ctttcctctg gtttgaaagc ctctggaaaa ctcaggattc tcacgactct      4380\n",
      "     accatgtcca gtggagttca gagatcgttc ctatacattt ctgttcatct taaggtggac      4440\n",
      "     tcgtttggtt accaatttaa ctagtcctgc agaggattta actgtgaacc tggagggcaa      4500\n",
      "     ggggtttcca cagttgctgc tcctttgggg caacgacggt ttcaaaccag gattttgtgt      4560\n",
      "     tttttcgttc cccccacccg cccccagcag atggaaagaa agcacctgtt tttacaaatt      4620\n",
      "     cttttttttt tttttttttt tttttttttg ctggtgtctg agcttcagta taaaagacaa      4680\n",
      "     aacttcctgt ttgtggaaca aaatttcgaa agaaaaaacc aaa                        4723\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(embl_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estrarre l'identificatore univoco e l'organismo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fare riferimento all'esercizio 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = re.search('^ID\\s+(\\w+).+\\s+(\\w+);', embl_str, re.M)\n",
    "(identifier, organism) = m.groups()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'M10051'"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HUM'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estrarre la sequenza nucleotidica (in lettere minuscole)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fare riferimento all'esercizio 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_row_list = re.findall('^\\s+(.+?)\\s+\\d+', embl_str, re.M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ggggggctgc gcggccgggt cggtgcgcac acgagaagga cgcgcggccc ccagcgctct',\n",
       " 'tgggggccgc ctcggagcat gacccccgcg ggccagcgcc gcgcgcctga tccgaggaga',\n",
       " 'ccccgcgctc ccgcagccat gggcaccggg ggccggcggg gggcggcggc cgcgccgctg',\n",
       " 'ctggtggcgg tggccgcgct gctactgggc gccgcgggcc acctgtaccc cggagaggtg',\n",
       " 'tgtcccggca tggatatccg gaacaacctc actaggttgc atgagctgga gaattgctct',\n",
       " 'gtcatcgaag gacacttgca gatactcttg atgttcaaaa cgaggcccga agatttccga',\n",
       " 'gacctcagtt tccccaaact catcatgatc actgattact tgctgctctt ccgggtctat',\n",
       " 'gggctcgaga gcctgaagga cctgttcccc aacctcacgg tcatccgggg atcacgactg',\n",
       " 'ttctttaact acgcgctggt catcttcgag atggttcacc tcaaggaact cggcctctac',\n",
       " 'aacctgatga acatcacccg gggttctgtc cgcatcgaga agaacaatga gctctgttac',\n",
       " 'ttggccacta tcgactggtc ccgtatcctg gattccgtgg aggataatca catcgtgttg',\n",
       " 'aacaaagatg acaacgagga gtgtggagac atctgtccgg gtaccgcgaa gggcaagacc',\n",
       " 'aactgccccg ccaccgtcat caacgggcag tttgtcgaac gatgttggac tcatagtcac',\n",
       " 'tgccagaaag tttgcccgac catctgtaag tcacacggct gcaccgccga aggcctctgt',\n",
       " 'tgccacagcg agtgcctggg caactgttct cagcccgacg accccaccaa gtgcgtggcc',\n",
       " 'tgccgcaact tctacctgga cggcaggtgt gtggagacct gcccgccccc gtactaccac',\n",
       " 'ttccaggact ggcgctgtgt gaacttcagc ttctgccagg acctgcacca caaatgcaag',\n",
       " 'aactcgcgga ggcagggctg ccaccaatac gtcattcaca acaacaagtg catccctgag',\n",
       " 'tgtccctccg ggtacacgat gaattccagc aacttgctgt gcaccccatg cctgggtccc',\n",
       " 'tgtcccaagg tgtgccacct cctagaaggc gagaagacca tcgactcggt gacgtctgcc',\n",
       " 'caggagctcc gaggatgcac cgtcatcaac gggagtctga tcatcaacat tcgaggaggc',\n",
       " 'aacaatctgg cagctgagct agaagccaac ctcggcctca ttgaagaaat ttcagggtat',\n",
       " 'ctaaaaatcc gccgatccta cgctctggtg tcactttcct tcttccggaa gttacgtctg',\n",
       " 'attcgaggag agaccttgga aattgggaac tactccttct atgccttgga caaccagaac',\n",
       " 'ctaaggcagc tctgggactg gagcaaacac aacctcacca ccactcaggg gaaactcttc',\n",
       " 'ttccactata accccaaact ctgcttgtca gaaatccaca agatggaaga agtttcagga',\n",
       " 'accaaggggc gccaggagag aaacgacatt gccctgaaga ccaatgggga caaggcatcc',\n",
       " 'tgtgaaaatg agttacttaa attttcttac attcggacat cttttgacaa gatcttgctg',\n",
       " 'agatgggagc cgtactggcc ccccgacttc cgagacctct tggggttcat gctgttctac',\n",
       " 'aaagaggccc cttatcagaa tgtgacggag ttcgatgggc aggatgcgtg tggttccaac',\n",
       " 'agttggacgg tggtagacat tgacccaccc ctgaggtcca acgaccccaa atcacagaac',\n",
       " 'cacccagggt ggctgatgcg gggtctcaag ccctggaccc agtatgccat ctttgtgaag',\n",
       " 'accctggtca ccttttcgga tgaacgccgg acctatgggg ccaagagtga catcatttat',\n",
       " 'gtccagacag atgccaccaa cccctctgtg cccctggatc caatctcagt gtctaactca',\n",
       " 'tcatcccaga ttattctgaa gtggaaacca ccctccgacc ccaatggcaa catcacccac',\n",
       " 'tacctggttt tctgggagag gcaggcggaa gacagtgagc tgttcgagct ggattattgc',\n",
       " 'ctcaaagggc tgaagctgcc ctcgaggacc tggtctccac cattcgagtc tgaagattct',\n",
       " 'cagaagcaca accagagtga gtatgaggat tcggccggcg aatgctgctc ctgtccaaag',\n",
       " 'acagactctc agatcctgaa ggagctggag gagtcctcgt ttaggaagac gtttgaggat',\n",
       " 'tacctgcaca acgtggtttt cgtccccaga aaaacctctt caggcactgg tgccgaggac',\n",
       " 'cctaggccat ctcggaaacg caggtccctt ggcgatgttg ggaatgtgac ggtggccgtg',\n",
       " 'cccacggtgg cagctttccc caacacttcc tcgaccagcg tgcccacgag tccggaggag',\n",
       " 'cacaggcctt ttgagaaggt ggtgaacaag gagtcgctgg tcatctccgg cttgcgacac',\n",
       " 'ttcacgggct atcgcatcga gctgcaggct tgcaaccagg acacccctga ggaacggtgc',\n",
       " 'agtgtggcag cctacgtcag tgcgaggacc atgcctgaag ccaaggctga tgacattgtt',\n",
       " 'ggccctgtga cgcatgaaat ctttgagaac aacgtcgtcc acttgatgtg gcaggagccg',\n",
       " 'aaggagccca atggtctgat cgtgctgtat gaagtgagtt atcggcgata tggtgatgag',\n",
       " 'gagctgcatc tctgcgtctc ccgcaagcac ttcgctctgg aacggggctg caggctgcgt',\n",
       " 'gggctgtcac cggggaacta cagcgtgcga atccgggcca cctcccttgc gggcaacggc',\n",
       " 'tcttggacgg aacccaccta tttctacgtg acagactatt tagacgtccc gtcaaatatt',\n",
       " 'gcaaaaatta tcatcggccc cctcatcttt gtctttctct tcagtgttgt gattggaagt',\n",
       " 'atttatctat tcctgagaaa gaggcagcca gatgggccgc tgggaccgct ttacgcttct',\n",
       " 'tcaaaccctg agtatctcag tgccagtgat gtgtttccat gctctgtgta cgtgccggac',\n",
       " 'gagtgggagg tgtctcgaga gaagatcacc ctccttcgag agctggggca gggctccttc',\n",
       " 'ggcatggtgt atgagggcaa tgccagggac atcatcaagg gtgaggcaga gacccgcgtg',\n",
       " 'gcggtgaaga cggtcaacga gtcagccagt ctccgagagc ggattgagtt cctcaatgag',\n",
       " 'gcctcggtca tgaagggctt cacctgccat cacgtggtgc gcctcctggg agtggtgtcc',\n",
       " 'aagggccagc ccacgctggt ggtgatggag ctgatggctc acggagacct gaagagctac',\n",
       " 'ctccgttctc tgcggccaga ggctgagaat aatcctggcc gccctccccc tacccttcaa',\n",
       " 'gagatgattc agatggcggc agagattgct gacgggatgg cctacctgaa cgccaagaag',\n",
       " 'tttgtgcatc gggacctggc agcgagaaac tgcatggtcg cccatgattt tactgtcaaa',\n",
       " 'attggagact ttggaatgac cagagacatc tatgaaacgg attactaccg gaaagggggc',\n",
       " 'aagggtctgc tccctgtacg gtggatggca ccggagtccc tgaaggatgg ggtcttcacc',\n",
       " 'acttcttctg acatgtggtc ctttggcgtg gtcctttggg aaatcaccag cttggcagaa',\n",
       " 'cagccttacc aaggcctgtc taatgaacag gtgttgaaat ttgtcatgga tggagggtat',\n",
       " 'ctggatcaac ccgacaactg tccagagaga gtcactgacc tcatgcgcat gtgctggcaa',\n",
       " 'ttcaacccca agatgaggcc aaccttcctg gagattgtca acctgctcaa ggacgacctg',\n",
       " 'caccccagct ttccagaggt gtcgttcttc cacagcgagg agaacaaggc tcccgagagt',\n",
       " 'gaggagctgg agatggagtt tgaggacatg gagaatgtgc ccctggaccg ttcctcgcac',\n",
       " 'tgtcagaggg aggaggcggg gggccgggat ggagggtcct cgctgggttt caagcggagc',\n",
       " 'tacgaggaac acatccctta cacacacatg aacggaggca agaaaaacgg gcggattctg',\n",
       " 'accttgcctc ggtccaatcc ttcctaacag tgcctaccgt ggcgggggcg ggcaggggtt',\n",
       " 'cccattttcg ctttcctctg gtttgaaagc ctctggaaaa ctcaggattc tcacgactct',\n",
       " 'accatgtcca gtggagttca gagatcgttc ctatacattt ctgttcatct taaggtggac',\n",
       " 'tcgtttggtt accaatttaa ctagtcctgc agaggattta actgtgaacc tggagggcaa',\n",
       " 'ggggtttcca cagttgctgc tcctttgggg caacgacggt ttcaaaccag gattttgtgt',\n",
       " 'tttttcgttc cccccacccg cccccagcag atggaaagaa agcacctgtt tttacaaatt',\n",
       " 'cttttttttt tttttttttt tttttttttg ctggtgtctg agcttcagta taaaagacaa',\n",
       " 'aacttcctgt ttgtggaaca aaatttcgaa agaaaaaacc aaa']"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_row_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotide_sequence = re.sub(r'\\s', '', ''.join(seq_row_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ggggggctgcgcggccgggtcggtgcgcacacgagaaggacgcgcggcccccagcgctcttgggggccgcctcggagcatgacccccgcgggccagcgccgcgcgcctgatccgaggagaccccgcgctcccgcagccatgggcaccgggggccggcggggggcggcggccgcgccgctgctggtggcggtggccgcgctgctactgggcgccgcgggccacctgtaccccggagaggtgtgtcccggcatggatatccggaacaacctcactaggttgcatgagctggagaattgctctgtcatcgaaggacacttgcagatactcttgatgttcaaaacgaggcccgaagatttccgagacctcagtttccccaaactcatcatgatcactgattacttgctgctcttccgggtctatgggctcgagagcctgaaggacctgttccccaacctcacggtcatccggggatcacgactgttctttaactacgcgctggtcatcttcgagatggttcacctcaaggaactcggcctctacaacctgatgaacatcacccggggttctgtccgcatcgagaagaacaatgagctctgttacttggccactatcgactggtcccgtatcctggattccgtggaggataatcacatcgtgttgaacaaagatgacaacgaggagtgtggagacatctgtccgggtaccgcgaagggcaagaccaactgccccgccaccgtcatcaacgggcagtttgtcgaacgatgttggactcatagtcactgccagaaagtttgcccgaccatctgtaagtcacacggctgcaccgccgaaggcctctgttgccacagcgagtgcctgggcaactgttctcagcccgacgaccccaccaagtgcgtggcctgccgcaacttctacctggacggcaggtgtgtggagacctgcccgcccccgtactaccacttccaggactggcgctgtgtgaacttcagcttctgccaggacctgcaccacaaatgcaagaactcgcggaggcagggctgccaccaatacgtcattcacaacaacaagtgcatccctgagtgtccctccgggtacacgatgaattccagcaacttgctgtgcaccccatgcctgggtccctgtcccaaggtgtgccacctcctagaaggcgagaagaccatcgactcggtgacgtctgcccaggagctccgaggatgcaccgtcatcaacgggagtctgatcatcaacattcgaggaggcaacaatctggcagctgagctagaagccaacctcggcctcattgaagaaatttcagggtatctaaaaatccgccgatcctacgctctggtgtcactttccttcttccggaagttacgtctgattcgaggagagaccttggaaattgggaactactccttctatgccttggacaaccagaacctaaggcagctctgggactggagcaaacacaacctcaccaccactcaggggaaactcttcttccactataaccccaaactctgcttgtcagaaatccacaagatggaagaagtttcaggaaccaaggggcgccaggagagaaacgacattgccctgaagaccaatggggacaaggcatcctgtgaaaatgagttacttaaattttcttacattcggacatcttttgacaagatcttgctgagatgggagccgtactggccccccgacttccgagacctcttggggttcatgctgttctacaaagaggccccttatcagaatgtgacggagttcgatgggcaggatgcgtgtggttccaacagttggacggtggtagacattgacccacccctgaggtccaacgaccccaaatcacagaaccacccagggtggctgatgcggggtctcaagccctggacccagtatgccatctttgtgaagaccctggtcaccttttcggatgaacgccggacctatggggccaagagtgacatcatttatgtccagacagatgccaccaacccctctgtgcccctggatccaatctcagtgtctaactcatcatcccagattattctgaagtggaaaccaccctccgaccccaatggcaacatcacccactacctggttttctgggagaggcaggcggaagacagtgagctgttcgagctggattattgcctcaaagggctgaagctgccctcgaggacctggtctccaccattcgagtctgaagattctcagaagcacaaccagagtgagtatgaggattcggccggcgaatgctgctcctgtccaaagacagactctcagatcctgaaggagctggaggagtcctcgtttaggaagacgtttgaggattacctgcacaacgtggttttcgtccccagaaaaacctcttcaggcactggtgccgaggaccctaggccatctcggaaacgcaggtcccttggcgatgttgggaatgtgacggtggccgtgcccacggtggcagctttccccaacacttcctcgaccagcgtgcccacgagtccggaggagcacaggccttttgagaaggtggtgaacaaggagtcgctggtcatctccggcttgcgacacttcacgggctatcgcatcgagctgcaggcttgcaaccaggacacccctgaggaacggtgcagtgtggcagcctacgtcagtgcgaggaccatgcctgaagccaaggctgatgacattgttggccctgtgacgcatgaaatctttgagaacaacgtcgtccacttgatgtggcaggagccgaaggagcccaatggtctgatcgtgctgtatgaagtgagttatcggcgatatggtgatgaggagctgcatctctgcgtctcccgcaagcacttcgctctggaacggggctgcaggctgcgtgggctgtcaccggggaactacagcgtgcgaatccgggccacctcccttgcgggcaacggctcttggacggaacccacctatttctacgtgacagactatttagacgtcccgtcaaatattgcaaaaattatcatcggccccctcatctttgtctttctcttcagtgttgtgattggaagtatttatctattcctgagaaagaggcagccagatgggccgctgggaccgctttacgcttcttcaaaccctgagtatctcagtgccagtgatgtgtttccatgctctgtgtacgtgccggacgagtgggaggtgtctcgagagaagatcaccctccttcgagagctggggcagggctccttcggcatggtgtatgagggcaatgccagggacatcatcaagggtgaggcagagacccgcgtggcggtgaagacggtcaacgagtcagccagtctccgagagcggattgagttcctcaatgaggcctcggtcatgaagggcttcacctgccatcacgtggtgcgcctcctgggagtggtgtccaagggccagcccacgctggtggtgatggagctgatggctcacggagacctgaagagctacctccgttctctgcggccagaggctgagaataatcctggccgccctccccctacccttcaagagatgattcagatggcggcagagattgctgacgggatggcctacctgaacgccaagaagtttgtgcatcgggacctggcagcgagaaactgcatggtcgcccatgattttactgtcaaaattggagactttggaatgaccagagacatctatgaaacggattactaccggaaagggggcaagggtctgctccctgtacggtggatggcaccggagtccctgaaggatggggtcttcaccacttcttctgacatgtggtcctttggcgtggtcctttgggaaatcaccagcttggcagaacagccttaccaaggcctgtctaatgaacaggtgttgaaatttgtcatggatggagggtatctggatcaacccgacaactgtccagagagagtcactgacctcatgcgcatgtgctggcaattcaaccccaagatgaggccaaccttcctggagattgtcaacctgctcaaggacgacctgcaccccagctttccagaggtgtcgttcttccacagcgaggagaacaaggctcccgagagtgaggagctggagatggagtttgaggacatggagaatgtgcccctggaccgttcctcgcactgtcagagggaggaggcggggggccgggatggagggtcctcgctgggtttcaagcggagctacgaggaacacatcccttacacacacatgaacggaggcaagaaaaacgggcggattctgaccttgcctcggtccaatccttcctaacagtgcctaccgtggcgggggcgggcaggggttcccattttcgctttcctctggtttgaaagcctctggaaaactcaggattctcacgactctaccatgtccagtggagttcagagatcgttcctatacatttctgttcatcttaaggtggactcgtttggttaccaatttaactagtcctgcagaggatttaactgtgaacctggagggcaaggggtttccacagttgctgctcctttggggcaacgacggtttcaaaccaggattttgtgttttttcgttccccccacccgcccccagcagatggaaagaaagcacctgtttttacaaattcttttttttttttttttttttttttttttgctggtgtctgagcttcagtataaaagacaaaacttcctgtttgtggaacaaaatttcgaaagaaaaaaccaaa'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucleotide_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estrarre la sequenza della proteina"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fare riferimento all'esercizio 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = re.search('^FT\\s+/translation=\\\"(\\w+)', embl_str, re.M)\n",
    "protein_prefix = s.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLT'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_prefix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_list = re.findall('^FT\\s+([A-Z]+)\"?$', embl_str, re.M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['RLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFP',\n",
       " 'NLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRI',\n",
       " 'LDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTI',\n",
       " 'CKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCV',\n",
       " 'NFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCH',\n",
       " 'LLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRR',\n",
       " 'SYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYN',\n",
       " 'PKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWE',\n",
       " 'PYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHP',\n",
       " 'GWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSS',\n",
       " 'SQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDS',\n",
       " 'QKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAE',\n",
       " 'DPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGL',\n",
       " 'RHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMW',\n",
       " 'QEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSL',\n",
       " 'AGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLFSVVIGSIYLFLRKRQPDGPLG',\n",
       " 'PLYASSNPEYLSASDVFPCSVYVPDEWEVSREKITLLRELGQGSFGMVYEGNARDIIKG',\n",
       " 'EAETRVAVKTVNESASLRERIEFLNEASVMKGFTCHHVVRLLGVVSKGQPTLVVMELMA',\n",
       " 'HGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMAAEIADGMAYLNAKKFVHRDLAARNCM',\n",
       " 'VAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPVRWMAPESLKDGVFTTSSDMWSFGVV',\n",
       " 'LWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDNCPERVTDLMRMCWQFNPKMRPTFL',\n",
       " 'EIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEMEFEDMENVPLDRSSHCQREEAGGR',\n",
       " 'DGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSNPS']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_sequence = ''.join([protein_prefix] + protein_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEYLSASDVFPCSVYVPDEWEVSREKITLLRELGQGSFGMVYEGNARDIIKGEAETRVAVKTVNESASLRERIEFLNEASVMKGFTCHHVVRLLGVVSKGQPTLVVMELMAHGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMAAEIADGMAYLNAKKFVHRDLAARNCMVAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPVRWMAPESLKDGVFTTSSDMWSFGVVLWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDNCPERVTDLMRMCWQFNPKMRPTFLEIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEMEFEDMENVPLDRSSHCQREEAGGRDGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSNPS'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estrarre la coding sequence (CDS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Estrarre dal *record*\n",
    "\n",
    "    FT   CDS             139..4287\n",
    "    \n",
    "lo start e l'end (1-based) della CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = re.search(r'^FT\\s+CDS\\s+(\\d+)..(\\d+)$', embl_str, re.M)\n",
    "(cds_start, cds_end) = map(int, m.groups())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4287"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Estrarre la sequenza della CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_sequence = nucleotide_sequence[cds_start-1:cds_end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'atgggcaccgggggccggcggggggcggcggccgcgccgctgctggtggcggtggccgcgctgctactgggcgccgcgggccacctgtaccccggagaggtgtgtcccggcatggatatccggaacaacctcactaggttgcatgagctggagaattgctctgtcatcgaaggacacttgcagatactcttgatgttcaaaacgaggcccgaagatttccgagacctcagtttccccaaactcatcatgatcactgattacttgctgctcttccgggtctatgggctcgagagcctgaaggacctgttccccaacctcacggtcatccggggatcacgactgttctttaactacgcgctggtcatcttcgagatggttcacctcaaggaactcggcctctacaacctgatgaacatcacccggggttctgtccgcatcgagaagaacaatgagctctgttacttggccactatcgactggtcccgtatcctggattccgtggaggataatcacatcgtgttgaacaaagatgacaacgaggagtgtggagacatctgtccgggtaccgcgaagggcaagaccaactgccccgccaccgtcatcaacgggcagtttgtcgaacgatgttggactcatagtcactgccagaaagtttgcccgaccatctgtaagtcacacggctgcaccgccgaaggcctctgttgccacagcgagtgcctgggcaactgttctcagcccgacgaccccaccaagtgcgtggcctgccgcaacttctacctggacggcaggtgtgtggagacctgcccgcccccgtactaccacttccaggactggcgctgtgtgaacttcagcttctgccaggacctgcaccacaaatgcaagaactcgcggaggcagggctgccaccaatacgtcattcacaacaacaagtgcatccctgagtgtccctccgggtacacgatgaattccagcaacttgctgtgcaccccatgcctgggtccctgtcccaaggtgtgccacctcctagaaggcgagaagaccatcgactcggtgacgtctgcccaggagctccgaggatgcaccgtcatcaacgggagtctgatcatcaacattcgaggaggcaacaatctggcagctgagctagaagccaacctcggcctcattgaagaaatttcagggtatctaaaaatccgccgatcctacgctctggtgtcactttccttcttccggaagttacgtctgattcgaggagagaccttggaaattgggaactactccttctatgccttggacaaccagaacctaaggcagctctgggactggagcaaacacaacctcaccaccactcaggggaaactcttcttccactataaccccaaactctgcttgtcagaaatccacaagatggaagaagtttcaggaaccaaggggcgccaggagagaaacgacattgccctgaagaccaatggggacaaggcatcctgtgaaaatgagttacttaaattttcttacattcggacatcttttgacaagatcttgctgagatgggagccgtactggccccccgacttccgagacctcttggggttcatgctgttctacaaagaggccccttatcagaatgtgacggagttcgatgggcaggatgcgtgtggttccaacagttggacggtggtagacattgacccacccctgaggtccaacgaccccaaatcacagaaccacccagggtggctgatgcggggtctcaagccctggacccagtatgccatctttgtgaagaccctggtcaccttttcggatgaacgccggacctatggggccaagagtgacatcatttatgtccagacagatgccaccaacccctctgtgcccctggatccaatctcagtgtctaactcatcatcccagattattctgaagtggaaaccaccctccgaccccaatggcaacatcacccactacctggttttctgggagaggcaggcggaagacagtgagctgttcgagctggattattgcctcaaagggctgaagctgccctcgaggacctggtctccaccattcgagtctgaagattctcagaagcacaaccagagtgagtatgaggattcggccggcgaatgctgctcctgtccaaagacagactctcagatcctgaaggagctggaggagtcctcgtttaggaagacgtttgaggattacctgcacaacgtggttttcgtccccagaaaaacctcttcaggcactggtgccgaggaccctaggccatctcggaaacgcaggtcccttggcgatgttgggaatgtgacggtggccgtgcccacggtggcagctttccccaacacttcctcgaccagcgtgcccacgagtccggaggagcacaggccttttgagaaggtggtgaacaaggagtcgctggtcatctccggcttgcgacacttcacgggctatcgcatcgagctgcaggcttgcaaccaggacacccctgaggaacggtgcagtgtggcagcctacgtcagtgcgaggaccatgcctgaagccaaggctgatgacattgttggccctgtgacgcatgaaatctttgagaacaacgtcgtccacttgatgtggcaggagccgaaggagcccaatggtctgatcgtgctgtatgaagtgagttatcggcgatatggtgatgaggagctgcatctctgcgtctcccgcaagcacttcgctctggaacggggctgcaggctgcgtgggctgtcaccggggaactacagcgtgcgaatccgggccacctcccttgcgggcaacggctcttggacggaacccacctatttctacgtgacagactatttagacgtcccgtcaaatattgcaaaaattatcatcggccccctcatctttgtctttctcttcagtgttgtgattggaagtatttatctattcctgagaaagaggcagccagatgggccgctgggaccgctttacgcttcttcaaaccctgagtatctcagtgccagtgatgtgtttccatgctctgtgtacgtgccggacgagtgggaggtgtctcgagagaagatcaccctccttcgagagctggggcagggctccttcggcatggtgtatgagggcaatgccagggacatcatcaagggtgaggcagagacccgcgtggcggtgaagacggtcaacgagtcagccagtctccgagagcggattgagttcctcaatgaggcctcggtcatgaagggcttcacctgccatcacgtggtgcgcctcctgggagtggtgtccaagggccagcccacgctggtggtgatggagctgatggctcacggagacctgaagagctacctccgttctctgcggccagaggctgagaataatcctggccgccctccccctacccttcaagagatgattcagatggcggcagagattgctgacgggatggcctacctgaacgccaagaagtttgtgcatcgggacctggcagcgagaaactgcatggtcgcccatgattttactgtcaaaattggagactttggaatgaccagagacatctatgaaacggattactaccggaaagggggcaagggtctgctccctgtacggtggatggcaccggagtccctgaaggatggggtcttcaccacttcttctgacatgtggtcctttggcgtggtcctttgggaaatcaccagcttggcagaacagccttaccaaggcctgtctaatgaacaggtgttgaaatttgtcatggatggagggtatctggatcaacccgacaactgtccagagagagtcactgacctcatgcgcatgtgctggcaattcaaccccaagatgaggccaaccttcctggagattgtcaacctgctcaaggacgacctgcaccccagctttccagaggtgtcgttcttccacagcgaggagaacaaggctcccgagagtgaggagctggagatggagtttgaggacatggagaatgtgcccctggaccgttcctcgcactgtcagagggaggaggcggggggccgggatggagggtcctcgctgggtttcaagcggagctacgaggaacacatcccttacacacacatgaacggaggcaagaaaaacgggcggattctgaccttgcctcggtccaatccttcctaa'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Salvare la coding sequence in formato `FASTA`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Produrre la sequenza della CDS in formato `FASTA` (separata in *record* di 60bp) con il seguente *header*:\n",
    "\n",
    "    >M10051-HUM; len = [length]; start = [yes|no]; end = [yes|no]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'>M10051-HUM; length = 4149; start = yes; stop = yes'"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header = '>' + identifier + '-' + organism + '; length = ' + str(len(cds_sequence))\n",
    "\n",
    "exist_codon_start = 'no'\n",
    "exist_codon_stop = 'no'\n",
    "\n",
    "if cds_sequence[:3] == 'atg':\n",
    "    exist_codon_start = 'yes'\n",
    "    \n",
    "if cds_sequence[-3:] in ['taa', 'tag', 'tga']:\n",
    "    exist_codon_stop = 'yes'\n",
    "\n",
    "header = header + '; start = ' + exist_codon_start + ';'\n",
    "header = header + ' stop = ' + exist_codon_stop\n",
    "header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_sequence_fasta = format_fasta(header, cds_sequence, 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">M10051-HUM; length = 4149; start = yes; stop = yes\n",
      "atgggcaccgggggccggcggggggcggcggccgcgccgctgctggtggcggtggccgcg\n",
      "ctgctactgggcgccgcgggccacctgtaccccggagaggtgtgtcccggcatggatatc\n",
      "cggaacaacctcactaggttgcatgagctggagaattgctctgtcatcgaaggacacttg\n",
      "cagatactcttgatgttcaaaacgaggcccgaagatttccgagacctcagtttccccaaa\n",
      "ctcatcatgatcactgattacttgctgctcttccgggtctatgggctcgagagcctgaag\n",
      "gacctgttccccaacctcacggtcatccggggatcacgactgttctttaactacgcgctg\n",
      "gtcatcttcgagatggttcacctcaaggaactcggcctctacaacctgatgaacatcacc\n",
      "cggggttctgtccgcatcgagaagaacaatgagctctgttacttggccactatcgactgg\n",
      "tcccgtatcctggattccgtggaggataatcacatcgtgttgaacaaagatgacaacgag\n",
      "gagtgtggagacatctgtccgggtaccgcgaagggcaagaccaactgccccgccaccgtc\n",
      "atcaacgggcagtttgtcgaacgatgttggactcatagtcactgccagaaagtttgcccg\n",
      "accatctgtaagtcacacggctgcaccgccgaaggcctctgttgccacagcgagtgcctg\n",
      "ggcaactgttctcagcccgacgaccccaccaagtgcgtggcctgccgcaacttctacctg\n",
      "gacggcaggtgtgtggagacctgcccgcccccgtactaccacttccaggactggcgctgt\n",
      "gtgaacttcagcttctgccaggacctgcaccacaaatgcaagaactcgcggaggcagggc\n",
      "tgccaccaatacgtcattcacaacaacaagtgcatccctgagtgtccctccgggtacacg\n",
      "atgaattccagcaacttgctgtgcaccccatgcctgggtccctgtcccaaggtgtgccac\n",
      "ctcctagaaggcgagaagaccatcgactcggtgacgtctgcccaggagctccgaggatgc\n",
      "accgtcatcaacgggagtctgatcatcaacattcgaggaggcaacaatctggcagctgag\n",
      "ctagaagccaacctcggcctcattgaagaaatttcagggtatctaaaaatccgccgatcc\n",
      "tacgctctggtgtcactttccttcttccggaagttacgtctgattcgaggagagaccttg\n",
      "gaaattgggaactactccttctatgccttggacaaccagaacctaaggcagctctgggac\n",
      "tggagcaaacacaacctcaccaccactcaggggaaactcttcttccactataaccccaaa\n",
      "ctctgcttgtcagaaatccacaagatggaagaagtttcaggaaccaaggggcgccaggag\n",
      "agaaacgacattgccctgaagaccaatggggacaaggcatcctgtgaaaatgagttactt\n",
      "aaattttcttacattcggacatcttttgacaagatcttgctgagatgggagccgtactgg\n",
      "ccccccgacttccgagacctcttggggttcatgctgttctacaaagaggccccttatcag\n",
      "aatgtgacggagttcgatgggcaggatgcgtgtggttccaacagttggacggtggtagac\n",
      "attgacccacccctgaggtccaacgaccccaaatcacagaaccacccagggtggctgatg\n",
      "cggggtctcaagccctggacccagtatgccatctttgtgaagaccctggtcaccttttcg\n",
      "gatgaacgccggacctatggggccaagagtgacatcatttatgtccagacagatgccacc\n",
      "aacccctctgtgcccctggatccaatctcagtgtctaactcatcatcccagattattctg\n",
      "aagtggaaaccaccctccgaccccaatggcaacatcacccactacctggttttctgggag\n",
      "aggcaggcggaagacagtgagctgttcgagctggattattgcctcaaagggctgaagctg\n",
      "ccctcgaggacctggtctccaccattcgagtctgaagattctcagaagcacaaccagagt\n",
      "gagtatgaggattcggccggcgaatgctgctcctgtccaaagacagactctcagatcctg\n",
      "aaggagctggaggagtcctcgtttaggaagacgtttgaggattacctgcacaacgtggtt\n",
      "ttcgtccccagaaaaacctcttcaggcactggtgccgaggaccctaggccatctcggaaa\n",
      "cgcaggtcccttggcgatgttgggaatgtgacggtggccgtgcccacggtggcagctttc\n",
      "cccaacacttcctcgaccagcgtgcccacgagtccggaggagcacaggccttttgagaag\n",
      "gtggtgaacaaggagtcgctggtcatctccggcttgcgacacttcacgggctatcgcatc\n",
      "gagctgcaggcttgcaaccaggacacccctgaggaacggtgcagtgtggcagcctacgtc\n",
      "agtgcgaggaccatgcctgaagccaaggctgatgacattgttggccctgtgacgcatgaa\n",
      "atctttgagaacaacgtcgtccacttgatgtggcaggagccgaaggagcccaatggtctg\n",
      "atcgtgctgtatgaagtgagttatcggcgatatggtgatgaggagctgcatctctgcgtc\n",
      "tcccgcaagcacttcgctctggaacggggctgcaggctgcgtgggctgtcaccggggaac\n",
      "tacagcgtgcgaatccgggccacctcccttgcgggcaacggctcttggacggaacccacc\n",
      "tatttctacgtgacagactatttagacgtcccgtcaaatattgcaaaaattatcatcggc\n",
      "cccctcatctttgtctttctcttcagtgttgtgattggaagtatttatctattcctgaga\n",
      "aagaggcagccagatgggccgctgggaccgctttacgcttcttcaaaccctgagtatctc\n",
      "agtgccagtgatgtgtttccatgctctgtgtacgtgccggacgagtgggaggtgtctcga\n",
      "gagaagatcaccctccttcgagagctggggcagggctccttcggcatggtgtatgagggc\n",
      "aatgccagggacatcatcaagggtgaggcagagacccgcgtggcggtgaagacggtcaac\n",
      "gagtcagccagtctccgagagcggattgagttcctcaatgaggcctcggtcatgaagggc\n",
      "ttcacctgccatcacgtggtgcgcctcctgggagtggtgtccaagggccagcccacgctg\n",
      "gtggtgatggagctgatggctcacggagacctgaagagctacctccgttctctgcggcca\n",
      "gaggctgagaataatcctggccgccctccccctacccttcaagagatgattcagatggcg\n",
      "gcagagattgctgacgggatggcctacctgaacgccaagaagtttgtgcatcgggacctg\n",
      "gcagcgagaaactgcatggtcgcccatgattttactgtcaaaattggagactttggaatg\n",
      "accagagacatctatgaaacggattactaccggaaagggggcaagggtctgctccctgta\n",
      "cggtggatggcaccggagtccctgaaggatggggtcttcaccacttcttctgacatgtgg\n",
      "tcctttggcgtggtcctttgggaaatcaccagcttggcagaacagccttaccaaggcctg\n",
      "tctaatgaacaggtgttgaaatttgtcatggatggagggtatctggatcaacccgacaac\n",
      "tgtccagagagagtcactgacctcatgcgcatgtgctggcaattcaaccccaagatgagg\n",
      "ccaaccttcctggagattgtcaacctgctcaaggacgacctgcaccccagctttccagag\n",
      "gtgtcgttcttccacagcgaggagaacaaggctcccgagagtgaggagctggagatggag\n",
      "tttgaggacatggagaatgtgcccctggaccgttcctcgcactgtcagagggaggaggcg\n",
      "gggggccgggatggagggtcctcgctgggtttcaagcggagctacgaggaacacatccct\n",
      "tacacacacatgaacggaggcaagaaaaacgggcggattctgaccttgcctcggtccaat\n",
      "ccttcctaa\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(cds_sequence_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Salvare la coding sequence in un file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./cds.fa', 'w') as output_file:\n",
    "    output_file.write(cds_sequence_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determinare le frequenze dei codoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Estrarre la lista dei codoni della CDS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativa1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_list = [cds_sequence[i:i+3] for i in range(0, len(cds_sequence), 3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativa2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_list = re.findall(r'.{3}', cds_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['atg',\n",
       " 'ggc',\n",
       " 'acc',\n",
       " 'ggg',\n",
       " 'ggc',\n",
       " 'cgg',\n",
       " 'cgg',\n",
       " 'ggg',\n",
       " 'gcg',\n",
       " 'gcg',\n",
       " 'gcc',\n",
       " 'gcg',\n",
       " 'ccg',\n",
       " 'ctg',\n",
       " 'ctg',\n",
       " 'gtg',\n",
       " 'gcg',\n",
       " 'gtg',\n",
       " 'gcc',\n",
       " 'gcg',\n",
       " 'ctg',\n",
       " 'cta',\n",
       " 'ctg',\n",
       " 'ggc',\n",
       " 'gcc',\n",
       " 'gcg',\n",
       " 'ggc',\n",
       " 'cac',\n",
       " 'ctg',\n",
       " 'tac',\n",
       " 'ccc',\n",
       " 'gga',\n",
       " 'gag',\n",
       " 'gtg',\n",
       " 'tgt',\n",
       " 'ccc',\n",
       " 'ggc',\n",
       " 'atg',\n",
       " 'gat',\n",
       " 'atc',\n",
       " 'cgg',\n",
       " 'aac',\n",
       " 'aac',\n",
       " 'ctc',\n",
       " 'act',\n",
       " 'agg',\n",
       " 'ttg',\n",
       " 'cat',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'gag',\n",
       " 'aat',\n",
       " 'tgc',\n",
       " 'tct',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'gaa',\n",
       " 'gga',\n",
       " 'cac',\n",
       " 'ttg',\n",
       " 'cag',\n",
       " 'ata',\n",
       " 'ctc',\n",
       " 'ttg',\n",
       " 'atg',\n",
       " 'ttc',\n",
       " 'aaa',\n",
       " 'acg',\n",
       " 'agg',\n",
       " 'ccc',\n",
       " 'gaa',\n",
       " 'gat',\n",
       " 'ttc',\n",
       " 'cga',\n",
       " 'gac',\n",
       " 'ctc',\n",
       " 'agt',\n",
       " 'ttc',\n",
       " 'ccc',\n",
       " 'aaa',\n",
       " 'ctc',\n",
       " 'atc',\n",
       " 'atg',\n",
       " 'atc',\n",
       " 'act',\n",
       " 'gat',\n",
       " 'tac',\n",
       " 'ttg',\n",
       " 'ctg',\n",
       " 'ctc',\n",
       " 'ttc',\n",
       " 'cgg',\n",
       " 'gtc',\n",
       " 'tat',\n",
       " 'ggg',\n",
       " 'ctc',\n",
       " 'gag',\n",
       " 'agc',\n",
       " 'ctg',\n",
       " 'aag',\n",
       " 'gac',\n",
       " 'ctg',\n",
       " 'ttc',\n",
       " 'ccc',\n",
       " 'aac',\n",
       " 'ctc',\n",
       " 'acg',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'cgg',\n",
       " 'gga',\n",
       " 'tca',\n",
       " 'cga',\n",
       " 'ctg',\n",
       " 'ttc',\n",
       " 'ttt',\n",
       " 'aac',\n",
       " 'tac',\n",
       " 'gcg',\n",
       " 'ctg',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'ttc',\n",
       " 'gag',\n",
       " 'atg',\n",
       " 'gtt',\n",
       " 'cac',\n",
       " 'ctc',\n",
       " 'aag',\n",
       " 'gaa',\n",
       " 'ctc',\n",
       " 'ggc',\n",
       " 'ctc',\n",
       " 'tac',\n",
       " 'aac',\n",
       " 'ctg',\n",
       " 'atg',\n",
       " 'aac',\n",
       " 'atc',\n",
       " 'acc',\n",
       " 'cgg',\n",
       " 'ggt',\n",
       " 'tct',\n",
       " 'gtc',\n",
       " 'cgc',\n",
       " 'atc',\n",
       " 'gag',\n",
       " 'aag',\n",
       " 'aac',\n",
       " 'aat',\n",
       " 'gag',\n",
       " 'ctc',\n",
       " 'tgt',\n",
       " 'tac',\n",
       " 'ttg',\n",
       " 'gcc',\n",
       " 'act',\n",
       " 'atc',\n",
       " 'gac',\n",
       " 'tgg',\n",
       " 'tcc',\n",
       " 'cgt',\n",
       " 'atc',\n",
       " 'ctg',\n",
       " 'gat',\n",
       " 'tcc',\n",
       " 'gtg',\n",
       " 'gag',\n",
       " 'gat',\n",
       " 'aat',\n",
       " 'cac',\n",
       " 'atc',\n",
       " 'gtg',\n",
       " 'ttg',\n",
       " 'aac',\n",
       " 'aaa',\n",
       " 'gat',\n",
       " 'gac',\n",
       " 'aac',\n",
       " 'gag',\n",
       " 'gag',\n",
       " 'tgt',\n",
       " 'gga',\n",
       " 'gac',\n",
       " 'atc',\n",
       " 'tgt',\n",
       " 'ccg',\n",
       " 'ggt',\n",
       " 'acc',\n",
       " 'gcg',\n",
       " 'aag',\n",
       " 'ggc',\n",
       " 'aag',\n",
       " 'acc',\n",
       " 'aac',\n",
       " 'tgc',\n",
       " 'ccc',\n",
       " 'gcc',\n",
       " 'acc',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'aac',\n",
       " 'ggg',\n",
       " 'cag',\n",
       " 'ttt',\n",
       " 'gtc',\n",
       " 'gaa',\n",
       " 'cga',\n",
       " 'tgt',\n",
       " 'tgg',\n",
       " 'act',\n",
       " 'cat',\n",
       " 'agt',\n",
       " 'cac',\n",
       " 'tgc',\n",
       " 'cag',\n",
       " 'aaa',\n",
       " 'gtt',\n",
       " 'tgc',\n",
       " 'ccg',\n",
       " 'acc',\n",
       " 'atc',\n",
       " 'tgt',\n",
       " 'aag',\n",
       " 'tca',\n",
       " 'cac',\n",
       " 'ggc',\n",
       " 'tgc',\n",
       " 'acc',\n",
       " 'gcc',\n",
       " 'gaa',\n",
       " 'ggc',\n",
       " 'ctc',\n",
       " 'tgt',\n",
       " 'tgc',\n",
       " 'cac',\n",
       " 'agc',\n",
       " 'gag',\n",
       " 'tgc',\n",
       " 'ctg',\n",
       " 'ggc',\n",
       " 'aac',\n",
       " 'tgt',\n",
       " 'tct',\n",
       " 'cag',\n",
       " 'ccc',\n",
       " 'gac',\n",
       " 'gac',\n",
       " 'ccc',\n",
       " 'acc',\n",
       " 'aag',\n",
       " 'tgc',\n",
       " 'gtg',\n",
       " 'gcc',\n",
       " 'tgc',\n",
       " 'cgc',\n",
       " 'aac',\n",
       " 'ttc',\n",
       " 'tac',\n",
       " 'ctg',\n",
       " 'gac',\n",
       " 'ggc',\n",
       " 'agg',\n",
       " 'tgt',\n",
       " 'gtg',\n",
       " 'gag',\n",
       " 'acc',\n",
       " 'tgc',\n",
       " 'ccg',\n",
       " 'ccc',\n",
       " 'ccg',\n",
       " 'tac',\n",
       " 'tac',\n",
       " 'cac',\n",
       " 'ttc',\n",
       " 'cag',\n",
       " 'gac',\n",
       " 'tgg',\n",
       " 'cgc',\n",
       " 'tgt',\n",
       " 'gtg',\n",
       " 'aac',\n",
       " 'ttc',\n",
       " 'agc',\n",
       " 'ttc',\n",
       " 'tgc',\n",
       " 'cag',\n",
       " 'gac',\n",
       " 'ctg',\n",
       " 'cac',\n",
       " 'cac',\n",
       " 'aaa',\n",
       " 'tgc',\n",
       " 'aag',\n",
       " 'aac',\n",
       " 'tcg',\n",
       " 'cgg',\n",
       " 'agg',\n",
       " 'cag',\n",
       " 'ggc',\n",
       " 'tgc',\n",
       " 'cac',\n",
       " 'caa',\n",
       " 'tac',\n",
       " 'gtc',\n",
       " 'att',\n",
       " 'cac',\n",
       " 'aac',\n",
       " 'aac',\n",
       " 'aag',\n",
       " 'tgc',\n",
       " 'atc',\n",
       " 'cct',\n",
       " 'gag',\n",
       " 'tgt',\n",
       " 'ccc',\n",
       " 'tcc',\n",
       " 'ggg',\n",
       " 'tac',\n",
       " 'acg',\n",
       " 'atg',\n",
       " 'aat',\n",
       " 'tcc',\n",
       " 'agc',\n",
       " 'aac',\n",
       " 'ttg',\n",
       " 'ctg',\n",
       " 'tgc',\n",
       " 'acc',\n",
       " 'cca',\n",
       " 'tgc',\n",
       " 'ctg',\n",
       " 'ggt',\n",
       " 'ccc',\n",
       " 'tgt',\n",
       " 'ccc',\n",
       " 'aag',\n",
       " 'gtg',\n",
       " 'tgc',\n",
       " 'cac',\n",
       " 'ctc',\n",
       " 'cta',\n",
       " 'gaa',\n",
       " 'ggc',\n",
       " 'gag',\n",
       " 'aag',\n",
       " 'acc',\n",
       " 'atc',\n",
       " 'gac',\n",
       " 'tcg',\n",
       " 'gtg',\n",
       " 'acg',\n",
       " 'tct',\n",
       " 'gcc',\n",
       " 'cag',\n",
       " 'gag',\n",
       " 'ctc',\n",
       " 'cga',\n",
       " 'gga',\n",
       " 'tgc',\n",
       " 'acc',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'aac',\n",
       " 'ggg',\n",
       " 'agt',\n",
       " 'ctg',\n",
       " 'atc',\n",
       " 'atc',\n",
       " 'aac',\n",
       " 'att',\n",
       " 'cga',\n",
       " 'gga',\n",
       " 'ggc',\n",
       " 'aac',\n",
       " 'aat',\n",
       " 'ctg',\n",
       " 'gca',\n",
       " 'gct',\n",
       " 'gag',\n",
       " 'cta',\n",
       " 'gaa',\n",
       " 'gcc',\n",
       " 'aac',\n",
       " 'ctc',\n",
       " 'ggc',\n",
       " 'ctc',\n",
       " 'att',\n",
       " 'gaa',\n",
       " 'gaa',\n",
       " 'att',\n",
       " 'tca',\n",
       " 'ggg',\n",
       " 'tat',\n",
       " 'cta',\n",
       " 'aaa',\n",
       " 'atc',\n",
       " 'cgc',\n",
       " 'cga',\n",
       " 'tcc',\n",
       " 'tac',\n",
       " 'gct',\n",
       " 'ctg',\n",
       " 'gtg',\n",
       " 'tca',\n",
       " 'ctt',\n",
       " 'tcc',\n",
       " 'ttc',\n",
       " 'ttc',\n",
       " 'cgg',\n",
       " 'aag',\n",
       " 'tta',\n",
       " 'cgt',\n",
       " 'ctg',\n",
       " 'att',\n",
       " 'cga',\n",
       " 'gga',\n",
       " 'gag',\n",
       " 'acc',\n",
       " 'ttg',\n",
       " 'gaa',\n",
       " 'att',\n",
       " 'ggg',\n",
       " 'aac',\n",
       " 'tac',\n",
       " 'tcc',\n",
       " 'ttc',\n",
       " 'tat',\n",
       " 'gcc',\n",
       " 'ttg',\n",
       " 'gac',\n",
       " 'aac',\n",
       " 'cag',\n",
       " 'aac',\n",
       " 'cta',\n",
       " 'agg',\n",
       " 'cag',\n",
       " 'ctc',\n",
       " 'tgg',\n",
       " 'gac',\n",
       " 'tgg',\n",
       " 'agc',\n",
       " 'aaa',\n",
       " 'cac',\n",
       " 'aac',\n",
       " 'ctc',\n",
       " 'acc',\n",
       " 'acc',\n",
       " 'act',\n",
       " 'cag',\n",
       " 'ggg',\n",
       " 'aaa',\n",
       " 'ctc',\n",
       " 'ttc',\n",
       " 'ttc',\n",
       " 'cac',\n",
       " 'tat',\n",
       " 'aac',\n",
       " 'ccc',\n",
       " 'aaa',\n",
       " 'ctc',\n",
       " 'tgc',\n",
       " 'ttg',\n",
       " 'tca',\n",
       " 'gaa',\n",
       " 'atc',\n",
       " 'cac',\n",
       " 'aag',\n",
       " 'atg',\n",
       " 'gaa',\n",
       " 'gaa',\n",
       " 'gtt',\n",
       " 'tca',\n",
       " 'gga',\n",
       " 'acc',\n",
       " 'aag',\n",
       " 'ggg',\n",
       " 'cgc',\n",
       " 'cag',\n",
       " 'gag',\n",
       " 'aga',\n",
       " 'aac',\n",
       " 'gac',\n",
       " 'att',\n",
       " 'gcc',\n",
       " 'ctg',\n",
       " 'aag',\n",
       " 'acc',\n",
       " 'aat',\n",
       " 'ggg',\n",
       " 'gac',\n",
       " 'aag',\n",
       " 'gca',\n",
       " 'tcc',\n",
       " 'tgt',\n",
       " 'gaa',\n",
       " 'aat',\n",
       " 'gag',\n",
       " 'tta',\n",
       " 'ctt',\n",
       " 'aaa',\n",
       " 'ttt',\n",
       " 'tct',\n",
       " 'tac',\n",
       " 'att',\n",
       " 'cgg',\n",
       " 'aca',\n",
       " 'tct',\n",
       " 'ttt',\n",
       " 'gac',\n",
       " 'aag',\n",
       " 'atc',\n",
       " 'ttg',\n",
       " 'ctg',\n",
       " 'aga',\n",
       " 'tgg',\n",
       " 'gag',\n",
       " 'ccg',\n",
       " 'tac',\n",
       " 'tgg',\n",
       " 'ccc',\n",
       " 'ccc',\n",
       " 'gac',\n",
       " 'ttc',\n",
       " 'cga',\n",
       " 'gac',\n",
       " 'ctc',\n",
       " 'ttg',\n",
       " 'ggg',\n",
       " 'ttc',\n",
       " 'atg',\n",
       " 'ctg',\n",
       " 'ttc',\n",
       " 'tac',\n",
       " 'aaa',\n",
       " 'gag',\n",
       " 'gcc',\n",
       " 'cct',\n",
       " 'tat',\n",
       " 'cag',\n",
       " 'aat',\n",
       " 'gtg',\n",
       " 'acg',\n",
       " 'gag',\n",
       " 'ttc',\n",
       " 'gat',\n",
       " 'ggg',\n",
       " 'cag',\n",
       " 'gat',\n",
       " 'gcg',\n",
       " 'tgt',\n",
       " 'ggt',\n",
       " 'tcc',\n",
       " 'aac',\n",
       " 'agt',\n",
       " 'tgg',\n",
       " 'acg',\n",
       " 'gtg',\n",
       " 'gta',\n",
       " 'gac',\n",
       " 'att',\n",
       " 'gac',\n",
       " 'cca',\n",
       " 'ccc',\n",
       " 'ctg',\n",
       " 'agg',\n",
       " 'tcc',\n",
       " 'aac',\n",
       " 'gac',\n",
       " 'ccc',\n",
       " 'aaa',\n",
       " 'tca',\n",
       " 'cag',\n",
       " 'aac',\n",
       " 'cac',\n",
       " 'cca',\n",
       " 'ggg',\n",
       " 'tgg',\n",
       " 'ctg',\n",
       " 'atg',\n",
       " 'cgg',\n",
       " 'ggt',\n",
       " 'ctc',\n",
       " 'aag',\n",
       " 'ccc',\n",
       " 'tgg',\n",
       " 'acc',\n",
       " 'cag',\n",
       " 'tat',\n",
       " 'gcc',\n",
       " 'atc',\n",
       " 'ttt',\n",
       " 'gtg',\n",
       " 'aag',\n",
       " 'acc',\n",
       " 'ctg',\n",
       " 'gtc',\n",
       " 'acc',\n",
       " 'ttt',\n",
       " 'tcg',\n",
       " 'gat',\n",
       " 'gaa',\n",
       " 'cgc',\n",
       " 'cgg',\n",
       " 'acc',\n",
       " 'tat',\n",
       " 'ggg',\n",
       " 'gcc',\n",
       " 'aag',\n",
       " 'agt',\n",
       " 'gac',\n",
       " 'atc',\n",
       " 'att',\n",
       " 'tat',\n",
       " 'gtc',\n",
       " 'cag',\n",
       " 'aca',\n",
       " 'gat',\n",
       " 'gcc',\n",
       " 'acc',\n",
       " 'aac',\n",
       " 'ccc',\n",
       " 'tct',\n",
       " 'gtg',\n",
       " 'ccc',\n",
       " 'ctg',\n",
       " 'gat',\n",
       " 'cca',\n",
       " 'atc',\n",
       " 'tca',\n",
       " 'gtg',\n",
       " 'tct',\n",
       " 'aac',\n",
       " 'tca',\n",
       " 'tca',\n",
       " 'tcc',\n",
       " 'cag',\n",
       " 'att',\n",
       " 'att',\n",
       " 'ctg',\n",
       " 'aag',\n",
       " 'tgg',\n",
       " 'aaa',\n",
       " 'cca',\n",
       " 'ccc',\n",
       " 'tcc',\n",
       " 'gac',\n",
       " 'ccc',\n",
       " 'aat',\n",
       " 'ggc',\n",
       " 'aac',\n",
       " 'atc',\n",
       " 'acc',\n",
       " 'cac',\n",
       " 'tac',\n",
       " 'ctg',\n",
       " 'gtt',\n",
       " 'ttc',\n",
       " 'tgg',\n",
       " 'gag',\n",
       " 'agg',\n",
       " 'cag',\n",
       " 'gcg',\n",
       " 'gaa',\n",
       " 'gac',\n",
       " 'agt',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'ttc',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'gat',\n",
       " 'tat',\n",
       " 'tgc',\n",
       " 'ctc',\n",
       " 'aaa',\n",
       " 'ggg',\n",
       " 'ctg',\n",
       " 'aag',\n",
       " 'ctg',\n",
       " 'ccc',\n",
       " 'tcg',\n",
       " 'agg',\n",
       " 'acc',\n",
       " 'tgg',\n",
       " 'tct',\n",
       " 'cca',\n",
       " 'cca',\n",
       " 'ttc',\n",
       " 'gag',\n",
       " 'tct',\n",
       " 'gaa',\n",
       " 'gat',\n",
       " 'tct',\n",
       " 'cag',\n",
       " 'aag',\n",
       " 'cac',\n",
       " 'aac',\n",
       " 'cag',\n",
       " 'agt',\n",
       " 'gag',\n",
       " 'tat',\n",
       " 'gag',\n",
       " 'gat',\n",
       " 'tcg',\n",
       " 'gcc',\n",
       " 'ggc',\n",
       " 'gaa',\n",
       " 'tgc',\n",
       " 'tgc',\n",
       " 'tcc',\n",
       " 'tgt',\n",
       " 'cca',\n",
       " 'aag',\n",
       " 'aca',\n",
       " 'gac',\n",
       " 'tct',\n",
       " 'cag',\n",
       " 'atc',\n",
       " 'ctg',\n",
       " 'aag',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'gag',\n",
       " 'gag',\n",
       " 'tcc',\n",
       " 'tcg',\n",
       " 'ttt',\n",
       " 'agg',\n",
       " 'aag',\n",
       " 'acg',\n",
       " 'ttt',\n",
       " 'gag',\n",
       " 'gat',\n",
       " 'tac',\n",
       " 'ctg',\n",
       " 'cac',\n",
       " 'aac',\n",
       " 'gtg',\n",
       " 'gtt',\n",
       " 'ttc',\n",
       " 'gtc',\n",
       " 'ccc',\n",
       " 'aga',\n",
       " 'aaa',\n",
       " 'acc',\n",
       " 'tct',\n",
       " 'tca',\n",
       " 'ggc',\n",
       " 'act',\n",
       " 'ggt',\n",
       " 'gcc',\n",
       " 'gag',\n",
       " 'gac',\n",
       " 'cct',\n",
       " 'agg',\n",
       " 'cca',\n",
       " 'tct',\n",
       " 'cgg',\n",
       " 'aaa',\n",
       " 'cgc',\n",
       " 'agg',\n",
       " 'tcc',\n",
       " 'ctt',\n",
       " 'ggc',\n",
       " 'gat',\n",
       " 'gtt',\n",
       " 'ggg',\n",
       " 'aat',\n",
       " 'gtg',\n",
       " 'acg',\n",
       " 'gtg',\n",
       " 'gcc',\n",
       " 'gtg',\n",
       " 'ccc',\n",
       " 'acg',\n",
       " 'gtg',\n",
       " 'gca',\n",
       " 'gct',\n",
       " 'ttc',\n",
       " 'ccc',\n",
       " 'aac',\n",
       " 'act',\n",
       " 'tcc',\n",
       " 'tcg',\n",
       " 'acc',\n",
       " 'agc',\n",
       " 'gtg',\n",
       " 'ccc',\n",
       " 'acg',\n",
       " 'agt',\n",
       " 'ccg',\n",
       " 'gag',\n",
       " 'gag',\n",
       " 'cac',\n",
       " 'agg',\n",
       " 'cct',\n",
       " 'ttt',\n",
       " 'gag',\n",
       " 'aag',\n",
       " 'gtg',\n",
       " 'gtg',\n",
       " 'aac',\n",
       " 'aag',\n",
       " 'gag',\n",
       " 'tcg',\n",
       " 'ctg',\n",
       " 'gtc',\n",
       " 'atc',\n",
       " 'tcc',\n",
       " 'ggc',\n",
       " 'ttg',\n",
       " 'cga',\n",
       " 'cac',\n",
       " 'ttc',\n",
       " 'acg',\n",
       " 'ggc',\n",
       " 'tat',\n",
       " 'cgc',\n",
       " 'atc',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'cag',\n",
       " 'gct',\n",
       " 'tgc',\n",
       " 'aac',\n",
       " 'cag',\n",
       " 'gac',\n",
       " 'acc',\n",
       " 'cct',\n",
       " 'gag',\n",
       " 'gaa',\n",
       " 'cgg',\n",
       " 'tgc',\n",
       " 'agt',\n",
       " 'gtg',\n",
       " 'gca',\n",
       " 'gcc',\n",
       " 'tac',\n",
       " 'gtc',\n",
       " 'agt',\n",
       " 'gcg',\n",
       " 'agg',\n",
       " 'acc',\n",
       " 'atg',\n",
       " 'cct',\n",
       " 'gaa',\n",
       " 'gcc',\n",
       " 'aag',\n",
       " 'gct',\n",
       " 'gat',\n",
       " 'gac',\n",
       " 'att',\n",
       " 'gtt',\n",
       " 'ggc',\n",
       " 'cct',\n",
       " 'gtg',\n",
       " 'acg',\n",
       " 'cat',\n",
       " 'gaa',\n",
       " 'atc',\n",
       " 'ttt',\n",
       " 'gag',\n",
       " 'aac',\n",
       " 'aac',\n",
       " 'gtc',\n",
       " 'gtc',\n",
       " 'cac',\n",
       " 'ttg',\n",
       " 'atg',\n",
       " 'tgg',\n",
       " 'cag',\n",
       " 'gag',\n",
       " 'ccg',\n",
       " 'aag',\n",
       " 'gag',\n",
       " 'ccc',\n",
       " 'aat',\n",
       " 'ggt',\n",
       " 'ctg',\n",
       " 'atc',\n",
       " 'gtg',\n",
       " 'ctg',\n",
       " 'tat',\n",
       " 'gaa',\n",
       " 'gtg',\n",
       " 'agt',\n",
       " 'tat',\n",
       " 'cgg',\n",
       " 'cga',\n",
       " 'tat',\n",
       " 'ggt',\n",
       " 'gat',\n",
       " 'gag',\n",
       " 'gag',\n",
       " 'ctg',\n",
       " 'cat',\n",
       " 'ctc',\n",
       " 'tgc',\n",
       " 'gtc',\n",
       " 'tcc',\n",
       " 'cgc',\n",
       " 'aag',\n",
       " 'cac',\n",
       " 'ttc',\n",
       " 'gct',\n",
       " 'ctg',\n",
       " 'gaa',\n",
       " 'cgg',\n",
       " 'ggc',\n",
       " 'tgc',\n",
       " 'agg',\n",
       " 'ctg',\n",
       " 'cgt',\n",
       " 'ggg',\n",
       " 'ctg',\n",
       " 'tca',\n",
       " 'ccg',\n",
       " 'ggg',\n",
       " 'aac',\n",
       " 'tac',\n",
       " 'agc',\n",
       " 'gtg',\n",
       " 'cga',\n",
       " 'atc',\n",
       " 'cgg',\n",
       " 'gcc',\n",
       " 'acc',\n",
       " 'tcc',\n",
       " 'ctt',\n",
       " 'gcg',\n",
       " 'ggc',\n",
       " 'aac',\n",
       " 'ggc',\n",
       " 'tct',\n",
       " 'tgg',\n",
       " 'acg',\n",
       " 'gaa',\n",
       " 'ccc',\n",
       " 'acc',\n",
       " 'tat',\n",
       " 'ttc',\n",
       " 'tac',\n",
       " 'gtg',\n",
       " 'aca',\n",
       " 'gac',\n",
       " 'tat',\n",
       " 'tta',\n",
       " 'gac',\n",
       " 'gtc',\n",
       " 'ccg',\n",
       " 'tca',\n",
       " 'aat',\n",
       " 'att',\n",
       " 'gca',\n",
       " 'aaa',\n",
       " 'att',\n",
       " 'atc',\n",
       " 'atc',\n",
       " 'ggc',\n",
       " 'ccc',\n",
       " 'ctc',\n",
       " 'atc',\n",
       " 'ttt',\n",
       " 'gtc',\n",
       " 'ttt',\n",
       " 'ctc',\n",
       " 'ttc',\n",
       " 'agt',\n",
       " 'gtt',\n",
       " 'gtg',\n",
       " 'att',\n",
       " 'gga',\n",
       " 'agt',\n",
       " 'att',\n",
       " 'tat',\n",
       " 'cta',\n",
       " 'ttc',\n",
       " 'ctg',\n",
       " 'aga',\n",
       " 'aag',\n",
       " 'agg',\n",
       " 'cag',\n",
       " 'cca',\n",
       " 'gat',\n",
       " 'ggg',\n",
       " 'ccg',\n",
       " 'ctg',\n",
       " 'gga',\n",
       " 'ccg',\n",
       " 'ctt',\n",
       " 'tac',\n",
       " 'gct',\n",
       " 'tct',\n",
       " 'tca',\n",
       " 'aac',\n",
       " 'cct',\n",
       " 'gag',\n",
       " 'tat',\n",
       " 'ctc',\n",
       " ...]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Costruire la lista di tuple *(codone, frequenza)* elencate per frequenze decrescenti."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_counter = Counter(codon_list)\n",
    "codon_frequencies = codon_counter.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('gag', 77),\n",
       " ('ctg', 67),\n",
       " ('aac', 53),\n",
       " ('gtg', 49),\n",
       " ('aag', 47),\n",
       " ('gac', 44),\n",
       " ('atc', 41),\n",
       " ('acc', 39),\n",
       " ('ttc', 39),\n",
       " ('ggc', 37),\n",
       " ('ccc', 37),\n",
       " ('ctc', 35),\n",
       " ('cag', 32),\n",
       " ('cac', 31),\n",
       " ('atg', 30),\n",
       " ('tgc', 30),\n",
       " ('tac', 29),\n",
       " ('gaa', 29),\n",
       " ('ggg', 28),\n",
       " ('gcc', 28),\n",
       " ('gtc', 28),\n",
       " ('tcc', 27),\n",
       " ('gat', 26),\n",
       " ('cgg', 25),\n",
       " ('att', 23),\n",
       " ('tct', 22),\n",
       " ('aaa', 21),\n",
       " ('tat', 21),\n",
       " ('ttt', 20),\n",
       " ('tgg', 20),\n",
       " ('aat', 19),\n",
       " ('agg', 18),\n",
       " ('gga', 17),\n",
       " ('tgt', 17),\n",
       " ('ttg', 17),\n",
       " ('agt', 17),\n",
       " ('gcg', 16),\n",
       " ('acg', 16),\n",
       " ('cct', 16),\n",
       " ('tca', 15),\n",
       " ('cca', 15),\n",
       " ('ccg', 14),\n",
       " ('cga', 14),\n",
       " ('cgc', 13),\n",
       " ('agc', 12),\n",
       " ('tcg', 12),\n",
       " ('ggt', 11),\n",
       " ('gct', 11),\n",
       " ('act', 10),\n",
       " ('gca', 10),\n",
       " ('gtt', 8),\n",
       " ('ctt', 8),\n",
       " ('cat', 7),\n",
       " ('aga', 7),\n",
       " ('cta', 6),\n",
       " ('cgt', 5),\n",
       " ('caa', 5),\n",
       " ('aca', 5),\n",
       " ('tta', 3),\n",
       " ('gta', 2),\n",
       " ('ata', 1),\n",
       " ('taa', 1)]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determinare le frequenze degli amminoacidi della proteina letta dall'*entry*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Estrarre la lista degli amminoacidi della proteina."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "ammino_list = list(protein_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M',\n",
       " 'G',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'R',\n",
       " 'R',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'P',\n",
       " 'L',\n",
       " 'L',\n",
       " 'V',\n",
       " 'A',\n",
       " 'V',\n",
       " 'A',\n",
       " 'A',\n",
       " 'L',\n",
       " 'L',\n",
       " 'L',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'H',\n",
       " 'L',\n",
       " 'Y',\n",
       " 'P',\n",
       " 'G',\n",
       " 'E',\n",
       " 'V',\n",
       " 'C',\n",
       " 'P',\n",
       " 'G',\n",
       " 'M',\n",
       " 'D',\n",
       " 'I',\n",
       " 'R',\n",
       " 'N',\n",
       " 'N',\n",
       " 'L',\n",
       " 'T',\n",
       " 'R',\n",
       " 'L',\n",
       " 'H',\n",
       " 'E',\n",
       " 'L',\n",
       " 'E',\n",
       " 'N',\n",
       " 'C',\n",
       " 'S',\n",
       " 'V',\n",
       " 'I',\n",
       " 'E',\n",
       " 'G',\n",
       " 'H',\n",
       " 'L',\n",
       " 'Q',\n",
       " 'I',\n",
       " 'L',\n",
       " 'L',\n",
       " 'M',\n",
       " 'F',\n",
       " 'K',\n",
       " 'T',\n",
       " 'R',\n",
       " 'P',\n",
       " 'E',\n",
       " 'D',\n",
       " 'F',\n",
       " 'R',\n",
       " 'D',\n",
       " 'L',\n",
       " 'S',\n",
       " 'F',\n",
       " 'P',\n",
       " 'K',\n",
       " 'L',\n",
       " 'I',\n",
       " 'M',\n",
       " 'I',\n",
       " 'T',\n",
       " 'D',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'L',\n",
       " 'L',\n",
       " 'F',\n",
       " 'R',\n",
       " 'V',\n",
       " 'Y',\n",
       " 'G',\n",
       " 'L',\n",
       " 'E',\n",
       " 'S',\n",
       " 'L',\n",
       " 'K',\n",
       " 'D',\n",
       " 'L',\n",
       " 'F',\n",
       " 'P',\n",
       " 'N',\n",
       " 'L',\n",
       " 'T',\n",
       " 'V',\n",
       " 'I',\n",
       " 'R',\n",
       " 'G',\n",
       " 'S',\n",
       " 'R',\n",
       " 'L',\n",
       " 'F',\n",
       " 'F',\n",
       " 'N',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'L',\n",
       " 'V',\n",
       " 'I',\n",
       " 'F',\n",
       " 'E',\n",
       " 'M',\n",
       " 'V',\n",
       " 'H',\n",
       " 'L',\n",
       " 'K',\n",
       " 'E',\n",
       " 'L',\n",
       " 'G',\n",
       " 'L',\n",
       " 'Y',\n",
       " 'N',\n",
       " 'L',\n",
       " 'M',\n",
       " 'N',\n",
       " 'I',\n",
       " 'T',\n",
       " 'R',\n",
       " 'G',\n",
       " 'S',\n",
       " 'V',\n",
       " 'R',\n",
       " 'I',\n",
       " 'E',\n",
       " 'K',\n",
       " 'N',\n",
       " 'N',\n",
       " 'E',\n",
       " 'L',\n",
       " 'C',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'A',\n",
       " 'T',\n",
       " 'I',\n",
       " 'D',\n",
       " 'W',\n",
       " 'S',\n",
       " 'R',\n",
       " 'I',\n",
       " 'L',\n",
       " 'D',\n",
       " 'S',\n",
       " 'V',\n",
       " 'E',\n",
       " 'D',\n",
       " 'N',\n",
       " 'H',\n",
       " 'I',\n",
       " 'V',\n",
       " 'L',\n",
       " 'N',\n",
       " 'K',\n",
       " 'D',\n",
       " 'D',\n",
       " 'N',\n",
       " 'E',\n",
       " 'E',\n",
       " 'C',\n",
       " 'G',\n",
       " 'D',\n",
       " 'I',\n",
       " 'C',\n",
       " 'P',\n",
       " 'G',\n",
       " 'T',\n",
       " 'A',\n",
       " 'K',\n",
       " 'G',\n",
       " 'K',\n",
       " 'T',\n",
       " 'N',\n",
       " 'C',\n",
       " 'P',\n",
       " 'A',\n",
       " 'T',\n",
       " 'V',\n",
       " 'I',\n",
       " 'N',\n",
       " 'G',\n",
       " 'Q',\n",
       " 'F',\n",
       " 'V',\n",
       " 'E',\n",
       " 'R',\n",
       " 'C',\n",
       " 'W',\n",
       " 'T',\n",
       " 'H',\n",
       " 'S',\n",
       " 'H',\n",
       " 'C',\n",
       " 'Q',\n",
       " 'K',\n",
       " 'V',\n",
       " 'C',\n",
       " 'P',\n",
       " 'T',\n",
       " 'I',\n",
       " 'C',\n",
       " 'K',\n",
       " 'S',\n",
       " 'H',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " 'A',\n",
       " 'E',\n",
       " 'G',\n",
       " 'L',\n",
       " 'C',\n",
       " 'C',\n",
       " 'H',\n",
       " 'S',\n",
       " 'E',\n",
       " 'C',\n",
       " 'L',\n",
       " 'G',\n",
       " 'N',\n",
       " 'C',\n",
       " 'S',\n",
       " 'Q',\n",
       " 'P',\n",
       " 'D',\n",
       " 'D',\n",
       " 'P',\n",
       " 'T',\n",
       " 'K',\n",
       " 'C',\n",
       " 'V',\n",
       " 'A',\n",
       " 'C',\n",
       " 'R',\n",
       " 'N',\n",
       " 'F',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'D',\n",
       " 'G',\n",
       " 'R',\n",
       " 'C',\n",
       " 'V',\n",
       " 'E',\n",
       " 'T',\n",
       " 'C',\n",
       " 'P',\n",
       " 'P',\n",
       " 'P',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'H',\n",
       " 'F',\n",
       " 'Q',\n",
       " 'D',\n",
       " 'W',\n",
       " 'R',\n",
       " 'C',\n",
       " 'V',\n",
       " 'N',\n",
       " 'F',\n",
       " 'S',\n",
       " 'F',\n",
       " 'C',\n",
       " 'Q',\n",
       " 'D',\n",
       " 'L',\n",
       " 'H',\n",
       " 'H',\n",
       " 'K',\n",
       " 'C',\n",
       " 'K',\n",
       " 'N',\n",
       " 'S',\n",
       " 'R',\n",
       " 'R',\n",
       " 'Q',\n",
       " 'G',\n",
       " 'C',\n",
       " 'H',\n",
       " 'Q',\n",
       " 'Y',\n",
       " 'V',\n",
       " 'I',\n",
       " 'H',\n",
       " 'N',\n",
       " 'N',\n",
       " 'K',\n",
       " 'C',\n",
       " 'I',\n",
       " 'P',\n",
       " 'E',\n",
       " 'C',\n",
       " 'P',\n",
       " 'S',\n",
       " 'G',\n",
       " 'Y',\n",
       " 'T',\n",
       " 'M',\n",
       " 'N',\n",
       " 'S',\n",
       " 'S',\n",
       " 'N',\n",
       " 'L',\n",
       " 'L',\n",
       " 'C',\n",
       " 'T',\n",
       " 'P',\n",
       " 'C',\n",
       " 'L',\n",
       " 'G',\n",
       " 'P',\n",
       " 'C',\n",
       " 'P',\n",
       " 'K',\n",
       " 'V',\n",
       " 'C',\n",
       " 'H',\n",
       " 'L',\n",
       " 'L',\n",
       " 'E',\n",
       " 'G',\n",
       " 'E',\n",
       " 'K',\n",
       " 'T',\n",
       " 'I',\n",
       " 'D',\n",
       " 'S',\n",
       " 'V',\n",
       " 'T',\n",
       " 'S',\n",
       " 'A',\n",
       " 'Q',\n",
       " 'E',\n",
       " 'L',\n",
       " 'R',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " 'V',\n",
       " 'I',\n",
       " 'N',\n",
       " 'G',\n",
       " 'S',\n",
       " 'L',\n",
       " 'I',\n",
       " 'I',\n",
       " 'N',\n",
       " 'I',\n",
       " 'R',\n",
       " 'G',\n",
       " 'G',\n",
       " 'N',\n",
       " 'N',\n",
       " 'L',\n",
       " 'A',\n",
       " 'A',\n",
       " 'E',\n",
       " 'L',\n",
       " 'E',\n",
       " 'A',\n",
       " 'N',\n",
       " 'L',\n",
       " 'G',\n",
       " 'L',\n",
       " 'I',\n",
       " 'E',\n",
       " 'E',\n",
       " 'I',\n",
       " 'S',\n",
       " 'G',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'K',\n",
       " 'I',\n",
       " 'R',\n",
       " 'R',\n",
       " 'S',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'L',\n",
       " 'V',\n",
       " 'S',\n",
       " 'L',\n",
       " 'S',\n",
       " 'F',\n",
       " 'F',\n",
       " 'R',\n",
       " 'K',\n",
       " 'L',\n",
       " 'R',\n",
       " 'L',\n",
       " 'I',\n",
       " 'R',\n",
       " 'G',\n",
       " 'E',\n",
       " 'T',\n",
       " 'L',\n",
       " 'E',\n",
       " 'I',\n",
       " 'G',\n",
       " 'N',\n",
       " 'Y',\n",
       " 'S',\n",
       " 'F',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'L',\n",
       " 'D',\n",
       " 'N',\n",
       " 'Q',\n",
       " 'N',\n",
       " 'L',\n",
       " 'R',\n",
       " 'Q',\n",
       " 'L',\n",
       " 'W',\n",
       " 'D',\n",
       " 'W',\n",
       " 'S',\n",
       " 'K',\n",
       " 'H',\n",
       " 'N',\n",
       " 'L',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'Q',\n",
       " 'G',\n",
       " 'K',\n",
       " 'L',\n",
       " 'F',\n",
       " 'F',\n",
       " 'H',\n",
       " 'Y',\n",
       " 'N',\n",
       " 'P',\n",
       " 'K',\n",
       " 'L',\n",
       " 'C',\n",
       " 'L',\n",
       " 'S',\n",
       " 'E',\n",
       " 'I',\n",
       " 'H',\n",
       " 'K',\n",
       " 'M',\n",
       " 'E',\n",
       " 'E',\n",
       " 'V',\n",
       " 'S',\n",
       " 'G',\n",
       " 'T',\n",
       " 'K',\n",
       " 'G',\n",
       " 'R',\n",
       " 'Q',\n",
       " 'E',\n",
       " 'R',\n",
       " 'N',\n",
       " 'D',\n",
       " 'I',\n",
       " 'A',\n",
       " 'L',\n",
       " 'K',\n",
       " 'T',\n",
       " 'N',\n",
       " 'G',\n",
       " 'D',\n",
       " 'K',\n",
       " 'A',\n",
       " 'S',\n",
       " 'C',\n",
       " 'E',\n",
       " 'N',\n",
       " 'E',\n",
       " 'L',\n",
       " 'L',\n",
       " 'K',\n",
       " 'F',\n",
       " 'S',\n",
       " 'Y',\n",
       " 'I',\n",
       " 'R',\n",
       " 'T',\n",
       " 'S',\n",
       " 'F',\n",
       " 'D',\n",
       " 'K',\n",
       " 'I',\n",
       " 'L',\n",
       " 'L',\n",
       " 'R',\n",
       " 'W',\n",
       " 'E',\n",
       " 'P',\n",
       " 'Y',\n",
       " 'W',\n",
       " 'P',\n",
       " 'P',\n",
       " 'D',\n",
       " 'F',\n",
       " 'R',\n",
       " 'D',\n",
       " 'L',\n",
       " 'L',\n",
       " 'G',\n",
       " 'F',\n",
       " 'M',\n",
       " 'L',\n",
       " 'F',\n",
       " 'Y',\n",
       " 'K',\n",
       " 'E',\n",
       " 'A',\n",
       " 'P',\n",
       " 'Y',\n",
       " 'Q',\n",
       " 'N',\n",
       " 'V',\n",
       " 'T',\n",
       " 'E',\n",
       " 'F',\n",
       " 'D',\n",
       " 'G',\n",
       " 'Q',\n",
       " 'D',\n",
       " 'A',\n",
       " 'C',\n",
       " 'G',\n",
       " 'S',\n",
       " 'N',\n",
       " 'S',\n",
       " 'W',\n",
       " 'T',\n",
       " 'V',\n",
       " 'V',\n",
       " 'D',\n",
       " 'I',\n",
       " 'D',\n",
       " 'P',\n",
       " 'P',\n",
       " 'L',\n",
       " 'R',\n",
       " 'S',\n",
       " 'N',\n",
       " 'D',\n",
       " 'P',\n",
       " 'K',\n",
       " 'S',\n",
       " 'Q',\n",
       " 'N',\n",
       " 'H',\n",
       " 'P',\n",
       " 'G',\n",
       " 'W',\n",
       " 'L',\n",
       " 'M',\n",
       " 'R',\n",
       " 'G',\n",
       " 'L',\n",
       " 'K',\n",
       " 'P',\n",
       " 'W',\n",
       " 'T',\n",
       " 'Q',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'I',\n",
       " 'F',\n",
       " 'V',\n",
       " 'K',\n",
       " 'T',\n",
       " 'L',\n",
       " 'V',\n",
       " 'T',\n",
       " 'F',\n",
       " 'S',\n",
       " 'D',\n",
       " 'E',\n",
       " 'R',\n",
       " 'R',\n",
       " 'T',\n",
       " 'Y',\n",
       " 'G',\n",
       " 'A',\n",
       " 'K',\n",
       " 'S',\n",
       " 'D',\n",
       " 'I',\n",
       " 'I',\n",
       " 'Y',\n",
       " 'V',\n",
       " 'Q',\n",
       " 'T',\n",
       " 'D',\n",
       " 'A',\n",
       " 'T',\n",
       " 'N',\n",
       " 'P',\n",
       " 'S',\n",
       " 'V',\n",
       " 'P',\n",
       " 'L',\n",
       " 'D',\n",
       " 'P',\n",
       " 'I',\n",
       " 'S',\n",
       " 'V',\n",
       " 'S',\n",
       " 'N',\n",
       " 'S',\n",
       " 'S',\n",
       " 'S',\n",
       " 'Q',\n",
       " 'I',\n",
       " 'I',\n",
       " 'L',\n",
       " 'K',\n",
       " 'W',\n",
       " 'K',\n",
       " 'P',\n",
       " 'P',\n",
       " 'S',\n",
       " 'D',\n",
       " 'P',\n",
       " 'N',\n",
       " 'G',\n",
       " 'N',\n",
       " 'I',\n",
       " 'T',\n",
       " 'H',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'V',\n",
       " 'F',\n",
       " 'W',\n",
       " 'E',\n",
       " 'R',\n",
       " 'Q',\n",
       " 'A',\n",
       " 'E',\n",
       " 'D',\n",
       " 'S',\n",
       " 'E',\n",
       " 'L',\n",
       " 'F',\n",
       " 'E',\n",
       " 'L',\n",
       " 'D',\n",
       " 'Y',\n",
       " 'C',\n",
       " 'L',\n",
       " 'K',\n",
       " 'G',\n",
       " 'L',\n",
       " 'K',\n",
       " 'L',\n",
       " 'P',\n",
       " 'S',\n",
       " 'R',\n",
       " 'T',\n",
       " 'W',\n",
       " 'S',\n",
       " 'P',\n",
       " 'P',\n",
       " 'F',\n",
       " 'E',\n",
       " 'S',\n",
       " 'E',\n",
       " 'D',\n",
       " 'S',\n",
       " 'Q',\n",
       " 'K',\n",
       " 'H',\n",
       " 'N',\n",
       " 'Q',\n",
       " 'S',\n",
       " 'E',\n",
       " 'Y',\n",
       " 'E',\n",
       " 'D',\n",
       " 'S',\n",
       " 'A',\n",
       " 'G',\n",
       " 'E',\n",
       " 'C',\n",
       " 'C',\n",
       " 'S',\n",
       " 'C',\n",
       " 'P',\n",
       " 'K',\n",
       " 'T',\n",
       " 'D',\n",
       " 'S',\n",
       " 'Q',\n",
       " 'I',\n",
       " 'L',\n",
       " 'K',\n",
       " 'E',\n",
       " 'L',\n",
       " 'E',\n",
       " 'E',\n",
       " 'S',\n",
       " 'S',\n",
       " 'F',\n",
       " 'R',\n",
       " 'K',\n",
       " 'T',\n",
       " 'F',\n",
       " 'E',\n",
       " 'D',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'H',\n",
       " 'N',\n",
       " 'V',\n",
       " 'V',\n",
       " 'F',\n",
       " 'V',\n",
       " 'P',\n",
       " 'R',\n",
       " 'K',\n",
       " 'T',\n",
       " 'S',\n",
       " 'S',\n",
       " 'G',\n",
       " 'T',\n",
       " 'G',\n",
       " 'A',\n",
       " 'E',\n",
       " 'D',\n",
       " 'P',\n",
       " 'R',\n",
       " 'P',\n",
       " 'S',\n",
       " 'R',\n",
       " 'K',\n",
       " 'R',\n",
       " 'R',\n",
       " 'S',\n",
       " 'L',\n",
       " 'G',\n",
       " 'D',\n",
       " 'V',\n",
       " 'G',\n",
       " 'N',\n",
       " 'V',\n",
       " 'T',\n",
       " 'V',\n",
       " 'A',\n",
       " 'V',\n",
       " 'P',\n",
       " 'T',\n",
       " 'V',\n",
       " 'A',\n",
       " 'A',\n",
       " 'F',\n",
       " 'P',\n",
       " 'N',\n",
       " 'T',\n",
       " 'S',\n",
       " 'S',\n",
       " 'T',\n",
       " 'S',\n",
       " 'V',\n",
       " 'P',\n",
       " 'T',\n",
       " 'S',\n",
       " 'P',\n",
       " 'E',\n",
       " 'E',\n",
       " 'H',\n",
       " 'R',\n",
       " 'P',\n",
       " 'F',\n",
       " 'E',\n",
       " 'K',\n",
       " 'V',\n",
       " 'V',\n",
       " 'N',\n",
       " 'K',\n",
       " 'E',\n",
       " 'S',\n",
       " 'L',\n",
       " 'V',\n",
       " 'I',\n",
       " 'S',\n",
       " 'G',\n",
       " 'L',\n",
       " 'R',\n",
       " 'H',\n",
       " 'F',\n",
       " 'T',\n",
       " 'G',\n",
       " 'Y',\n",
       " 'R',\n",
       " 'I',\n",
       " 'E',\n",
       " 'L',\n",
       " 'Q',\n",
       " 'A',\n",
       " 'C',\n",
       " 'N',\n",
       " 'Q',\n",
       " 'D',\n",
       " 'T',\n",
       " 'P',\n",
       " 'E',\n",
       " 'E',\n",
       " 'R',\n",
       " 'C',\n",
       " 'S',\n",
       " 'V',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'V',\n",
       " 'S',\n",
       " 'A',\n",
       " 'R',\n",
       " 'T',\n",
       " 'M',\n",
       " 'P',\n",
       " 'E',\n",
       " 'A',\n",
       " 'K',\n",
       " 'A',\n",
       " 'D',\n",
       " 'D',\n",
       " 'I',\n",
       " 'V',\n",
       " 'G',\n",
       " 'P',\n",
       " 'V',\n",
       " 'T',\n",
       " 'H',\n",
       " 'E',\n",
       " 'I',\n",
       " 'F',\n",
       " 'E',\n",
       " 'N',\n",
       " 'N',\n",
       " 'V',\n",
       " 'V',\n",
       " 'H',\n",
       " 'L',\n",
       " 'M',\n",
       " 'W',\n",
       " 'Q',\n",
       " 'E',\n",
       " 'P',\n",
       " 'K',\n",
       " 'E',\n",
       " 'P',\n",
       " 'N',\n",
       " 'G',\n",
       " 'L',\n",
       " 'I',\n",
       " 'V',\n",
       " 'L',\n",
       " 'Y',\n",
       " 'E',\n",
       " 'V',\n",
       " 'S',\n",
       " 'Y',\n",
       " 'R',\n",
       " 'R',\n",
       " 'Y',\n",
       " 'G',\n",
       " 'D',\n",
       " 'E',\n",
       " 'E',\n",
       " 'L',\n",
       " 'H',\n",
       " 'L',\n",
       " 'C',\n",
       " 'V',\n",
       " 'S',\n",
       " 'R',\n",
       " 'K',\n",
       " 'H',\n",
       " 'F',\n",
       " 'A',\n",
       " 'L',\n",
       " 'E',\n",
       " 'R',\n",
       " 'G',\n",
       " 'C',\n",
       " 'R',\n",
       " 'L',\n",
       " 'R',\n",
       " 'G',\n",
       " 'L',\n",
       " 'S',\n",
       " 'P',\n",
       " 'G',\n",
       " 'N',\n",
       " 'Y',\n",
       " 'S',\n",
       " 'V',\n",
       " 'R',\n",
       " 'I',\n",
       " 'R',\n",
       " 'A',\n",
       " 'T',\n",
       " 'S',\n",
       " 'L',\n",
       " 'A',\n",
       " 'G',\n",
       " 'N',\n",
       " 'G',\n",
       " 'S',\n",
       " 'W',\n",
       " 'T',\n",
       " 'E',\n",
       " 'P',\n",
       " 'T',\n",
       " 'Y',\n",
       " 'F',\n",
       " 'Y',\n",
       " 'V',\n",
       " 'T',\n",
       " 'D',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'D',\n",
       " 'V',\n",
       " 'P',\n",
       " 'S',\n",
       " 'N',\n",
       " 'I',\n",
       " 'A',\n",
       " 'K',\n",
       " 'I',\n",
       " 'I',\n",
       " 'I',\n",
       " 'G',\n",
       " 'P',\n",
       " 'L',\n",
       " 'I',\n",
       " 'F',\n",
       " 'V',\n",
       " 'F',\n",
       " 'L',\n",
       " 'F',\n",
       " 'S',\n",
       " 'V',\n",
       " 'V',\n",
       " 'I',\n",
       " 'G',\n",
       " 'S',\n",
       " 'I',\n",
       " 'Y',\n",
       " 'L',\n",
       " 'F',\n",
       " 'L',\n",
       " 'R',\n",
       " 'K',\n",
       " 'R',\n",
       " 'Q',\n",
       " 'P',\n",
       " 'D',\n",
       " 'G',\n",
       " 'P',\n",
       " 'L',\n",
       " 'G',\n",
       " 'P',\n",
       " 'L',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'S',\n",
       " 'S',\n",
       " 'N',\n",
       " 'P',\n",
       " 'E',\n",
       " 'Y',\n",
       " 'L',\n",
       " ...]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ammino_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Costruire la lista di tuple *(amminoacido, frequenza)* elencate per frequenza decrescente."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "ammino_counter = Counter(ammino_list)\n",
    "ammino_frequencies = ammino_counter.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('L', 136),\n",
       " ('E', 106),\n",
       " ('S', 105),\n",
       " ('G', 93),\n",
       " ('V', 87),\n",
       " ('R', 82),\n",
       " ('P', 82),\n",
       " ('N', 72),\n",
       " ('T', 70),\n",
       " ('D', 70),\n",
       " ('K', 68),\n",
       " ('A', 65),\n",
       " ('I', 65),\n",
       " ('F', 59),\n",
       " ('Y', 50),\n",
       " ('C', 47),\n",
       " ('H', 38),\n",
       " ('Q', 37),\n",
       " ('M', 30),\n",
       " ('W', 20)]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ammino_frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Produrre i diagrammi a barre delle frequenze dei codoni e degli amminoacidi.\n",
    "\n",
    "Importare il package `matplotlib`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package matplotlib:\n",
      "\n",
      "NAME\n",
      "    matplotlib - An object-oriented plotting library.\n",
      "\n",
      "DESCRIPTION\n",
      "    A procedural interface is provided by the companion pyplot module,\n",
      "    which may be imported directly, e.g.::\n",
      "    \n",
      "        import matplotlib.pyplot as plt\n",
      "    \n",
      "    or using ipython::\n",
      "    \n",
      "        ipython\n",
      "    \n",
      "    at your terminal, followed by::\n",
      "    \n",
      "        In [1]: %matplotlib\n",
      "        In [2]: import matplotlib.pyplot as plt\n",
      "    \n",
      "    at the ipython shell prompt.\n",
      "    \n",
      "    For the most part, direct use of the explicit object-oriented library is\n",
      "    encouraged when programming; the implicit pyplot interface is primarily for\n",
      "    working interactively. The exceptions to this suggestion are the pyplot\n",
      "    functions `.pyplot.figure`, `.pyplot.subplot`, `.pyplot.subplots`, and\n",
      "    `.pyplot.savefig`, which can greatly simplify scripting.  See\n",
      "    :ref:`api_interfaces` for an explanation of the tradeoffs between the implicit\n",
      "    and explicit interfaces.\n",
      "    \n",
      "    Modules include:\n",
      "    \n",
      "        :mod:`matplotlib.axes`\n",
      "            The `~.axes.Axes` class.  Most pyplot functions are wrappers for\n",
      "            `~.axes.Axes` methods.  The axes module is the highest level of OO\n",
      "            access to the library.\n",
      "    \n",
      "        :mod:`matplotlib.figure`\n",
      "            The `.Figure` class.\n",
      "    \n",
      "        :mod:`matplotlib.artist`\n",
      "            The `.Artist` base class for all classes that draw things.\n",
      "    \n",
      "        :mod:`matplotlib.lines`\n",
      "            The `.Line2D` class for drawing lines and markers.\n",
      "    \n",
      "        :mod:`matplotlib.patches`\n",
      "            Classes for drawing polygons.\n",
      "    \n",
      "        :mod:`matplotlib.text`\n",
      "            The `.Text` and `.Annotation` classes.\n",
      "    \n",
      "        :mod:`matplotlib.image`\n",
      "            The `.AxesImage` and `.FigureImage` classes.\n",
      "    \n",
      "        :mod:`matplotlib.collections`\n",
      "            Classes for efficient drawing of groups of lines or polygons.\n",
      "    \n",
      "        :mod:`matplotlib.colors`\n",
      "            Color specifications and making colormaps.\n",
      "    \n",
      "        :mod:`matplotlib.cm`\n",
      "            Colormaps, and the `.ScalarMappable` mixin class for providing color\n",
      "            mapping functionality to other classes.\n",
      "    \n",
      "        :mod:`matplotlib.ticker`\n",
      "            Calculation of tick mark locations and formatting of tick labels.\n",
      "    \n",
      "        :mod:`matplotlib.backends`\n",
      "            A subpackage with modules for various GUI libraries and output formats.\n",
      "    \n",
      "    The base matplotlib namespace includes:\n",
      "    \n",
      "        `~matplotlib.rcParams`\n",
      "            Default configuration settings; their defaults may be overridden using\n",
      "            a :file:`matplotlibrc` file.\n",
      "    \n",
      "        `~matplotlib.use`\n",
      "            Setting the Matplotlib backend.  This should be called before any\n",
      "            figure is created, because it is not possible to switch between\n",
      "            different GUI backends after that.\n",
      "    \n",
      "    Matplotlib was initially written by John D. Hunter (1968-2012) and is now\n",
      "    developed and maintained by a host of others.\n",
      "    \n",
      "    Occasionally the internal documentation (python docstrings) will refer\n",
      "    to MATLAB®, a registered trademark of The MathWorks, Inc.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _afm\n",
      "    _animation_data\n",
      "    _api (package)\n",
      "    _blocking_input\n",
      "    _c_internal_utils\n",
      "    _cm\n",
      "    _cm_listed\n",
      "    _color_data\n",
      "    _constrained_layout\n",
      "    _docstring\n",
      "    _enums\n",
      "    _fontconfig_pattern\n",
      "    _image\n",
      "    _internal_utils\n",
      "    _layoutgrid\n",
      "    _mathtext\n",
      "    _mathtext_data\n",
      "    _path\n",
      "    _pylab_helpers\n",
      "    _qhull\n",
      "    _text_helpers\n",
      "    _tight_bbox\n",
      "    _tight_layout\n",
      "    _tri\n",
      "    _ttconv\n",
      "    _type1font\n",
      "    _version\n",
      "    afm\n",
      "    animation\n",
      "    artist\n",
      "    axes (package)\n",
      "    axis\n",
      "    backend_bases\n",
      "    backend_managers\n",
      "    backend_tools\n",
      "    backends (package)\n",
      "    bezier\n",
      "    blocking_input\n",
      "    category\n",
      "    cbook (package)\n",
      "    cm\n",
      "    collections\n",
      "    colorbar\n",
      "    colors\n",
      "    container\n",
      "    contour\n",
      "    dates\n",
      "    docstring\n",
      "    dviread\n",
      "    figure\n",
      "    font_manager\n",
      "    fontconfig_pattern\n",
      "    ft2font\n",
      "    gridspec\n",
      "    hatch\n",
      "    image\n",
      "    layout_engine\n",
      "    legend\n",
      "    legend_handler\n",
      "    lines\n",
      "    markers\n",
      "    mathtext\n",
      "    mlab\n",
      "    offsetbox\n",
      "    patches\n",
      "    path\n",
      "    patheffects\n",
      "    projections (package)\n",
      "    pylab\n",
      "    pyplot\n",
      "    quiver\n",
      "    rcsetup\n",
      "    sankey\n",
      "    scale\n",
      "    sphinxext (package)\n",
      "    spines\n",
      "    stackplot\n",
      "    streamplot\n",
      "    style (package)\n",
      "    table\n",
      "    testing (package)\n",
      "    tests (package)\n",
      "    texmanager\n",
      "    text\n",
      "    textpath\n",
      "    ticker\n",
      "    tight_bbox\n",
      "    tight_layout\n",
      "    transforms\n",
      "    tri (package)\n",
      "    type1font\n",
      "    units\n",
      "    widgets\n",
      "\n",
      "CLASSES\n",
      "    builtins.FileNotFoundError(builtins.OSError)\n",
      "        ExecutableNotFoundError\n",
      "    builtins.dict(builtins.object)\n",
      "        RcParams(collections.abc.MutableMapping, builtins.dict)\n",
      "    collections.abc.MutableMapping(collections.abc.Mapping)\n",
      "        RcParams(collections.abc.MutableMapping, builtins.dict)\n",
      "    \n",
      "    class ExecutableNotFoundError(builtins.FileNotFoundError)\n",
      "     |  Error raised when an executable that Matplotlib optionally\n",
      "     |  depends on can't be found.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      ExecutableNotFoundError\n",
      "     |      builtins.FileNotFoundError\n",
      "     |      builtins.OSError\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.FileNotFoundError:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.OSError:\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.OSError:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.OSError:\n",
      "     |  \n",
      "     |  characters_written\n",
      "     |  \n",
      "     |  errno\n",
      "     |      POSIX exception code\n",
      "     |  \n",
      "     |  filename\n",
      "     |      exception filename\n",
      "     |  \n",
      "     |  filename2\n",
      "     |      second exception filename\n",
      "     |  \n",
      "     |  strerror\n",
      "     |      exception strerror\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class RcParams(collections.abc.MutableMapping, builtins.dict)\n",
      "     |  RcParams(*args, **kwargs)\n",
      "     |  \n",
      "     |  A dictionary object including validation.\n",
      "     |  \n",
      "     |  Validating functions are defined and associated with rc parameters in\n",
      "     |  :mod:`matplotlib.rcsetup`.\n",
      "     |  \n",
      "     |  The list of rcParams is:\n",
      "     |  \n",
      "     |  - _internal.classic_mode\n",
      "     |  - agg.path.chunksize\n",
      "     |  - animation.bitrate\n",
      "     |  - animation.codec\n",
      "     |  - animation.convert_args\n",
      "     |  - animation.convert_path\n",
      "     |  - animation.embed_limit\n",
      "     |  - animation.ffmpeg_args\n",
      "     |  - animation.ffmpeg_path\n",
      "     |  - animation.frame_format\n",
      "     |  - animation.html\n",
      "     |  - animation.writer\n",
      "     |  - axes.autolimit_mode\n",
      "     |  - axes.axisbelow\n",
      "     |  - axes.edgecolor\n",
      "     |  - axes.facecolor\n",
      "     |  - axes.formatter.limits\n",
      "     |  - axes.formatter.min_exponent\n",
      "     |  - axes.formatter.offset_threshold\n",
      "     |  - axes.formatter.use_locale\n",
      "     |  - axes.formatter.use_mathtext\n",
      "     |  - axes.formatter.useoffset\n",
      "     |  - axes.grid\n",
      "     |  - axes.grid.axis\n",
      "     |  - axes.grid.which\n",
      "     |  - axes.labelcolor\n",
      "     |  - axes.labelpad\n",
      "     |  - axes.labelsize\n",
      "     |  - axes.labelweight\n",
      "     |  - axes.linewidth\n",
      "     |  - axes.prop_cycle\n",
      "     |  - axes.spines.bottom\n",
      "     |  - axes.spines.left\n",
      "     |  - axes.spines.right\n",
      "     |  - axes.spines.top\n",
      "     |  - axes.titlecolor\n",
      "     |  - axes.titlelocation\n",
      "     |  - axes.titlepad\n",
      "     |  - axes.titlesize\n",
      "     |  - axes.titleweight\n",
      "     |  - axes.titley\n",
      "     |  - axes.unicode_minus\n",
      "     |  - axes.xmargin\n",
      "     |  - axes.ymargin\n",
      "     |  - axes.zmargin\n",
      "     |  - axes3d.grid\n",
      "     |  - backend\n",
      "     |  - backend_fallback\n",
      "     |  - boxplot.bootstrap\n",
      "     |  - boxplot.boxprops.color\n",
      "     |  - boxplot.boxprops.linestyle\n",
      "     |  - boxplot.boxprops.linewidth\n",
      "     |  - boxplot.capprops.color\n",
      "     |  - boxplot.capprops.linestyle\n",
      "     |  - boxplot.capprops.linewidth\n",
      "     |  - boxplot.flierprops.color\n",
      "     |  - boxplot.flierprops.linestyle\n",
      "     |  - boxplot.flierprops.linewidth\n",
      "     |  - boxplot.flierprops.marker\n",
      "     |  - boxplot.flierprops.markeredgecolor\n",
      "     |  - boxplot.flierprops.markeredgewidth\n",
      "     |  - boxplot.flierprops.markerfacecolor\n",
      "     |  - boxplot.flierprops.markersize\n",
      "     |  - boxplot.meanline\n",
      "     |  - boxplot.meanprops.color\n",
      "     |  - boxplot.meanprops.linestyle\n",
      "     |  - boxplot.meanprops.linewidth\n",
      "     |  - boxplot.meanprops.marker\n",
      "     |  - boxplot.meanprops.markeredgecolor\n",
      "     |  - boxplot.meanprops.markerfacecolor\n",
      "     |  - boxplot.meanprops.markersize\n",
      "     |  - boxplot.medianprops.color\n",
      "     |  - boxplot.medianprops.linestyle\n",
      "     |  - boxplot.medianprops.linewidth\n",
      "     |  - boxplot.notch\n",
      "     |  - boxplot.patchartist\n",
      "     |  - boxplot.showbox\n",
      "     |  - boxplot.showcaps\n",
      "     |  - boxplot.showfliers\n",
      "     |  - boxplot.showmeans\n",
      "     |  - boxplot.vertical\n",
      "     |  - boxplot.whiskerprops.color\n",
      "     |  - boxplot.whiskerprops.linestyle\n",
      "     |  - boxplot.whiskerprops.linewidth\n",
      "     |  - boxplot.whiskers\n",
      "     |  - contour.algorithm\n",
      "     |  - contour.corner_mask\n",
      "     |  - contour.linewidth\n",
      "     |  - contour.negative_linestyle\n",
      "     |  - date.autoformatter.day\n",
      "     |  - date.autoformatter.hour\n",
      "     |  - date.autoformatter.microsecond\n",
      "     |  - date.autoformatter.minute\n",
      "     |  - date.autoformatter.month\n",
      "     |  - date.autoformatter.second\n",
      "     |  - date.autoformatter.year\n",
      "     |  - date.converter\n",
      "     |  - date.epoch\n",
      "     |  - date.interval_multiples\n",
      "     |  - docstring.hardcopy\n",
      "     |  - errorbar.capsize\n",
      "     |  - figure.autolayout\n",
      "     |  - figure.constrained_layout.h_pad\n",
      "     |  - figure.constrained_layout.hspace\n",
      "     |  - figure.constrained_layout.use\n",
      "     |  - figure.constrained_layout.w_pad\n",
      "     |  - figure.constrained_layout.wspace\n",
      "     |  - figure.dpi\n",
      "     |  - figure.edgecolor\n",
      "     |  - figure.facecolor\n",
      "     |  - figure.figsize\n",
      "     |  - figure.frameon\n",
      "     |  - figure.labelsize\n",
      "     |  - figure.labelweight\n",
      "     |  - figure.max_open_warning\n",
      "     |  - figure.raise_window\n",
      "     |  - figure.subplot.bottom\n",
      "     |  - figure.subplot.hspace\n",
      "     |  - figure.subplot.left\n",
      "     |  - figure.subplot.right\n",
      "     |  - figure.subplot.top\n",
      "     |  - figure.subplot.wspace\n",
      "     |  - figure.titlesize\n",
      "     |  - figure.titleweight\n",
      "     |  - font.cursive\n",
      "     |  - font.family\n",
      "     |  - font.fantasy\n",
      "     |  - font.monospace\n",
      "     |  - font.sans-serif\n",
      "     |  - font.serif\n",
      "     |  - font.size\n",
      "     |  - font.stretch\n",
      "     |  - font.style\n",
      "     |  - font.variant\n",
      "     |  - font.weight\n",
      "     |  - grid.alpha\n",
      "     |  - grid.color\n",
      "     |  - grid.linestyle\n",
      "     |  - grid.linewidth\n",
      "     |  - hatch.color\n",
      "     |  - hatch.linewidth\n",
      "     |  - hist.bins\n",
      "     |  - image.aspect\n",
      "     |  - image.cmap\n",
      "     |  - image.composite_image\n",
      "     |  - image.interpolation\n",
      "     |  - image.lut\n",
      "     |  - image.origin\n",
      "     |  - image.resample\n",
      "     |  - interactive\n",
      "     |  - keymap.back\n",
      "     |  - keymap.copy\n",
      "     |  - keymap.forward\n",
      "     |  - keymap.fullscreen\n",
      "     |  - keymap.grid\n",
      "     |  - keymap.grid_minor\n",
      "     |  - keymap.help\n",
      "     |  - keymap.home\n",
      "     |  - keymap.pan\n",
      "     |  - keymap.quit\n",
      "     |  - keymap.quit_all\n",
      "     |  - keymap.save\n",
      "     |  - keymap.xscale\n",
      "     |  - keymap.yscale\n",
      "     |  - keymap.zoom\n",
      "     |  - legend.borderaxespad\n",
      "     |  - legend.borderpad\n",
      "     |  - legend.columnspacing\n",
      "     |  - legend.edgecolor\n",
      "     |  - legend.facecolor\n",
      "     |  - legend.fancybox\n",
      "     |  - legend.fontsize\n",
      "     |  - legend.framealpha\n",
      "     |  - legend.frameon\n",
      "     |  - legend.handleheight\n",
      "     |  - legend.handlelength\n",
      "     |  - legend.handletextpad\n",
      "     |  - legend.labelcolor\n",
      "     |  - legend.labelspacing\n",
      "     |  - legend.loc\n",
      "     |  - legend.markerscale\n",
      "     |  - legend.numpoints\n",
      "     |  - legend.scatterpoints\n",
      "     |  - legend.shadow\n",
      "     |  - legend.title_fontsize\n",
      "     |  - lines.antialiased\n",
      "     |  - lines.color\n",
      "     |  - lines.dash_capstyle\n",
      "     |  - lines.dash_joinstyle\n",
      "     |  - lines.dashdot_pattern\n",
      "     |  - lines.dashed_pattern\n",
      "     |  - lines.dotted_pattern\n",
      "     |  - lines.linestyle\n",
      "     |  - lines.linewidth\n",
      "     |  - lines.marker\n",
      "     |  - lines.markeredgecolor\n",
      "     |  - lines.markeredgewidth\n",
      "     |  - lines.markerfacecolor\n",
      "     |  - lines.markersize\n",
      "     |  - lines.scale_dashes\n",
      "     |  - lines.solid_capstyle\n",
      "     |  - lines.solid_joinstyle\n",
      "     |  - markers.fillstyle\n",
      "     |  - mathtext.bf\n",
      "     |  - mathtext.cal\n",
      "     |  - mathtext.default\n",
      "     |  - mathtext.fallback\n",
      "     |  - mathtext.fontset\n",
      "     |  - mathtext.it\n",
      "     |  - mathtext.rm\n",
      "     |  - mathtext.sf\n",
      "     |  - mathtext.tt\n",
      "     |  - patch.antialiased\n",
      "     |  - patch.edgecolor\n",
      "     |  - patch.facecolor\n",
      "     |  - patch.force_edgecolor\n",
      "     |  - patch.linewidth\n",
      "     |  - path.effects\n",
      "     |  - path.simplify\n",
      "     |  - path.simplify_threshold\n",
      "     |  - path.sketch\n",
      "     |  - path.snap\n",
      "     |  - pcolor.shading\n",
      "     |  - pcolormesh.snap\n",
      "     |  - pdf.compression\n",
      "     |  - pdf.fonttype\n",
      "     |  - pdf.inheritcolor\n",
      "     |  - pdf.use14corefonts\n",
      "     |  - pgf.preamble\n",
      "     |  - pgf.rcfonts\n",
      "     |  - pgf.texsystem\n",
      "     |  - polaraxes.grid\n",
      "     |  - ps.distiller.res\n",
      "     |  - ps.fonttype\n",
      "     |  - ps.papersize\n",
      "     |  - ps.useafm\n",
      "     |  - ps.usedistiller\n",
      "     |  - savefig.bbox\n",
      "     |  - savefig.directory\n",
      "     |  - savefig.dpi\n",
      "     |  - savefig.edgecolor\n",
      "     |  - savefig.facecolor\n",
      "     |  - savefig.format\n",
      "     |  - savefig.orientation\n",
      "     |  - savefig.pad_inches\n",
      "     |  - savefig.transparent\n",
      "     |  - scatter.edgecolors\n",
      "     |  - scatter.marker\n",
      "     |  - svg.fonttype\n",
      "     |  - svg.hashsalt\n",
      "     |  - svg.image_inline\n",
      "     |  - text.antialiased\n",
      "     |  - text.color\n",
      "     |  - text.hinting\n",
      "     |  - text.hinting_factor\n",
      "     |  - text.kerning_factor\n",
      "     |  - text.latex.preamble\n",
      "     |  - text.parse_math\n",
      "     |  - text.usetex\n",
      "     |  - timezone\n",
      "     |  - tk.window_focus\n",
      "     |  - toolbar\n",
      "     |  - webagg.address\n",
      "     |  - webagg.open_in_browser\n",
      "     |  - webagg.port\n",
      "     |  - webagg.port_retries\n",
      "     |  - xaxis.labellocation\n",
      "     |  - xtick.alignment\n",
      "     |  - xtick.bottom\n",
      "     |  - xtick.color\n",
      "     |  - xtick.direction\n",
      "     |  - xtick.labelbottom\n",
      "     |  - xtick.labelcolor\n",
      "     |  - xtick.labelsize\n",
      "     |  - xtick.labeltop\n",
      "     |  - xtick.major.bottom\n",
      "     |  - xtick.major.pad\n",
      "     |  - xtick.major.size\n",
      "     |  - xtick.major.top\n",
      "     |  - xtick.major.width\n",
      "     |  - xtick.minor.bottom\n",
      "     |  - xtick.minor.pad\n",
      "     |  - xtick.minor.size\n",
      "     |  - xtick.minor.top\n",
      "     |  - xtick.minor.visible\n",
      "     |  - xtick.minor.width\n",
      "     |  - xtick.top\n",
      "     |  - yaxis.labellocation\n",
      "     |  - ytick.alignment\n",
      "     |  - ytick.color\n",
      "     |  - ytick.direction\n",
      "     |  - ytick.labelcolor\n",
      "     |  - ytick.labelleft\n",
      "     |  - ytick.labelright\n",
      "     |  - ytick.labelsize\n",
      "     |  - ytick.left\n",
      "     |  - ytick.major.left\n",
      "     |  - ytick.major.pad\n",
      "     |  - ytick.major.right\n",
      "     |  - ytick.major.size\n",
      "     |  - ytick.major.width\n",
      "     |  - ytick.minor.left\n",
      "     |  - ytick.minor.pad\n",
      "     |  - ytick.minor.right\n",
      "     |  - ytick.minor.size\n",
      "     |  - ytick.minor.visible\n",
      "     |  - ytick.minor.width\n",
      "     |  - ytick.right\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  :ref:`customizing-with-matplotlibrc-files`\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      RcParams\n",
      "     |      collections.abc.MutableMapping\n",
      "     |      collections.abc.Mapping\n",
      "     |      collections.abc.Collection\n",
      "     |      collections.abc.Sized\n",
      "     |      collections.abc.Iterable\n",
      "     |      collections.abc.Container\n",
      "     |      builtins.dict\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getitem__(self, key)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __init__(self, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __iter__(self)\n",
      "     |      Yield sorted list of keys.\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return len(self).\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setitem__(self, key, val)\n",
      "     |      Set self[key] to value.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  copy(self)\n",
      "     |      Copy this RcParams instance.\n",
      "     |  \n",
      "     |  find_all(self, pattern)\n",
      "     |      Return the subset of this RcParams dictionary whose keys match,\n",
      "     |      using :func:`re.search`, the given ``pattern``.\n",
      "     |      \n",
      "     |      .. note::\n",
      "     |      \n",
      "     |          Changes to the returned dictionary are *not* propagated to\n",
      "     |          the parent RcParams dictionary.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes defined here:\n",
      "     |  \n",
      "     |  __abstractmethods__ = frozenset({'__delitem__'})\n",
      "     |  \n",
      "     |  validate = {'_internal.classic_mode': <function validate_bool>, 'agg.p...\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from collections.abc.MutableMapping:\n",
      "     |  \n",
      "     |  __delitem__(self, key)\n",
      "     |  \n",
      "     |  clear(self)\n",
      "     |      D.clear() -> None.  Remove all items from D.\n",
      "     |  \n",
      "     |  pop(self, key, default=<object object at 0x1096e3160>)\n",
      "     |      D.pop(k[,d]) -> v, remove specified key and return the corresponding value.\n",
      "     |      If key is not found, d is returned if given, otherwise KeyError is raised.\n",
      "     |  \n",
      "     |  popitem(self)\n",
      "     |      D.popitem() -> (k, v), remove and return some (key, value) pair\n",
      "     |      as a 2-tuple; but raise KeyError if D is empty.\n",
      "     |  \n",
      "     |  setdefault(self, key, default=None)\n",
      "     |      D.setdefault(k[,d]) -> D.get(k,d), also set D[k]=d if k not in D\n",
      "     |  \n",
      "     |  update(self, other=(), /, **kwds)\n",
      "     |      D.update([E, ]**F) -> None.  Update D from mapping/iterable E and F.\n",
      "     |      If E present and has a .keys() method, does:     for k in E: D[k] = E[k]\n",
      "     |      If E present and lacks .keys() method, does:     for (k, v) in E: D[k] = v\n",
      "     |      In either case, this is followed by: for k, v in F.items(): D[k] = v\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from collections.abc.Mapping:\n",
      "     |  \n",
      "     |  __contains__(self, key)\n",
      "     |  \n",
      "     |  __eq__(self, other)\n",
      "     |      Return self==value.\n",
      "     |  \n",
      "     |  get(self, key, default=None)\n",
      "     |      D.get(k[,d]) -> D[k] if k in D, else d.  d defaults to None.\n",
      "     |  \n",
      "     |  items(self)\n",
      "     |      D.items() -> a set-like object providing a view on D's items\n",
      "     |  \n",
      "     |  keys(self)\n",
      "     |      D.keys() -> a set-like object providing a view on D's keys\n",
      "     |  \n",
      "     |  values(self)\n",
      "     |      D.values() -> an object providing a view on D's values\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from collections.abc.Mapping:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "     |  \n",
      "     |  __reversed__ = None\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from collections.abc.Collection:\n",
      "     |  \n",
      "     |  __subclasshook__(C) from abc.ABCMeta\n",
      "     |      Abstract classes can override this to customize issubclass().\n",
      "     |      \n",
      "     |      This is invoked early on by abc.ABCMeta.__subclasscheck__().\n",
      "     |      It should return True, False or NotImplemented.  If it returns\n",
      "     |      NotImplemented, the normal algorithm is used.  Otherwise, it\n",
      "     |      overrides the normal algorithm (and the outcome is cached).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from collections.abc.Iterable:\n",
      "     |  \n",
      "     |  __class_getitem__ = GenericAlias(...) from abc.ABCMeta\n",
      "     |      Represent a PEP 585 generic type\n",
      "     |      \n",
      "     |      E.g. for t = list[int], t.__origin__ is list and t.__args__ is (int,).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __ge__(self, value, /)\n",
      "     |      Return self>=value.\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __gt__(self, value, /)\n",
      "     |      Return self>value.\n",
      "     |  \n",
      "     |  __ior__(self, value, /)\n",
      "     |      Return self|=value.\n",
      "     |  \n",
      "     |  __le__(self, value, /)\n",
      "     |      Return self<=value.\n",
      "     |  \n",
      "     |  __lt__(self, value, /)\n",
      "     |      Return self<value.\n",
      "     |  \n",
      "     |  __ne__(self, value, /)\n",
      "     |      Return self!=value.\n",
      "     |  \n",
      "     |  __or__(self, value, /)\n",
      "     |      Return self|value.\n",
      "     |  \n",
      "     |  __ror__(self, value, /)\n",
      "     |      Return value|self.\n",
      "     |  \n",
      "     |  __sizeof__(...)\n",
      "     |      D.__sizeof__() -> size of D in memory, in bytes\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  fromkeys(iterable, value=None, /) from abc.ABCMeta\n",
      "     |      Create a new dictionary with keys from iterable and values set to value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "\n",
      "FUNCTIONS\n",
      "    checkdep_usetex(s)\n",
      "        [*Deprecated*] \n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. deprecated:: 3.6\n",
      "           Use a vendored copy of this function instead.\\\n",
      "    \n",
      "    get_backend()\n",
      "        Return the name of the current backend.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.use\n",
      "    \n",
      "    get_cachedir()\n",
      "        Return the string path of the cache directory.\n",
      "        \n",
      "        The procedure used to find the directory is the same as for\n",
      "        _get_config_dir, except using ``$XDG_CACHE_HOME``/``$HOME/.cache`` instead.\n",
      "    \n",
      "    get_configdir()\n",
      "        Return the string path of the configuration directory.\n",
      "        \n",
      "        The directory is chosen as follows:\n",
      "        \n",
      "        1. If the MPLCONFIGDIR environment variable is supplied, choose that.\n",
      "        2. On Linux, follow the XDG specification and look first in\n",
      "           ``$XDG_CONFIG_HOME``, if defined, or ``$HOME/.config``.  On other\n",
      "           platforms, choose ``$HOME/.matplotlib``.\n",
      "        3. If the chosen directory exists and is writable, use that as the\n",
      "           configuration directory.\n",
      "        4. Else, create a temporary directory, and use it as the configuration\n",
      "           directory.\n",
      "    \n",
      "    get_data_path()\n",
      "        Return the path to Matplotlib data.\n",
      "    \n",
      "    interactive(b)\n",
      "        Set whether to redraw after every plotting command (e.g. `.pyplot.xlabel`).\n",
      "    \n",
      "    is_interactive()\n",
      "        Return whether to redraw after every plotting command.\n",
      "        \n",
      "        .. note::\n",
      "        \n",
      "            This function is only intended for use in backends. End users should\n",
      "            use `.pyplot.isinteractive` instead.\n",
      "    \n",
      "    is_url(filename)\n",
      "        [*Deprecated*] Return whether *filename* is an http, https, ftp, or file URL path.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. deprecated:: 3.5\n",
      "    \n",
      "    matplotlib_fname()\n",
      "        Get the location of the config file.\n",
      "        \n",
      "        The file location is determined in the following order\n",
      "        \n",
      "        - ``$PWD/matplotlibrc``\n",
      "        - ``$MATPLOTLIBRC`` if it is not a directory\n",
      "        - ``$MATPLOTLIBRC/matplotlibrc``\n",
      "        - ``$MPLCONFIGDIR/matplotlibrc``\n",
      "        - On Linux,\n",
      "            - ``$XDG_CONFIG_HOME/matplotlib/matplotlibrc`` (if ``$XDG_CONFIG_HOME``\n",
      "              is defined)\n",
      "            - or ``$HOME/.config/matplotlib/matplotlibrc`` (if ``$XDG_CONFIG_HOME``\n",
      "              is not defined)\n",
      "        - On other platforms,\n",
      "          - ``$HOME/.matplotlib/matplotlibrc`` if ``$HOME`` is defined\n",
      "        - Lastly, it looks in ``$MATPLOTLIBDATA/matplotlibrc``, which should always\n",
      "          exist.\n",
      "    \n",
      "    rc(group, **kwargs)\n",
      "        Set the current `.rcParams`.  *group* is the grouping for the rc, e.g.,\n",
      "        for ``lines.linewidth`` the group is ``lines``, for\n",
      "        ``axes.facecolor``, the group is ``axes``, and so on.  Group may\n",
      "        also be a list or tuple of group names, e.g., (*xtick*, *ytick*).\n",
      "        *kwargs* is a dictionary attribute name/value pairs, e.g.,::\n",
      "        \n",
      "          rc('lines', linewidth=2, color='r')\n",
      "        \n",
      "        sets the current `.rcParams` and is equivalent to::\n",
      "        \n",
      "          rcParams['lines.linewidth'] = 2\n",
      "          rcParams['lines.color'] = 'r'\n",
      "        \n",
      "        The following aliases are available to save typing for interactive users:\n",
      "        \n",
      "        =====   =================\n",
      "        Alias   Property\n",
      "        =====   =================\n",
      "        'lw'    'linewidth'\n",
      "        'ls'    'linestyle'\n",
      "        'c'     'color'\n",
      "        'fc'    'facecolor'\n",
      "        'ec'    'edgecolor'\n",
      "        'mew'   'markeredgewidth'\n",
      "        'aa'    'antialiased'\n",
      "        =====   =================\n",
      "        \n",
      "        Thus you could abbreviate the above call as::\n",
      "        \n",
      "              rc('lines', lw=2, c='r')\n",
      "        \n",
      "        Note you can use python's kwargs dictionary facility to store\n",
      "        dictionaries of default parameters.  e.g., you can customize the\n",
      "        font rc as follows::\n",
      "        \n",
      "          font = {'family' : 'monospace',\n",
      "                  'weight' : 'bold',\n",
      "                  'size'   : 'larger'}\n",
      "          rc('font', **font)  # pass in the font dict as kwargs\n",
      "        \n",
      "        This enables you to easily switch between several configurations.  Use\n",
      "        ``matplotlib.style.use('default')`` or :func:`~matplotlib.rcdefaults` to\n",
      "        restore the default `.rcParams` after changes.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Similar functionality is available by using the normal dict interface, i.e.\n",
      "        ``rcParams.update({\"lines.linewidth\": 2, ...})`` (but ``rcParams.update``\n",
      "        does not support abbreviations or grouping).\n",
      "    \n",
      "    rc_context(rc=None, fname=None)\n",
      "        Return a context manager for temporarily changing rcParams.\n",
      "        \n",
      "        The :rc:`backend` will not be reset by the context manager.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rc : dict\n",
      "            The rcParams to temporarily set.\n",
      "        fname : str or path-like\n",
      "            A file with Matplotlib rc settings. If both *fname* and *rc* are given,\n",
      "            settings from *rc* take precedence.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        :ref:`customizing-with-matplotlibrc-files`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Passing explicit values via a dict::\n",
      "        \n",
      "            with mpl.rc_context({'interactive': False}):\n",
      "                fig, ax = plt.subplots()\n",
      "                ax.plot(range(3), range(3))\n",
      "                fig.savefig('example.png')\n",
      "                plt.close(fig)\n",
      "        \n",
      "        Loading settings from a file::\n",
      "        \n",
      "             with mpl.rc_context(fname='print.rc'):\n",
      "                 plt.plot(x, y)  # uses 'print.rc'\n",
      "    \n",
      "    rc_file(fname, *, use_default_template=True)\n",
      "        Update `.rcParams` from file.\n",
      "        \n",
      "        Style-blacklisted `.rcParams` (defined in\n",
      "        ``matplotlib.style.core.STYLE_BLACKLIST``) are not updated.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fname : str or path-like\n",
      "            A file with Matplotlib rc settings.\n",
      "        \n",
      "        use_default_template : bool\n",
      "            If True, initialize with default parameters before updating with those\n",
      "            in the given file. If False, the current configuration persists\n",
      "            and only the parameters specified in the file are updated.\n",
      "    \n",
      "    rc_file_defaults()\n",
      "        Restore the `.rcParams` from the original rc file loaded by Matplotlib.\n",
      "        \n",
      "        Style-blacklisted `.rcParams` (defined in\n",
      "        ``matplotlib.style.core.STYLE_BLACKLIST``) are not updated.\n",
      "    \n",
      "    rc_params(fail_on_error=False)\n",
      "        Construct a `RcParams` instance from the default Matplotlib rc file.\n",
      "    \n",
      "    rc_params_from_file(fname, fail_on_error=False, use_default_template=True)\n",
      "        Construct a `RcParams` from file *fname*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fname : str or path-like\n",
      "            A file with Matplotlib rc settings.\n",
      "        fail_on_error : bool\n",
      "            If True, raise an error when the parser fails to convert a parameter.\n",
      "        use_default_template : bool\n",
      "            If True, initialize with default parameters before updating with those\n",
      "            in the given file. If False, the configuration class only contains the\n",
      "            parameters specified in the file. (Useful for updating dicts.)\n",
      "    \n",
      "    rcdefaults()\n",
      "        Restore the `.rcParams` from Matplotlib's internal default style.\n",
      "        \n",
      "        Style-blacklisted `.rcParams` (defined in\n",
      "        ``matplotlib.style.core.STYLE_BLACKLIST``) are not updated.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.rc_file_defaults\n",
      "            Restore the `.rcParams` from the rc file originally loaded by\n",
      "            Matplotlib.\n",
      "        matplotlib.style.use\n",
      "            Use a specific style file.  Call ``style.use('default')`` to restore\n",
      "            the default style.\n",
      "    \n",
      "    set_loglevel(level)\n",
      "        Set Matplotlib's root logger and root logger handler level, creating\n",
      "        the handler if it does not exist yet.\n",
      "        \n",
      "        Typically, one should call ``set_loglevel(\"info\")`` or\n",
      "        ``set_loglevel(\"debug\")`` to get additional debugging information.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        level : {\"notset\", \"debug\", \"info\", \"warning\", \"error\", \"critical\"}\n",
      "            The log level of the handler.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The first time this function is called, an additional handler is attached\n",
      "        to Matplotlib's root handler; this handler is reused every time and this\n",
      "        function simply manipulates the logger and handler's level.\n",
      "    \n",
      "    test(verbosity=None, coverage=False, **kwargs)\n",
      "        [*Deprecated*] Run the matplotlib test suite.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. deprecated:: 3.5\n",
      "           Use pytest instead.\n",
      "    \n",
      "    use(backend, *, force=True)\n",
      "        Select the backend used for rendering and GUI integration.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        backend : str\n",
      "            The backend to switch to.  This can either be one of the standard\n",
      "            backend names, which are case-insensitive:\n",
      "        \n",
      "            - interactive backends:\n",
      "              GTK3Agg, GTK3Cairo, GTK4Agg, GTK4Cairo, MacOSX, nbAgg, QtAgg,\n",
      "              QtCairo, TkAgg, TkCairo, WebAgg, WX, WXAgg, WXCairo, Qt5Agg, Qt5Cairo\n",
      "        \n",
      "            - non-interactive backends:\n",
      "              agg, cairo, pdf, pgf, ps, svg, template\n",
      "        \n",
      "            or a string of the form: ``module://my.module.name``.\n",
      "        \n",
      "            Switching to an interactive backend is not possible if an unrelated\n",
      "            event loop has already been started (e.g., switching to GTK3Agg if a\n",
      "            TkAgg window has already been opened).  Switching to a non-interactive\n",
      "            backend is always possible.\n",
      "        \n",
      "        force : bool, default: True\n",
      "            If True (the default), raise an `ImportError` if the backend cannot be\n",
      "            set up (either because it fails to import, or because an incompatible\n",
      "            GUI interactive framework is already running); if False, silently\n",
      "            ignore the failure.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        :ref:`backends`\n",
      "        matplotlib.get_backend\n",
      "\n",
      "DATA\n",
      "    __bibtex__ = '@Article{Hunter:2007,\\n  Author    = {Hunter, J. ...ishe...\n",
      "    color_sequences = <matplotlib.colors.ColorSequenceRegistry object>\n",
      "    colormaps = <matplotlib.cm.ColormapRegistry object>\n",
      "    defaultParams = {'_internal.classic_mode': [False, <function validate_...\n",
      "    default_test_modules = ['matplotlib.tests', 'mpl_toolkits.tests']\n",
      "    rcParams = RcParams({'_internal.classic_mode': False,\n",
      "         ...nor.widt...\n",
      "    rcParamsDefault = RcParams({'_internal.classic_mode': False,\n",
      "         ...n...\n",
      "    rcParamsOrig = RcParams({'_internal.classic_mode': False,\n",
      "         ...nor....\n",
      "\n",
      "VERSION\n",
      "    3.6.2\n",
      "\n",
      "FILE\n",
      "    /opt/miniconda3/lib/python3.9/site-packages/matplotlib/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(matplotlib)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importare il modulo `pyplot`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module matplotlib.pyplot in matplotlib:\n",
      "\n",
      "NAME\n",
      "    matplotlib.pyplot\n",
      "\n",
      "DESCRIPTION\n",
      "    `matplotlib.pyplot` is a state-based interface to matplotlib. It provides\n",
      "    an implicit,  MATLAB-like, way of plotting.  It also opens figures on your\n",
      "    screen, and acts as the figure GUI manager.\n",
      "    \n",
      "    pyplot is mainly intended for interactive plots and simple cases of\n",
      "    programmatic plot generation::\n",
      "    \n",
      "        import numpy as np\n",
      "        import matplotlib.pyplot as plt\n",
      "    \n",
      "        x = np.arange(0, 5, 0.1)\n",
      "        y = np.sin(x)\n",
      "        plt.plot(x, y)\n",
      "    \n",
      "    The explicit object-oriented API is recommended for complex plots, though\n",
      "    pyplot is still usually used to create the figure and often the axes in the\n",
      "    figure. See `.pyplot.figure`, `.pyplot.subplots`, and\n",
      "    `.pyplot.subplot_mosaic` to create figures, and\n",
      "    :doc:`Axes API </api/axes_api>` for the plotting methods on an Axes::\n",
      "    \n",
      "        import numpy as np\n",
      "        import matplotlib.pyplot as plt\n",
      "    \n",
      "        x = np.arange(0, 5, 0.1)\n",
      "        y = np.sin(x)\n",
      "        fig, ax = plt.subplots()\n",
      "        ax.plot(x, y)\n",
      "    \n",
      "    \n",
      "    See :ref:`api_interfaces` for an explanation of the tradeoffs between the\n",
      "    implicit and explicit interfaces.\n",
      "\n",
      "FUNCTIONS\n",
      "    acorr(x, *, data=None, **kwargs)\n",
      "        Plot the autocorrelation of *x*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array-like\n",
      "        \n",
      "        detrend : callable, default: `.mlab.detrend_none` (no detrending)\n",
      "            A detrending function applied to *x*.  It must have the\n",
      "            signature ::\n",
      "        \n",
      "                detrend(x: np.ndarray) -> np.ndarray\n",
      "        \n",
      "        normed : bool, default: True\n",
      "            If ``True``, input vectors are normalised to unit length.\n",
      "        \n",
      "        usevlines : bool, default: True\n",
      "            Determines the plot style.\n",
      "        \n",
      "            If ``True``, vertical lines are plotted from 0 to the acorr value\n",
      "            using `.Axes.vlines`. Additionally, a horizontal line is plotted\n",
      "            at y=0 using `.Axes.axhline`.\n",
      "        \n",
      "            If ``False``, markers are plotted at the acorr values using\n",
      "            `.Axes.plot`.\n",
      "        \n",
      "        maxlags : int, default: 10\n",
      "            Number of lags to show. If ``None``, will return all\n",
      "            ``2 * len(x) - 1`` lags.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lags : array (length ``2*maxlags+1``)\n",
      "            The lag vector.\n",
      "        c : array  (length ``2*maxlags+1``)\n",
      "            The auto correlation vector.\n",
      "        line : `.LineCollection` or `.Line2D`\n",
      "            `.Artist` added to the Axes of the correlation:\n",
      "        \n",
      "            - `.LineCollection` if *usevlines* is True.\n",
      "            - `.Line2D` if *usevlines* is False.\n",
      "        b : `.Line2D` or None\n",
      "            Horizontal line at 0 if *usevlines* is True\n",
      "            None *usevlines* is False.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        linestyle : `.Line2D` property, optional\n",
      "            The linestyle for plotting the data points.\n",
      "            Only used if *usevlines* is ``False``.\n",
      "        \n",
      "        marker : str, default: 'o'\n",
      "            The marker for plotting the data points.\n",
      "            Only used if *usevlines* is ``False``.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Additional parameters are passed to `.Axes.vlines` and\n",
      "            `.Axes.axhline` if *usevlines* is ``True``; otherwise they are\n",
      "            passed to `.Axes.plot`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The cross correlation is performed with `numpy.correlate` with\n",
      "        ``mode = \"full\"``.\n",
      "    \n",
      "    angle_spectrum(x, Fs=None, Fc=None, window=None, pad_to=None, sides=None, *, data=None, **kwargs)\n",
      "        Plot the angle spectrum.\n",
      "        \n",
      "        Compute the angle spectrum (wrapped phase spectrum) of *x*.\n",
      "        Data is padded to a length of *pad_to* and the windowing function\n",
      "        *window* is applied to the signal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1-D array or sequence\n",
      "            Array or sequence containing the data.\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  While not increasing the actual resolution of the spectrum (the\n",
      "            minimum distance between resolvable peaks), this can give more points in\n",
      "            the plot, allowing for more detail. This corresponds to the *n* parameter\n",
      "            in the call to `~numpy.fft.fft`.  The default is None, which sets *pad_to*\n",
      "            equal to the length of the input signal (i.e. no padding).\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        spectrum : 1-D array\n",
      "            The values for the angle spectrum in radians (real valued).\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the elements in *spectrum*.\n",
      "        \n",
      "        line : `~matplotlib.lines.Line2D`\n",
      "            The line created by this function.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        magnitude_spectrum\n",
      "            Plots the magnitudes of the corresponding frequencies.\n",
      "        phase_spectrum\n",
      "            Plots the unwrapped version of this function.\n",
      "        specgram\n",
      "            Can plot the angle spectrum of segments within the signal in a\n",
      "            colormap.\n",
      "    \n",
      "    annotate(text, xy, xytext=None, xycoords='data', textcoords=None, arrowprops=None, annotation_clip=None, **kwargs)\n",
      "        Annotate the point *xy* with text *text*.\n",
      "        \n",
      "        In the simplest form, the text is placed at *xy*.\n",
      "        \n",
      "        Optionally, the text can be displayed in another position *xytext*.\n",
      "        An arrow pointing from the text to the annotated point *xy* can then\n",
      "        be added by defining *arrowprops*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        text : str\n",
      "            The text of the annotation.\n",
      "        \n",
      "        xy : (float, float)\n",
      "            The point *(x, y)* to annotate. The coordinate system is determined\n",
      "            by *xycoords*.\n",
      "        \n",
      "        xytext : (float, float), default: *xy*\n",
      "            The position *(x, y)* to place the text at. The coordinate system\n",
      "            is determined by *textcoords*.\n",
      "        \n",
      "        xycoords : str or `.Artist` or `.Transform` or callable or (float, float), default: 'data'\n",
      "        \n",
      "            The coordinate system that *xy* is given in. The following types\n",
      "            of values are supported:\n",
      "        \n",
      "            - One of the following strings:\n",
      "        \n",
      "              ==================== ============================================\n",
      "              Value                Description\n",
      "              ==================== ============================================\n",
      "              'figure points'      Points from the lower left of the figure\n",
      "              'figure pixels'      Pixels from the lower left of the figure\n",
      "              'figure fraction'    Fraction of figure from lower left\n",
      "              'subfigure points'   Points from the lower left of the subfigure\n",
      "              'subfigure pixels'   Pixels from the lower left of the subfigure\n",
      "              'subfigure fraction' Fraction of subfigure from lower left\n",
      "              'axes points'        Points from lower left corner of axes\n",
      "              'axes pixels'        Pixels from lower left corner of axes\n",
      "              'axes fraction'      Fraction of axes from lower left\n",
      "              'data'               Use the coordinate system of the object\n",
      "                                   being annotated (default)\n",
      "              'polar'              *(theta, r)* if not native 'data'\n",
      "                                   coordinates\n",
      "              ==================== ============================================\n",
      "        \n",
      "              Note that 'subfigure pixels' and 'figure pixels' are the same\n",
      "              for the parent figure, so users who want code that is usable in\n",
      "              a subfigure can use 'subfigure pixels'.\n",
      "        \n",
      "            - An `.Artist`: *xy* is interpreted as a fraction of the artist's\n",
      "              `~matplotlib.transforms.Bbox`. E.g. *(0, 0)* would be the lower\n",
      "              left corner of the bounding box and *(0.5, 1)* would be the\n",
      "              center top of the bounding box.\n",
      "        \n",
      "            - A `.Transform` to transform *xy* to screen coordinates.\n",
      "        \n",
      "            - A function with one of the following signatures::\n",
      "        \n",
      "                def transform(renderer) -> Bbox\n",
      "                def transform(renderer) -> Transform\n",
      "        \n",
      "              where *renderer* is a `.RendererBase` subclass.\n",
      "        \n",
      "              The result of the function is interpreted like the `.Artist` and\n",
      "              `.Transform` cases above.\n",
      "        \n",
      "            - A tuple *(xcoords, ycoords)* specifying separate coordinate\n",
      "              systems for *x* and *y*. *xcoords* and *ycoords* must each be\n",
      "              of one of the above described types.\n",
      "        \n",
      "            See :ref:`plotting-guide-annotation` for more details.\n",
      "        \n",
      "        textcoords : str or `.Artist` or `.Transform` or callable or (float, float), default: value of *xycoords*\n",
      "            The coordinate system that *xytext* is given in.\n",
      "        \n",
      "            All *xycoords* values are valid as well as the following\n",
      "            strings:\n",
      "        \n",
      "            =================   =========================================\n",
      "            Value               Description\n",
      "            =================   =========================================\n",
      "            'offset points'     Offset (in points) from the *xy* value\n",
      "            'offset pixels'     Offset (in pixels) from the *xy* value\n",
      "            =================   =========================================\n",
      "        \n",
      "        arrowprops : dict, optional\n",
      "            The properties used to draw a `.FancyArrowPatch` arrow between the\n",
      "            positions *xy* and *xytext*.  Defaults to None, i.e. no arrow is\n",
      "            drawn.\n",
      "        \n",
      "            For historical reasons there are two different ways to specify\n",
      "            arrows, \"simple\" and \"fancy\":\n",
      "        \n",
      "            **Simple arrow:**\n",
      "        \n",
      "            If *arrowprops* does not contain the key 'arrowstyle' the\n",
      "            allowed keys are:\n",
      "        \n",
      "            ==========   ======================================================\n",
      "            Key          Description\n",
      "            ==========   ======================================================\n",
      "            width        The width of the arrow in points\n",
      "            headwidth    The width of the base of the arrow head in points\n",
      "            headlength   The length of the arrow head in points\n",
      "            shrink       Fraction of total length to shrink from both ends\n",
      "            ?            Any key to :class:`matplotlib.patches.FancyArrowPatch`\n",
      "            ==========   ======================================================\n",
      "        \n",
      "            The arrow is attached to the edge of the text box, the exact\n",
      "            position (corners or centers) depending on where it's pointing to.\n",
      "        \n",
      "            **Fancy arrow:**\n",
      "        \n",
      "            This is used if 'arrowstyle' is provided in the *arrowprops*.\n",
      "        \n",
      "            Valid keys are the following `~matplotlib.patches.FancyArrowPatch`\n",
      "            parameters:\n",
      "        \n",
      "            ===============  ==================================================\n",
      "            Key              Description\n",
      "            ===============  ==================================================\n",
      "            arrowstyle       the arrow style\n",
      "            connectionstyle  the connection style\n",
      "            relpos           see below; default is (0.5, 0.5)\n",
      "            patchA           default is bounding box of the text\n",
      "            patchB           default is None\n",
      "            shrinkA          default is 2 points\n",
      "            shrinkB          default is 2 points\n",
      "            mutation_scale   default is text size (in points)\n",
      "            mutation_aspect  default is 1.\n",
      "            ?                any key for :class:`matplotlib.patches.PathPatch`\n",
      "            ===============  ==================================================\n",
      "        \n",
      "            The exact starting point position of the arrow is defined by\n",
      "            *relpos*. It's a tuple of relative coordinates of the text box,\n",
      "            where (0, 0) is the lower left corner and (1, 1) is the upper\n",
      "            right corner. Values <0 and >1 are supported and specify points\n",
      "            outside the text box. By default (0.5, 0.5) the starting point is\n",
      "            centered in the text box.\n",
      "        \n",
      "        annotation_clip : bool or None, default: None\n",
      "            Whether to clip (i.e. not draw) the annotation when the annotation\n",
      "            point *xy* is outside the axes area.\n",
      "        \n",
      "            - If *True*, the annotation will be clipped when *xy* is outside\n",
      "              the axes.\n",
      "            - If *False*, the annotation will always be drawn.\n",
      "            - If *None*, the annotation will be clipped when *xy* is outside\n",
      "              the axes and *xycoords* is 'data'.\n",
      "        \n",
      "        **kwargs\n",
      "            Additional kwargs are passed to `~matplotlib.text.Text`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.Annotation`\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        :ref:`plotting-guide-annotation`\n",
      "    \n",
      "    arrow(x, y, dx, dy, **kwargs)\n",
      "        Add an arrow to the Axes.\n",
      "        \n",
      "        This draws an arrow from ``(x, y)`` to ``(x+dx, y+dy)``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : float\n",
      "            The x and y coordinates of the arrow base.\n",
      "        \n",
      "        dx, dy : float\n",
      "            The length of the arrow along x and y direction.\n",
      "        \n",
      "        width : float, default: 0.001\n",
      "            Width of full arrow tail.\n",
      "        \n",
      "        length_includes_head : bool, default: False\n",
      "            True if head is to be counted in calculating the length.\n",
      "        \n",
      "        head_width : float or None, default: 3*width\n",
      "            Total width of the full arrow head.\n",
      "        \n",
      "        head_length : float or None, default: 1.5*head_width\n",
      "            Length of arrow head.\n",
      "        \n",
      "        shape : {'full', 'left', 'right'}, default: 'full'\n",
      "            Draw the left-half, right-half, or full arrow.\n",
      "        \n",
      "        overhang : float, default: 0\n",
      "            Fraction that the arrow is swept back (0 overhang means\n",
      "            triangular shape). Can be negative or greater than one.\n",
      "        \n",
      "        head_starts_at_zero : bool, default: False\n",
      "            If True, the head starts being drawn at coordinate 0\n",
      "            instead of ending at coordinate 0.\n",
      "        \n",
      "        **kwargs\n",
      "            `.Patch` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: unknown\n",
      "            animated: bool\n",
      "            antialiased or aa: bool or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color: color\n",
      "            edgecolor or ec: color or None\n",
      "            facecolor or fc: color or None\n",
      "            figure: `.Figure`\n",
      "            fill: bool\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float or None\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.FancyArrow`\n",
      "            The created `.FancyArrow` object.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The resulting arrow is affected by the Axes aspect ratio and limits.\n",
      "        This may produce an arrow whose head is not square with its stem. To\n",
      "        create an arrow whose head is square with its stem,\n",
      "        use :meth:`annotate` for example:\n",
      "        \n",
      "        >>> ax.annotate(\"\", xy=(0.5, 0.5), xytext=(0, 0),\n",
      "        ...             arrowprops=dict(arrowstyle=\"->\"))\n",
      "    \n",
      "    autoscale(enable=True, axis='both', tight=None)\n",
      "        Autoscale the axis view to the data (toggle).\n",
      "        \n",
      "        Convenience method for simple axis view autoscaling.\n",
      "        It turns autoscaling on or off, and then,\n",
      "        if autoscaling for either axis is on, it performs\n",
      "        the autoscaling on the specified axis or Axes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        enable : bool or None, default: True\n",
      "            True turns autoscaling on, False turns it off.\n",
      "            None leaves the autoscaling state unchanged.\n",
      "        axis : {'both', 'x', 'y'}, default: 'both'\n",
      "            The axis on which to operate.  (For 3D Axes, *axis* can also be set\n",
      "            to 'z', and 'both' refers to all three axes.)\n",
      "        tight : bool or None, default: None\n",
      "            If True, first set the margins to zero.  Then, this argument is\n",
      "            forwarded to `~.axes.Axes.autoscale_view` (regardless of\n",
      "            its value); see the description of its behavior there.\n",
      "    \n",
      "    autumn()\n",
      "        Set the colormap to 'autumn'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    axes(arg=None, **kwargs)\n",
      "        Add an Axes to the current figure and make it the current Axes.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            plt.axes()\n",
      "            plt.axes(rect, projection=None, polar=False, **kwargs)\n",
      "            plt.axes(ax)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        arg : None or 4-tuple\n",
      "            The exact behavior of this function depends on the type:\n",
      "        \n",
      "            - *None*: A new full window Axes is added using\n",
      "              ``subplot(**kwargs)``.\n",
      "            - 4-tuple of floats *rect* = ``[left, bottom, width, height]``.\n",
      "              A new Axes is added with dimensions *rect* in normalized\n",
      "              (0, 1) units using `~.Figure.add_axes` on the current figure.\n",
      "        \n",
      "        projection : {None, 'aitoff', 'hammer', 'lambert', 'mollweide', 'polar', 'rectilinear', str}, optional\n",
      "            The projection type of the `~.axes.Axes`. *str* is the name of\n",
      "            a custom projection, see `~matplotlib.projections`. The default\n",
      "            None results in a 'rectilinear' projection.\n",
      "        \n",
      "        polar : bool, default: False\n",
      "            If True, equivalent to projection='polar'.\n",
      "        \n",
      "        sharex, sharey : `~.axes.Axes`, optional\n",
      "            Share the x or y `~matplotlib.axis` with sharex and/or sharey.\n",
      "            The axis will have the same limits, ticks, and scale as the axis\n",
      "            of the shared Axes.\n",
      "        \n",
      "        label : str\n",
      "            A label for the returned Axes.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~.axes.Axes`, or a subclass of `~.axes.Axes`\n",
      "            The returned axes class depends on the projection used. It is\n",
      "            `~.axes.Axes` if rectilinear projection is used and\n",
      "            `.projections.polar.PolarAxes` if polar projection is used.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            This method also takes the keyword arguments for\n",
      "            the returned Axes class. The keyword arguments for the\n",
      "            rectilinear Axes class `~.axes.Axes` can be found in\n",
      "            the following table but there might also be other keyword\n",
      "            arguments if another projection is used, see the actual Axes\n",
      "            class.\n",
      "        \n",
      "            Properties:\n",
      "            adjustable: {'box', 'datalim'}\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            anchor: (float, float) or {'C', 'SW', 'S', 'SE', 'E', 'NE', ...}\n",
      "            animated: bool\n",
      "            aspect: {'auto', 'equal'} or float\n",
      "            autoscale_on: bool\n",
      "            autoscalex_on: unknown\n",
      "            autoscaley_on: unknown\n",
      "            axes_locator: Callable[[Axes, Renderer], Bbox]\n",
      "            axisbelow: bool or 'line'\n",
      "            box_aspect: float or None\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            facecolor or fc: color\n",
      "            figure: `.Figure`\n",
      "            frame_on: bool\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            mouseover: bool\n",
      "            navigate: bool\n",
      "            navigate_mode: unknown\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            position: [left, bottom, width, height] or `~matplotlib.transforms.Bbox`\n",
      "            prop_cycle: unknown\n",
      "            rasterization_zorder: float or None\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            title: str\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xbound: unknown\n",
      "            xlabel: str\n",
      "            xlim: (bottom: float, top: float)\n",
      "            xmargin: float greater than -0.5\n",
      "            xscale: unknown\n",
      "            xticklabels: unknown\n",
      "            xticks: unknown\n",
      "            ybound: unknown\n",
      "            ylabel: str\n",
      "            ylim: (bottom: float, top: float)\n",
      "            ymargin: float greater than -0.5\n",
      "            yscale: unknown\n",
      "            yticklabels: unknown\n",
      "            yticks: unknown\n",
      "            zorder: float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        If the figure already has an Axes with key (*args*,\n",
      "        *kwargs*) then it will simply make that axes current and\n",
      "        return it.  This behavior is deprecated. Meanwhile, if you do\n",
      "        not want this behavior (i.e., you want to force the creation of a\n",
      "        new axes), you must use a unique set of args and kwargs.  The Axes\n",
      "        *label* attribute has been exposed for this purpose: if you want\n",
      "        two Axes that are otherwise identical to be added to the figure,\n",
      "        make sure you give them unique labels.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Figure.add_axes\n",
      "        .pyplot.subplot\n",
      "        .Figure.add_subplot\n",
      "        .Figure.subplots\n",
      "        .pyplot.subplots\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            # Creating a new full window Axes\n",
      "            plt.axes()\n",
      "        \n",
      "            # Creating a new Axes with specified dimensions and a grey background\n",
      "            plt.axes((left, bottom, width, height), facecolor='grey')\n",
      "    \n",
      "    axhline(y=0, xmin=0, xmax=1, **kwargs)\n",
      "        Add a horizontal line across the Axes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : float, default: 0\n",
      "            y position in data coordinates of the horizontal line.\n",
      "        \n",
      "        xmin : float, default: 0\n",
      "            Should be between 0 and 1, 0 being the far left of the plot, 1 the\n",
      "            far right of the plot.\n",
      "        \n",
      "        xmax : float, default: 1\n",
      "            Should be between 0 and 1, 0 being the far left of the plot, 1 the\n",
      "            far right of the plot.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.lines.Line2D`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            Valid keyword arguments are `.Line2D` properties, with the\n",
      "            exception of 'transform':\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        hlines : Add horizontal lines in data coordinates.\n",
      "        axhspan : Add a horizontal span (rectangle) across the axis.\n",
      "        axline : Add a line with an arbitrary slope.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        * draw a thick red hline at 'y' = 0 that spans the xrange::\n",
      "        \n",
      "            >>> axhline(linewidth=4, color='r')\n",
      "        \n",
      "        * draw a default hline at 'y' = 1 that spans the xrange::\n",
      "        \n",
      "            >>> axhline(y=1)\n",
      "        \n",
      "        * draw a default hline at 'y' = .5 that spans the middle half of\n",
      "          the xrange::\n",
      "        \n",
      "            >>> axhline(y=.5, xmin=0.25, xmax=0.75)\n",
      "    \n",
      "    axhspan(ymin, ymax, xmin=0, xmax=1, **kwargs)\n",
      "        Add a horizontal span (rectangle) across the Axes.\n",
      "        \n",
      "        The rectangle spans from *ymin* to *ymax* vertically, and, by default,\n",
      "        the whole x-axis horizontally.  The x-span can be set using *xmin*\n",
      "        (default: 0) and *xmax* (default: 1) which are in axis units; e.g.\n",
      "        ``xmin = 0.5`` always refers to the middle of the x-axis regardless of\n",
      "        the limits set by `~.Axes.set_xlim`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        ymin : float\n",
      "            Lower y-coordinate of the span, in data units.\n",
      "        ymax : float\n",
      "            Upper y-coordinate of the span, in data units.\n",
      "        xmin : float, default: 0\n",
      "            Lower x-coordinate of the span, in x-axis (0-1) units.\n",
      "        xmax : float, default: 1\n",
      "            Upper x-coordinate of the span, in x-axis (0-1) units.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.patches.Polygon`\n",
      "            Horizontal span (rectangle) from (xmin, ymin) to (xmax, ymax).\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.patches.Polygon` properties\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            closed: bool\n",
      "            color: color\n",
      "            edgecolor or ec: color or None\n",
      "            facecolor or fc: color or None\n",
      "            figure: `.Figure`\n",
      "            fill: bool\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float or None\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xy: (N, 2) array-like\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        axvspan : Add a vertical span across the Axes.\n",
      "    \n",
      "    axis(*args, emit=True, **kwargs)\n",
      "        Convenience method to get or set some axis properties.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "          xmin, xmax, ymin, ymax = axis()\n",
      "          xmin, xmax, ymin, ymax = axis([xmin, xmax, ymin, ymax])\n",
      "          xmin, xmax, ymin, ymax = axis(option)\n",
      "          xmin, xmax, ymin, ymax = axis(**kwargs)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xmin, xmax, ymin, ymax : float, optional\n",
      "            The axis limits to be set.  This can also be achieved using ::\n",
      "        \n",
      "                ax.set(xlim=(xmin, xmax), ylim=(ymin, ymax))\n",
      "        \n",
      "        option : bool or str\n",
      "            If a bool, turns axis lines and labels on or off. If a string,\n",
      "            possible values are:\n",
      "        \n",
      "            ======== ==========================================================\n",
      "            Value    Description\n",
      "            ======== ==========================================================\n",
      "            'on'     Turn on axis lines and labels. Same as ``True``.\n",
      "            'off'    Turn off axis lines and labels. Same as ``False``.\n",
      "            'equal'  Set equal scaling (i.e., make circles circular) by\n",
      "                     changing axis limits. This is the same as\n",
      "                     ``ax.set_aspect('equal', adjustable='datalim')``.\n",
      "                     Explicit data limits may not be respected in this case.\n",
      "            'scaled' Set equal scaling (i.e., make circles circular) by\n",
      "                     changing dimensions of the plot box. This is the same as\n",
      "                     ``ax.set_aspect('equal', adjustable='box', anchor='C')``.\n",
      "                     Additionally, further autoscaling will be disabled.\n",
      "            'tight'  Set limits just large enough to show all data, then\n",
      "                     disable further autoscaling.\n",
      "            'auto'   Automatic scaling (fill plot box with data).\n",
      "            'image'  'scaled' with axis limits equal to data limits.\n",
      "            'square' Square plot; similar to 'scaled', but initially forcing\n",
      "                     ``xmax-xmin == ymax-ymin``.\n",
      "            ======== ==========================================================\n",
      "        \n",
      "        emit : bool, default: True\n",
      "            Whether observers are notified of the axis limit change.\n",
      "            This option is passed on to `~.Axes.set_xlim` and\n",
      "            `~.Axes.set_ylim`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xmin, xmax, ymin, ymax : float\n",
      "            The axis limits.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.axes.Axes.set_xlim\n",
      "        matplotlib.axes.Axes.set_ylim\n",
      "    \n",
      "    axline(xy1, xy2=None, *, slope=None, **kwargs)\n",
      "        Add an infinitely long straight line.\n",
      "        \n",
      "        The line can be defined either by two points *xy1* and *xy2*, or\n",
      "        by one point *xy1* and a *slope*.\n",
      "        \n",
      "        This draws a straight line \"on the screen\", regardless of the x and y\n",
      "        scales, and is thus also suitable for drawing exponential decays in\n",
      "        semilog plots, power laws in loglog plots, etc. However, *slope*\n",
      "        should only be used with linear scales; It has no clear meaning for\n",
      "        all other scales, and thus the behavior is undefined. Please specify\n",
      "        the line using the points *xy1*, *xy2* for non-linear scales.\n",
      "        \n",
      "        The *transform* keyword argument only applies to the points *xy1*,\n",
      "        *xy2*. The *slope* (if given) is always in data coordinates. This can\n",
      "        be used e.g. with ``ax.transAxes`` for drawing grid lines with a fixed\n",
      "        slope.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xy1, xy2 : (float, float)\n",
      "            Points for the line to pass through.\n",
      "            Either *xy2* or *slope* has to be given.\n",
      "        slope : float, optional\n",
      "            The slope of the line. Either *xy2* or *slope* has to be given.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.Line2D`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            Valid kwargs are `.Line2D` properties\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        axhline : for horizontal lines\n",
      "        axvline : for vertical lines\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Draw a thick red line passing through (0, 0) and (1, 1)::\n",
      "        \n",
      "            >>> axline((0, 0), (1, 1), linewidth=4, color='r')\n",
      "    \n",
      "    axvline(x=0, ymin=0, ymax=1, **kwargs)\n",
      "        Add a vertical line across the Axes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : float, default: 0\n",
      "            x position in data coordinates of the vertical line.\n",
      "        \n",
      "        ymin : float, default: 0\n",
      "            Should be between 0 and 1, 0 being the bottom of the plot, 1 the\n",
      "            top of the plot.\n",
      "        \n",
      "        ymax : float, default: 1\n",
      "            Should be between 0 and 1, 0 being the bottom of the plot, 1 the\n",
      "            top of the plot.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.lines.Line2D`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            Valid keyword arguments are `.Line2D` properties, with the\n",
      "            exception of 'transform':\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        vlines : Add vertical lines in data coordinates.\n",
      "        axvspan : Add a vertical span (rectangle) across the axis.\n",
      "        axline : Add a line with an arbitrary slope.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        * draw a thick red vline at *x* = 0 that spans the yrange::\n",
      "        \n",
      "            >>> axvline(linewidth=4, color='r')\n",
      "        \n",
      "        * draw a default vline at *x* = 1 that spans the yrange::\n",
      "        \n",
      "            >>> axvline(x=1)\n",
      "        \n",
      "        * draw a default vline at *x* = .5 that spans the middle half of\n",
      "          the yrange::\n",
      "        \n",
      "            >>> axvline(x=.5, ymin=0.25, ymax=0.75)\n",
      "    \n",
      "    axvspan(xmin, xmax, ymin=0, ymax=1, **kwargs)\n",
      "        Add a vertical span (rectangle) across the Axes.\n",
      "        \n",
      "        The rectangle spans from *xmin* to *xmax* horizontally, and, by\n",
      "        default, the whole y-axis vertically.  The y-span can be set using\n",
      "        *ymin* (default: 0) and *ymax* (default: 1) which are in axis units;\n",
      "        e.g. ``ymin = 0.5`` always refers to the middle of the y-axis\n",
      "        regardless of the limits set by `~.Axes.set_ylim`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xmin : float\n",
      "            Lower x-coordinate of the span, in data units.\n",
      "        xmax : float\n",
      "            Upper x-coordinate of the span, in data units.\n",
      "        ymin : float, default: 0\n",
      "            Lower y-coordinate of the span, in y-axis units (0-1).\n",
      "        ymax : float, default: 1\n",
      "            Upper y-coordinate of the span, in y-axis units (0-1).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.patches.Polygon`\n",
      "            Vertical span (rectangle) from (xmin, ymin) to (xmax, ymax).\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.patches.Polygon` properties\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            closed: bool\n",
      "            color: color\n",
      "            edgecolor or ec: color or None\n",
      "            facecolor or fc: color or None\n",
      "            figure: `.Figure`\n",
      "            fill: bool\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float or None\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xy: (N, 2) array-like\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        axhspan : Add a horizontal span across the Axes.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Draw a vertical, green, translucent rectangle from x = 1.25 to\n",
      "        x = 1.55 that spans the yrange of the Axes.\n",
      "        \n",
      "        >>> axvspan(1.25, 1.55, facecolor='g', alpha=0.5)\n",
      "    \n",
      "    bar(x, height, width=0.8, bottom=None, *, align='center', data=None, **kwargs)\n",
      "        Make a bar plot.\n",
      "        \n",
      "        The bars are positioned at *x* with the given *align*\\ment. Their\n",
      "        dimensions are given by *height* and *width*. The vertical baseline\n",
      "        is *bottom* (default 0).\n",
      "        \n",
      "        Many parameters can take either a single value applying to all bars\n",
      "        or a sequence of values, one for each bar.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : float or array-like\n",
      "            The x coordinates of the bars. See also *align* for the\n",
      "            alignment of the bars to the coordinates.\n",
      "        \n",
      "        height : float or array-like\n",
      "            The height(s) of the bars.\n",
      "        \n",
      "        width : float or array-like, default: 0.8\n",
      "            The width(s) of the bars.\n",
      "        \n",
      "        bottom : float or array-like, default: 0\n",
      "            The y coordinate(s) of the bottom side(s) of the bars.\n",
      "        \n",
      "        align : {'center', 'edge'}, default: 'center'\n",
      "            Alignment of the bars to the *x* coordinates:\n",
      "        \n",
      "            - 'center': Center the base on the *x* positions.\n",
      "            - 'edge': Align the left edges of the bars with the *x* positions.\n",
      "        \n",
      "            To align the bars on the right edge pass a negative *width* and\n",
      "            ``align='edge'``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.BarContainer`\n",
      "            Container with all the bars and optionally errorbars.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        color : color or list of color, optional\n",
      "            The colors of the bar faces.\n",
      "        \n",
      "        edgecolor : color or list of color, optional\n",
      "            The colors of the bar edges.\n",
      "        \n",
      "        linewidth : float or array-like, optional\n",
      "            Width of the bar edge(s). If 0, don't draw edges.\n",
      "        \n",
      "        tick_label : str or list of str, optional\n",
      "            The tick labels of the bars.\n",
      "            Default: None (Use default numeric labels.)\n",
      "        \n",
      "        label : str or list of str, optional\n",
      "            A single label is attached to the resulting `.BarContainer` as a\n",
      "            label for the whole dataset.\n",
      "            If a list is provided, it must be the same length as *x* and\n",
      "            labels the individual bars. Repeated labels are not de-duplicated\n",
      "            and will cause repeated label entries, so this is best used when\n",
      "            bars also differ in style (e.g., by passing a list to *color*.)\n",
      "        \n",
      "        xerr, yerr : float or array-like of shape(N,) or shape(2, N), optional\n",
      "            If not *None*, add horizontal / vertical errorbars to the bar tips.\n",
      "            The values are +/- sizes relative to the data:\n",
      "        \n",
      "            - scalar: symmetric +/- values for all bars\n",
      "            - shape(N,): symmetric +/- values for each bar\n",
      "            - shape(2, N): Separate - and + values for each bar. First row\n",
      "              contains the lower errors, the second row contains the upper\n",
      "              errors.\n",
      "            - *None*: No errorbar. (Default)\n",
      "        \n",
      "            See :doc:`/gallery/statistics/errorbar_features` for an example on\n",
      "            the usage of *xerr* and *yerr*.\n",
      "        \n",
      "        ecolor : color or list of color, default: 'black'\n",
      "            The line color of the errorbars.\n",
      "        \n",
      "        capsize : float, default: :rc:`errorbar.capsize`\n",
      "           The length of the error bar caps in points.\n",
      "        \n",
      "        error_kw : dict, optional\n",
      "            Dictionary of keyword arguments to be passed to the\n",
      "            `~.Axes.errorbar` method. Values of *ecolor* or *capsize* defined\n",
      "            here take precedence over the independent keyword arguments.\n",
      "        \n",
      "        log : bool, default: False\n",
      "            If *True*, set the y-axis to be log scale.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs : `.Rectangle` properties\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            angle: unknown\n",
      "            animated: bool\n",
      "            antialiased or aa: bool or None\n",
      "            bounds: (left, bottom, width, height)\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color: color\n",
      "            edgecolor or ec: color or None\n",
      "            facecolor or fc: color or None\n",
      "            figure: `.Figure`\n",
      "            fill: bool\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            height: unknown\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float or None\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            width: unknown\n",
      "            x: unknown\n",
      "            xy: (float, float)\n",
      "            y: unknown\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        barh : Plot a horizontal bar plot.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Stacked bars can be achieved by passing individual *bottom* values per\n",
      "        bar. See :doc:`/gallery/lines_bars_and_markers/bar_stacked`.\n",
      "    \n",
      "    bar_label(container, labels=None, *, fmt='%g', label_type='edge', padding=0, **kwargs)\n",
      "        Label a bar plot.\n",
      "        \n",
      "        Adds labels to bars in the given `.BarContainer`.\n",
      "        You may need to adjust the axis limits to fit the labels.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        container : `.BarContainer`\n",
      "            Container with all the bars and optionally errorbars, likely\n",
      "            returned from `.bar` or `.barh`.\n",
      "        \n",
      "        labels : array-like, optional\n",
      "            A list of label texts, that should be displayed. If not given, the\n",
      "            label texts will be the data values formatted with *fmt*.\n",
      "        \n",
      "        fmt : str, default: '%g'\n",
      "            A format string for the label.\n",
      "        \n",
      "        label_type : {'edge', 'center'}, default: 'edge'\n",
      "            The label type. Possible values:\n",
      "        \n",
      "            - 'edge': label placed at the end-point of the bar segment, and the\n",
      "              value displayed will be the position of that end-point.\n",
      "            - 'center': label placed in the center of the bar segment, and the\n",
      "              value displayed will be the length of that segment.\n",
      "              (useful for stacked bars, i.e.,\n",
      "              :doc:`/gallery/lines_bars_and_markers/bar_label_demo`)\n",
      "        \n",
      "        padding : float, default: 0\n",
      "            Distance of label from the end of the bar, in points.\n",
      "        \n",
      "        **kwargs\n",
      "            Any remaining keyword arguments are passed through to\n",
      "            `.Axes.annotate`. The alignment parameters (\n",
      "            *horizontalalignment* / *ha*, *verticalalignment* / *va*) are\n",
      "            not supported because the labels are automatically aligned to\n",
      "            the bars.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Text`\n",
      "            A list of `.Text` instances for the labels.\n",
      "    \n",
      "    barbs(*args, data=None, **kwargs)\n",
      "        Plot a 2D field of barbs.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "          barbs([X, Y], U, V, [C], **kwargs)\n",
      "        \n",
      "        Where *X*, *Y* define the barb locations, *U*, *V* define the barb\n",
      "        directions, and *C* optionally sets the color.\n",
      "        \n",
      "        All arguments may be 1D or 2D. *U*, *V*, *C* may be masked arrays, but masked\n",
      "        *X*, *Y* are not supported at present.\n",
      "        \n",
      "        Barbs are traditionally used in meteorology as a way to plot the speed\n",
      "        and direction of wind observations, but can technically be used to\n",
      "        plot any two dimensional vector quantity.  As opposed to arrows, which\n",
      "        give vector magnitude by the length of the arrow, the barbs give more\n",
      "        quantitative information about the vector magnitude by putting slanted\n",
      "        lines or a triangle for various increments in magnitude, as show\n",
      "        schematically below::\n",
      "        \n",
      "          :                   /\\    \\\n",
      "          :                  /  \\    \\\n",
      "          :                 /    \\    \\    \\\n",
      "          :                /      \\    \\    \\\n",
      "          :               ------------------------------\n",
      "        \n",
      "        The largest increment is given by a triangle (or \"flag\"). After those\n",
      "        come full lines (barbs). The smallest increment is a half line.  There\n",
      "        is only, of course, ever at most 1 half line.  If the magnitude is\n",
      "        small and only needs a single half-line and no full lines or\n",
      "        triangles, the half-line is offset from the end of the barb so that it\n",
      "        can be easily distinguished from barbs with a single full line.  The\n",
      "        magnitude for the barb shown above would nominally be 65, using the\n",
      "        standard increments of 50, 10, and 5.\n",
      "        \n",
      "        See also https://en.wikipedia.org/wiki/Wind_barb.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X, Y : 1D or 2D array-like, optional\n",
      "            The x and y coordinates of the barb locations. See *pivot* for how the\n",
      "            barbs are drawn to the x, y positions.\n",
      "        \n",
      "            If not given, they will be generated as a uniform integer meshgrid based\n",
      "            on the dimensions of *U* and *V*.\n",
      "        \n",
      "            If *X* and *Y* are 1D but *U*, *V* are 2D, *X*, *Y* are expanded to 2D\n",
      "            using ``X, Y = np.meshgrid(X, Y)``. In this case ``len(X)`` and ``len(Y)``\n",
      "            must match the column and row dimensions of *U* and *V*.\n",
      "        \n",
      "        U, V : 1D or 2D array-like\n",
      "            The x and y components of the barb shaft.\n",
      "        \n",
      "        C : 1D or 2D array-like, optional\n",
      "            Numeric data that defines the barb colors by colormapping via *norm* and\n",
      "            *cmap*.\n",
      "        \n",
      "            This does not support explicit colors. If you want to set colors directly,\n",
      "            use *barbcolor* instead.\n",
      "        \n",
      "        length : float, default: 7\n",
      "            Length of the barb in points; the other parts of the barb\n",
      "            are scaled against this.\n",
      "        \n",
      "        pivot : {'tip', 'middle'} or float, default: 'tip'\n",
      "            The part of the arrow that is anchored to the *X*, *Y* grid. The barb\n",
      "            rotates about this point. This can also be a number, which shifts the\n",
      "            start of the barb that many points away from grid point.\n",
      "        \n",
      "        barbcolor : color or color sequence\n",
      "            The color of all parts of the barb except for the flags.  This parameter\n",
      "            is analogous to the *edgecolor* parameter for polygons, which can be used\n",
      "            instead. However this parameter will override facecolor.\n",
      "        \n",
      "        flagcolor : color or color sequence\n",
      "            The color of any flags on the barb.  This parameter is analogous to the\n",
      "            *facecolor* parameter for polygons, which can be used instead. However,\n",
      "            this parameter will override facecolor.  If this is not set (and *C* has\n",
      "            not either) then *flagcolor* will be set to match *barbcolor* so that the\n",
      "            barb has a uniform color. If *C* has been set, *flagcolor* has no effect.\n",
      "        \n",
      "        sizes : dict, optional\n",
      "            A dictionary of coefficients specifying the ratio of a given\n",
      "            feature to the length of the barb. Only those values one wishes to\n",
      "            override need to be included.  These features include:\n",
      "        \n",
      "            - 'spacing' - space between features (flags, full/half barbs)\n",
      "            - 'height' - height (distance from shaft to top) of a flag or full barb\n",
      "            - 'width' - width of a flag, twice the width of a full barb\n",
      "            - 'emptybarb' - radius of the circle used for low magnitudes\n",
      "        \n",
      "        fill_empty : bool, default: False\n",
      "            Whether the empty barbs (circles) that are drawn should be filled with\n",
      "            the flag color.  If they are not filled, the center is transparent.\n",
      "        \n",
      "        rounding : bool, default: True\n",
      "            Whether the vector magnitude should be rounded when allocating barb\n",
      "            components.  If True, the magnitude is rounded to the nearest multiple\n",
      "            of the half-barb increment.  If False, the magnitude is simply truncated\n",
      "            to the next lowest multiple.\n",
      "        \n",
      "        barb_increments : dict, optional\n",
      "            A dictionary of increments specifying values to associate with\n",
      "            different parts of the barb. Only those values one wishes to\n",
      "            override need to be included.\n",
      "        \n",
      "            - 'half' - half barbs (Default is 5)\n",
      "            - 'full' - full barbs (Default is 10)\n",
      "            - 'flag' - flags (default is 50)\n",
      "        \n",
      "        flip_barb : bool or array-like of bool, default: False\n",
      "            Whether the lines and flags should point opposite to normal.\n",
      "            Normal behavior is for the barbs and lines to point right (comes from wind\n",
      "            barbs having these features point towards low pressure in the Northern\n",
      "            Hemisphere).\n",
      "        \n",
      "            A single value is applied to all barbs. Individual barbs can be flipped by\n",
      "            passing a bool array of the same size as *U* and *V*.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        barbs : `~matplotlib.quiver.Barbs`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs\n",
      "            The barbs can further be customized using `.PolyCollection` keyword\n",
      "            arguments:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "    \n",
      "    barh(y, width, height=0.8, left=None, *, align='center', data=None, **kwargs)\n",
      "        Make a horizontal bar plot.\n",
      "        \n",
      "        The bars are positioned at *y* with the given *align*\\ment. Their\n",
      "        dimensions are given by *width* and *height*. The horizontal baseline\n",
      "        is *left* (default 0).\n",
      "        \n",
      "        Many parameters can take either a single value applying to all bars\n",
      "        or a sequence of values, one for each bar.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : float or array-like\n",
      "            The y coordinates of the bars. See also *align* for the\n",
      "            alignment of the bars to the coordinates.\n",
      "        \n",
      "        width : float or array-like\n",
      "            The width(s) of the bars.\n",
      "        \n",
      "        height : float or array-like, default: 0.8\n",
      "            The heights of the bars.\n",
      "        \n",
      "        left : float or array-like, default: 0\n",
      "            The x coordinates of the left side(s) of the bars.\n",
      "        \n",
      "        align : {'center', 'edge'}, default: 'center'\n",
      "            Alignment of the base to the *y* coordinates*:\n",
      "        \n",
      "            - 'center': Center the bars on the *y* positions.\n",
      "            - 'edge': Align the bottom edges of the bars with the *y*\n",
      "              positions.\n",
      "        \n",
      "            To align the bars on the top edge pass a negative *height* and\n",
      "            ``align='edge'``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.BarContainer`\n",
      "            Container with all the bars and optionally errorbars.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        color : color or list of color, optional\n",
      "            The colors of the bar faces.\n",
      "        \n",
      "        edgecolor : color or list of color, optional\n",
      "            The colors of the bar edges.\n",
      "        \n",
      "        linewidth : float or array-like, optional\n",
      "            Width of the bar edge(s). If 0, don't draw edges.\n",
      "        \n",
      "        tick_label : str or list of str, optional\n",
      "            The tick labels of the bars.\n",
      "            Default: None (Use default numeric labels.)\n",
      "        \n",
      "        label : str or list of str, optional\n",
      "            A single label is attached to the resulting `.BarContainer` as a\n",
      "            label for the whole dataset.\n",
      "            If a list is provided, it must be the same length as *y* and\n",
      "            labels the individual bars. Repeated labels are not de-duplicated\n",
      "            and will cause repeated label entries, so this is best used when\n",
      "            bars also differ in style (e.g., by passing a list to *color*.)\n",
      "        \n",
      "        xerr, yerr : float or array-like of shape(N,) or shape(2, N), optional\n",
      "            If not *None*, add horizontal / vertical errorbars to the bar tips.\n",
      "            The values are +/- sizes relative to the data:\n",
      "        \n",
      "            - scalar: symmetric +/- values for all bars\n",
      "            - shape(N,): symmetric +/- values for each bar\n",
      "            - shape(2, N): Separate - and + values for each bar. First row\n",
      "              contains the lower errors, the second row contains the upper\n",
      "              errors.\n",
      "            - *None*: No errorbar. (default)\n",
      "        \n",
      "            See :doc:`/gallery/statistics/errorbar_features` for an example on\n",
      "            the usage of *xerr* and *yerr*.\n",
      "        \n",
      "        ecolor : color or list of color, default: 'black'\n",
      "            The line color of the errorbars.\n",
      "        \n",
      "        capsize : float, default: :rc:`errorbar.capsize`\n",
      "           The length of the error bar caps in points.\n",
      "        \n",
      "        error_kw : dict, optional\n",
      "            Dictionary of keyword arguments to be passed to the\n",
      "            `~.Axes.errorbar` method. Values of *ecolor* or *capsize* defined\n",
      "            here take precedence over the independent keyword arguments.\n",
      "        \n",
      "        log : bool, default: False\n",
      "            If ``True``, set the x-axis to be log scale.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs : `.Rectangle` properties\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            angle: unknown\n",
      "            animated: bool\n",
      "            antialiased or aa: bool or None\n",
      "            bounds: (left, bottom, width, height)\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color: color\n",
      "            edgecolor or ec: color or None\n",
      "            facecolor or fc: color or None\n",
      "            figure: `.Figure`\n",
      "            fill: bool\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            height: unknown\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float or None\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            width: unknown\n",
      "            x: unknown\n",
      "            xy: (float, float)\n",
      "            y: unknown\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        bar : Plot a vertical bar plot.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Stacked bars can be achieved by passing individual *left* values per\n",
      "        bar. See\n",
      "        :doc:`/gallery/lines_bars_and_markers/horizontal_barchart_distribution`.\n",
      "    \n",
      "    bone()\n",
      "        Set the colormap to 'bone'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    box(on=None)\n",
      "        Turn the axes box on or off on the current axes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        on : bool or None\n",
      "            The new `~matplotlib.axes.Axes` box state. If ``None``, toggle\n",
      "            the state.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        :meth:`matplotlib.axes.Axes.set_frame_on`\n",
      "        :meth:`matplotlib.axes.Axes.get_frame_on`\n",
      "    \n",
      "    boxplot(x, notch=None, sym=None, vert=None, whis=None, positions=None, widths=None, patch_artist=None, bootstrap=None, usermedians=None, conf_intervals=None, meanline=None, showmeans=None, showcaps=None, showbox=None, showfliers=None, boxprops=None, labels=None, flierprops=None, medianprops=None, meanprops=None, capprops=None, whiskerprops=None, manage_ticks=True, autorange=False, zorder=None, capwidths=None, *, data=None)\n",
      "        Draw a box and whisker plot.\n",
      "        \n",
      "        The box extends from the first quartile (Q1) to the third\n",
      "        quartile (Q3) of the data, with a line at the median.  The\n",
      "        whiskers extend from the box by 1.5x the inter-quartile range\n",
      "        (IQR).  Flier points are those past the end of the whiskers.\n",
      "        See https://en.wikipedia.org/wiki/Box_plot for reference.\n",
      "        \n",
      "        .. code-block:: none\n",
      "        \n",
      "                  Q1-1.5IQR   Q1   median  Q3   Q3+1.5IQR\n",
      "                               |-----:-----|\n",
      "               o      |--------|     :     |--------|    o  o\n",
      "                               |-----:-----|\n",
      "             flier             <----------->            fliers\n",
      "                                    IQR\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : Array or a sequence of vectors.\n",
      "            The input data.  If a 2D array, a boxplot is drawn for each column\n",
      "            in *x*.  If a sequence of 1D arrays, a boxplot is drawn for each\n",
      "            array in *x*.\n",
      "        \n",
      "        notch : bool, default: False\n",
      "            Whether to draw a notched boxplot (`True`), or a rectangular\n",
      "            boxplot (`False`).  The notches represent the confidence interval\n",
      "            (CI) around the median.  The documentation for *bootstrap*\n",
      "            describes how the locations of the notches are computed by\n",
      "            default, but their locations may also be overridden by setting the\n",
      "            *conf_intervals* parameter.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "                In cases where the values of the CI are less than the\n",
      "                lower quartile or greater than the upper quartile, the\n",
      "                notches will extend beyond the box, giving it a\n",
      "                distinctive \"flipped\" appearance. This is expected\n",
      "                behavior and consistent with other statistical\n",
      "                visualization packages.\n",
      "        \n",
      "        sym : str, optional\n",
      "            The default symbol for flier points.  An empty string ('') hides\n",
      "            the fliers.  If `None`, then the fliers default to 'b+'.  More\n",
      "            control is provided by the *flierprops* parameter.\n",
      "        \n",
      "        vert : bool, default: True\n",
      "            If `True`, draws vertical boxes.\n",
      "            If `False`, draw horizontal boxes.\n",
      "        \n",
      "        whis : float or (float, float), default: 1.5\n",
      "            The position of the whiskers.\n",
      "        \n",
      "            If a float, the lower whisker is at the lowest datum above\n",
      "            ``Q1 - whis*(Q3-Q1)``, and the upper whisker at the highest datum\n",
      "            below ``Q3 + whis*(Q3-Q1)``, where Q1 and Q3 are the first and\n",
      "            third quartiles.  The default value of ``whis = 1.5`` corresponds\n",
      "            to Tukey's original definition of boxplots.\n",
      "        \n",
      "            If a pair of floats, they indicate the percentiles at which to\n",
      "            draw the whiskers (e.g., (5, 95)).  In particular, setting this to\n",
      "            (0, 100) results in whiskers covering the whole range of the data.\n",
      "        \n",
      "            In the edge case where ``Q1 == Q3``, *whis* is automatically set\n",
      "            to (0, 100) (cover the whole range of the data) if *autorange* is\n",
      "            True.\n",
      "        \n",
      "            Beyond the whiskers, data are considered outliers and are plotted\n",
      "            as individual points.\n",
      "        \n",
      "        bootstrap : int, optional\n",
      "            Specifies whether to bootstrap the confidence intervals\n",
      "            around the median for notched boxplots. If *bootstrap* is\n",
      "            None, no bootstrapping is performed, and notches are\n",
      "            calculated using a Gaussian-based asymptotic approximation\n",
      "            (see McGill, R., Tukey, J.W., and Larsen, W.A., 1978, and\n",
      "            Kendall and Stuart, 1967). Otherwise, bootstrap specifies\n",
      "            the number of times to bootstrap the median to determine its\n",
      "            95% confidence intervals. Values between 1000 and 10000 are\n",
      "            recommended.\n",
      "        \n",
      "        usermedians : 1D array-like, optional\n",
      "            A 1D array-like of length ``len(x)``.  Each entry that is not\n",
      "            `None` forces the value of the median for the corresponding\n",
      "            dataset.  For entries that are `None`, the medians are computed\n",
      "            by Matplotlib as normal.\n",
      "        \n",
      "        conf_intervals : array-like, optional\n",
      "            A 2D array-like of shape ``(len(x), 2)``.  Each entry that is not\n",
      "            None forces the location of the corresponding notch (which is\n",
      "            only drawn if *notch* is `True`).  For entries that are `None`,\n",
      "            the notches are computed by the method specified by the other\n",
      "            parameters (e.g., *bootstrap*).\n",
      "        \n",
      "        positions : array-like, optional\n",
      "            The positions of the boxes. The ticks and limits are\n",
      "            automatically set to match the positions. Defaults to\n",
      "            ``range(1, N+1)`` where N is the number of boxes to be drawn.\n",
      "        \n",
      "        widths : float or array-like\n",
      "            The widths of the boxes.  The default is 0.5, or ``0.15*(distance\n",
      "            between extreme positions)``, if that is smaller.\n",
      "        \n",
      "        patch_artist : bool, default: False\n",
      "            If `False` produces boxes with the Line2D artist. Otherwise,\n",
      "            boxes are drawn with Patch artists.\n",
      "        \n",
      "        labels : sequence, optional\n",
      "            Labels for each dataset (one per dataset).\n",
      "        \n",
      "        manage_ticks : bool, default: True\n",
      "            If True, the tick locations and labels will be adjusted to match\n",
      "            the boxplot positions.\n",
      "        \n",
      "        autorange : bool, default: False\n",
      "            When `True` and the data are distributed such that the 25th and\n",
      "            75th percentiles are equal, *whis* is set to (0, 100) such\n",
      "            that the whisker ends are at the minimum and maximum of the data.\n",
      "        \n",
      "        meanline : bool, default: False\n",
      "            If `True` (and *showmeans* is `True`), will try to render the\n",
      "            mean as a line spanning the full width of the box according to\n",
      "            *meanprops* (see below).  Not recommended if *shownotches* is also\n",
      "            True.  Otherwise, means will be shown as points.\n",
      "        \n",
      "        zorder : float, default: ``Line2D.zorder = 2``\n",
      "            The zorder of the boxplot.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dict\n",
      "          A dictionary mapping each component of the boxplot to a list\n",
      "          of the `.Line2D` instances created. That dictionary has the\n",
      "          following keys (assuming vertical boxplots):\n",
      "        \n",
      "          - ``boxes``: the main body of the boxplot showing the\n",
      "            quartiles and the median's confidence intervals if\n",
      "            enabled.\n",
      "        \n",
      "          - ``medians``: horizontal lines at the median of each box.\n",
      "        \n",
      "          - ``whiskers``: the vertical lines extending to the most\n",
      "            extreme, non-outlier data points.\n",
      "        \n",
      "          - ``caps``: the horizontal lines at the ends of the\n",
      "            whiskers.\n",
      "        \n",
      "          - ``fliers``: points representing data that extend beyond\n",
      "            the whiskers (fliers).\n",
      "        \n",
      "          - ``means``: points or lines representing the means.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        showcaps : bool, default: True\n",
      "            Show the caps on the ends of whiskers.\n",
      "        showbox : bool, default: True\n",
      "            Show the central box.\n",
      "        showfliers : bool, default: True\n",
      "            Show the outliers beyond the caps.\n",
      "        showmeans : bool, default: False\n",
      "            Show the arithmetic means.\n",
      "        capprops : dict, default: None\n",
      "            The style of the caps.\n",
      "        capwidths : float or array, default: None\n",
      "            The widths of the caps.\n",
      "        boxprops : dict, default: None\n",
      "            The style of the box.\n",
      "        whiskerprops : dict, default: None\n",
      "            The style of the whiskers.\n",
      "        flierprops : dict, default: None\n",
      "            The style of the fliers.\n",
      "        medianprops : dict, default: None\n",
      "            The style of the median.\n",
      "        meanprops : dict, default: None\n",
      "            The style of the mean.\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        violinplot : Draw an estimate of the probability density function.\n",
      "    \n",
      "    broken_barh(xranges, yrange, *, data=None, **kwargs)\n",
      "        Plot a horizontal sequence of rectangles.\n",
      "        \n",
      "        A rectangle is drawn for each element of *xranges*. All rectangles\n",
      "        have the same vertical position and size defined by *yrange*.\n",
      "        \n",
      "        This is a convenience function for instantiating a\n",
      "        `.BrokenBarHCollection`, adding it to the Axes and autoscaling the\n",
      "        view.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xranges : sequence of tuples (*xmin*, *xwidth*)\n",
      "            The x-positions and extends of the rectangles. For each tuple\n",
      "            (*xmin*, *xwidth*) a rectangle is drawn from *xmin* to *xmin* +\n",
      "            *xwidth*.\n",
      "        yrange : (*ymin*, *yheight*)\n",
      "            The y-position and extend for all the rectangles.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~.collections.BrokenBarHCollection`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        **kwargs : `.BrokenBarHCollection` properties\n",
      "        \n",
      "            Each *kwarg* can be either a single argument applying to all\n",
      "            rectangles, e.g.::\n",
      "        \n",
      "                facecolors='black'\n",
      "        \n",
      "            or a sequence of arguments over which is cycled, e.g.::\n",
      "        \n",
      "                facecolors=('black', 'blue')\n",
      "        \n",
      "            would create interleaving black and blue rectangles.\n",
      "        \n",
      "            Supported keywords:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "    \n",
      "    cla()\n",
      "        Clear the current axes.\n",
      "    \n",
      "    clabel(CS, levels=None, **kwargs)\n",
      "        Label a contour plot.\n",
      "        \n",
      "        Adds labels to line contours in given `.ContourSet`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        CS : `.ContourSet` instance\n",
      "            Line contours to label.\n",
      "        \n",
      "        levels : array-like, optional\n",
      "            A list of level values, that should be labeled. The list must be\n",
      "            a subset of ``CS.levels``. If not given, all levels are labeled.\n",
      "        \n",
      "        **kwargs\n",
      "            All other parameters are documented in `~.ContourLabeler.clabel`.\n",
      "    \n",
      "    clf()\n",
      "        Clear the current figure.\n",
      "    \n",
      "    clim(vmin=None, vmax=None)\n",
      "        Set the color limits of the current image.\n",
      "        \n",
      "        If either *vmin* or *vmax* is None, the image min/max respectively\n",
      "        will be used for color scaling.\n",
      "        \n",
      "        If you want to set the clim of multiple images, use\n",
      "        `~.ScalarMappable.set_clim` on every image, for example::\n",
      "        \n",
      "          for im in gca().get_images():\n",
      "              im.set_clim(0, 0.5)\n",
      "    \n",
      "    close(fig=None)\n",
      "        Close a figure window.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fig : None or int or str or `.Figure`\n",
      "            The figure to close. There are a number of ways to specify this:\n",
      "        \n",
      "            - *None*: the current figure\n",
      "            - `.Figure`: the given `.Figure` instance\n",
      "            - ``int``: a figure number\n",
      "            - ``str``: a figure name\n",
      "            - 'all': all figures\n",
      "    \n",
      "    cohere(x, y, NFFT=256, Fs=2, Fc=0, detrend=<function detrend_none at 0x127a34dc0>, window=<function window_hanning at 0x127a34af0>, noverlap=0, pad_to=None, sides='default', scale_by_freq=None, *, data=None, **kwargs)\n",
      "        Plot the coherence between *x* and *y*.\n",
      "        \n",
      "        Coherence is the normalized cross spectral density:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "          C_{xy} = \\frac{|P_{xy}|^2}{P_{xx}P_{yy}}\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  This can be different from *NFFT*, which specifies the number\n",
      "            of data points used.  While not increasing the actual resolution of the\n",
      "            spectrum (the minimum distance between resolvable peaks), this can give\n",
      "            more points in the plot, allowing for more detail. This corresponds to\n",
      "            the *n* parameter in the call to `~numpy.fft.fft`. The default is None,\n",
      "            which sets *pad_to* equal to *NFFT*\n",
      "        \n",
      "        NFFT : int, default: 256\n",
      "            The number of data points used in each block for the FFT.  A power 2 is\n",
      "            most efficient.  This should *NOT* be used to get zero padding, or the\n",
      "            scaling of the result will be incorrect; use *pad_to* for this instead.\n",
      "        \n",
      "        detrend : {'none', 'mean', 'linear'} or callable, default: 'none'\n",
      "            The function applied to each segment before fft-ing, designed to remove\n",
      "            the mean or linear trend.  Unlike in MATLAB, where the *detrend* parameter\n",
      "            is a vector, in Matplotlib it is a function.  The :mod:`~matplotlib.mlab`\n",
      "            module defines `.detrend_none`, `.detrend_mean`, and `.detrend_linear`,\n",
      "            but you can use a custom function as well.  You can also use a string to\n",
      "            choose one of the functions: 'none' calls `.detrend_none`. 'mean' calls\n",
      "            `.detrend_mean`. 'linear' calls `.detrend_linear`.\n",
      "        \n",
      "        scale_by_freq : bool, default: True\n",
      "            Whether the resulting density values should be scaled by the scaling\n",
      "            frequency, which gives density in units of 1/Hz.  This allows for\n",
      "            integration over the returned frequency values.  The default is True for\n",
      "            MATLAB compatibility.\n",
      "        \n",
      "        noverlap : int, default: 0 (no overlap)\n",
      "            The number of points of overlap between blocks.\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        Cxy : 1-D array\n",
      "            The coherence vector.\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies for the elements in *Cxy*.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Bendat & Piersol -- Random Data: Analysis and Measurement Procedures,\n",
      "        John Wiley & Sons (1986)\n",
      "    \n",
      "    colorbar(mappable=None, cax=None, ax=None, **kwargs)\n",
      "        Add a colorbar to a plot.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        mappable\n",
      "            The `matplotlib.cm.ScalarMappable` (i.e., `.AxesImage`,\n",
      "            `.ContourSet`, etc.) described by this colorbar.  This argument is\n",
      "            mandatory for the `.Figure.colorbar` method but optional for the\n",
      "            `.pyplot.colorbar` function, which sets the default to the current\n",
      "            image.\n",
      "        \n",
      "            Note that one can create a `.ScalarMappable` \"on-the-fly\" to\n",
      "            generate colorbars not attached to a previously drawn artist, e.g.\n",
      "            ::\n",
      "        \n",
      "                fig.colorbar(cm.ScalarMappable(norm=norm, cmap=cmap), ax=ax)\n",
      "        \n",
      "        cax : `~matplotlib.axes.Axes`, optional\n",
      "            Axes into which the colorbar will be drawn.\n",
      "        \n",
      "        ax : `~.axes.Axes` or list or `numpy.ndarray` of Axes, optional\n",
      "            One or more parent axes from which space for a new colorbar axes\n",
      "            will be stolen, if *cax* is None.  This has no effect if *cax* is\n",
      "            set.\n",
      "        \n",
      "        use_gridspec : bool, optional\n",
      "            If *cax* is ``None``, a new *cax* is created as an instance of\n",
      "            Axes.  If *ax* is an instance of Subplot and *use_gridspec* is\n",
      "            ``True``, *cax* is created as an instance of Subplot using the\n",
      "            :mod:`.gridspec` module.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        colorbar : `~matplotlib.colorbar.Colorbar`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        \n",
      "        location : None or {'left', 'right', 'top', 'bottom'}\n",
      "            The location, relative to the parent axes, where the colorbar axes\n",
      "            is created.  It also determines the *orientation* of the colorbar\n",
      "            (colorbars on the left and right are vertical, colorbars at the top\n",
      "            and bottom are horizontal).  If None, the location will come from the\n",
      "            *orientation* if it is set (vertical colorbars on the right, horizontal\n",
      "            ones at the bottom), or default to 'right' if *orientation* is unset.\n",
      "        \n",
      "        orientation : None or {'vertical', 'horizontal'}\n",
      "            The orientation of the colorbar.  It is preferable to set the *location*\n",
      "            of the colorbar, as that also determines the *orientation*; passing\n",
      "            incompatible values for *location* and *orientation* raises an exception.\n",
      "        \n",
      "        fraction : float, default: 0.15\n",
      "            Fraction of original axes to use for colorbar.\n",
      "        \n",
      "        shrink : float, default: 1.0\n",
      "            Fraction by which to multiply the size of the colorbar.\n",
      "        \n",
      "        aspect : float, default: 20\n",
      "            Ratio of long to short dimensions.\n",
      "        \n",
      "        pad : float, default: 0.05 if vertical, 0.15 if horizontal\n",
      "            Fraction of original axes between colorbar and new image axes.\n",
      "        \n",
      "        anchor : (float, float), optional\n",
      "            The anchor point of the colorbar axes.\n",
      "            Defaults to (0.0, 0.5) if vertical; (0.5, 1.0) if horizontal.\n",
      "        \n",
      "        panchor : (float, float), or *False*, optional\n",
      "            The anchor point of the colorbar parent axes. If *False*, the parent\n",
      "            axes' anchor will be unchanged.\n",
      "            Defaults to (1.0, 0.5) if vertical; (0.5, 0.0) if horizontal.\n",
      "        \n",
      "        extend : {'neither', 'both', 'min', 'max'}\n",
      "            Make pointed end(s) for out-of-range values (unless 'neither').  These are\n",
      "            set for a given colormap using the colormap set_under and set_over methods.\n",
      "        \n",
      "        extendfrac : {*None*, 'auto', length, lengths}\n",
      "            If set to *None*, both the minimum and maximum triangular colorbar\n",
      "            extensions will have a length of 5% of the interior colorbar length (this\n",
      "            is the default setting).\n",
      "        \n",
      "            If set to 'auto', makes the triangular colorbar extensions the same lengths\n",
      "            as the interior boxes (when *spacing* is set to 'uniform') or the same\n",
      "            lengths as the respective adjacent interior boxes (when *spacing* is set to\n",
      "            'proportional').\n",
      "        \n",
      "            If a scalar, indicates the length of both the minimum and maximum\n",
      "            triangular colorbar extensions as a fraction of the interior colorbar\n",
      "            length.  A two-element sequence of fractions may also be given, indicating\n",
      "            the lengths of the minimum and maximum colorbar extensions respectively as\n",
      "            a fraction of the interior colorbar length.\n",
      "        \n",
      "        extendrect : bool\n",
      "            If *False* the minimum and maximum colorbar extensions will be triangular\n",
      "            (the default).  If *True* the extensions will be rectangular.\n",
      "        \n",
      "        spacing : {'uniform', 'proportional'}\n",
      "            For discrete colorbars (`.BoundaryNorm` or contours), 'uniform' gives each\n",
      "            color the same space; 'proportional' makes the space proportional to the\n",
      "            data interval.\n",
      "        \n",
      "        ticks : None or list of ticks or Locator\n",
      "            If None, ticks are determined automatically from the input.\n",
      "        \n",
      "        format : None or str or Formatter\n",
      "            If None, `~.ticker.ScalarFormatter` is used.\n",
      "            Format strings, e.g., ``\"%4.2e\"`` or ``\"{x:.2e}\"``, are supported.\n",
      "            An alternative `~.ticker.Formatter` may be given instead.\n",
      "        \n",
      "        drawedges : bool\n",
      "            Whether to draw lines at color boundaries.\n",
      "        \n",
      "        label : str\n",
      "            The label on the colorbar's long axis.\n",
      "        \n",
      "        boundaries, values : None or a sequence\n",
      "            If unset, the colormap will be displayed on a 0-1 scale.\n",
      "            If sequences, *values* must have a length 1 less than *boundaries*.  For\n",
      "            each region delimited by adjacent entries in *boundaries*, the color mapped\n",
      "            to the corresponding value in values will be used.\n",
      "            Normally only useful for indexed colors (i.e. ``norm=NoNorm()``) or other\n",
      "            unusual circumstances.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        If *mappable* is a `~.contour.ContourSet`, its *extend* kwarg is\n",
      "        included automatically.\n",
      "        \n",
      "        The *shrink* kwarg provides a simple way to scale the colorbar with\n",
      "        respect to the axes. Note that if *cax* is specified, it determines the\n",
      "        size of the colorbar and *shrink* and *aspect* kwargs are ignored.\n",
      "        \n",
      "        For more precise control, you can manually specify the positions of the\n",
      "        axes objects in which the mappable and the colorbar are drawn.  In this\n",
      "        case, do not use any of the axes properties kwargs.\n",
      "        \n",
      "        It is known that some vector graphics viewers (svg and pdf) renders\n",
      "        white gaps between segments of the colorbar.  This is due to bugs in\n",
      "        the viewers, not Matplotlib.  As a workaround, the colorbar can be\n",
      "        rendered with overlapping segments::\n",
      "        \n",
      "            cbar = colorbar()\n",
      "            cbar.solids.set_edgecolor(\"face\")\n",
      "            draw()\n",
      "        \n",
      "        However this has negative consequences in other circumstances, e.g.\n",
      "        with semi-transparent images (alpha < 1) and colorbar extensions;\n",
      "        therefore, this workaround is not used by default (see issue #1188).\n",
      "    \n",
      "    connect(s, func)\n",
      "        Bind function *func* to event *s*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        s : str\n",
      "            One of the following events ids:\n",
      "        \n",
      "            - 'button_press_event'\n",
      "            - 'button_release_event'\n",
      "            - 'draw_event'\n",
      "            - 'key_press_event'\n",
      "            - 'key_release_event'\n",
      "            - 'motion_notify_event'\n",
      "            - 'pick_event'\n",
      "            - 'resize_event'\n",
      "            - 'scroll_event'\n",
      "            - 'figure_enter_event',\n",
      "            - 'figure_leave_event',\n",
      "            - 'axes_enter_event',\n",
      "            - 'axes_leave_event'\n",
      "            - 'close_event'.\n",
      "        \n",
      "        func : callable\n",
      "            The callback function to be executed, which must have the\n",
      "            signature::\n",
      "        \n",
      "                def func(event: Event) -> Any\n",
      "        \n",
      "            For the location events (button and key press/release), if the\n",
      "            mouse is over the Axes, the ``inaxes`` attribute of the event will\n",
      "            be set to the `~matplotlib.axes.Axes` the event occurs is over, and\n",
      "            additionally, the variables ``xdata`` and ``ydata`` attributes will\n",
      "            be set to the mouse location in data coordinates.  See `.KeyEvent`\n",
      "            and `.MouseEvent` for more info.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        cid\n",
      "            A connection id that can be used with\n",
      "            `.FigureCanvasBase.mpl_disconnect`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            def on_press(event):\n",
      "                print('you pressed', event.button, event.xdata, event.ydata)\n",
      "        \n",
      "            cid = canvas.mpl_connect('button_press_event', on_press)\n",
      "    \n",
      "    contour(*args, data=None, **kwargs)\n",
      "        Plot contour lines.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "            contour([X, Y,] Z, [levels], **kwargs)\n",
      "        \n",
      "        `.contour` and `.contourf` draw contour lines and filled contours,\n",
      "        respectively.  Except as noted, function signatures and return values\n",
      "        are the same for both versions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X, Y : array-like, optional\n",
      "            The coordinates of the values in *Z*.\n",
      "        \n",
      "            *X* and *Y* must both be 2D with the same shape as *Z* (e.g.\n",
      "            created via `numpy.meshgrid`), or they must both be 1-D such\n",
      "            that ``len(X) == N`` is the number of columns in *Z* and\n",
      "            ``len(Y) == M`` is the number of rows in *Z*.\n",
      "        \n",
      "            *X* and *Y* must both be ordered monotonically.\n",
      "        \n",
      "            If not given, they are assumed to be integer indices, i.e.\n",
      "            ``X = range(N)``, ``Y = range(M)``.\n",
      "        \n",
      "        Z : (M, N) array-like\n",
      "            The height values over which the contour is drawn.  Color-mapping is\n",
      "            controlled by *cmap*, *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        levels : int or array-like, optional\n",
      "            Determines the number and positions of the contour lines / regions.\n",
      "        \n",
      "            If an int *n*, use `~matplotlib.ticker.MaxNLocator`, which tries\n",
      "            to automatically choose no more than *n+1* \"nice\" contour levels\n",
      "            between minimum and maximum numeric values of *Z*.\n",
      "        \n",
      "            If array-like, draw contour lines at the specified levels.\n",
      "            The values must be in increasing order.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~.contour.QuadContourSet`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        corner_mask : bool, default: :rc:`contour.corner_mask`\n",
      "            Enable/disable corner masking, which only has an effect if *Z* is\n",
      "            a masked array.  If ``False``, any quad touching a masked point is\n",
      "            masked out.  If ``True``, only the triangular corners of quads\n",
      "            nearest those points are always masked out, other triangular\n",
      "            corners comprising three unmasked points are contoured as usual.\n",
      "        \n",
      "        colors : color string or sequence of colors, optional\n",
      "            The colors of the levels, i.e. the lines for `.contour` and the\n",
      "            areas for `.contourf`.\n",
      "        \n",
      "            The sequence is cycled for the levels in ascending order. If the\n",
      "            sequence is shorter than the number of levels, it's repeated.\n",
      "        \n",
      "            As a shortcut, single color strings may be used in place of\n",
      "            one-element lists, i.e. ``'red'`` instead of ``['red']`` to color\n",
      "            all levels with the same color. This shortcut does only work for\n",
      "            color strings, not for other ways of specifying colors.\n",
      "        \n",
      "            By default (value *None*), the colormap specified by *cmap*\n",
      "            will be used.\n",
      "        \n",
      "        alpha : float, default: 1\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            If *vmin* or *vmax* are not given, the default color scaling is based on\n",
      "            *levels*.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        origin : {*None*, 'upper', 'lower', 'image'}, default: None\n",
      "            Determines the orientation and exact position of *Z* by specifying\n",
      "            the position of ``Z[0, 0]``.  This is only relevant, if *X*, *Y*\n",
      "            are not given.\n",
      "        \n",
      "            - *None*: ``Z[0, 0]`` is at X=0, Y=0 in the lower left corner.\n",
      "            - 'lower': ``Z[0, 0]`` is at X=0.5, Y=0.5 in the lower left corner.\n",
      "            - 'upper': ``Z[0, 0]`` is at X=N+0.5, Y=0.5 in the upper left\n",
      "              corner.\n",
      "            - 'image': Use the value from :rc:`image.origin`.\n",
      "        \n",
      "        extent : (x0, x1, y0, y1), optional\n",
      "            If *origin* is not *None*, then *extent* is interpreted as in\n",
      "            `.imshow`: it gives the outer pixel boundaries. In this case, the\n",
      "            position of Z[0, 0] is the center of the pixel, not a corner. If\n",
      "            *origin* is *None*, then (*x0*, *y0*) is the position of Z[0, 0],\n",
      "            and (*x1*, *y1*) is the position of Z[-1, -1].\n",
      "        \n",
      "            This argument is ignored if *X* and *Y* are specified in the call\n",
      "            to contour.\n",
      "        \n",
      "        locator : ticker.Locator subclass, optional\n",
      "            The locator is used to determine the contour levels if they\n",
      "            are not given explicitly via *levels*.\n",
      "            Defaults to `~.ticker.MaxNLocator`.\n",
      "        \n",
      "        extend : {'neither', 'both', 'min', 'max'}, default: 'neither'\n",
      "            Determines the ``contourf``-coloring of values that are outside the\n",
      "            *levels* range.\n",
      "        \n",
      "            If 'neither', values outside the *levels* range are not colored.\n",
      "            If 'min', 'max' or 'both', color the values below, above or below\n",
      "            and above the *levels* range.\n",
      "        \n",
      "            Values below ``min(levels)`` and above ``max(levels)`` are mapped\n",
      "            to the under/over values of the `.Colormap`. Note that most\n",
      "            colormaps do not have dedicated colors for these by default, so\n",
      "            that the over and under values are the edge values of the colormap.\n",
      "            You may want to set these values explicitly using\n",
      "            `.Colormap.set_under` and `.Colormap.set_over`.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "                An existing `.QuadContourSet` does not get notified if\n",
      "                properties of its colormap are changed. Therefore, an explicit\n",
      "                call `.QuadContourSet.changed()` is needed after modifying the\n",
      "                colormap. The explicit call can be left out, if a colorbar is\n",
      "                assigned to the `.QuadContourSet` because it internally calls\n",
      "                `.QuadContourSet.changed()`.\n",
      "        \n",
      "            Example::\n",
      "        \n",
      "                x = np.arange(1, 10)\n",
      "                y = x.reshape(-1, 1)\n",
      "                h = x * y\n",
      "        \n",
      "                cs = plt.contourf(h, levels=[10, 30, 50],\n",
      "                    colors=['#808080', '#A0A0A0', '#C0C0C0'], extend='both')\n",
      "                cs.cmap.set_over('red')\n",
      "                cs.cmap.set_under('blue')\n",
      "                cs.changed()\n",
      "        \n",
      "        xunits, yunits : registered units, optional\n",
      "            Override axis units by specifying an instance of a\n",
      "            :class:`matplotlib.units.ConversionInterface`.\n",
      "        \n",
      "        antialiased : bool, optional\n",
      "            Enable antialiasing, overriding the defaults.  For\n",
      "            filled contours, the default is *True*.  For line contours,\n",
      "            it is taken from :rc:`lines.antialiased`.\n",
      "        \n",
      "        nchunk : int >= 0, optional\n",
      "            If 0, no subdivision of the domain.  Specify a positive integer to\n",
      "            divide the domain into subdomains of *nchunk* by *nchunk* quads.\n",
      "            Chunking reduces the maximum length of polygons generated by the\n",
      "            contouring algorithm which reduces the rendering workload passed\n",
      "            on to the backend and also requires slightly less RAM.  It can\n",
      "            however introduce rendering artifacts at chunk boundaries depending\n",
      "            on the backend, the *antialiased* flag and value of *alpha*.\n",
      "        \n",
      "        linewidths : float or array-like, default: :rc:`contour.linewidth`\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            The line width of the contour lines.\n",
      "        \n",
      "            If a number, all levels will be plotted with this linewidth.\n",
      "        \n",
      "            If a sequence, the levels in ascending order will be plotted with\n",
      "            the linewidths in the order specified.\n",
      "        \n",
      "            If None, this falls back to :rc:`lines.linewidth`.\n",
      "        \n",
      "        linestyles : {*None*, 'solid', 'dashed', 'dashdot', 'dotted'}, optional\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            If *linestyles* is *None*, the default is 'solid' unless the lines are\n",
      "            monochrome. In that case, negative contours will instead take their\n",
      "            linestyle from the *negative_linestyles* argument.\n",
      "        \n",
      "            *linestyles* can also be an iterable of the above strings specifying a set\n",
      "            of linestyles to be used. If this iterable is shorter than the number of\n",
      "            contour levels it will be repeated as necessary.\n",
      "        \n",
      "        negative_linestyles : {*None*, 'solid', 'dashed', 'dashdot', 'dotted'},                        optional\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            If *linestyles* is *None* and the lines are monochrome, this argument\n",
      "            specifies the line style for negative contours.\n",
      "        \n",
      "            If *negative_linestyles* is *None*, the default is taken from\n",
      "            :rc:`contour.negative_linestyles`.\n",
      "        \n",
      "            *negative_linestyles* can also be an iterable of the above strings\n",
      "            specifying a set of linestyles to be used. If this iterable is shorter than\n",
      "            the number of contour levels it will be repeated as necessary.\n",
      "        \n",
      "        hatches : list[str], optional\n",
      "            *Only applies to* `.contourf`.\n",
      "        \n",
      "            A list of cross hatch patterns to use on the filled areas.\n",
      "            If None, no hatching will be added to the contour.\n",
      "            Hatching is supported in the PostScript, PDF, SVG and Agg\n",
      "            backends only.\n",
      "        \n",
      "        algorithm : {'mpl2005', 'mpl2014', 'serial', 'threaded'}, optional\n",
      "            Which contouring algorithm to use to calculate the contour lines and\n",
      "            polygons. The algorithms are implemented in\n",
      "            `ContourPy <https://github.com/contourpy/contourpy>`_, consult the\n",
      "            `ContourPy documentation <https://contourpy.readthedocs.io>`_ for\n",
      "            further information.\n",
      "        \n",
      "            The default is taken from :rc:`contour.algorithm`.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        1. `.contourf` differs from the MATLAB version in that it does not draw\n",
      "           the polygon edges. To draw edges, add line contours with calls to\n",
      "           `.contour`.\n",
      "        \n",
      "        2. `.contourf` fills intervals that are closed at the top; that is, for\n",
      "           boundaries *z1* and *z2*, the filled region is::\n",
      "        \n",
      "              z1 < Z <= z2\n",
      "        \n",
      "           except for the lowest interval, which is closed on both sides (i.e.\n",
      "           it includes the lowest value).\n",
      "        \n",
      "        3. `.contour` and `.contourf` use a `marching squares\n",
      "           <https://en.wikipedia.org/wiki/Marching_squares>`_ algorithm to\n",
      "           compute contour locations.  More information can be found in\n",
      "           `ContourPy documentation <https://contourpy.readthedocs.io>`_.\n",
      "    \n",
      "    contourf(*args, data=None, **kwargs)\n",
      "        Plot filled contours.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "            contourf([X, Y,] Z, [levels], **kwargs)\n",
      "        \n",
      "        `.contour` and `.contourf` draw contour lines and filled contours,\n",
      "        respectively.  Except as noted, function signatures and return values\n",
      "        are the same for both versions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X, Y : array-like, optional\n",
      "            The coordinates of the values in *Z*.\n",
      "        \n",
      "            *X* and *Y* must both be 2D with the same shape as *Z* (e.g.\n",
      "            created via `numpy.meshgrid`), or they must both be 1-D such\n",
      "            that ``len(X) == N`` is the number of columns in *Z* and\n",
      "            ``len(Y) == M`` is the number of rows in *Z*.\n",
      "        \n",
      "            *X* and *Y* must both be ordered monotonically.\n",
      "        \n",
      "            If not given, they are assumed to be integer indices, i.e.\n",
      "            ``X = range(N)``, ``Y = range(M)``.\n",
      "        \n",
      "        Z : (M, N) array-like\n",
      "            The height values over which the contour is drawn.  Color-mapping is\n",
      "            controlled by *cmap*, *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        levels : int or array-like, optional\n",
      "            Determines the number and positions of the contour lines / regions.\n",
      "        \n",
      "            If an int *n*, use `~matplotlib.ticker.MaxNLocator`, which tries\n",
      "            to automatically choose no more than *n+1* \"nice\" contour levels\n",
      "            between minimum and maximum numeric values of *Z*.\n",
      "        \n",
      "            If array-like, draw contour lines at the specified levels.\n",
      "            The values must be in increasing order.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~.contour.QuadContourSet`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        corner_mask : bool, default: :rc:`contour.corner_mask`\n",
      "            Enable/disable corner masking, which only has an effect if *Z* is\n",
      "            a masked array.  If ``False``, any quad touching a masked point is\n",
      "            masked out.  If ``True``, only the triangular corners of quads\n",
      "            nearest those points are always masked out, other triangular\n",
      "            corners comprising three unmasked points are contoured as usual.\n",
      "        \n",
      "        colors : color string or sequence of colors, optional\n",
      "            The colors of the levels, i.e. the lines for `.contour` and the\n",
      "            areas for `.contourf`.\n",
      "        \n",
      "            The sequence is cycled for the levels in ascending order. If the\n",
      "            sequence is shorter than the number of levels, it's repeated.\n",
      "        \n",
      "            As a shortcut, single color strings may be used in place of\n",
      "            one-element lists, i.e. ``'red'`` instead of ``['red']`` to color\n",
      "            all levels with the same color. This shortcut does only work for\n",
      "            color strings, not for other ways of specifying colors.\n",
      "        \n",
      "            By default (value *None*), the colormap specified by *cmap*\n",
      "            will be used.\n",
      "        \n",
      "        alpha : float, default: 1\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            If *vmin* or *vmax* are not given, the default color scaling is based on\n",
      "            *levels*.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        origin : {*None*, 'upper', 'lower', 'image'}, default: None\n",
      "            Determines the orientation and exact position of *Z* by specifying\n",
      "            the position of ``Z[0, 0]``.  This is only relevant, if *X*, *Y*\n",
      "            are not given.\n",
      "        \n",
      "            - *None*: ``Z[0, 0]`` is at X=0, Y=0 in the lower left corner.\n",
      "            - 'lower': ``Z[0, 0]`` is at X=0.5, Y=0.5 in the lower left corner.\n",
      "            - 'upper': ``Z[0, 0]`` is at X=N+0.5, Y=0.5 in the upper left\n",
      "              corner.\n",
      "            - 'image': Use the value from :rc:`image.origin`.\n",
      "        \n",
      "        extent : (x0, x1, y0, y1), optional\n",
      "            If *origin* is not *None*, then *extent* is interpreted as in\n",
      "            `.imshow`: it gives the outer pixel boundaries. In this case, the\n",
      "            position of Z[0, 0] is the center of the pixel, not a corner. If\n",
      "            *origin* is *None*, then (*x0*, *y0*) is the position of Z[0, 0],\n",
      "            and (*x1*, *y1*) is the position of Z[-1, -1].\n",
      "        \n",
      "            This argument is ignored if *X* and *Y* are specified in the call\n",
      "            to contour.\n",
      "        \n",
      "        locator : ticker.Locator subclass, optional\n",
      "            The locator is used to determine the contour levels if they\n",
      "            are not given explicitly via *levels*.\n",
      "            Defaults to `~.ticker.MaxNLocator`.\n",
      "        \n",
      "        extend : {'neither', 'both', 'min', 'max'}, default: 'neither'\n",
      "            Determines the ``contourf``-coloring of values that are outside the\n",
      "            *levels* range.\n",
      "        \n",
      "            If 'neither', values outside the *levels* range are not colored.\n",
      "            If 'min', 'max' or 'both', color the values below, above or below\n",
      "            and above the *levels* range.\n",
      "        \n",
      "            Values below ``min(levels)`` and above ``max(levels)`` are mapped\n",
      "            to the under/over values of the `.Colormap`. Note that most\n",
      "            colormaps do not have dedicated colors for these by default, so\n",
      "            that the over and under values are the edge values of the colormap.\n",
      "            You may want to set these values explicitly using\n",
      "            `.Colormap.set_under` and `.Colormap.set_over`.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "                An existing `.QuadContourSet` does not get notified if\n",
      "                properties of its colormap are changed. Therefore, an explicit\n",
      "                call `.QuadContourSet.changed()` is needed after modifying the\n",
      "                colormap. The explicit call can be left out, if a colorbar is\n",
      "                assigned to the `.QuadContourSet` because it internally calls\n",
      "                `.QuadContourSet.changed()`.\n",
      "        \n",
      "            Example::\n",
      "        \n",
      "                x = np.arange(1, 10)\n",
      "                y = x.reshape(-1, 1)\n",
      "                h = x * y\n",
      "        \n",
      "                cs = plt.contourf(h, levels=[10, 30, 50],\n",
      "                    colors=['#808080', '#A0A0A0', '#C0C0C0'], extend='both')\n",
      "                cs.cmap.set_over('red')\n",
      "                cs.cmap.set_under('blue')\n",
      "                cs.changed()\n",
      "        \n",
      "        xunits, yunits : registered units, optional\n",
      "            Override axis units by specifying an instance of a\n",
      "            :class:`matplotlib.units.ConversionInterface`.\n",
      "        \n",
      "        antialiased : bool, optional\n",
      "            Enable antialiasing, overriding the defaults.  For\n",
      "            filled contours, the default is *True*.  For line contours,\n",
      "            it is taken from :rc:`lines.antialiased`.\n",
      "        \n",
      "        nchunk : int >= 0, optional\n",
      "            If 0, no subdivision of the domain.  Specify a positive integer to\n",
      "            divide the domain into subdomains of *nchunk* by *nchunk* quads.\n",
      "            Chunking reduces the maximum length of polygons generated by the\n",
      "            contouring algorithm which reduces the rendering workload passed\n",
      "            on to the backend and also requires slightly less RAM.  It can\n",
      "            however introduce rendering artifacts at chunk boundaries depending\n",
      "            on the backend, the *antialiased* flag and value of *alpha*.\n",
      "        \n",
      "        linewidths : float or array-like, default: :rc:`contour.linewidth`\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            The line width of the contour lines.\n",
      "        \n",
      "            If a number, all levels will be plotted with this linewidth.\n",
      "        \n",
      "            If a sequence, the levels in ascending order will be plotted with\n",
      "            the linewidths in the order specified.\n",
      "        \n",
      "            If None, this falls back to :rc:`lines.linewidth`.\n",
      "        \n",
      "        linestyles : {*None*, 'solid', 'dashed', 'dashdot', 'dotted'}, optional\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            If *linestyles* is *None*, the default is 'solid' unless the lines are\n",
      "            monochrome. In that case, negative contours will instead take their\n",
      "            linestyle from the *negative_linestyles* argument.\n",
      "        \n",
      "            *linestyles* can also be an iterable of the above strings specifying a set\n",
      "            of linestyles to be used. If this iterable is shorter than the number of\n",
      "            contour levels it will be repeated as necessary.\n",
      "        \n",
      "        negative_linestyles : {*None*, 'solid', 'dashed', 'dashdot', 'dotted'},                        optional\n",
      "            *Only applies to* `.contour`.\n",
      "        \n",
      "            If *linestyles* is *None* and the lines are monochrome, this argument\n",
      "            specifies the line style for negative contours.\n",
      "        \n",
      "            If *negative_linestyles* is *None*, the default is taken from\n",
      "            :rc:`contour.negative_linestyles`.\n",
      "        \n",
      "            *negative_linestyles* can also be an iterable of the above strings\n",
      "            specifying a set of linestyles to be used. If this iterable is shorter than\n",
      "            the number of contour levels it will be repeated as necessary.\n",
      "        \n",
      "        hatches : list[str], optional\n",
      "            *Only applies to* `.contourf`.\n",
      "        \n",
      "            A list of cross hatch patterns to use on the filled areas.\n",
      "            If None, no hatching will be added to the contour.\n",
      "            Hatching is supported in the PostScript, PDF, SVG and Agg\n",
      "            backends only.\n",
      "        \n",
      "        algorithm : {'mpl2005', 'mpl2014', 'serial', 'threaded'}, optional\n",
      "            Which contouring algorithm to use to calculate the contour lines and\n",
      "            polygons. The algorithms are implemented in\n",
      "            `ContourPy <https://github.com/contourpy/contourpy>`_, consult the\n",
      "            `ContourPy documentation <https://contourpy.readthedocs.io>`_ for\n",
      "            further information.\n",
      "        \n",
      "            The default is taken from :rc:`contour.algorithm`.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        1. `.contourf` differs from the MATLAB version in that it does not draw\n",
      "           the polygon edges. To draw edges, add line contours with calls to\n",
      "           `.contour`.\n",
      "        \n",
      "        2. `.contourf` fills intervals that are closed at the top; that is, for\n",
      "           boundaries *z1* and *z2*, the filled region is::\n",
      "        \n",
      "              z1 < Z <= z2\n",
      "        \n",
      "           except for the lowest interval, which is closed on both sides (i.e.\n",
      "           it includes the lowest value).\n",
      "        \n",
      "        3. `.contour` and `.contourf` use a `marching squares\n",
      "           <https://en.wikipedia.org/wiki/Marching_squares>`_ algorithm to\n",
      "           compute contour locations.  More information can be found in\n",
      "           `ContourPy documentation <https://contourpy.readthedocs.io>`_.\n",
      "    \n",
      "    cool()\n",
      "        Set the colormap to 'cool'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    copper()\n",
      "        Set the colormap to 'copper'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    csd(x, y, NFFT=None, Fs=None, Fc=None, detrend=None, window=None, noverlap=None, pad_to=None, sides=None, scale_by_freq=None, return_line=None, *, data=None, **kwargs)\n",
      "        Plot the cross-spectral density.\n",
      "        \n",
      "        The cross spectral density :math:`P_{xy}` by Welch's average\n",
      "        periodogram method.  The vectors *x* and *y* are divided into\n",
      "        *NFFT* length segments.  Each segment is detrended by function\n",
      "        *detrend* and windowed by function *window*.  *noverlap* gives\n",
      "        the length of the overlap between segments.  The product of\n",
      "        the direct FFTs of *x* and *y* are averaged over each segment\n",
      "        to compute :math:`P_{xy}`, with a scaling to correct for power\n",
      "        loss due to windowing.\n",
      "        \n",
      "        If len(*x*) < *NFFT* or len(*y*) < *NFFT*, they will be zero\n",
      "        padded to *NFFT*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : 1-D arrays or sequences\n",
      "            Arrays or sequences containing the data.\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  This can be different from *NFFT*, which specifies the number\n",
      "            of data points used.  While not increasing the actual resolution of the\n",
      "            spectrum (the minimum distance between resolvable peaks), this can give\n",
      "            more points in the plot, allowing for more detail. This corresponds to\n",
      "            the *n* parameter in the call to `~numpy.fft.fft`. The default is None,\n",
      "            which sets *pad_to* equal to *NFFT*\n",
      "        \n",
      "        NFFT : int, default: 256\n",
      "            The number of data points used in each block for the FFT.  A power 2 is\n",
      "            most efficient.  This should *NOT* be used to get zero padding, or the\n",
      "            scaling of the result will be incorrect; use *pad_to* for this instead.\n",
      "        \n",
      "        detrend : {'none', 'mean', 'linear'} or callable, default: 'none'\n",
      "            The function applied to each segment before fft-ing, designed to remove\n",
      "            the mean or linear trend.  Unlike in MATLAB, where the *detrend* parameter\n",
      "            is a vector, in Matplotlib it is a function.  The :mod:`~matplotlib.mlab`\n",
      "            module defines `.detrend_none`, `.detrend_mean`, and `.detrend_linear`,\n",
      "            but you can use a custom function as well.  You can also use a string to\n",
      "            choose one of the functions: 'none' calls `.detrend_none`. 'mean' calls\n",
      "            `.detrend_mean`. 'linear' calls `.detrend_linear`.\n",
      "        \n",
      "        scale_by_freq : bool, default: True\n",
      "            Whether the resulting density values should be scaled by the scaling\n",
      "            frequency, which gives density in units of 1/Hz.  This allows for\n",
      "            integration over the returned frequency values.  The default is True for\n",
      "            MATLAB compatibility.\n",
      "        \n",
      "        noverlap : int, default: 0 (no overlap)\n",
      "            The number of points of overlap between segments.\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        return_line : bool, default: False\n",
      "            Whether to include the line object plotted in the returned values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        Pxy : 1-D array\n",
      "            The values for the cross spectrum :math:`P_{xy}` before scaling\n",
      "            (complex valued).\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the elements in *Pxy*.\n",
      "        \n",
      "        line : `~matplotlib.lines.Line2D`\n",
      "            The line created by this function.\n",
      "            Only returned if *return_line* is True.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        psd : is equivalent to setting ``y = x``.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For plotting, the power is plotted as\n",
      "        :math:`10 \\log_{10}(P_{xy})` for decibels, though :math:`P_{xy}` itself\n",
      "        is returned.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Bendat & Piersol -- Random Data: Analysis and Measurement Procedures,\n",
      "        John Wiley & Sons (1986)\n",
      "    \n",
      "    delaxes(ax=None)\n",
      "        Remove an `~.axes.Axes` (defaulting to the current axes) from its figure.\n",
      "    \n",
      "    disconnect(cid)\n",
      "        Disconnect the callback with id *cid*.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            cid = canvas.mpl_connect('button_press_event', on_press)\n",
      "            # ... later\n",
      "            canvas.mpl_disconnect(cid)\n",
      "    \n",
      "    draw()\n",
      "        Redraw the current figure.\n",
      "        \n",
      "        This is used to update a figure that has been altered, but not\n",
      "        automatically re-drawn.  If interactive mode is on (via `.ion()`), this\n",
      "        should be only rarely needed, but there may be ways to modify the state of\n",
      "        a figure without marking it as \"stale\".  Please report these cases as bugs.\n",
      "        \n",
      "        This is equivalent to calling ``fig.canvas.draw_idle()``, where ``fig`` is\n",
      "        the current figure.\n",
      "    \n",
      "    draw_if_interactive(*args, **kwargs)\n",
      "        Redraw the current figure if in interactive mode.\n",
      "        \n",
      "        .. warning::\n",
      "        \n",
      "            End users will typically not have to call this function because the\n",
      "            the interactive mode takes care of this.\n",
      "    \n",
      "    errorbar(x, y, yerr=None, xerr=None, fmt='', ecolor=None, elinewidth=None, capsize=None, barsabove=False, lolims=False, uplims=False, xlolims=False, xuplims=False, errorevery=1, capthick=None, *, data=None, **kwargs)\n",
      "        Plot y versus x as lines and/or markers with attached errorbars.\n",
      "        \n",
      "        *x*, *y* define the data locations, *xerr*, *yerr* define the errorbar\n",
      "        sizes. By default, this draws the data markers/lines as well the\n",
      "        errorbars. Use fmt='none' to draw errorbars without any data markers.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : float or array-like\n",
      "            The data positions.\n",
      "        \n",
      "        xerr, yerr : float or array-like, shape(N,) or shape(2, N), optional\n",
      "            The errorbar sizes:\n",
      "        \n",
      "            - scalar: Symmetric +/- values for all data points.\n",
      "            - shape(N,): Symmetric +/-values for each data point.\n",
      "            - shape(2, N): Separate - and + values for each bar. First row\n",
      "              contains the lower errors, the second row contains the upper\n",
      "              errors.\n",
      "            - *None*: No errorbar.\n",
      "        \n",
      "            All values must be >= 0.\n",
      "        \n",
      "            See :doc:`/gallery/statistics/errorbar_features`\n",
      "            for an example on the usage of ``xerr`` and ``yerr``.\n",
      "        \n",
      "        fmt : str, default: ''\n",
      "            The format for the data points / data lines. See `.plot` for\n",
      "            details.\n",
      "        \n",
      "            Use 'none' (case insensitive) to plot errorbars without any data\n",
      "            markers.\n",
      "        \n",
      "        ecolor : color, default: None\n",
      "            The color of the errorbar lines.  If None, use the color of the\n",
      "            line connecting the markers.\n",
      "        \n",
      "        elinewidth : float, default: None\n",
      "            The linewidth of the errorbar lines. If None, the linewidth of\n",
      "            the current style is used.\n",
      "        \n",
      "        capsize : float, default: :rc:`errorbar.capsize`\n",
      "            The length of the error bar caps in points.\n",
      "        \n",
      "        capthick : float, default: None\n",
      "            An alias to the keyword argument *markeredgewidth* (a.k.a. *mew*).\n",
      "            This setting is a more sensible name for the property that\n",
      "            controls the thickness of the error bar cap in points. For\n",
      "            backwards compatibility, if *mew* or *markeredgewidth* are given,\n",
      "            then they will over-ride *capthick*. This may change in future\n",
      "            releases.\n",
      "        \n",
      "        barsabove : bool, default: False\n",
      "            If True, will plot the errorbars above the plot\n",
      "            symbols. Default is below.\n",
      "        \n",
      "        lolims, uplims, xlolims, xuplims : bool, default: False\n",
      "            These arguments can be used to indicate that a value gives only\n",
      "            upper/lower limits.  In that case a caret symbol is used to\n",
      "            indicate this. *lims*-arguments may be scalars, or array-likes of\n",
      "            the same length as *xerr* and *yerr*.  To use limits with inverted\n",
      "            axes, `~.Axes.set_xlim` or `~.Axes.set_ylim` must be called before\n",
      "            :meth:`errorbar`.  Note the tricky parameter names: setting e.g.\n",
      "            *lolims* to True means that the y-value is a *lower* limit of the\n",
      "            True value, so, only an *upward*-pointing arrow will be drawn!\n",
      "        \n",
      "        errorevery : int or (int, int), default: 1\n",
      "            draws error bars on a subset of the data. *errorevery* =N draws\n",
      "            error bars on the points (x[::N], y[::N]).\n",
      "            *errorevery* =(start, N) draws error bars on the points\n",
      "            (x[start::N], y[start::N]). e.g. errorevery=(6, 3)\n",
      "            adds error bars to the data at (x[6], x[9], x[12], x[15], ...).\n",
      "            Used to avoid overlapping error bars when two series share x-axis\n",
      "            values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.ErrorbarContainer`\n",
      "            The container contains:\n",
      "        \n",
      "            - plotline: `.Line2D` instance of x, y plot markers and/or line.\n",
      "            - caplines: A tuple of `.Line2D` instances of the error bar caps.\n",
      "            - barlinecols: A tuple of `.LineCollection` with the horizontal and\n",
      "              vertical error ranges.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*, *xerr*, *yerr*\n",
      "        \n",
      "        **kwargs\n",
      "            All other keyword arguments are passed on to the `~.Axes.plot` call\n",
      "            drawing the markers. For example, this code makes big red squares\n",
      "            with thick green edges::\n",
      "        \n",
      "                x, y, yerr = rand(3, 10)\n",
      "                errorbar(x, y, yerr, marker='s', mfc='red',\n",
      "                         mec='green', ms=20, mew=4)\n",
      "        \n",
      "            where *mfc*, *mec*, *ms* and *mew* are aliases for the longer\n",
      "            property names, *markerfacecolor*, *markeredgecolor*, *markersize*\n",
      "            and *markeredgewidth*.\n",
      "        \n",
      "            Valid kwargs for the marker properties are:\n",
      "        \n",
      "            - *dashes*\n",
      "            - *dash_capstyle*\n",
      "            - *dash_joinstyle*\n",
      "            - *drawstyle*\n",
      "            - *fillstyle*\n",
      "            - *linestyle*\n",
      "            - *marker*\n",
      "            - *markeredgecolor*\n",
      "            - *markeredgewidth*\n",
      "            - *markerfacecolor*\n",
      "            - *markerfacecoloralt*\n",
      "            - *markersize*\n",
      "            - *markevery*\n",
      "            - *solid_capstyle*\n",
      "            - *solid_joinstyle*\n",
      "        \n",
      "            Refer to the corresponding `.Line2D` property for more details:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "    \n",
      "    eventplot(positions, orientation='horizontal', lineoffsets=1, linelengths=1, linewidths=None, colors=None, linestyles='solid', *, data=None, **kwargs)\n",
      "        Plot identical parallel lines at the given positions.\n",
      "        \n",
      "        This type of plot is commonly used in neuroscience for representing\n",
      "        neural events, where it is usually called a spike raster, dot raster,\n",
      "        or raster plot.\n",
      "        \n",
      "        However, it is useful in any situation where you wish to show the\n",
      "        timing or position of multiple sets of discrete events, such as the\n",
      "        arrival times of people to a business on each day of the month or the\n",
      "        date of hurricanes each year of the last century.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        positions : array-like or list of array-like\n",
      "            A 1D array-like defines the positions of one sequence of events.\n",
      "        \n",
      "            Multiple groups of events may be passed as a list of array-likes.\n",
      "            Each group can be styled independently by passing lists of values\n",
      "            to *lineoffsets*, *linelengths*, *linewidths*, *colors* and\n",
      "            *linestyles*.\n",
      "        \n",
      "            Note that *positions* can be a 2D array, but in practice different\n",
      "            event groups usually have different counts so that one will use a\n",
      "            list of different-length arrays rather than a 2D array.\n",
      "        \n",
      "        orientation : {'horizontal', 'vertical'}, default: 'horizontal'\n",
      "            The direction of the event sequence:\n",
      "        \n",
      "            - 'horizontal': the events are arranged horizontally.\n",
      "              The indicator lines are vertical.\n",
      "            - 'vertical': the events are arranged vertically.\n",
      "              The indicator lines are horizontal.\n",
      "        \n",
      "        lineoffsets : float or array-like, default: 1\n",
      "            The offset of the center of the lines from the origin, in the\n",
      "            direction orthogonal to *orientation*.\n",
      "        \n",
      "            If *positions* is 2D, this can be a sequence with length matching\n",
      "            the length of *positions*.\n",
      "        \n",
      "        linelengths : float or array-like, default: 1\n",
      "            The total height of the lines (i.e. the lines stretches from\n",
      "            ``lineoffset - linelength/2`` to ``lineoffset + linelength/2``).\n",
      "        \n",
      "            If *positions* is 2D, this can be a sequence with length matching\n",
      "            the length of *positions*.\n",
      "        \n",
      "        linewidths : float or array-like, default: :rc:`lines.linewidth`\n",
      "            The line width(s) of the event lines, in points.\n",
      "        \n",
      "            If *positions* is 2D, this can be a sequence with length matching\n",
      "            the length of *positions*.\n",
      "        \n",
      "        colors : color or list of colors, default: :rc:`lines.color`\n",
      "            The color(s) of the event lines.\n",
      "        \n",
      "            If *positions* is 2D, this can be a sequence with length matching\n",
      "            the length of *positions*.\n",
      "        \n",
      "        linestyles : str or tuple or list of such values, default: 'solid'\n",
      "            Default is 'solid'. Valid strings are ['solid', 'dashed',\n",
      "            'dashdot', 'dotted', '-', '--', '-.', ':']. Dash tuples\n",
      "            should be of the form::\n",
      "        \n",
      "                (offset, onoffseq),\n",
      "        \n",
      "            where *onoffseq* is an even length tuple of on and off ink\n",
      "            in points.\n",
      "        \n",
      "            If *positions* is 2D, this can be a sequence with length matching\n",
      "            the length of *positions*.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *positions*, *lineoffsets*, *linelengths*, *linewidths*, *colors*, *linestyles*\n",
      "        \n",
      "        **kwargs\n",
      "            Other keyword arguments are line collection properties.  See\n",
      "            `.LineCollection` for a list of the valid properties.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.EventCollection`\n",
      "            The `.EventCollection` that were added.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For *linelengths*, *linewidths*, *colors*, and *linestyles*, if only\n",
      "        a single value is given, that value is applied to all lines.  If an\n",
      "        array-like is given, it must have the same length as *positions*, and\n",
      "        each value will be applied to the corresponding row of the array.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        .. plot:: gallery/lines_bars_and_markers/eventplot_demo.py\n",
      "    \n",
      "    figimage(X, xo=0, yo=0, alpha=None, norm=None, cmap=None, vmin=None, vmax=None, origin=None, resize=False, **kwargs)\n",
      "        Add a non-resampled image to the figure.\n",
      "        \n",
      "        The image is attached to the lower or upper left corner depending on\n",
      "        *origin*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X\n",
      "            The image data. This is an array of one of the following shapes:\n",
      "        \n",
      "            - (M, N): an image with scalar data.  Color-mapping is controlled\n",
      "              by *cmap*, *norm*, *vmin*, and *vmax*.\n",
      "            - (M, N, 3): an image with RGB values (0-1 float or 0-255 int).\n",
      "            - (M, N, 4): an image with RGBA values (0-1 float or 0-255 int),\n",
      "              i.e. including transparency.\n",
      "        \n",
      "        xo, yo : int\n",
      "            The *x*/*y* image offset in pixels.\n",
      "        \n",
      "        alpha : None or float\n",
      "            The alpha blending value.\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        origin : {'upper', 'lower'}, default: :rc:`image.origin`\n",
      "            Indicates where the [0, 0] index of the array is in the upper left\n",
      "            or lower left corner of the axes.\n",
      "        \n",
      "        resize : bool\n",
      "            If *True*, resize the figure to match the given image size.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `matplotlib.image.FigureImage`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            Additional kwargs are `.Artist` kwargs passed on to `.FigureImage`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        figimage complements the Axes image (`~matplotlib.axes.Axes.imshow`)\n",
      "        which will be resampled to fit the current Axes.  If you want\n",
      "        a resampled image to fill the entire figure, you can define an\n",
      "        `~matplotlib.axes.Axes` with extent [0, 0, 1, 1].\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            f = plt.figure()\n",
      "            nx = int(f.get_figwidth() * f.dpi)\n",
      "            ny = int(f.get_figheight() * f.dpi)\n",
      "            data = np.random.random((ny, nx))\n",
      "            f.figimage(data)\n",
      "            plt.show()\n",
      "    \n",
      "    figlegend(*args, **kwargs)\n",
      "        Place a legend on the figure.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            figlegend()\n",
      "            figlegend(handles, labels)\n",
      "            figlegend(handles=handles)\n",
      "            figlegend(labels)\n",
      "        \n",
      "        The call signatures correspond to the following different ways to use\n",
      "        this method:\n",
      "        \n",
      "        **1. Automatic detection of elements to be shown in the legend**\n",
      "        \n",
      "        The elements to be added to the legend are automatically determined,\n",
      "        when you do not pass in any extra arguments.\n",
      "        \n",
      "        In this case, the labels are taken from the artist. You can specify\n",
      "        them either at artist creation or by calling the\n",
      "        :meth:`~.Artist.set_label` method on the artist::\n",
      "        \n",
      "            ax.plot([1, 2, 3], label='Inline label')\n",
      "            fig.figlegend()\n",
      "        \n",
      "        or::\n",
      "        \n",
      "            line, = ax.plot([1, 2, 3])\n",
      "            line.set_label('Label via method')\n",
      "            fig.figlegend()\n",
      "        \n",
      "        Specific lines can be excluded from the automatic legend element\n",
      "        selection by defining a label starting with an underscore.\n",
      "        This is default for all artists, so calling `.Figure.legend` without\n",
      "        any arguments and without setting the labels manually will result in\n",
      "        no legend being drawn.\n",
      "        \n",
      "        \n",
      "        **2. Explicitly listing the artists and labels in the legend**\n",
      "        \n",
      "        For full control of which artists have a legend entry, it is possible\n",
      "        to pass an iterable of legend artists followed by an iterable of\n",
      "        legend labels respectively::\n",
      "        \n",
      "            fig.figlegend([line1, line2, line3], ['label1', 'label2', 'label3'])\n",
      "        \n",
      "        \n",
      "        **3. Explicitly listing the artists in the legend**\n",
      "        \n",
      "        This is similar to 2, but the labels are taken from the artists'\n",
      "        label properties. Example::\n",
      "        \n",
      "            line1, = ax1.plot([1, 2, 3], label='label1')\n",
      "            line2, = ax2.plot([1, 2, 3], label='label2')\n",
      "            fig.figlegend(handles=[line1, line2])\n",
      "        \n",
      "        \n",
      "        **4. Labeling existing plot elements**\n",
      "        \n",
      "        .. admonition:: Discouraged\n",
      "        \n",
      "            This call signature is discouraged, because the relation between\n",
      "            plot elements and labels is only implicit by their order and can\n",
      "            easily be mixed up.\n",
      "        \n",
      "        To make a legend for all artists on all Axes, call this function with\n",
      "        an iterable of strings, one for each legend item. For example::\n",
      "        \n",
      "            fig, (ax1, ax2) = plt.subplots(1, 2)\n",
      "            ax1.plot([1, 3, 5], color='blue')\n",
      "            ax2.plot([2, 4, 6], color='red')\n",
      "            fig.figlegend(['the blues', 'the reds'])\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        handles : list of `.Artist`, optional\n",
      "            A list of Artists (lines, patches) to be added to the legend.\n",
      "            Use this together with *labels*, if you need full control on what\n",
      "            is shown in the legend and the automatic mechanism described above\n",
      "            is not sufficient.\n",
      "        \n",
      "            The length of handles and labels should be the same in this\n",
      "            case. If they are not, they are truncated to the smaller length.\n",
      "        \n",
      "        labels : list of str, optional\n",
      "            A list of labels to show next to the artists.\n",
      "            Use this together with *handles*, if you need full control on what\n",
      "            is shown in the legend and the automatic mechanism described above\n",
      "            is not sufficient.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.legend.Legend`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        \n",
      "        loc : str or pair of floats, default: :rc:`legend.loc` ('best' for axes, 'upper right' for figures)\n",
      "            The location of the legend.\n",
      "        \n",
      "            The strings\n",
      "            ``'upper left', 'upper right', 'lower left', 'lower right'``\n",
      "            place the legend at the corresponding corner of the axes/figure.\n",
      "        \n",
      "            The strings\n",
      "            ``'upper center', 'lower center', 'center left', 'center right'``\n",
      "            place the legend at the center of the corresponding edge of the\n",
      "            axes/figure.\n",
      "        \n",
      "            The string ``'center'`` places the legend at the center of the axes/figure.\n",
      "        \n",
      "            The string ``'best'`` places the legend at the location, among the nine\n",
      "            locations defined so far, with the minimum overlap with other drawn\n",
      "            artists.  This option can be quite slow for plots with large amounts of\n",
      "            data; your plotting speed may benefit from providing a specific location.\n",
      "        \n",
      "            The location can also be a 2-tuple giving the coordinates of the lower-left\n",
      "            corner of the legend in axes coordinates (in which case *bbox_to_anchor*\n",
      "            will be ignored).\n",
      "        \n",
      "            For back-compatibility, ``'center right'`` (but no other location) can also\n",
      "            be spelled ``'right'``, and each \"string\" locations can also be given as a\n",
      "            numeric value:\n",
      "        \n",
      "                ===============   =============\n",
      "                Location String   Location Code\n",
      "                ===============   =============\n",
      "                'best'            0\n",
      "                'upper right'     1\n",
      "                'upper left'      2\n",
      "                'lower left'      3\n",
      "                'lower right'     4\n",
      "                'right'           5\n",
      "                'center left'     6\n",
      "                'center right'    7\n",
      "                'lower center'    8\n",
      "                'upper center'    9\n",
      "                'center'          10\n",
      "                ===============   =============\n",
      "        \n",
      "        bbox_to_anchor : `.BboxBase`, 2-tuple, or 4-tuple of floats\n",
      "            Box that is used to position the legend in conjunction with *loc*.\n",
      "            Defaults to `axes.bbox` (if called as a method to `.Axes.legend`) or\n",
      "            `figure.bbox` (if `.Figure.legend`).  This argument allows arbitrary\n",
      "            placement of the legend.\n",
      "        \n",
      "            Bbox coordinates are interpreted in the coordinate system given by\n",
      "            *bbox_transform*, with the default transform\n",
      "            Axes or Figure coordinates, depending on which ``legend`` is called.\n",
      "        \n",
      "            If a 4-tuple or `.BboxBase` is given, then it specifies the bbox\n",
      "            ``(x, y, width, height)`` that the legend is placed in.\n",
      "            To put the legend in the best location in the bottom right\n",
      "            quadrant of the axes (or figure)::\n",
      "        \n",
      "                loc='best', bbox_to_anchor=(0.5, 0., 0.5, 0.5)\n",
      "        \n",
      "            A 2-tuple ``(x, y)`` places the corner of the legend specified by *loc* at\n",
      "            x, y.  For example, to put the legend's upper right-hand corner in the\n",
      "            center of the axes (or figure) the following keywords can be used::\n",
      "        \n",
      "                loc='upper right', bbox_to_anchor=(0.5, 0.5)\n",
      "        \n",
      "        ncols : int, default: 1\n",
      "            The number of columns that the legend has.\n",
      "        \n",
      "            For backward compatibility, the spelling *ncol* is also supported\n",
      "            but it is discouraged. If both are given, *ncols* takes precedence.\n",
      "        \n",
      "        prop : None or `matplotlib.font_manager.FontProperties` or dict\n",
      "            The font properties of the legend. If None (default), the current\n",
      "            :data:`matplotlib.rcParams` will be used.\n",
      "        \n",
      "        fontsize : int or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}\n",
      "            The font size of the legend. If the value is numeric the size will be the\n",
      "            absolute font size in points. String values are relative to the current\n",
      "            default font size. This argument is only used if *prop* is not specified.\n",
      "        \n",
      "        labelcolor : str or list, default: :rc:`legend.labelcolor`\n",
      "            The color of the text in the legend. Either a valid color string\n",
      "            (for example, 'red'), or a list of color strings. The labelcolor can\n",
      "            also be made to match the color of the line or marker using 'linecolor',\n",
      "            'markerfacecolor' (or 'mfc'), or 'markeredgecolor' (or 'mec').\n",
      "        \n",
      "            Labelcolor can be set globally using :rc:`legend.labelcolor`. If None,\n",
      "            use :rc:`text.color`.\n",
      "        \n",
      "        numpoints : int, default: :rc:`legend.numpoints`\n",
      "            The number of marker points in the legend when creating a legend\n",
      "            entry for a `.Line2D` (line).\n",
      "        \n",
      "        scatterpoints : int, default: :rc:`legend.scatterpoints`\n",
      "            The number of marker points in the legend when creating\n",
      "            a legend entry for a `.PathCollection` (scatter plot).\n",
      "        \n",
      "        scatteryoffsets : iterable of floats, default: ``[0.375, 0.5, 0.3125]``\n",
      "            The vertical offset (relative to the font size) for the markers\n",
      "            created for a scatter plot legend entry. 0.0 is at the base the\n",
      "            legend text, and 1.0 is at the top. To draw all markers at the\n",
      "            same height, set to ``[0.5]``.\n",
      "        \n",
      "        markerscale : float, default: :rc:`legend.markerscale`\n",
      "            The relative size of legend markers compared with the originally\n",
      "            drawn ones.\n",
      "        \n",
      "        markerfirst : bool, default: True\n",
      "            If *True*, legend marker is placed to the left of the legend label.\n",
      "            If *False*, legend marker is placed to the right of the legend label.\n",
      "        \n",
      "        frameon : bool, default: :rc:`legend.frameon`\n",
      "            Whether the legend should be drawn on a patch (frame).\n",
      "        \n",
      "        fancybox : bool, default: :rc:`legend.fancybox`\n",
      "            Whether round edges should be enabled around the `.FancyBboxPatch` which\n",
      "            makes up the legend's background.\n",
      "        \n",
      "        shadow : bool, default: :rc:`legend.shadow`\n",
      "            Whether to draw a shadow behind the legend.\n",
      "        \n",
      "        framealpha : float, default: :rc:`legend.framealpha`\n",
      "            The alpha transparency of the legend's background.\n",
      "            If *shadow* is activated and *framealpha* is ``None``, the default value is\n",
      "            ignored.\n",
      "        \n",
      "        facecolor : \"inherit\" or color, default: :rc:`legend.facecolor`\n",
      "            The legend's background color.\n",
      "            If ``\"inherit\"``, use :rc:`axes.facecolor`.\n",
      "        \n",
      "        edgecolor : \"inherit\" or color, default: :rc:`legend.edgecolor`\n",
      "            The legend's background patch edge color.\n",
      "            If ``\"inherit\"``, use take :rc:`axes.edgecolor`.\n",
      "        \n",
      "        mode : {\"expand\", None}\n",
      "            If *mode* is set to ``\"expand\"`` the legend will be horizontally\n",
      "            expanded to fill the axes area (or *bbox_to_anchor* if defines\n",
      "            the legend's size).\n",
      "        \n",
      "        bbox_transform : None or `matplotlib.transforms.Transform`\n",
      "            The transform for the bounding box (*bbox_to_anchor*). For a value\n",
      "            of ``None`` (default) the Axes'\n",
      "            :data:`~matplotlib.axes.Axes.transAxes` transform will be used.\n",
      "        \n",
      "        title : str or None\n",
      "            The legend's title. Default is no title (``None``).\n",
      "        \n",
      "        title_fontproperties : None or `matplotlib.font_manager.FontProperties` or dict\n",
      "            The font properties of the legend's title. If None (default), the\n",
      "            *title_fontsize* argument will be used if present; if *title_fontsize* is\n",
      "            also None, the current :rc:`legend.title_fontsize` will be used.\n",
      "        \n",
      "        title_fontsize : int or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}, default: :rc:`legend.title_fontsize`\n",
      "            The font size of the legend's title.\n",
      "            Note: This cannot be combined with *title_fontproperties*. If you want\n",
      "            to set the fontsize alongside other font properties, use the *size*\n",
      "            parameter in *title_fontproperties*.\n",
      "        \n",
      "        alignment : {'center', 'left', 'right'}, default: 'center'\n",
      "            The alignment of the legend title and the box of entries. The entries\n",
      "            are aligned as a single block, so that markers always lined up.\n",
      "        \n",
      "        borderpad : float, default: :rc:`legend.borderpad`\n",
      "            The fractional whitespace inside the legend border, in font-size units.\n",
      "        \n",
      "        labelspacing : float, default: :rc:`legend.labelspacing`\n",
      "            The vertical space between the legend entries, in font-size units.\n",
      "        \n",
      "        handlelength : float, default: :rc:`legend.handlelength`\n",
      "            The length of the legend handles, in font-size units.\n",
      "        \n",
      "        handleheight : float, default: :rc:`legend.handleheight`\n",
      "            The height of the legend handles, in font-size units.\n",
      "        \n",
      "        handletextpad : float, default: :rc:`legend.handletextpad`\n",
      "            The pad between the legend handle and text, in font-size units.\n",
      "        \n",
      "        borderaxespad : float, default: :rc:`legend.borderaxespad`\n",
      "            The pad between the axes and legend border, in font-size units.\n",
      "        \n",
      "        columnspacing : float, default: :rc:`legend.columnspacing`\n",
      "            The spacing between columns, in font-size units.\n",
      "        \n",
      "        handler_map : dict or None\n",
      "            The custom dictionary mapping instances or types to a legend\n",
      "            handler. This *handler_map* updates the default handler map\n",
      "            found at `matplotlib.legend.Legend.get_legend_handler_map`.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Axes.legend\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Some artists are not supported by this function.  See\n",
      "        :doc:`/tutorials/intermediate/legend_guide` for details.\n",
      "    \n",
      "    fignum_exists(num)\n",
      "        Return whether the figure with the given id exists.\n",
      "    \n",
      "    figtext(x, y, s, fontdict=None, **kwargs)\n",
      "        Add text to figure.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : float\n",
      "            The position to place the text. By default, this is in figure\n",
      "            coordinates, floats in [0, 1]. The coordinate system can be changed\n",
      "            using the *transform* keyword.\n",
      "        \n",
      "        s : str\n",
      "            The text string.\n",
      "        \n",
      "        fontdict : dict, optional\n",
      "            A dictionary to override the default text properties. If not given,\n",
      "            the defaults are determined by :rc:`font.*`. Properties passed as\n",
      "            *kwargs* override the corresponding ones given in *fontdict*.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~.text.Text`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.text.Text` properties\n",
      "            Other miscellaneous text parameters.\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            backgroundcolor: color\n",
      "            bbox: dict with properties for `.patches.FancyBboxPatch`\n",
      "            clip_box: unknown\n",
      "            clip_on: unknown\n",
      "            clip_path: unknown\n",
      "            color or c: color\n",
      "            figure: `.Figure`\n",
      "            fontfamily or family: {FONTNAME, 'serif', 'sans-serif', 'cursive', 'fantasy', 'monospace'}\n",
      "            fontproperties or font or font_properties: `.font_manager.FontProperties` or `str` or `pathlib.Path`\n",
      "            fontsize or size: float or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}\n",
      "            fontstretch or stretch: {a numeric value in range 0-1000, 'ultra-condensed', 'extra-condensed', 'condensed', 'semi-condensed', 'normal', 'semi-expanded', 'expanded', 'extra-expanded', 'ultra-expanded'}\n",
      "            fontstyle or style: {'normal', 'italic', 'oblique'}\n",
      "            fontvariant or variant: {'normal', 'small-caps'}\n",
      "            fontweight or weight: {a numeric value in range 0-1000, 'ultralight', 'light', 'normal', 'regular', 'book', 'medium', 'roman', 'semibold', 'demibold', 'demi', 'bold', 'heavy', 'extra bold', 'black'}\n",
      "            gid: str\n",
      "            horizontalalignment or ha: {'left', 'center', 'right'}\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linespacing: float (multiple of font size)\n",
      "            math_fontfamily: str\n",
      "            mouseover: bool\n",
      "            multialignment or ma: {'left', 'right', 'center'}\n",
      "            parse_math: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            position: (float, float)\n",
      "            rasterized: bool\n",
      "            rotation: float or {'vertical', 'horizontal'}\n",
      "            rotation_mode: {None, 'default', 'anchor'}\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            text: object\n",
      "            transform: `.Transform`\n",
      "            transform_rotates_text: bool\n",
      "            url: str\n",
      "            usetex: bool or None\n",
      "            verticalalignment or va: {'bottom', 'baseline', 'center', 'center_baseline', 'top'}\n",
      "            visible: bool\n",
      "            wrap: bool\n",
      "            x: float\n",
      "            y: float\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Axes.text\n",
      "        .pyplot.text\n",
      "    \n",
      "    figure(num=None, figsize=None, dpi=None, *, facecolor=None, edgecolor=None, frameon=True, FigureClass=<class 'matplotlib.figure.Figure'>, clear=False, **kwargs)\n",
      "        Create a new figure, or activate an existing figure.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        num : int or str or `.Figure` or `.SubFigure`, optional\n",
      "            A unique identifier for the figure.\n",
      "        \n",
      "            If a figure with that identifier already exists, this figure is made\n",
      "            active and returned. An integer refers to the ``Figure.number``\n",
      "            attribute, a string refers to the figure label.\n",
      "        \n",
      "            If there is no figure with the identifier or *num* is not given, a new\n",
      "            figure is created, made active and returned.  If *num* is an int, it\n",
      "            will be used for the ``Figure.number`` attribute, otherwise, an\n",
      "            auto-generated integer value is used (starting at 1 and incremented\n",
      "            for each new figure). If *num* is a string, the figure label and the\n",
      "            window title is set to this value.  If num is a ``SubFigure``, its\n",
      "            parent ``Figure`` is activated.\n",
      "        \n",
      "        figsize : (float, float), default: :rc:`figure.figsize`\n",
      "            Width, height in inches.\n",
      "        \n",
      "        dpi : float, default: :rc:`figure.dpi`\n",
      "            The resolution of the figure in dots-per-inch.\n",
      "        \n",
      "        facecolor : color, default: :rc:`figure.facecolor`\n",
      "            The background color.\n",
      "        \n",
      "        edgecolor : color, default: :rc:`figure.edgecolor`\n",
      "            The border color.\n",
      "        \n",
      "        frameon : bool, default: True\n",
      "            If False, suppress drawing the figure frame.\n",
      "        \n",
      "        FigureClass : subclass of `~matplotlib.figure.Figure`\n",
      "            If set, an instance of this subclass will be created, rather than a\n",
      "            plain `.Figure`.\n",
      "        \n",
      "        clear : bool, default: False\n",
      "            If True and the figure already exists, then it is cleared.\n",
      "        \n",
      "        layout : {'constrained', 'tight', `.LayoutEngine`, None}, default: None\n",
      "            The layout mechanism for positioning of plot elements to avoid\n",
      "            overlapping Axes decorations (labels, ticks, etc). Note that layout\n",
      "            managers can measurably slow down figure display. Defaults to *None*\n",
      "            (but see the documentation of the `.Figure` constructor regarding the\n",
      "            interaction with rcParams).\n",
      "        \n",
      "        **kwargs\n",
      "            Additional keyword arguments are passed to the `.Figure` constructor.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.figure.Figure`\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Newly created figures are passed to the `~.FigureCanvasBase.new_manager`\n",
      "        method or the `new_figure_manager` function provided by the current\n",
      "        backend, which install a canvas and a manager on the figure.\n",
      "        \n",
      "        If you are creating many figures, make sure you explicitly call\n",
      "        `.pyplot.close` on the figures you are not using, because this will\n",
      "        enable pyplot to properly clean up the memory.\n",
      "        \n",
      "        `~matplotlib.rcParams` defines the default values, which can be modified\n",
      "        in the matplotlibrc file.\n",
      "    \n",
      "    fill(*args, data=None, **kwargs)\n",
      "        Plot filled polygons.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        *args : sequence of x, y, [color]\n",
      "            Each polygon is defined by the lists of *x* and *y* positions of\n",
      "            its nodes, optionally followed by a *color* specifier. See\n",
      "            :mod:`matplotlib.colors` for supported color specifiers. The\n",
      "            standard color cycle is used for polygons without a color\n",
      "            specifier.\n",
      "        \n",
      "            You can plot multiple polygons by providing multiple *x*, *y*,\n",
      "            *[color]* groups.\n",
      "        \n",
      "            For example, each of the following is legal::\n",
      "        \n",
      "                ax.fill(x, y)                    # a polygon with default color\n",
      "                ax.fill(x, y, \"b\")               # a blue polygon\n",
      "                ax.fill(x, y, x2, y2)            # two polygons\n",
      "                ax.fill(x, y, \"b\", x2, y2, \"r\")  # a blue and a red polygon\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            An object with labelled data. If given, provide the label names to\n",
      "            plot in *x* and *y*, e.g.::\n",
      "        \n",
      "                ax.fill(\"time\", \"signal\",\n",
      "                        data={\"time\": [0, 1, 2], \"signal\": [0, 1, 0]})\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `~matplotlib.patches.Polygon`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.patches.Polygon` properties\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Use :meth:`fill_between` if you would like to fill the region between\n",
      "        two curves.\n",
      "    \n",
      "    fill_between(x, y1, y2=0, where=None, interpolate=False, step=None, *, data=None, **kwargs)\n",
      "        Fill the area between two horizontal curves.\n",
      "        \n",
      "        The curves are defined by the points (*x*, *y1*) and (*x*,\n",
      "        *y2*).  This creates one or multiple polygons describing the filled\n",
      "        area.\n",
      "        \n",
      "        You may exclude some horizontal sections from filling using *where*.\n",
      "        \n",
      "        By default, the edges connect the given points directly.  Use *step*\n",
      "        if the filling should be a step function, i.e. constant in between\n",
      "        *x*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array (length N)\n",
      "            The x coordinates of the nodes defining the curves.\n",
      "        \n",
      "        y1 : array (length N) or scalar\n",
      "            The y coordinates of the nodes defining the first curve.\n",
      "        \n",
      "        y2 : array (length N) or scalar, default: 0\n",
      "            The y coordinates of the nodes defining the second curve.\n",
      "        \n",
      "        where : array of bool (length N), optional\n",
      "            Define *where* to exclude some horizontal regions from being filled.\n",
      "            The filled regions are defined by the coordinates ``x[where]``.\n",
      "            More precisely, fill between ``x[i]`` and ``x[i+1]`` if\n",
      "            ``where[i] and where[i+1]``.  Note that this definition implies\n",
      "            that an isolated *True* value between two *False* values in *where*\n",
      "            will not result in filling.  Both sides of the *True* position\n",
      "            remain unfilled due to the adjacent *False* values.\n",
      "        \n",
      "        interpolate : bool, default: False\n",
      "            This option is only relevant if *where* is used and the two curves\n",
      "            are crossing each other.\n",
      "        \n",
      "            Semantically, *where* is often used for *y1* > *y2* or\n",
      "            similar.  By default, the nodes of the polygon defining the filled\n",
      "            region will only be placed at the positions in the *x* array.\n",
      "            Such a polygon cannot describe the above semantics close to the\n",
      "            intersection.  The x-sections containing the intersection are\n",
      "            simply clipped.\n",
      "        \n",
      "            Setting *interpolate* to *True* will calculate the actual\n",
      "            intersection point and extend the filled region up to this point.\n",
      "        \n",
      "        step : {'pre', 'post', 'mid'}, optional\n",
      "            Define *step* if the filling should be a step function,\n",
      "            i.e. constant in between *x*.  The value determines where the\n",
      "            step will occur:\n",
      "        \n",
      "            - 'pre': The y value is continued constantly to the left from\n",
      "              every *x* position, i.e. the interval ``(x[i-1], x[i]]`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'post': The y value is continued constantly to the right from\n",
      "              every *x* position, i.e. the interval ``[x[i], x[i+1])`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'mid': Steps occur half-way between the *x* positions.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.PolyCollection`\n",
      "            A `.PolyCollection` containing the plotted polygons.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y1*, *y2*, *where*\n",
      "        \n",
      "        **kwargs\n",
      "            All other keyword arguments are passed on to `.PolyCollection`.\n",
      "            They control the `.Polygon` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fill_between : Fill between two sets of y-values.\n",
      "        fill_betweenx : Fill between two sets of x-values.\n",
      "    \n",
      "    fill_betweenx(y, x1, x2=0, where=None, step=None, interpolate=False, *, data=None, **kwargs)\n",
      "        Fill the area between two vertical curves.\n",
      "        \n",
      "        The curves are defined by the points (*y*, *x1*) and (*y*,\n",
      "        *x2*).  This creates one or multiple polygons describing the filled\n",
      "        area.\n",
      "        \n",
      "        You may exclude some vertical sections from filling using *where*.\n",
      "        \n",
      "        By default, the edges connect the given points directly.  Use *step*\n",
      "        if the filling should be a step function, i.e. constant in between\n",
      "        *y*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : array (length N)\n",
      "            The y coordinates of the nodes defining the curves.\n",
      "        \n",
      "        x1 : array (length N) or scalar\n",
      "            The x coordinates of the nodes defining the first curve.\n",
      "        \n",
      "        x2 : array (length N) or scalar, default: 0\n",
      "            The x coordinates of the nodes defining the second curve.\n",
      "        \n",
      "        where : array of bool (length N), optional\n",
      "            Define *where* to exclude some vertical regions from being filled.\n",
      "            The filled regions are defined by the coordinates ``y[where]``.\n",
      "            More precisely, fill between ``y[i]`` and ``y[i+1]`` if\n",
      "            ``where[i] and where[i+1]``.  Note that this definition implies\n",
      "            that an isolated *True* value between two *False* values in *where*\n",
      "            will not result in filling.  Both sides of the *True* position\n",
      "            remain unfilled due to the adjacent *False* values.\n",
      "        \n",
      "        interpolate : bool, default: False\n",
      "            This option is only relevant if *where* is used and the two curves\n",
      "            are crossing each other.\n",
      "        \n",
      "            Semantically, *where* is often used for *x1* > *x2* or\n",
      "            similar.  By default, the nodes of the polygon defining the filled\n",
      "            region will only be placed at the positions in the *y* array.\n",
      "            Such a polygon cannot describe the above semantics close to the\n",
      "            intersection.  The y-sections containing the intersection are\n",
      "            simply clipped.\n",
      "        \n",
      "            Setting *interpolate* to *True* will calculate the actual\n",
      "            intersection point and extend the filled region up to this point.\n",
      "        \n",
      "        step : {'pre', 'post', 'mid'}, optional\n",
      "            Define *step* if the filling should be a step function,\n",
      "            i.e. constant in between *y*.  The value determines where the\n",
      "            step will occur:\n",
      "        \n",
      "            - 'pre': The y value is continued constantly to the left from\n",
      "              every *x* position, i.e. the interval ``(x[i-1], x[i]]`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'post': The y value is continued constantly to the right from\n",
      "              every *x* position, i.e. the interval ``[x[i], x[i+1])`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'mid': Steps occur half-way between the *x* positions.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.PolyCollection`\n",
      "            A `.PolyCollection` containing the plotted polygons.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *y*, *x1*, *x2*, *where*\n",
      "        \n",
      "        **kwargs\n",
      "            All other keyword arguments are passed on to `.PolyCollection`.\n",
      "            They control the `.Polygon` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fill_between : Fill between two sets of y-values.\n",
      "        fill_betweenx : Fill between two sets of x-values.\n",
      "    \n",
      "    findobj(o=None, match=None, include_self=True)\n",
      "        Find artist objects.\n",
      "        \n",
      "        Recursively find all `.Artist` instances contained in the artist.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        match\n",
      "            A filter criterion for the matches. This can be\n",
      "        \n",
      "            - *None*: Return all objects contained in artist.\n",
      "            - A function with signature ``def match(artist: Artist) -> bool``.\n",
      "              The result will only contain artists for which the function\n",
      "              returns *True*.\n",
      "            - A class instance: e.g., `.Line2D`. The result will only contain\n",
      "              artists of this class or its subclasses (``isinstance`` check).\n",
      "        \n",
      "        include_self : bool\n",
      "            Include *self* in the list to be checked for a match.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Artist`\n",
      "    \n",
      "    flag()\n",
      "        Set the colormap to 'flag'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    gca()\n",
      "        Get the current Axes.\n",
      "        \n",
      "        If there is currently no Axes on this Figure, a new one is created\n",
      "        using `.Figure.add_subplot`.  (To test whether there is currently an\n",
      "        Axes on a Figure, check whether ``figure.axes`` is empty.  To test\n",
      "        whether there is currently a Figure on the pyplot figure stack, check\n",
      "        whether `.pyplot.get_fignums()` is empty.)\n",
      "    \n",
      "    gcf()\n",
      "        Get the current figure.\n",
      "        \n",
      "        If there is currently no figure on the pyplot figure stack, a new one is\n",
      "        created using `~.pyplot.figure()`.  (To test whether there is currently a\n",
      "        figure on the pyplot figure stack, check whether `~.pyplot.get_fignums()`\n",
      "        is empty.)\n",
      "    \n",
      "    gci()\n",
      "        Get the current colorable artist.\n",
      "        \n",
      "        Specifically, returns the current `.ScalarMappable` instance (`.Image`\n",
      "        created by `imshow` or `figimage`, `.Collection` created by `pcolor` or\n",
      "        `scatter`, etc.), or *None* if no such instance has been defined.\n",
      "        \n",
      "        The current image is an attribute of the current Axes, or the nearest\n",
      "        earlier Axes in the current figure that contains an image.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Historically, the only colorable artists were images; hence the name\n",
      "        ``gci`` (get current image).\n",
      "    \n",
      "    get(obj, *args, **kwargs)\n",
      "        Return the value of an `.Artist`'s *property*, or print all of them.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        obj : `.Artist`\n",
      "            The queried artist; e.g., a `.Line2D`, a `.Text`, or an `~.axes.Axes`.\n",
      "        \n",
      "        property : str or None, default: None\n",
      "            If *property* is 'somename', this function returns\n",
      "            ``obj.get_somename()``.\n",
      "        \n",
      "            If it's None (or unset), it *prints* all gettable properties from\n",
      "            *obj*.  Many properties have aliases for shorter typing, e.g. 'lw' is\n",
      "            an alias for 'linewidth'.  In the output, aliases and full property\n",
      "            names will be listed as:\n",
      "        \n",
      "              property or alias = value\n",
      "        \n",
      "            e.g.:\n",
      "        \n",
      "              linewidth or lw = 2\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        setp\n",
      "    \n",
      "    get_cmap(name=None, lut=None)\n",
      "        Get a colormap instance, defaulting to rc values if *name* is None.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        name : `matplotlib.colors.Colormap` or str or None, default: None\n",
      "            If a `.Colormap` instance, it will be returned. Otherwise, the name of\n",
      "            a colormap known to Matplotlib, which will be resampled by *lut*. The\n",
      "            default, None, means :rc:`image.cmap`.\n",
      "        lut : int or None, default: None\n",
      "            If *name* is not already a Colormap instance and *lut* is not None, the\n",
      "            colormap will be resampled to have *lut* entries in the lookup table.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        Colormap\n",
      "    \n",
      "    get_current_fig_manager()\n",
      "        Return the figure manager of the current figure.\n",
      "        \n",
      "        The figure manager is a container for the actual backend-depended window\n",
      "        that displays the figure on screen.\n",
      "        \n",
      "        If no current figure exists, a new one is created, and its figure\n",
      "        manager is returned.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.FigureManagerBase` or backend-dependent subclass thereof\n",
      "    \n",
      "    get_figlabels()\n",
      "        Return a list of existing figure labels.\n",
      "    \n",
      "    get_fignums()\n",
      "        Return a list of existing figure numbers.\n",
      "    \n",
      "    get_plot_commands()\n",
      "        Get a sorted list of all of the plotting commands.\n",
      "    \n",
      "    getp(obj, *args, **kwargs)\n",
      "        Return the value of an `.Artist`'s *property*, or print all of them.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        obj : `.Artist`\n",
      "            The queried artist; e.g., a `.Line2D`, a `.Text`, or an `~.axes.Axes`.\n",
      "        \n",
      "        property : str or None, default: None\n",
      "            If *property* is 'somename', this function returns\n",
      "            ``obj.get_somename()``.\n",
      "        \n",
      "            If it's None (or unset), it *prints* all gettable properties from\n",
      "            *obj*.  Many properties have aliases for shorter typing, e.g. 'lw' is\n",
      "            an alias for 'linewidth'.  In the output, aliases and full property\n",
      "            names will be listed as:\n",
      "        \n",
      "              property or alias = value\n",
      "        \n",
      "            e.g.:\n",
      "        \n",
      "              linewidth or lw = 2\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        setp\n",
      "    \n",
      "    ginput(n=1, timeout=30, show_clicks=True, mouse_add=<MouseButton.LEFT: 1>, mouse_pop=<MouseButton.RIGHT: 3>, mouse_stop=<MouseButton.MIDDLE: 2>)\n",
      "        Blocking call to interact with a figure.\n",
      "        \n",
      "        Wait until the user clicks *n* times on the figure, and return the\n",
      "        coordinates of each click in a list.\n",
      "        \n",
      "        There are three possible interactions:\n",
      "        \n",
      "        - Add a point.\n",
      "        - Remove the most recently added point.\n",
      "        - Stop the interaction and return the points added so far.\n",
      "        \n",
      "        The actions are assigned to mouse buttons via the arguments\n",
      "        *mouse_add*, *mouse_pop* and *mouse_stop*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        n : int, default: 1\n",
      "            Number of mouse clicks to accumulate. If negative, accumulate\n",
      "            clicks until the input is terminated manually.\n",
      "        timeout : float, default: 30 seconds\n",
      "            Number of seconds to wait before timing out. If zero or negative\n",
      "            will never timeout.\n",
      "        show_clicks : bool, default: True\n",
      "            If True, show a red cross at the location of each click.\n",
      "        mouse_add : `.MouseButton` or None, default: `.MouseButton.LEFT`\n",
      "            Mouse button used to add points.\n",
      "        mouse_pop : `.MouseButton` or None, default: `.MouseButton.RIGHT`\n",
      "            Mouse button used to remove the most recently added point.\n",
      "        mouse_stop : `.MouseButton` or None, default: `.MouseButton.MIDDLE`\n",
      "            Mouse button used to stop input.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of tuples\n",
      "            A list of the clicked (x, y) coordinates.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The keyboard can also be used to select points in case your mouse\n",
      "        does not have one or more of the buttons.  The delete and backspace\n",
      "        keys act like right clicking (i.e., remove last point), the enter key\n",
      "        terminates input and any other key (not already used by the window\n",
      "        manager) selects a point.\n",
      "    \n",
      "    gray()\n",
      "        Set the colormap to 'gray'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    grid(visible=None, which='major', axis='both', **kwargs)\n",
      "        Configure the grid lines.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        visible : bool or None, optional\n",
      "            Whether to show the grid lines.  If any *kwargs* are supplied, it\n",
      "            is assumed you want the grid on and *visible* will be set to True.\n",
      "        \n",
      "            If *visible* is *None* and there are no *kwargs*, this toggles the\n",
      "            visibility of the lines.\n",
      "        \n",
      "        which : {'major', 'minor', 'both'}, optional\n",
      "            The grid lines to apply the changes on.\n",
      "        \n",
      "        axis : {'both', 'x', 'y'}, optional\n",
      "            The axis to apply the changes on.\n",
      "        \n",
      "        **kwargs : `.Line2D` properties\n",
      "            Define the line properties of the grid, e.g.::\n",
      "        \n",
      "                grid(color='r', linestyle='-', linewidth=2)\n",
      "        \n",
      "            Valid keyword arguments are:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The axis is drawn as a unit, so the effective zorder for drawing the\n",
      "        grid is determined by the zorder of each axis, not by the zorder of the\n",
      "        `.Line2D` objects comprising the grid.  Therefore, to set grid zorder,\n",
      "        use `.set_axisbelow` or, for more control, call the\n",
      "        `~.Artist.set_zorder` method of each axis.\n",
      "    \n",
      "    hexbin(x, y, C=None, gridsize=100, bins=None, xscale='linear', yscale='linear', extent=None, cmap=None, norm=None, vmin=None, vmax=None, alpha=None, linewidths=None, edgecolors='face', reduce_C_function=<function mean at 0x10d861d70>, mincnt=None, marginals=False, *, data=None, **kwargs)\n",
      "        Make a 2D hexagonal binning plot of points *x*, *y*.\n",
      "        \n",
      "        If *C* is *None*, the value of the hexagon is determined by the number\n",
      "        of points in the hexagon. Otherwise, *C* specifies values at the\n",
      "        coordinate (x[i], y[i]). For each hexagon, these values are reduced\n",
      "        using *reduce_C_function*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like\n",
      "            The data positions. *x* and *y* must be of the same length.\n",
      "        \n",
      "        C : array-like, optional\n",
      "            If given, these values are accumulated in the bins. Otherwise,\n",
      "            every point has a value of 1. Must be of the same length as *x*\n",
      "            and *y*.\n",
      "        \n",
      "        gridsize : int or (int, int), default: 100\n",
      "            If a single int, the number of hexagons in the *x*-direction.\n",
      "            The number of hexagons in the *y*-direction is chosen such that\n",
      "            the hexagons are approximately regular.\n",
      "        \n",
      "            Alternatively, if a tuple (*nx*, *ny*), the number of hexagons\n",
      "            in the *x*-direction and the *y*-direction. In the\n",
      "            *y*-direction, counting is done along vertically aligned\n",
      "            hexagons, not along the zig-zag chains of hexagons; see the\n",
      "            following illustration.\n",
      "        \n",
      "            .. plot::\n",
      "        \n",
      "               import numpy\n",
      "               import matplotlib.pyplot as plt\n",
      "        \n",
      "               np.random.seed(19680801)\n",
      "               n= 300\n",
      "               x = np.random.standard_normal(n)\n",
      "               y = np.random.standard_normal(n)\n",
      "        \n",
      "               fig, ax = plt.subplots(figsize=(4, 4))\n",
      "               h = ax.hexbin(x, y, gridsize=(5, 3))\n",
      "               hx, hy = h.get_offsets().T\n",
      "               ax.plot(hx[24::3], hy[24::3], 'ro-')\n",
      "               ax.plot(hx[-3:], hy[-3:], 'ro-')\n",
      "               ax.set_title('gridsize=(5, 3)')\n",
      "               ax.axis('off')\n",
      "        \n",
      "            To get approximately regular hexagons, choose\n",
      "            :math:`n_x = \\sqrt{3}\\,n_y`.\n",
      "        \n",
      "        bins : 'log' or int or sequence, default: None\n",
      "            Discretization of the hexagon values.\n",
      "        \n",
      "            - If *None*, no binning is applied; the color of each hexagon\n",
      "              directly corresponds to its count value.\n",
      "            - If 'log', use a logarithmic scale for the colormap.\n",
      "              Internally, :math:`log_{10}(i+1)` is used to determine the\n",
      "              hexagon color. This is equivalent to ``norm=LogNorm()``.\n",
      "            - If an integer, divide the counts in the specified number\n",
      "              of bins, and color the hexagons accordingly.\n",
      "            - If a sequence of values, the values of the lower bound of\n",
      "              the bins to be used.\n",
      "        \n",
      "        xscale : {'linear', 'log'}, default: 'linear'\n",
      "            Use a linear or log10 scale on the horizontal axis.\n",
      "        \n",
      "        yscale : {'linear', 'log'}, default: 'linear'\n",
      "            Use a linear or log10 scale on the vertical axis.\n",
      "        \n",
      "        mincnt : int > 0, default: *None*\n",
      "            If not *None*, only display cells with more than *mincnt*\n",
      "            number of points in the cell.\n",
      "        \n",
      "        marginals : bool, default: *False*\n",
      "            If marginals is *True*, plot the marginal density as\n",
      "            colormapped rectangles along the bottom of the x-axis and\n",
      "            left of the y-axis.\n",
      "        \n",
      "        extent : 4-tuple of float, default: *None*\n",
      "            The limits of the bins (xmin, xmax, ymin, ymax).\n",
      "            The default assigns the limits based on\n",
      "            *gridsize*, *x*, *y*, *xscale* and *yscale*.\n",
      "        \n",
      "            If *xscale* or *yscale* is set to 'log', the limits are\n",
      "            expected to be the exponent for a power of 10. E.g. for\n",
      "            x-limits of 1 and 50 in 'linear' scale and y-limits\n",
      "            of 10 and 1000 in 'log' scale, enter (1, 50, 1, 3).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.collections.PolyCollection`\n",
      "            A `.PolyCollection` defining the hexagonal bins.\n",
      "        \n",
      "            - `.PolyCollection.get_offsets` contains a Mx2 array containing\n",
      "              the x, y positions of the M hexagon centers.\n",
      "            - `.PolyCollection.get_array` contains the values of the M\n",
      "              hexagons.\n",
      "        \n",
      "            If *marginals* is *True*, horizontal\n",
      "            bar and vertical bar (both PolyCollections) will be attached\n",
      "            to the return collection as attributes *hbar* and *vbar*.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "        alpha : float between 0 and 1, optional\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        linewidths : float, default: *None*\n",
      "            If *None*, defaults to 1.0.\n",
      "        \n",
      "        edgecolors : {'face', 'none', *None*} or color, default: 'face'\n",
      "            The color of the hexagon edges. Possible values are:\n",
      "        \n",
      "            - 'face': Draw the edges in the same color as the fill color.\n",
      "            - 'none': No edges are drawn. This can sometimes lead to unsightly\n",
      "              unpainted pixels between the hexagons.\n",
      "            - *None*: Draw outlines in the default color.\n",
      "            - An explicit color.\n",
      "        \n",
      "        reduce_C_function : callable, default: `numpy.mean`\n",
      "            The function to aggregate *C* within the bins. It is ignored if\n",
      "            *C* is not given. This must have the signature::\n",
      "        \n",
      "                def reduce_C_function(C: array) -> float\n",
      "        \n",
      "            Commonly used functions are:\n",
      "        \n",
      "            - `numpy.mean`: average of the points\n",
      "            - `numpy.sum`: integral of the point values\n",
      "            - `numpy.amax`: value taken from the largest point\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*, *C*\n",
      "        \n",
      "        **kwargs : `~matplotlib.collections.PolyCollection` properties\n",
      "            All other keyword arguments are passed on to `.PolyCollection`:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        hist2d : 2D histogram rectangular bins\n",
      "    \n",
      "    hist(x, bins=None, range=None, density=False, weights=None, cumulative=False, bottom=None, histtype='bar', align='mid', orientation='vertical', rwidth=None, log=False, color=None, label=None, stacked=False, *, data=None, **kwargs)\n",
      "        Compute and plot a histogram.\n",
      "        \n",
      "        This method uses `numpy.histogram` to bin the data in *x* and count the\n",
      "        number of values in each bin, then draws the distribution either as a\n",
      "        `.BarContainer` or `.Polygon`. The *bins*, *range*, *density*, and\n",
      "        *weights* parameters are forwarded to `numpy.histogram`.\n",
      "        \n",
      "        If the data has already been binned and counted, use `~.bar` or\n",
      "        `~.stairs` to plot the distribution::\n",
      "        \n",
      "            counts, bins = np.histogram(x)\n",
      "            plt.stairs(counts, bins)\n",
      "        \n",
      "        Alternatively, plot pre-computed bins and counts using ``hist()`` by\n",
      "        treating each bin as a single point with a weight equal to its count::\n",
      "        \n",
      "            plt.hist(bins[:-1], bins, weights=counts)\n",
      "        \n",
      "        The data input *x* can be a singular array, a list of datasets of\n",
      "        potentially different lengths ([*x0*, *x1*, ...]), or a 2D ndarray in\n",
      "        which each column is a dataset. Note that the ndarray form is\n",
      "        transposed relative to the list form. If the input is an array, then\n",
      "        the return value is a tuple (*n*, *bins*, *patches*); if the input is a\n",
      "        sequence of arrays, then the return value is a tuple\n",
      "        ([*n0*, *n1*, ...], *bins*, [*patches0*, *patches1*, ...]).\n",
      "        \n",
      "        Masked arrays are not supported.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (n,) array or sequence of (n,) arrays\n",
      "            Input values, this takes either a single array or a sequence of\n",
      "            arrays which are not required to be of the same length.\n",
      "        \n",
      "        bins : int or sequence or str, default: :rc:`hist.bins`\n",
      "            If *bins* is an integer, it defines the number of equal-width bins\n",
      "            in the range.\n",
      "        \n",
      "            If *bins* is a sequence, it defines the bin edges, including the\n",
      "            left edge of the first bin and the right edge of the last bin;\n",
      "            in this case, bins may be unequally spaced.  All but the last\n",
      "            (righthand-most) bin is half-open.  In other words, if *bins* is::\n",
      "        \n",
      "                [1, 2, 3, 4]\n",
      "        \n",
      "            then the first bin is ``[1, 2)`` (including 1, but excluding 2) and\n",
      "            the second ``[2, 3)``.  The last bin, however, is ``[3, 4]``, which\n",
      "            *includes* 4.\n",
      "        \n",
      "            If *bins* is a string, it is one of the binning strategies\n",
      "            supported by `numpy.histogram_bin_edges`: 'auto', 'fd', 'doane',\n",
      "            'scott', 'stone', 'rice', 'sturges', or 'sqrt'.\n",
      "        \n",
      "        range : tuple or None, default: None\n",
      "            The lower and upper range of the bins. Lower and upper outliers\n",
      "            are ignored. If not provided, *range* is ``(x.min(), x.max())``.\n",
      "            Range has no effect if *bins* is a sequence.\n",
      "        \n",
      "            If *bins* is a sequence or *range* is specified, autoscaling\n",
      "            is based on the specified bin range instead of the\n",
      "            range of x.\n",
      "        \n",
      "        density : bool, default: False\n",
      "            If ``True``, draw and return a probability density: each bin\n",
      "            will display the bin's raw count divided by the total number of\n",
      "            counts *and the bin width*\n",
      "            (``density = counts / (sum(counts) * np.diff(bins))``),\n",
      "            so that the area under the histogram integrates to 1\n",
      "            (``np.sum(density * np.diff(bins)) == 1``).\n",
      "        \n",
      "            If *stacked* is also ``True``, the sum of the histograms is\n",
      "            normalized to 1.\n",
      "        \n",
      "        weights : (n,) array-like or None, default: None\n",
      "            An array of weights, of the same shape as *x*.  Each value in\n",
      "            *x* only contributes its associated weight towards the bin count\n",
      "            (instead of 1).  If *density* is ``True``, the weights are\n",
      "            normalized, so that the integral of the density over the range\n",
      "            remains 1.\n",
      "        \n",
      "        cumulative : bool or -1, default: False\n",
      "            If ``True``, then a histogram is computed where each bin gives the\n",
      "            counts in that bin plus all bins for smaller values. The last bin\n",
      "            gives the total number of datapoints.\n",
      "        \n",
      "            If *density* is also ``True`` then the histogram is normalized such\n",
      "            that the last bin equals 1.\n",
      "        \n",
      "            If *cumulative* is a number less than 0 (e.g., -1), the direction\n",
      "            of accumulation is reversed.  In this case, if *density* is also\n",
      "            ``True``, then the histogram is normalized such that the first bin\n",
      "            equals 1.\n",
      "        \n",
      "        bottom : array-like, scalar, or None, default: None\n",
      "            Location of the bottom of each bin, ie. bins are drawn from\n",
      "            ``bottom`` to ``bottom + hist(x, bins)`` If a scalar, the bottom\n",
      "            of each bin is shifted by the same amount. If an array, each bin\n",
      "            is shifted independently and the length of bottom must match the\n",
      "            number of bins. If None, defaults to 0.\n",
      "        \n",
      "        histtype : {'bar', 'barstacked', 'step', 'stepfilled'}, default: 'bar'\n",
      "            The type of histogram to draw.\n",
      "        \n",
      "            - 'bar' is a traditional bar-type histogram.  If multiple data\n",
      "              are given the bars are arranged side by side.\n",
      "            - 'barstacked' is a bar-type histogram where multiple\n",
      "              data are stacked on top of each other.\n",
      "            - 'step' generates a lineplot that is by default unfilled.\n",
      "            - 'stepfilled' generates a lineplot that is by default filled.\n",
      "        \n",
      "        align : {'left', 'mid', 'right'}, default: 'mid'\n",
      "            The horizontal alignment of the histogram bars.\n",
      "        \n",
      "            - 'left': bars are centered on the left bin edges.\n",
      "            - 'mid': bars are centered between the bin edges.\n",
      "            - 'right': bars are centered on the right bin edges.\n",
      "        \n",
      "        orientation : {'vertical', 'horizontal'}, default: 'vertical'\n",
      "            If 'horizontal', `~.Axes.barh` will be used for bar-type histograms\n",
      "            and the *bottom* kwarg will be the left edges.\n",
      "        \n",
      "        rwidth : float or None, default: None\n",
      "            The relative width of the bars as a fraction of the bin width.  If\n",
      "            ``None``, automatically compute the width.\n",
      "        \n",
      "            Ignored if *histtype* is 'step' or 'stepfilled'.\n",
      "        \n",
      "        log : bool, default: False\n",
      "            If ``True``, the histogram axis will be set to a log scale.\n",
      "        \n",
      "        color : color or array-like of colors or None, default: None\n",
      "            Color or sequence of colors, one per dataset.  Default (``None``)\n",
      "            uses the standard line color sequence.\n",
      "        \n",
      "        label : str or None, default: None\n",
      "            String, or sequence of strings to match multiple datasets.  Bar\n",
      "            charts yield multiple patches per dataset, but only the first gets\n",
      "            the label, so that `~.Axes.legend` will work as expected.\n",
      "        \n",
      "        stacked : bool, default: False\n",
      "            If ``True``, multiple data are stacked on top of each other If\n",
      "            ``False`` multiple data are arranged side by side if histtype is\n",
      "            'bar' or on top of each other if histtype is 'step'\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        n : array or list of arrays\n",
      "            The values of the histogram bins. See *density* and *weights* for a\n",
      "            description of the possible semantics.  If input *x* is an array,\n",
      "            then this is an array of length *nbins*. If input is a sequence of\n",
      "            arrays ``[data1, data2, ...]``, then this is a list of arrays with\n",
      "            the values of the histograms for each of the arrays in the same\n",
      "            order.  The dtype of the array *n* (or of its element arrays) will\n",
      "            always be float even if no weighting or normalization is used.\n",
      "        \n",
      "        bins : array\n",
      "            The edges of the bins. Length nbins + 1 (nbins left edges and right\n",
      "            edge of last bin).  Always a single array even when multiple data\n",
      "            sets are passed in.\n",
      "        \n",
      "        patches : `.BarContainer` or list of a single `.Polygon` or list of such objects\n",
      "            Container of individual artists used to create the histogram\n",
      "            or list of such containers if there are multiple input datasets.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *weights*\n",
      "        \n",
      "        **kwargs\n",
      "            `~matplotlib.patches.Patch` properties\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        hist2d : 2D histogram with rectangular bins\n",
      "        hexbin : 2D histogram with hexagonal bins\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For large numbers of bins (>1000), plotting can be significantly faster\n",
      "        if *histtype* is set to 'step' or 'stepfilled' rather than 'bar' or\n",
      "        'barstacked'.\n",
      "    \n",
      "    hist2d(x, y, bins=10, range=None, density=False, weights=None, cmin=None, cmax=None, *, data=None, **kwargs)\n",
      "        Make a 2D histogram plot.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like, shape (n, )\n",
      "            Input values\n",
      "        \n",
      "        bins : None or int or [int, int] or array-like or [array, array]\n",
      "        \n",
      "            The bin specification:\n",
      "        \n",
      "            - If int, the number of bins for the two dimensions\n",
      "              (nx=ny=bins).\n",
      "            - If ``[int, int]``, the number of bins in each dimension\n",
      "              (nx, ny = bins).\n",
      "            - If array-like, the bin edges for the two dimensions\n",
      "              (x_edges=y_edges=bins).\n",
      "            - If ``[array, array]``, the bin edges in each dimension\n",
      "              (x_edges, y_edges = bins).\n",
      "        \n",
      "            The default value is 10.\n",
      "        \n",
      "        range : array-like shape(2, 2), optional\n",
      "            The leftmost and rightmost edges of the bins along each dimension\n",
      "            (if not specified explicitly in the bins parameters): ``[[xmin,\n",
      "            xmax], [ymin, ymax]]``. All values outside of this range will be\n",
      "            considered outliers and not tallied in the histogram.\n",
      "        \n",
      "        density : bool, default: False\n",
      "            Normalize histogram.  See the documentation for the *density*\n",
      "            parameter of `~.Axes.hist` for more details.\n",
      "        \n",
      "        weights : array-like, shape (n, ), optional\n",
      "            An array of values w_i weighing each sample (x_i, y_i).\n",
      "        \n",
      "        cmin, cmax : float, default: None\n",
      "            All bins that has count less than *cmin* or more than *cmax* will\n",
      "            not be displayed (set to NaN before passing to imshow) and these\n",
      "            count values in the return value count histogram will also be set\n",
      "            to nan upon return.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        h : 2D array\n",
      "            The bi-dimensional histogram of samples x and y. Values in x are\n",
      "            histogrammed along the first dimension and values in y are\n",
      "            histogrammed along the second dimension.\n",
      "        xedges : 1D array\n",
      "            The bin edges along the x axis.\n",
      "        yedges : 1D array\n",
      "            The bin edges along the y axis.\n",
      "        image : `~.matplotlib.collections.QuadMesh`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "        alpha : ``0 <= scalar <= 1`` or ``None``, optional\n",
      "            The alpha blending value.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*, *weights*\n",
      "        \n",
      "        **kwargs\n",
      "            Additional parameters are passed along to the\n",
      "            `~.Axes.pcolormesh` method and `~matplotlib.collections.QuadMesh`\n",
      "            constructor.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        hist : 1D histogram plotting\n",
      "        hexbin : 2D histogram with hexagonal bins\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        - Currently ``hist2d`` calculates its own axis limits, and any limits\n",
      "          previously set are ignored.\n",
      "        - Rendering the histogram with a logarithmic color scale is\n",
      "          accomplished by passing a `.colors.LogNorm` instance to the *norm*\n",
      "          keyword argument. Likewise, power-law normalization (similar\n",
      "          in effect to gamma correction) can be accomplished with\n",
      "          `.colors.PowerNorm`.\n",
      "    \n",
      "    hlines(y, xmin, xmax, colors=None, linestyles='solid', label='', *, data=None, **kwargs)\n",
      "        Plot horizontal lines at each *y* from *xmin* to *xmax*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : float or array-like\n",
      "            y-indexes where to plot the lines.\n",
      "        \n",
      "        xmin, xmax : float or array-like\n",
      "            Respective beginning and end of each line. If scalars are\n",
      "            provided, all lines will have the same length.\n",
      "        \n",
      "        colors : list of colors, default: :rc:`lines.color`\n",
      "        \n",
      "        linestyles : {'solid', 'dashed', 'dashdot', 'dotted'}, optional\n",
      "        \n",
      "        label : str, default: ''\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.collections.LineCollection`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *y*, *xmin*, *xmax*, *colors*\n",
      "        **kwargs :  `~matplotlib.collections.LineCollection` properties.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        vlines : vertical lines\n",
      "        axhline : horizontal line across the Axes\n",
      "    \n",
      "    hot()\n",
      "        Set the colormap to 'hot'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    hsv()\n",
      "        Set the colormap to 'hsv'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    imread(fname, format=None)\n",
      "        Read an image from a file into an array.\n",
      "        \n",
      "        .. note::\n",
      "        \n",
      "            This function exists for historical reasons.  It is recommended to\n",
      "            use `PIL.Image.open` instead for loading images.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fname : str or file-like\n",
      "            The image file to read: a filename, a URL or a file-like object opened\n",
      "            in read-binary mode.\n",
      "        \n",
      "            Passing a URL is deprecated.  Please open the URL\n",
      "            for reading and pass the result to Pillow, e.g. with\n",
      "            ``np.array(PIL.Image.open(urllib.request.urlopen(url)))``.\n",
      "        format : str, optional\n",
      "            The image file format assumed for reading the data.  The image is\n",
      "            loaded as a PNG file if *format* is set to \"png\", if *fname* is a path\n",
      "            or opened file with a \".png\" extension, or if it is an URL.  In all\n",
      "            other cases, *format* is ignored and the format is auto-detected by\n",
      "            `PIL.Image.open`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `numpy.array`\n",
      "            The image data. The returned array has shape\n",
      "        \n",
      "            - (M, N) for grayscale images.\n",
      "            - (M, N, 3) for RGB images.\n",
      "            - (M, N, 4) for RGBA images.\n",
      "        \n",
      "            PNG images are returned as float arrays (0-1).  All other formats are\n",
      "            returned as int arrays, with a bit depth determined by the file's\n",
      "            contents.\n",
      "    \n",
      "    imsave(fname, arr, **kwargs)\n",
      "        Colormap and save an array as an image file.\n",
      "        \n",
      "        RGB(A) images are passed through.  Single channel images will be\n",
      "        colormapped according to *cmap* and *norm*.\n",
      "        \n",
      "        .. note ::\n",
      "        \n",
      "           If you want to save a single channel image as gray scale please use an\n",
      "           image I/O library (such as pillow, tifffile, or imageio) directly.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fname : str or path-like or file-like\n",
      "            A path or a file-like object to store the image in.\n",
      "            If *format* is not set, then the output format is inferred from the\n",
      "            extension of *fname*, if any, and from :rc:`savefig.format` otherwise.\n",
      "            If *format* is set, it determines the output format.\n",
      "        arr : array-like\n",
      "            The image data. The shape can be one of\n",
      "            MxN (luminance), MxNx3 (RGB) or MxNx4 (RGBA).\n",
      "        vmin, vmax : float, optional\n",
      "            *vmin* and *vmax* set the color scaling for the image by fixing the\n",
      "            values that map to the colormap color limits. If either *vmin*\n",
      "            or *vmax* is None, that limit is determined from the *arr*\n",
      "            min/max value.\n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            A Colormap instance or registered colormap name. The colormap\n",
      "            maps scalar data to colors. It is ignored for RGB(A) data.\n",
      "        format : str, optional\n",
      "            The file format, e.g. 'png', 'pdf', 'svg', ...  The behavior when this\n",
      "            is unset is documented under *fname*.\n",
      "        origin : {'upper', 'lower'}, default: :rc:`image.origin`\n",
      "            Indicates whether the ``(0, 0)`` index of the array is in the upper\n",
      "            left or lower left corner of the axes.\n",
      "        dpi : float\n",
      "            The DPI to store in the metadata of the file.  This does not affect the\n",
      "            resolution of the output image.  Depending on file format, this may be\n",
      "            rounded to the nearest integer.\n",
      "        metadata : dict, optional\n",
      "            Metadata in the image file.  The supported keys depend on the output\n",
      "            format, see the documentation of the respective backends for more\n",
      "            information.\n",
      "        pil_kwargs : dict, optional\n",
      "            Keyword arguments passed to `PIL.Image.Image.save`.  If the 'pnginfo'\n",
      "            key is present, it completely overrides *metadata*, including the\n",
      "            default 'Software' key.\n",
      "    \n",
      "    imshow(X, cmap=None, norm=None, *, aspect=None, interpolation=None, alpha=None, vmin=None, vmax=None, origin=None, extent=None, interpolation_stage=None, filternorm=True, filterrad=4.0, resample=None, url=None, data=None, **kwargs)\n",
      "        Display data as an image, i.e., on a 2D regular raster.\n",
      "        \n",
      "        The input may either be actual RGB(A) data, or 2D scalar data, which\n",
      "        will be rendered as a pseudocolor image. For displaying a grayscale\n",
      "        image set up the colormapping using the parameters\n",
      "        ``cmap='gray', vmin=0, vmax=255``.\n",
      "        \n",
      "        The number of pixels used to render an image is set by the Axes size\n",
      "        and the *dpi* of the figure. This can lead to aliasing artifacts when\n",
      "        the image is resampled because the displayed image size will usually\n",
      "        not match the size of *X* (see\n",
      "        :doc:`/gallery/images_contours_and_fields/image_antialiasing`).\n",
      "        The resampling can be controlled via the *interpolation* parameter\n",
      "        and/or :rc:`image.interpolation`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X : array-like or PIL image\n",
      "            The image data. Supported array shapes are:\n",
      "        \n",
      "            - (M, N): an image with scalar data. The values are mapped to\n",
      "              colors using normalization and a colormap. See parameters *norm*,\n",
      "              *cmap*, *vmin*, *vmax*.\n",
      "            - (M, N, 3): an image with RGB values (0-1 float or 0-255 int).\n",
      "            - (M, N, 4): an image with RGBA values (0-1 float or 0-255 int),\n",
      "              i.e. including transparency.\n",
      "        \n",
      "            The first two dimensions (M, N) define the rows and columns of\n",
      "            the image.\n",
      "        \n",
      "            Out-of-range RGB(A) values are clipped.\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            This parameter is ignored if *X* is RGB(A).\n",
      "        \n",
      "        aspect : {'equal', 'auto'} or float, default: :rc:`image.aspect`\n",
      "            The aspect ratio of the Axes.  This parameter is particularly\n",
      "            relevant for images since it determines whether data pixels are\n",
      "            square.\n",
      "        \n",
      "            This parameter is a shortcut for explicitly calling\n",
      "            `.Axes.set_aspect`. See there for further details.\n",
      "        \n",
      "            - 'equal': Ensures an aspect ratio of 1. Pixels will be square\n",
      "              (unless pixel sizes are explicitly made non-square in data\n",
      "              coordinates using *extent*).\n",
      "            - 'auto': The Axes is kept fixed and the aspect is adjusted so\n",
      "              that the data fit in the Axes. In general, this will result in\n",
      "              non-square pixels.\n",
      "        \n",
      "        interpolation : str, default: :rc:`image.interpolation`\n",
      "            The interpolation method used.\n",
      "        \n",
      "            Supported values are 'none', 'antialiased', 'nearest', 'bilinear',\n",
      "            'bicubic', 'spline16', 'spline36', 'hanning', 'hamming', 'hermite',\n",
      "            'kaiser', 'quadric', 'catrom', 'gaussian', 'bessel', 'mitchell',\n",
      "            'sinc', 'lanczos', 'blackman'.\n",
      "        \n",
      "            If *interpolation* is 'none', then no interpolation is performed\n",
      "            on the Agg, ps, pdf and svg backends. Other backends will fall back\n",
      "            to 'nearest'. Note that most SVG renderers perform interpolation at\n",
      "            rendering and that the default interpolation method they implement\n",
      "            may differ.\n",
      "        \n",
      "            If *interpolation* is the default 'antialiased', then 'nearest'\n",
      "            interpolation is used if the image is upsampled by more than a\n",
      "            factor of three (i.e. the number of display pixels is at least\n",
      "            three times the size of the data array).  If the upsampling rate is\n",
      "            smaller than 3, or the image is downsampled, then 'hanning'\n",
      "            interpolation is used to act as an anti-aliasing filter, unless the\n",
      "            image happens to be upsampled by exactly a factor of two or one.\n",
      "        \n",
      "            See\n",
      "            :doc:`/gallery/images_contours_and_fields/interpolation_methods`\n",
      "            for an overview of the supported interpolation methods, and\n",
      "            :doc:`/gallery/images_contours_and_fields/image_antialiasing` for\n",
      "            a discussion of image antialiasing.\n",
      "        \n",
      "            Some interpolation methods require an additional radius parameter,\n",
      "            which can be set by *filterrad*. Additionally, the antigrain image\n",
      "            resize filter is controlled by the parameter *filternorm*.\n",
      "        \n",
      "        interpolation_stage : {'data', 'rgba'}, default: 'data'\n",
      "            If 'data', interpolation\n",
      "            is carried out on the data provided by the user.  If 'rgba', the\n",
      "            interpolation is carried out after the colormapping has been\n",
      "            applied (visual interpolation).\n",
      "        \n",
      "        alpha : float or array-like, optional\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "            If *alpha* is an array, the alpha blending values are applied pixel\n",
      "            by pixel, and *alpha* must have the same shape as *X*.\n",
      "        \n",
      "        origin : {'upper', 'lower'}, default: :rc:`image.origin`\n",
      "            Place the [0, 0] index of the array in the upper left or lower\n",
      "            left corner of the Axes. The convention (the default) 'upper' is\n",
      "            typically used for matrices and images.\n",
      "        \n",
      "            Note that the vertical axis points upward for 'lower'\n",
      "            but downward for 'upper'.\n",
      "        \n",
      "            See the :doc:`/tutorials/intermediate/imshow_extent` tutorial for\n",
      "            examples and a more detailed description.\n",
      "        \n",
      "        extent : floats (left, right, bottom, top), optional\n",
      "            The bounding box in data coordinates that the image will fill.\n",
      "            The image is stretched individually along x and y to fill the box.\n",
      "        \n",
      "            The default extent is determined by the following conditions.\n",
      "            Pixels have unit size in data coordinates. Their centers are on\n",
      "            integer coordinates, and their center coordinates range from 0 to\n",
      "            columns-1 horizontally and from 0 to rows-1 vertically.\n",
      "        \n",
      "            Note that the direction of the vertical axis and thus the default\n",
      "            values for top and bottom depend on *origin*:\n",
      "        \n",
      "            - For ``origin == 'upper'`` the default is\n",
      "              ``(-0.5, numcols-0.5, numrows-0.5, -0.5)``.\n",
      "            - For ``origin == 'lower'`` the default is\n",
      "              ``(-0.5, numcols-0.5, -0.5, numrows-0.5)``.\n",
      "        \n",
      "            See the :doc:`/tutorials/intermediate/imshow_extent` tutorial for\n",
      "            examples and a more detailed description.\n",
      "        \n",
      "        filternorm : bool, default: True\n",
      "            A parameter for the antigrain image resize filter (see the\n",
      "            antigrain documentation).  If *filternorm* is set, the filter\n",
      "            normalizes integer values and corrects the rounding errors. It\n",
      "            doesn't do anything with the source floating point values, it\n",
      "            corrects only integers according to the rule of 1.0 which means\n",
      "            that any sum of pixel weights must be equal to 1.0.  So, the\n",
      "            filter function must produce a graph of the proper shape.\n",
      "        \n",
      "        filterrad : float > 0, default: 4.0\n",
      "            The filter radius for filters that have a radius parameter, i.e.\n",
      "            when interpolation is one of: 'sinc', 'lanczos' or 'blackman'.\n",
      "        \n",
      "        resample : bool, default: :rc:`image.resample`\n",
      "            When *True*, use a full resampling method.  When *False*, only\n",
      "            resample when the output image is larger than the input image.\n",
      "        \n",
      "        url : str, optional\n",
      "            Set the url of the created `.AxesImage`. See `.Artist.set_url`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.image.AxesImage`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs : `~matplotlib.artist.Artist` properties\n",
      "            These parameters are passed on to the constructor of the\n",
      "            `.AxesImage` artist.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matshow : Plot a matrix or an array as an image.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Unless *extent* is used, pixel centers will be located at integer\n",
      "        coordinates. In other words: the origin will coincide with the center\n",
      "        of pixel (0, 0).\n",
      "        \n",
      "        There are two common representations for RGB images with an alpha\n",
      "        channel:\n",
      "        \n",
      "        -   Straight (unassociated) alpha: R, G, and B channels represent the\n",
      "            color of the pixel, disregarding its opacity.\n",
      "        -   Premultiplied (associated) alpha: R, G, and B channels represent\n",
      "            the color of the pixel, adjusted for its opacity by multiplication.\n",
      "        \n",
      "        `~matplotlib.pyplot.imshow` expects RGB images adopting the straight\n",
      "        (unassociated) alpha representation.\n",
      "    \n",
      "    inferno()\n",
      "        Set the colormap to 'inferno'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    install_repl_displayhook()\n",
      "        Connect to the display hook of the current shell.\n",
      "        \n",
      "        The display hook gets called when the read-evaluate-print-loop (REPL) of\n",
      "        the shell has finished the execution of a command. We use this callback\n",
      "        to be able to automatically update a figure in interactive mode.\n",
      "        \n",
      "        This works both with IPython and with vanilla python shells.\n",
      "    \n",
      "    ioff()\n",
      "        Disable interactive mode.\n",
      "        \n",
      "        See `.pyplot.isinteractive` for more details.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ion : Enable interactive mode.\n",
      "        isinteractive : Whether interactive mode is enabled.\n",
      "        show : Show all figures (and maybe block).\n",
      "        pause : Show all figures, and block for a time.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For a temporary change, this can be used as a context manager::\n",
      "        \n",
      "            # if interactive mode is on\n",
      "            # then figures will be shown on creation\n",
      "            plt.ion()\n",
      "            # This figure will be shown immediately\n",
      "            fig = plt.figure()\n",
      "        \n",
      "            with plt.ioff():\n",
      "                # interactive mode will be off\n",
      "                # figures will not automatically be shown\n",
      "                fig2 = plt.figure()\n",
      "                # ...\n",
      "        \n",
      "        To enable optional usage as a context manager, this function returns a\n",
      "        `~contextlib.ExitStack` object, which is not intended to be stored or\n",
      "        accessed by the user.\n",
      "    \n",
      "    ion()\n",
      "        Enable interactive mode.\n",
      "        \n",
      "        See `.pyplot.isinteractive` for more details.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ioff : Disable interactive mode.\n",
      "        isinteractive : Whether interactive mode is enabled.\n",
      "        show : Show all figures (and maybe block).\n",
      "        pause : Show all figures, and block for a time.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For a temporary change, this can be used as a context manager::\n",
      "        \n",
      "            # if interactive mode is off\n",
      "            # then figures will not be shown on creation\n",
      "            plt.ioff()\n",
      "            # This figure will not be shown immediately\n",
      "            fig = plt.figure()\n",
      "        \n",
      "            with plt.ion():\n",
      "                # interactive mode will be on\n",
      "                # figures will automatically be shown\n",
      "                fig2 = plt.figure()\n",
      "                # ...\n",
      "        \n",
      "        To enable optional usage as a context manager, this function returns a\n",
      "        `~contextlib.ExitStack` object, which is not intended to be stored or\n",
      "        accessed by the user.\n",
      "    \n",
      "    isinteractive()\n",
      "        Return whether plots are updated after every plotting command.\n",
      "        \n",
      "        The interactive mode is mainly useful if you build plots from the command\n",
      "        line and want to see the effect of each command while you are building the\n",
      "        figure.\n",
      "        \n",
      "        In interactive mode:\n",
      "        \n",
      "        - newly created figures will be shown immediately;\n",
      "        - figures will automatically redraw on change;\n",
      "        - `.pyplot.show` will not block by default.\n",
      "        \n",
      "        In non-interactive mode:\n",
      "        \n",
      "        - newly created figures and changes to figures will not be reflected until\n",
      "          explicitly asked to be;\n",
      "        - `.pyplot.show` will block by default.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ion : Enable interactive mode.\n",
      "        ioff : Disable interactive mode.\n",
      "        show : Show all figures (and maybe block).\n",
      "        pause : Show all figures, and block for a time.\n",
      "    \n",
      "    jet()\n",
      "        Set the colormap to 'jet'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    legend(*args, **kwargs)\n",
      "        Place a legend on the Axes.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            legend()\n",
      "            legend(handles, labels)\n",
      "            legend(handles=handles)\n",
      "            legend(labels)\n",
      "        \n",
      "        The call signatures correspond to the following different ways to use\n",
      "        this method:\n",
      "        \n",
      "        **1. Automatic detection of elements to be shown in the legend**\n",
      "        \n",
      "        The elements to be added to the legend are automatically determined,\n",
      "        when you do not pass in any extra arguments.\n",
      "        \n",
      "        In this case, the labels are taken from the artist. You can specify\n",
      "        them either at artist creation or by calling the\n",
      "        :meth:`~.Artist.set_label` method on the artist::\n",
      "        \n",
      "            ax.plot([1, 2, 3], label='Inline label')\n",
      "            ax.legend()\n",
      "        \n",
      "        or::\n",
      "        \n",
      "            line, = ax.plot([1, 2, 3])\n",
      "            line.set_label('Label via method')\n",
      "            ax.legend()\n",
      "        \n",
      "        .. note::\n",
      "            Specific artists can be excluded from the automatic legend element\n",
      "            selection by using a label starting with an underscore, \"_\".\n",
      "            A string starting with an underscore is the default label for all\n",
      "            artists, so calling `.Axes.legend` without any arguments and\n",
      "            without setting the labels manually will result in no legend being\n",
      "            drawn.\n",
      "        \n",
      "        \n",
      "        **2. Explicitly listing the artists and labels in the legend**\n",
      "        \n",
      "        For full control of which artists have a legend entry, it is possible\n",
      "        to pass an iterable of legend artists followed by an iterable of\n",
      "        legend labels respectively::\n",
      "        \n",
      "            ax.legend([line1, line2, line3], ['label1', 'label2', 'label3'])\n",
      "        \n",
      "        \n",
      "        **3. Explicitly listing the artists in the legend**\n",
      "        \n",
      "        This is similar to 2, but the labels are taken from the artists'\n",
      "        label properties. Example::\n",
      "        \n",
      "            line1, = ax.plot([1, 2, 3], label='label1')\n",
      "            line2, = ax.plot([1, 2, 3], label='label2')\n",
      "            ax.legend(handles=[line1, line2])\n",
      "        \n",
      "        \n",
      "        **4. Labeling existing plot elements**\n",
      "        \n",
      "        .. admonition:: Discouraged\n",
      "        \n",
      "            This call signature is discouraged, because the relation between\n",
      "            plot elements and labels is only implicit by their order and can\n",
      "            easily be mixed up.\n",
      "        \n",
      "        To make a legend for all artists on an Axes, call this function with\n",
      "        an iterable of strings, one for each legend item. For example::\n",
      "        \n",
      "            ax.plot([1, 2, 3])\n",
      "            ax.plot([5, 6, 7])\n",
      "            ax.legend(['First line', 'Second line'])\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        handles : sequence of `.Artist`, optional\n",
      "            A list of Artists (lines, patches) to be added to the legend.\n",
      "            Use this together with *labels*, if you need full control on what\n",
      "            is shown in the legend and the automatic mechanism described above\n",
      "            is not sufficient.\n",
      "        \n",
      "            The length of handles and labels should be the same in this\n",
      "            case. If they are not, they are truncated to the smaller length.\n",
      "        \n",
      "        labels : list of str, optional\n",
      "            A list of labels to show next to the artists.\n",
      "            Use this together with *handles*, if you need full control on what\n",
      "            is shown in the legend and the automatic mechanism described above\n",
      "            is not sufficient.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.legend.Legend`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        \n",
      "        loc : str or pair of floats, default: :rc:`legend.loc` ('best' for axes, 'upper right' for figures)\n",
      "            The location of the legend.\n",
      "        \n",
      "            The strings\n",
      "            ``'upper left', 'upper right', 'lower left', 'lower right'``\n",
      "            place the legend at the corresponding corner of the axes/figure.\n",
      "        \n",
      "            The strings\n",
      "            ``'upper center', 'lower center', 'center left', 'center right'``\n",
      "            place the legend at the center of the corresponding edge of the\n",
      "            axes/figure.\n",
      "        \n",
      "            The string ``'center'`` places the legend at the center of the axes/figure.\n",
      "        \n",
      "            The string ``'best'`` places the legend at the location, among the nine\n",
      "            locations defined so far, with the minimum overlap with other drawn\n",
      "            artists.  This option can be quite slow for plots with large amounts of\n",
      "            data; your plotting speed may benefit from providing a specific location.\n",
      "        \n",
      "            The location can also be a 2-tuple giving the coordinates of the lower-left\n",
      "            corner of the legend in axes coordinates (in which case *bbox_to_anchor*\n",
      "            will be ignored).\n",
      "        \n",
      "            For back-compatibility, ``'center right'`` (but no other location) can also\n",
      "            be spelled ``'right'``, and each \"string\" locations can also be given as a\n",
      "            numeric value:\n",
      "        \n",
      "                ===============   =============\n",
      "                Location String   Location Code\n",
      "                ===============   =============\n",
      "                'best'            0\n",
      "                'upper right'     1\n",
      "                'upper left'      2\n",
      "                'lower left'      3\n",
      "                'lower right'     4\n",
      "                'right'           5\n",
      "                'center left'     6\n",
      "                'center right'    7\n",
      "                'lower center'    8\n",
      "                'upper center'    9\n",
      "                'center'          10\n",
      "                ===============   =============\n",
      "        \n",
      "        bbox_to_anchor : `.BboxBase`, 2-tuple, or 4-tuple of floats\n",
      "            Box that is used to position the legend in conjunction with *loc*.\n",
      "            Defaults to `axes.bbox` (if called as a method to `.Axes.legend`) or\n",
      "            `figure.bbox` (if `.Figure.legend`).  This argument allows arbitrary\n",
      "            placement of the legend.\n",
      "        \n",
      "            Bbox coordinates are interpreted in the coordinate system given by\n",
      "            *bbox_transform*, with the default transform\n",
      "            Axes or Figure coordinates, depending on which ``legend`` is called.\n",
      "        \n",
      "            If a 4-tuple or `.BboxBase` is given, then it specifies the bbox\n",
      "            ``(x, y, width, height)`` that the legend is placed in.\n",
      "            To put the legend in the best location in the bottom right\n",
      "            quadrant of the axes (or figure)::\n",
      "        \n",
      "                loc='best', bbox_to_anchor=(0.5, 0., 0.5, 0.5)\n",
      "        \n",
      "            A 2-tuple ``(x, y)`` places the corner of the legend specified by *loc* at\n",
      "            x, y.  For example, to put the legend's upper right-hand corner in the\n",
      "            center of the axes (or figure) the following keywords can be used::\n",
      "        \n",
      "                loc='upper right', bbox_to_anchor=(0.5, 0.5)\n",
      "        \n",
      "        ncols : int, default: 1\n",
      "            The number of columns that the legend has.\n",
      "        \n",
      "            For backward compatibility, the spelling *ncol* is also supported\n",
      "            but it is discouraged. If both are given, *ncols* takes precedence.\n",
      "        \n",
      "        prop : None or `matplotlib.font_manager.FontProperties` or dict\n",
      "            The font properties of the legend. If None (default), the current\n",
      "            :data:`matplotlib.rcParams` will be used.\n",
      "        \n",
      "        fontsize : int or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}\n",
      "            The font size of the legend. If the value is numeric the size will be the\n",
      "            absolute font size in points. String values are relative to the current\n",
      "            default font size. This argument is only used if *prop* is not specified.\n",
      "        \n",
      "        labelcolor : str or list, default: :rc:`legend.labelcolor`\n",
      "            The color of the text in the legend. Either a valid color string\n",
      "            (for example, 'red'), or a list of color strings. The labelcolor can\n",
      "            also be made to match the color of the line or marker using 'linecolor',\n",
      "            'markerfacecolor' (or 'mfc'), or 'markeredgecolor' (or 'mec').\n",
      "        \n",
      "            Labelcolor can be set globally using :rc:`legend.labelcolor`. If None,\n",
      "            use :rc:`text.color`.\n",
      "        \n",
      "        numpoints : int, default: :rc:`legend.numpoints`\n",
      "            The number of marker points in the legend when creating a legend\n",
      "            entry for a `.Line2D` (line).\n",
      "        \n",
      "        scatterpoints : int, default: :rc:`legend.scatterpoints`\n",
      "            The number of marker points in the legend when creating\n",
      "            a legend entry for a `.PathCollection` (scatter plot).\n",
      "        \n",
      "        scatteryoffsets : iterable of floats, default: ``[0.375, 0.5, 0.3125]``\n",
      "            The vertical offset (relative to the font size) for the markers\n",
      "            created for a scatter plot legend entry. 0.0 is at the base the\n",
      "            legend text, and 1.0 is at the top. To draw all markers at the\n",
      "            same height, set to ``[0.5]``.\n",
      "        \n",
      "        markerscale : float, default: :rc:`legend.markerscale`\n",
      "            The relative size of legend markers compared with the originally\n",
      "            drawn ones.\n",
      "        \n",
      "        markerfirst : bool, default: True\n",
      "            If *True*, legend marker is placed to the left of the legend label.\n",
      "            If *False*, legend marker is placed to the right of the legend label.\n",
      "        \n",
      "        frameon : bool, default: :rc:`legend.frameon`\n",
      "            Whether the legend should be drawn on a patch (frame).\n",
      "        \n",
      "        fancybox : bool, default: :rc:`legend.fancybox`\n",
      "            Whether round edges should be enabled around the `.FancyBboxPatch` which\n",
      "            makes up the legend's background.\n",
      "        \n",
      "        shadow : bool, default: :rc:`legend.shadow`\n",
      "            Whether to draw a shadow behind the legend.\n",
      "        \n",
      "        framealpha : float, default: :rc:`legend.framealpha`\n",
      "            The alpha transparency of the legend's background.\n",
      "            If *shadow* is activated and *framealpha* is ``None``, the default value is\n",
      "            ignored.\n",
      "        \n",
      "        facecolor : \"inherit\" or color, default: :rc:`legend.facecolor`\n",
      "            The legend's background color.\n",
      "            If ``\"inherit\"``, use :rc:`axes.facecolor`.\n",
      "        \n",
      "        edgecolor : \"inherit\" or color, default: :rc:`legend.edgecolor`\n",
      "            The legend's background patch edge color.\n",
      "            If ``\"inherit\"``, use take :rc:`axes.edgecolor`.\n",
      "        \n",
      "        mode : {\"expand\", None}\n",
      "            If *mode* is set to ``\"expand\"`` the legend will be horizontally\n",
      "            expanded to fill the axes area (or *bbox_to_anchor* if defines\n",
      "            the legend's size).\n",
      "        \n",
      "        bbox_transform : None or `matplotlib.transforms.Transform`\n",
      "            The transform for the bounding box (*bbox_to_anchor*). For a value\n",
      "            of ``None`` (default) the Axes'\n",
      "            :data:`~matplotlib.axes.Axes.transAxes` transform will be used.\n",
      "        \n",
      "        title : str or None\n",
      "            The legend's title. Default is no title (``None``).\n",
      "        \n",
      "        title_fontproperties : None or `matplotlib.font_manager.FontProperties` or dict\n",
      "            The font properties of the legend's title. If None (default), the\n",
      "            *title_fontsize* argument will be used if present; if *title_fontsize* is\n",
      "            also None, the current :rc:`legend.title_fontsize` will be used.\n",
      "        \n",
      "        title_fontsize : int or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}, default: :rc:`legend.title_fontsize`\n",
      "            The font size of the legend's title.\n",
      "            Note: This cannot be combined with *title_fontproperties*. If you want\n",
      "            to set the fontsize alongside other font properties, use the *size*\n",
      "            parameter in *title_fontproperties*.\n",
      "        \n",
      "        alignment : {'center', 'left', 'right'}, default: 'center'\n",
      "            The alignment of the legend title and the box of entries. The entries\n",
      "            are aligned as a single block, so that markers always lined up.\n",
      "        \n",
      "        borderpad : float, default: :rc:`legend.borderpad`\n",
      "            The fractional whitespace inside the legend border, in font-size units.\n",
      "        \n",
      "        labelspacing : float, default: :rc:`legend.labelspacing`\n",
      "            The vertical space between the legend entries, in font-size units.\n",
      "        \n",
      "        handlelength : float, default: :rc:`legend.handlelength`\n",
      "            The length of the legend handles, in font-size units.\n",
      "        \n",
      "        handleheight : float, default: :rc:`legend.handleheight`\n",
      "            The height of the legend handles, in font-size units.\n",
      "        \n",
      "        handletextpad : float, default: :rc:`legend.handletextpad`\n",
      "            The pad between the legend handle and text, in font-size units.\n",
      "        \n",
      "        borderaxespad : float, default: :rc:`legend.borderaxespad`\n",
      "            The pad between the axes and legend border, in font-size units.\n",
      "        \n",
      "        columnspacing : float, default: :rc:`legend.columnspacing`\n",
      "            The spacing between columns, in font-size units.\n",
      "        \n",
      "        handler_map : dict or None\n",
      "            The custom dictionary mapping instances or types to a legend\n",
      "            handler. This *handler_map* updates the default handler map\n",
      "            found at `matplotlib.legend.Legend.get_legend_handler_map`.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Figure.legend\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Some artists are not supported by this function.  See\n",
      "        :doc:`/tutorials/intermediate/legend_guide` for details.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        .. plot:: gallery/text_labels_and_annotations/legend.py\n",
      "    \n",
      "    locator_params(axis='both', tight=None, **kwargs)\n",
      "        Control behavior of major tick locators.\n",
      "        \n",
      "        Because the locator is involved in autoscaling, `~.Axes.autoscale_view`\n",
      "        is called automatically after the parameters are changed.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        axis : {'both', 'x', 'y'}, default: 'both'\n",
      "            The axis on which to operate.  (For 3D Axes, *axis* can also be\n",
      "            set to 'z', and 'both' refers to all three axes.)\n",
      "        tight : bool or None, optional\n",
      "            Parameter passed to `~.Axes.autoscale_view`.\n",
      "            Default is None, for no change.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            Remaining keyword arguments are passed to directly to the\n",
      "            ``set_params()`` method of the locator. Supported keywords depend\n",
      "            on the type of the locator. See for example\n",
      "            `~.ticker.MaxNLocator.set_params` for the `.ticker.MaxNLocator`\n",
      "            used by default for linear.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        When plotting small subplots, one might want to reduce the maximum\n",
      "        number of ticks and use tight bounds, for example::\n",
      "        \n",
      "            ax.locator_params(tight=True, nbins=4)\n",
      "    \n",
      "    loglog(*args, **kwargs)\n",
      "        Make a plot with log scaling on both the x and y axis.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            loglog([x], y, [fmt], data=None, **kwargs)\n",
      "            loglog([x], y, [fmt], [x2], y2, [fmt2], ..., **kwargs)\n",
      "        \n",
      "        This is just a thin wrapper around `.plot` which additionally changes\n",
      "        both the x-axis and the y-axis to log scaling. All of the concepts and\n",
      "        parameters of plot can be used here as well.\n",
      "        \n",
      "        The additional parameters *base*, *subs* and *nonpositive* control the\n",
      "        x/y-axis properties. They are just forwarded to `.Axes.set_xscale` and\n",
      "        `.Axes.set_yscale`. To use different properties on the x-axis and the\n",
      "        y-axis, use e.g.\n",
      "        ``ax.set_xscale(\"log\", base=10); ax.set_yscale(\"log\", base=2)``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        base : float, default: 10\n",
      "            Base of the logarithm.\n",
      "        \n",
      "        subs : sequence, optional\n",
      "            The location of the minor ticks. If *None*, reasonable locations\n",
      "            are automatically chosen depending on the number of decades in the\n",
      "            plot. See `.Axes.set_xscale`/`.Axes.set_yscale` for details.\n",
      "        \n",
      "        nonpositive : {'mask', 'clip'}, default: 'mask'\n",
      "            Non-positive values can be masked as invalid, or clipped to a very\n",
      "            small positive number.\n",
      "        \n",
      "        **kwargs\n",
      "            All parameters supported by `.plot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            Objects representing the plotted data.\n",
      "    \n",
      "    magma()\n",
      "        Set the colormap to 'magma'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    magnitude_spectrum(x, Fs=None, Fc=None, window=None, pad_to=None, sides=None, scale=None, *, data=None, **kwargs)\n",
      "        Plot the magnitude spectrum.\n",
      "        \n",
      "        Compute the magnitude spectrum of *x*.  Data is padded to a\n",
      "        length of *pad_to* and the windowing function *window* is applied to\n",
      "        the signal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1-D array or sequence\n",
      "            Array or sequence containing the data.\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  While not increasing the actual resolution of the spectrum (the\n",
      "            minimum distance between resolvable peaks), this can give more points in\n",
      "            the plot, allowing for more detail. This corresponds to the *n* parameter\n",
      "            in the call to `~numpy.fft.fft`.  The default is None, which sets *pad_to*\n",
      "            equal to the length of the input signal (i.e. no padding).\n",
      "        \n",
      "        scale : {'default', 'linear', 'dB'}\n",
      "            The scaling of the values in the *spec*.  'linear' is no scaling.\n",
      "            'dB' returns the values in dB scale, i.e., the dB amplitude\n",
      "            (20 * log10). 'default' is 'linear'.\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        spectrum : 1-D array\n",
      "            The values for the magnitude spectrum before scaling (real valued).\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the elements in *spectrum*.\n",
      "        \n",
      "        line : `~matplotlib.lines.Line2D`\n",
      "            The line created by this function.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        psd\n",
      "            Plots the power spectral density.\n",
      "        angle_spectrum\n",
      "            Plots the angles of the corresponding frequencies.\n",
      "        phase_spectrum\n",
      "            Plots the phase (unwrapped angle) of the corresponding frequencies.\n",
      "        specgram\n",
      "            Can plot the magnitude spectrum of segments within the signal in a\n",
      "            colormap.\n",
      "    \n",
      "    margins(*margins, x=None, y=None, tight=True)\n",
      "        Set or retrieve autoscaling margins.\n",
      "        \n",
      "        The padding added to each limit of the Axes is the *margin*\n",
      "        times the data interval. All input parameters must be floats\n",
      "        within the range [0, 1]. Passing both positional and keyword\n",
      "        arguments is invalid and will raise a TypeError. If no\n",
      "        arguments (positional or otherwise) are provided, the current\n",
      "        margins will remain in place and simply be returned.\n",
      "        \n",
      "        Specifying any margin changes only the autoscaling; for example,\n",
      "        if *xmargin* is not None, then *xmargin* times the X data\n",
      "        interval will be added to each end of that interval before\n",
      "        it is used in autoscaling.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        *margins : float, optional\n",
      "            If a single positional argument is provided, it specifies\n",
      "            both margins of the x-axis and y-axis limits. If two\n",
      "            positional arguments are provided, they will be interpreted\n",
      "            as *xmargin*, *ymargin*. If setting the margin on a single\n",
      "            axis is desired, use the keyword arguments described below.\n",
      "        \n",
      "        x, y : float, optional\n",
      "            Specific margin values for the x-axis and y-axis,\n",
      "            respectively. These cannot be used with positional\n",
      "            arguments, but can be used individually to alter on e.g.,\n",
      "            only the y-axis.\n",
      "        \n",
      "        tight : bool or None, default: True\n",
      "            The *tight* parameter is passed to `~.axes.Axes.autoscale_view`,\n",
      "            which is executed after a margin is changed; the default\n",
      "            here is *True*, on the assumption that when margins are\n",
      "            specified, no additional padding to match tick marks is\n",
      "            usually desired.  Setting *tight* to *None* preserves\n",
      "            the previous setting.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xmargin, ymargin : float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        If a previously used Axes method such as :meth:`pcolor` has set\n",
      "        :attr:`use_sticky_edges` to `True`, only the limits not set by\n",
      "        the \"sticky artists\" will be modified. To force all of the\n",
      "        margins to be set, set :attr:`use_sticky_edges` to `False`\n",
      "        before calling :meth:`margins`.\n",
      "    \n",
      "    matshow(A, fignum=None, **kwargs)\n",
      "        Display an array as a matrix in a new figure window.\n",
      "        \n",
      "        The origin is set at the upper left hand corner and rows (first\n",
      "        dimension of the array) are displayed horizontally.  The aspect\n",
      "        ratio of the figure window is that of the array, unless this would\n",
      "        make an excessively short or narrow figure.\n",
      "        \n",
      "        Tick labels for the xaxis are placed on top.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        A : 2D array-like\n",
      "            The matrix to be displayed.\n",
      "        \n",
      "        fignum : None or int or False\n",
      "            If *None*, create a new figure window with automatic numbering.\n",
      "        \n",
      "            If a nonzero integer, draw into the figure with the given number\n",
      "            (create it if it does not exist).\n",
      "        \n",
      "            If 0, use the current axes (or create one if it does not exist).\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "               Because of how `.Axes.matshow` tries to set the figure aspect\n",
      "               ratio to be the one of the array, strange things may happen if you\n",
      "               reuse an existing figure.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.image.AxesImage`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.axes.Axes.imshow` arguments\n",
      "    \n",
      "    minorticks_off()\n",
      "        Remove minor ticks from the Axes.\n",
      "    \n",
      "    minorticks_on()\n",
      "        Display minor ticks on the Axes.\n",
      "        \n",
      "        Displaying minor ticks may reduce performance; you may turn them off\n",
      "        using `minorticks_off()` if drawing speed is a problem.\n",
      "    \n",
      "    new_figure_manager(*args, **kwargs)\n",
      "        Create a new figure manager instance.\n",
      "    \n",
      "    nipy_spectral()\n",
      "        Set the colormap to 'nipy_spectral'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    pause(interval)\n",
      "        Run the GUI event loop for *interval* seconds.\n",
      "        \n",
      "        If there is an active figure, it will be updated and displayed before the\n",
      "        pause, and the GUI event loop (if any) will run during the pause.\n",
      "        \n",
      "        This can be used for crude animation.  For more complex animation use\n",
      "        :mod:`matplotlib.animation`.\n",
      "        \n",
      "        If there is no active figure, sleep for *interval* seconds instead.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.animation : Proper animations\n",
      "        show : Show all figures and optional block until all figures are closed.\n",
      "    \n",
      "    pcolor(*args, shading=None, alpha=None, norm=None, cmap=None, vmin=None, vmax=None, data=None, **kwargs)\n",
      "        Create a pseudocolor plot with a non-regular rectangular grid.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "            pcolor([X, Y,] C, **kwargs)\n",
      "        \n",
      "        *X* and *Y* can be used to specify the corners of the quadrilaterals.\n",
      "        \n",
      "        .. hint::\n",
      "        \n",
      "            ``pcolor()`` can be very slow for large arrays. In most\n",
      "            cases you should use the similar but much faster\n",
      "            `~.Axes.pcolormesh` instead. See\n",
      "            :ref:`Differences between pcolor() and pcolormesh()\n",
      "            <differences-pcolor-pcolormesh>` for a discussion of the\n",
      "            differences.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        C : 2D array-like\n",
      "            The color-mapped values.  Color-mapping is controlled by *cmap*,\n",
      "            *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        X, Y : array-like, optional\n",
      "            The coordinates of the corners of quadrilaterals of a pcolormesh::\n",
      "        \n",
      "                (X[i+1, j], Y[i+1, j])       (X[i+1, j+1], Y[i+1, j+1])\n",
      "                                      +-----+\n",
      "                                      |     |\n",
      "                                      +-----+\n",
      "                    (X[i, j], Y[i, j])       (X[i, j+1], Y[i, j+1])\n",
      "        \n",
      "            Note that the column index corresponds to the x-coordinate, and\n",
      "            the row index corresponds to y. For details, see the\n",
      "            :ref:`Notes <axes-pcolormesh-grid-orientation>` section below.\n",
      "        \n",
      "            If ``shading='flat'`` the dimensions of *X* and *Y* should be one\n",
      "            greater than those of *C*, and the quadrilateral is colored due\n",
      "            to the value at ``C[i, j]``.  If *X*, *Y* and *C* have equal\n",
      "            dimensions, a warning will be raised and the last row and column\n",
      "            of *C* will be ignored.\n",
      "        \n",
      "            If ``shading='nearest'``, the dimensions of *X* and *Y* should be\n",
      "            the same as those of *C* (if not, a ValueError will be raised). The\n",
      "            color ``C[i, j]`` will be centered on ``(X[i, j], Y[i, j])``.\n",
      "        \n",
      "            If *X* and/or *Y* are 1-D arrays or column vectors they will be\n",
      "            expanded as needed into the appropriate 2D arrays, making a\n",
      "            rectangular grid.\n",
      "        \n",
      "        shading : {'flat', 'nearest', 'auto'}, default: :rc:`pcolor.shading`\n",
      "            The fill style for the quadrilateral. Possible values:\n",
      "        \n",
      "            - 'flat': A solid color is used for each quad. The color of the\n",
      "              quad (i, j), (i+1, j), (i, j+1), (i+1, j+1) is given by\n",
      "              ``C[i, j]``. The dimensions of *X* and *Y* should be\n",
      "              one greater than those of *C*; if they are the same as *C*,\n",
      "              then a deprecation warning is raised, and the last row\n",
      "              and column of *C* are dropped.\n",
      "            - 'nearest': Each grid point will have a color centered on it,\n",
      "              extending halfway between the adjacent grid centers.  The\n",
      "              dimensions of *X* and *Y* must be the same as *C*.\n",
      "            - 'auto': Choose 'flat' if dimensions of *X* and *Y* are one\n",
      "              larger than *C*.  Choose 'nearest' if dimensions are the same.\n",
      "        \n",
      "            See :doc:`/gallery/images_contours_and_fields/pcolormesh_grids`\n",
      "            for more description.\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "        edgecolors : {'none', None, 'face', color, color sequence}, optional\n",
      "            The color of the edges. Defaults to 'none'. Possible values:\n",
      "        \n",
      "            - 'none' or '': No edge.\n",
      "            - *None*: :rc:`patch.edgecolor` will be used. Note that currently\n",
      "              :rc:`patch.force_edgecolor` has to be True for this to work.\n",
      "            - 'face': Use the adjacent face color.\n",
      "            - A color or sequence of colors will set the edge color.\n",
      "        \n",
      "            The singular form *edgecolor* works as an alias.\n",
      "        \n",
      "        alpha : float, default: None\n",
      "            The alpha blending value of the face color, between 0 (transparent)\n",
      "            and 1 (opaque). Note: The edgecolor is currently not affected by\n",
      "            this.\n",
      "        \n",
      "        snap : bool, default: False\n",
      "            Whether to snap the mesh to pixel boundaries.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `matplotlib.collections.Collection`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        antialiaseds : bool, default: False\n",
      "            The default *antialiaseds* is False if the default\n",
      "            *edgecolors*\\ =\"none\" is used.  This eliminates artificial lines\n",
      "            at patch boundaries, and works regardless of the value of alpha.\n",
      "            If *edgecolors* is not \"none\", then the default *antialiaseds*\n",
      "            is taken from :rc:`patch.antialiased`.\n",
      "            Stroking the edges may be preferred if *alpha* is 1, but will\n",
      "            cause artifacts otherwise.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs\n",
      "            Additionally, the following arguments are allowed. They are passed\n",
      "            along to the `~matplotlib.collections.PolyCollection` constructor:\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        pcolormesh : for an explanation of the differences between\n",
      "            pcolor and pcolormesh.\n",
      "        imshow : If *X* and *Y* are each equidistant, `~.Axes.imshow` can be a\n",
      "            faster alternative.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        **Masked arrays**\n",
      "        \n",
      "        *X*, *Y* and *C* may be masked arrays. If either ``C[i, j]``, or one\n",
      "        of the vertices surrounding ``C[i, j]`` (*X* or *Y* at\n",
      "        ``[i, j], [i+1, j], [i, j+1], [i+1, j+1]``) is masked, nothing is\n",
      "        plotted.\n",
      "        \n",
      "        .. _axes-pcolor-grid-orientation:\n",
      "        \n",
      "        **Grid orientation**\n",
      "        \n",
      "        The grid orientation follows the standard matrix convention: An array\n",
      "        *C* with shape (nrows, ncolumns) is plotted with the column number as\n",
      "        *X* and the row number as *Y*.\n",
      "    \n",
      "    pcolormesh(*args, alpha=None, norm=None, cmap=None, vmin=None, vmax=None, shading=None, antialiased=False, data=None, **kwargs)\n",
      "        Create a pseudocolor plot with a non-regular rectangular grid.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "            pcolormesh([X, Y,] C, **kwargs)\n",
      "        \n",
      "        *X* and *Y* can be used to specify the corners of the quadrilaterals.\n",
      "        \n",
      "        .. hint::\n",
      "        \n",
      "           `~.Axes.pcolormesh` is similar to `~.Axes.pcolor`. It is much faster\n",
      "           and preferred in most cases. For a detailed discussion on the\n",
      "           differences see :ref:`Differences between pcolor() and pcolormesh()\n",
      "           <differences-pcolor-pcolormesh>`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        C : 2D array-like\n",
      "            The color-mapped values.  Color-mapping is controlled by *cmap*,\n",
      "            *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        X, Y : array-like, optional\n",
      "            The coordinates of the corners of quadrilaterals of a pcolormesh::\n",
      "        \n",
      "                (X[i+1, j], Y[i+1, j])       (X[i+1, j+1], Y[i+1, j+1])\n",
      "                                      +-----+\n",
      "                                      |     |\n",
      "                                      +-----+\n",
      "                    (X[i, j], Y[i, j])       (X[i, j+1], Y[i, j+1])\n",
      "        \n",
      "            Note that the column index corresponds to the x-coordinate, and\n",
      "            the row index corresponds to y. For details, see the\n",
      "            :ref:`Notes <axes-pcolormesh-grid-orientation>` section below.\n",
      "        \n",
      "            If ``shading='flat'`` the dimensions of *X* and *Y* should be one\n",
      "            greater than those of *C*, and the quadrilateral is colored due\n",
      "            to the value at ``C[i, j]``.  If *X*, *Y* and *C* have equal\n",
      "            dimensions, a warning will be raised and the last row and column\n",
      "            of *C* will be ignored.\n",
      "        \n",
      "            If ``shading='nearest'`` or ``'gouraud'``, the dimensions of *X*\n",
      "            and *Y* should be the same as those of *C* (if not, a ValueError\n",
      "            will be raised).  For ``'nearest'`` the color ``C[i, j]`` is\n",
      "            centered on ``(X[i, j], Y[i, j])``.  For ``'gouraud'``, a smooth\n",
      "            interpolation is caried out between the quadrilateral corners.\n",
      "        \n",
      "            If *X* and/or *Y* are 1-D arrays or column vectors they will be\n",
      "            expanded as needed into the appropriate 2D arrays, making a\n",
      "            rectangular grid.\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "        edgecolors : {'none', None, 'face', color, color sequence}, optional\n",
      "            The color of the edges. Defaults to 'none'. Possible values:\n",
      "        \n",
      "            - 'none' or '': No edge.\n",
      "            - *None*: :rc:`patch.edgecolor` will be used. Note that currently\n",
      "              :rc:`patch.force_edgecolor` has to be True for this to work.\n",
      "            - 'face': Use the adjacent face color.\n",
      "            - A color or sequence of colors will set the edge color.\n",
      "        \n",
      "            The singular form *edgecolor* works as an alias.\n",
      "        \n",
      "        alpha : float, default: None\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        shading : {'flat', 'nearest', 'gouraud', 'auto'}, optional\n",
      "            The fill style for the quadrilateral; defaults to\n",
      "            'flat' or :rc:`pcolor.shading`. Possible values:\n",
      "        \n",
      "            - 'flat': A solid color is used for each quad. The color of the\n",
      "              quad (i, j), (i+1, j), (i, j+1), (i+1, j+1) is given by\n",
      "              ``C[i, j]``. The dimensions of *X* and *Y* should be\n",
      "              one greater than those of *C*; if they are the same as *C*,\n",
      "              then a deprecation warning is raised, and the last row\n",
      "              and column of *C* are dropped.\n",
      "            - 'nearest': Each grid point will have a color centered on it,\n",
      "              extending halfway between the adjacent grid centers.  The\n",
      "              dimensions of *X* and *Y* must be the same as *C*.\n",
      "            - 'gouraud': Each quad will be Gouraud shaded: The color of the\n",
      "              corners (i', j') are given by ``C[i', j']``. The color values of\n",
      "              the area in between is interpolated from the corner values.\n",
      "              The dimensions of *X* and *Y* must be the same as *C*. When\n",
      "              Gouraud shading is used, *edgecolors* is ignored.\n",
      "            - 'auto': Choose 'flat' if dimensions of *X* and *Y* are one\n",
      "              larger than *C*.  Choose 'nearest' if dimensions are the same.\n",
      "        \n",
      "            See :doc:`/gallery/images_contours_and_fields/pcolormesh_grids`\n",
      "            for more description.\n",
      "        \n",
      "        snap : bool, default: False\n",
      "            Whether to snap the mesh to pixel boundaries.\n",
      "        \n",
      "        rasterized : bool, optional\n",
      "            Rasterize the pcolormesh when drawing vector graphics.  This can\n",
      "            speed up rendering and produce smaller files for large data sets.\n",
      "            See also :doc:`/gallery/misc/rasterization_demo`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `matplotlib.collections.QuadMesh`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs\n",
      "            Additionally, the following arguments are allowed. They are passed\n",
      "            along to the `~matplotlib.collections.QuadMesh` constructor:\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: (M, N) array-like or M*N array-like\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        pcolor : An alternative implementation with slightly different\n",
      "            features. For a detailed discussion on the differences see\n",
      "            :ref:`Differences between pcolor() and pcolormesh()\n",
      "            <differences-pcolor-pcolormesh>`.\n",
      "        imshow : If *X* and *Y* are each equidistant, `~.Axes.imshow` can be a\n",
      "            faster alternative.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        **Masked arrays**\n",
      "        \n",
      "        *C* may be a masked array. If ``C[i, j]`` is masked, the corresponding\n",
      "        quadrilateral will be transparent. Masking of *X* and *Y* is not\n",
      "        supported. Use `~.Axes.pcolor` if you need this functionality.\n",
      "        \n",
      "        .. _axes-pcolormesh-grid-orientation:\n",
      "        \n",
      "        **Grid orientation**\n",
      "        \n",
      "        The grid orientation follows the standard matrix convention: An array\n",
      "        *C* with shape (nrows, ncolumns) is plotted with the column number as\n",
      "        *X* and the row number as *Y*.\n",
      "        \n",
      "        .. _differences-pcolor-pcolormesh:\n",
      "        \n",
      "        **Differences between pcolor() and pcolormesh()**\n",
      "        \n",
      "        Both methods are used to create a pseudocolor plot of a 2D array\n",
      "        using quadrilaterals.\n",
      "        \n",
      "        The main difference lies in the created object and internal data\n",
      "        handling:\n",
      "        While `~.Axes.pcolor` returns a `.PolyCollection`, `~.Axes.pcolormesh`\n",
      "        returns a `.QuadMesh`. The latter is more specialized for the given\n",
      "        purpose and thus is faster. It should almost always be preferred.\n",
      "        \n",
      "        There is also a slight difference in the handling of masked arrays.\n",
      "        Both `~.Axes.pcolor` and `~.Axes.pcolormesh` support masked arrays\n",
      "        for *C*. However, only `~.Axes.pcolor` supports masked arrays for *X*\n",
      "        and *Y*. The reason lies in the internal handling of the masked values.\n",
      "        `~.Axes.pcolor` leaves out the respective polygons from the\n",
      "        PolyCollection. `~.Axes.pcolormesh` sets the facecolor of the masked\n",
      "        elements to transparent. You can see the difference when using\n",
      "        edgecolors. While all edges are drawn irrespective of masking in a\n",
      "        QuadMesh, the edge between two adjacent masked quadrilaterals in\n",
      "        `~.Axes.pcolor` is not drawn as the corresponding polygons do not\n",
      "        exist in the PolyCollection.\n",
      "        \n",
      "        Another difference is the support of Gouraud shading in\n",
      "        `~.Axes.pcolormesh`, which is not available with `~.Axes.pcolor`.\n",
      "    \n",
      "    phase_spectrum(x, Fs=None, Fc=None, window=None, pad_to=None, sides=None, *, data=None, **kwargs)\n",
      "        Plot the phase spectrum.\n",
      "        \n",
      "        Compute the phase spectrum (unwrapped angle spectrum) of *x*.\n",
      "        Data is padded to a length of *pad_to* and the windowing function\n",
      "        *window* is applied to the signal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1-D array or sequence\n",
      "            Array or sequence containing the data\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  While not increasing the actual resolution of the spectrum (the\n",
      "            minimum distance between resolvable peaks), this can give more points in\n",
      "            the plot, allowing for more detail. This corresponds to the *n* parameter\n",
      "            in the call to `~numpy.fft.fft`.  The default is None, which sets *pad_to*\n",
      "            equal to the length of the input signal (i.e. no padding).\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        spectrum : 1-D array\n",
      "            The values for the phase spectrum in radians (real valued).\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the elements in *spectrum*.\n",
      "        \n",
      "        line : `~matplotlib.lines.Line2D`\n",
      "            The line created by this function.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        magnitude_spectrum\n",
      "            Plots the magnitudes of the corresponding frequencies.\n",
      "        angle_spectrum\n",
      "            Plots the wrapped version of this function.\n",
      "        specgram\n",
      "            Can plot the phase spectrum of segments within the signal in a\n",
      "            colormap.\n",
      "    \n",
      "    pie(x, explode=None, labels=None, colors=None, autopct=None, pctdistance=0.6, shadow=False, labeldistance=1.1, startangle=0, radius=1, counterclock=True, wedgeprops=None, textprops=None, center=(0, 0), frame=False, rotatelabels=False, *, normalize=True, data=None)\n",
      "        Plot a pie chart.\n",
      "        \n",
      "        Make a pie chart of array *x*.  The fractional area of each wedge is\n",
      "        given by ``x/sum(x)``.\n",
      "        \n",
      "        The wedges are plotted counterclockwise, by default starting from the\n",
      "        x-axis.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1D array-like\n",
      "            The wedge sizes.\n",
      "        \n",
      "        explode : array-like, default: None\n",
      "            If not *None*, is a ``len(x)`` array which specifies the fraction\n",
      "            of the radius with which to offset each wedge.\n",
      "        \n",
      "        labels : list, default: None\n",
      "            A sequence of strings providing the labels for each wedge\n",
      "        \n",
      "        colors : array-like, default: None\n",
      "            A sequence of colors through which the pie chart will cycle.  If\n",
      "            *None*, will use the colors in the currently active cycle.\n",
      "        \n",
      "        autopct : None or str or callable, default: None\n",
      "            If not *None*, is a string or function used to label the wedges\n",
      "            with their numeric value.  The label will be placed inside the\n",
      "            wedge.  If it is a format string, the label will be ``fmt % pct``.\n",
      "            If it is a function, it will be called.\n",
      "        \n",
      "        pctdistance : float, default: 0.6\n",
      "            The ratio between the center of each pie slice and the start of\n",
      "            the text generated by *autopct*.  Ignored if *autopct* is *None*.\n",
      "        \n",
      "        shadow : bool, default: False\n",
      "            Draw a shadow beneath the pie.\n",
      "        \n",
      "        normalize : bool, default: True\n",
      "            When *True*, always make a full pie by normalizing x so that\n",
      "            ``sum(x) == 1``. *False* makes a partial pie if ``sum(x) <= 1``\n",
      "            and raises a `ValueError` for ``sum(x) > 1``.\n",
      "        \n",
      "        labeldistance : float or None, default: 1.1\n",
      "            The radial distance at which the pie labels are drawn.\n",
      "            If set to ``None``, label are not drawn, but are stored for use in\n",
      "            ``legend()``\n",
      "        \n",
      "        startangle : float, default: 0 degrees\n",
      "            The angle by which the start of the pie is rotated,\n",
      "            counterclockwise from the x-axis.\n",
      "        \n",
      "        radius : float, default: 1\n",
      "            The radius of the pie.\n",
      "        \n",
      "        counterclock : bool, default: True\n",
      "            Specify fractions direction, clockwise or counterclockwise.\n",
      "        \n",
      "        wedgeprops : dict, default: None\n",
      "            Dict of arguments passed to the wedge objects making the pie.\n",
      "            For example, you can pass in ``wedgeprops = {'linewidth': 3}``\n",
      "            to set the width of the wedge border lines equal to 3.\n",
      "            For more details, look at the doc/arguments of the wedge object.\n",
      "            By default ``clip_on=False``.\n",
      "        \n",
      "        textprops : dict, default: None\n",
      "            Dict of arguments to pass to the text objects.\n",
      "        \n",
      "        center : (float, float), default: (0, 0)\n",
      "            The coordinates of the center of the chart.\n",
      "        \n",
      "        frame : bool, default: False\n",
      "            Plot Axes frame with the chart if true.\n",
      "        \n",
      "        rotatelabels : bool, default: False\n",
      "            Rotate each label to the angle of the corresponding slice if true.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *explode*, *labels*, *colors*\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        patches : list\n",
      "            A sequence of `matplotlib.patches.Wedge` instances\n",
      "        \n",
      "        texts : list\n",
      "            A list of the label `.Text` instances.\n",
      "        \n",
      "        autotexts : list\n",
      "            A list of `.Text` instances for the numeric labels. This will only\n",
      "            be returned if the parameter *autopct* is not *None*.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The pie chart will probably look best if the figure and Axes are\n",
      "        square, or the Axes aspect is equal.\n",
      "        This method sets the aspect ratio of the axis to \"equal\".\n",
      "        The Axes aspect ratio can be controlled with `.Axes.set_aspect`.\n",
      "    \n",
      "    pink()\n",
      "        Set the colormap to 'pink'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    plasma()\n",
      "        Set the colormap to 'plasma'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    plot(*args, scalex=True, scaley=True, data=None, **kwargs)\n",
      "        Plot y versus x as lines and/or markers.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            plot([x], y, [fmt], *, data=None, **kwargs)\n",
      "            plot([x], y, [fmt], [x2], y2, [fmt2], ..., **kwargs)\n",
      "        \n",
      "        The coordinates of the points or line nodes are given by *x*, *y*.\n",
      "        \n",
      "        The optional parameter *fmt* is a convenient way for defining basic\n",
      "        formatting like color, marker and linestyle. It's a shortcut string\n",
      "        notation described in the *Notes* section below.\n",
      "        \n",
      "        >>> plot(x, y)        # plot x and y using default line style and color\n",
      "        >>> plot(x, y, 'bo')  # plot x and y using blue circle markers\n",
      "        >>> plot(y)           # plot y using x as index array 0..N-1\n",
      "        >>> plot(y, 'r+')     # ditto, but with red plusses\n",
      "        \n",
      "        You can use `.Line2D` properties as keyword arguments for more\n",
      "        control on the appearance. Line properties and *fmt* can be mixed.\n",
      "        The following two calls yield identical results:\n",
      "        \n",
      "        >>> plot(x, y, 'go--', linewidth=2, markersize=12)\n",
      "        >>> plot(x, y, color='green', marker='o', linestyle='dashed',\n",
      "        ...      linewidth=2, markersize=12)\n",
      "        \n",
      "        When conflicting with *fmt*, keyword arguments take precedence.\n",
      "        \n",
      "        \n",
      "        **Plotting labelled data**\n",
      "        \n",
      "        There's a convenient way for plotting objects with labelled data (i.e.\n",
      "        data that can be accessed by index ``obj['y']``). Instead of giving\n",
      "        the data in *x* and *y*, you can provide the object in the *data*\n",
      "        parameter and just give the labels for *x* and *y*::\n",
      "        \n",
      "        >>> plot('xlabel', 'ylabel', data=obj)\n",
      "        \n",
      "        All indexable objects are supported. This could e.g. be a `dict`, a\n",
      "        `pandas.DataFrame` or a structured numpy array.\n",
      "        \n",
      "        \n",
      "        **Plotting multiple sets of data**\n",
      "        \n",
      "        There are various ways to plot multiple sets of data.\n",
      "        \n",
      "        - The most straight forward way is just to call `plot` multiple times.\n",
      "          Example:\n",
      "        \n",
      "          >>> plot(x1, y1, 'bo')\n",
      "          >>> plot(x2, y2, 'go')\n",
      "        \n",
      "        - If *x* and/or *y* are 2D arrays a separate data set will be drawn\n",
      "          for every column. If both *x* and *y* are 2D, they must have the\n",
      "          same shape. If only one of them is 2D with shape (N, m) the other\n",
      "          must have length N and will be used for every data set m.\n",
      "        \n",
      "          Example:\n",
      "        \n",
      "          >>> x = [1, 2, 3]\n",
      "          >>> y = np.array([[1, 2], [3, 4], [5, 6]])\n",
      "          >>> plot(x, y)\n",
      "        \n",
      "          is equivalent to:\n",
      "        \n",
      "          >>> for col in range(y.shape[1]):\n",
      "          ...     plot(x, y[:, col])\n",
      "        \n",
      "        - The third way is to specify multiple sets of *[x]*, *y*, *[fmt]*\n",
      "          groups::\n",
      "        \n",
      "          >>> plot(x1, y1, 'g^', x2, y2, 'g-')\n",
      "        \n",
      "          In this case, any additional keyword argument applies to all\n",
      "          datasets. Also this syntax cannot be combined with the *data*\n",
      "          parameter.\n",
      "        \n",
      "        By default, each line is assigned a different style specified by a\n",
      "        'style cycle'. The *fmt* and line property parameters are only\n",
      "        necessary if you want explicit deviations from these defaults.\n",
      "        Alternatively, you can also change the style cycle using\n",
      "        :rc:`axes.prop_cycle`.\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like or scalar\n",
      "            The horizontal / vertical coordinates of the data points.\n",
      "            *x* values are optional and default to ``range(len(y))``.\n",
      "        \n",
      "            Commonly, these parameters are 1D arrays.\n",
      "        \n",
      "            They can also be scalars, or two-dimensional (in that case, the\n",
      "            columns represent separate data sets).\n",
      "        \n",
      "            These arguments cannot be passed as keywords.\n",
      "        \n",
      "        fmt : str, optional\n",
      "            A format string, e.g. 'ro' for red circles. See the *Notes*\n",
      "            section for a full description of the format strings.\n",
      "        \n",
      "            Format strings are just an abbreviation for quickly setting\n",
      "            basic line properties. All of these and more can also be\n",
      "            controlled by keyword arguments.\n",
      "        \n",
      "            This argument cannot be passed as keyword.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            An object with labelled data. If given, provide the label names to\n",
      "            plot in *x* and *y*.\n",
      "        \n",
      "            .. note::\n",
      "                Technically there's a slight ambiguity in calls where the\n",
      "                second label is a valid *fmt*. ``plot('n', 'o', data=obj)``\n",
      "                could be ``plt(x, y)`` or ``plt(y, fmt)``. In such cases,\n",
      "                the former interpretation is chosen, but a warning is issued.\n",
      "                You may suppress the warning by adding an empty format string\n",
      "                ``plot('n', 'o', '', data=obj)``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            A list of lines representing the plotted data.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        scalex, scaley : bool, default: True\n",
      "            These parameters determine if the view limits are adapted to the\n",
      "            data limits. The values are passed on to\n",
      "            `~.axes.Axes.autoscale_view`.\n",
      "        \n",
      "        **kwargs : `.Line2D` properties, optional\n",
      "            *kwargs* are used to specify properties like a line label (for\n",
      "            auto legends), linewidth, antialiasing, marker face color.\n",
      "            Example::\n",
      "        \n",
      "            >>> plot([1, 2, 3], [1, 2, 3], 'go-', label='line 1', linewidth=2)\n",
      "            >>> plot([1, 2, 3], [1, 4, 9], 'rs', label='line 2')\n",
      "        \n",
      "            If you specify multiple lines with one plot call, the kwargs apply\n",
      "            to all those lines. In case the label object is iterable, each\n",
      "            element is used as labels for each set of data.\n",
      "        \n",
      "            Here is a list of available `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scatter : XY scatter plot with markers of varying size and/or color (\n",
      "            sometimes also called bubble chart).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        **Format Strings**\n",
      "        \n",
      "        A format string consists of a part for color, marker and line::\n",
      "        \n",
      "            fmt = '[marker][line][color]'\n",
      "        \n",
      "        Each of them is optional. If not provided, the value from the style\n",
      "        cycle is used. Exception: If ``line`` is given, but no ``marker``,\n",
      "        the data will be a line without markers.\n",
      "        \n",
      "        Other combinations such as ``[color][marker][line]`` are also\n",
      "        supported, but note that their parsing may be ambiguous.\n",
      "        \n",
      "        **Markers**\n",
      "        \n",
      "        =============   ===============================\n",
      "        character       description\n",
      "        =============   ===============================\n",
      "        ``'.'``         point marker\n",
      "        ``','``         pixel marker\n",
      "        ``'o'``         circle marker\n",
      "        ``'v'``         triangle_down marker\n",
      "        ``'^'``         triangle_up marker\n",
      "        ``'<'``         triangle_left marker\n",
      "        ``'>'``         triangle_right marker\n",
      "        ``'1'``         tri_down marker\n",
      "        ``'2'``         tri_up marker\n",
      "        ``'3'``         tri_left marker\n",
      "        ``'4'``         tri_right marker\n",
      "        ``'8'``         octagon marker\n",
      "        ``'s'``         square marker\n",
      "        ``'p'``         pentagon marker\n",
      "        ``'P'``         plus (filled) marker\n",
      "        ``'*'``         star marker\n",
      "        ``'h'``         hexagon1 marker\n",
      "        ``'H'``         hexagon2 marker\n",
      "        ``'+'``         plus marker\n",
      "        ``'x'``         x marker\n",
      "        ``'X'``         x (filled) marker\n",
      "        ``'D'``         diamond marker\n",
      "        ``'d'``         thin_diamond marker\n",
      "        ``'|'``         vline marker\n",
      "        ``'_'``         hline marker\n",
      "        =============   ===============================\n",
      "        \n",
      "        **Line Styles**\n",
      "        \n",
      "        =============    ===============================\n",
      "        character        description\n",
      "        =============    ===============================\n",
      "        ``'-'``          solid line style\n",
      "        ``'--'``         dashed line style\n",
      "        ``'-.'``         dash-dot line style\n",
      "        ``':'``          dotted line style\n",
      "        =============    ===============================\n",
      "        \n",
      "        Example format strings::\n",
      "        \n",
      "            'b'    # blue markers with default shape\n",
      "            'or'   # red circles\n",
      "            '-g'   # green solid line\n",
      "            '--'   # dashed line with default color\n",
      "            '^k:'  # black triangle_up markers connected by a dotted line\n",
      "        \n",
      "        **Colors**\n",
      "        \n",
      "        The supported color abbreviations are the single letter codes\n",
      "        \n",
      "        =============    ===============================\n",
      "        character        color\n",
      "        =============    ===============================\n",
      "        ``'b'``          blue\n",
      "        ``'g'``          green\n",
      "        ``'r'``          red\n",
      "        ``'c'``          cyan\n",
      "        ``'m'``          magenta\n",
      "        ``'y'``          yellow\n",
      "        ``'k'``          black\n",
      "        ``'w'``          white\n",
      "        =============    ===============================\n",
      "        \n",
      "        and the ``'CN'`` colors that index into the default property cycle.\n",
      "        \n",
      "        If the color is the only part of the format string, you can\n",
      "        additionally use any  `matplotlib.colors` spec, e.g. full names\n",
      "        (``'green'``) or hex strings (``'#008000'``).\n",
      "    \n",
      "    plot_date(x, y, fmt='o', tz=None, xdate=True, ydate=False, *, data=None, **kwargs)\n",
      "        [*Discouraged*] Plot coercing the axis to treat floats as dates.\n",
      "        \n",
      "        .. admonition:: Discouraged\n",
      "        \n",
      "            This method exists for historic reasons and will be deprecated in\n",
      "            the future.\n",
      "        \n",
      "            - ``datetime``-like data should directly be plotted using\n",
      "              `~.Axes.plot`.\n",
      "            -  If you need to plot plain numeric data as :ref:`date-format` or\n",
      "               need to set a timezone, call ``ax.xaxis.axis_date`` /\n",
      "               ``ax.yaxis.axis_date`` before `~.Axes.plot`. See\n",
      "               `.Axis.axis_date`.\n",
      "        \n",
      "        Similar to `.plot`, this plots *y* vs. *x* as lines or markers.\n",
      "        However, the axis labels are formatted as dates depending on *xdate*\n",
      "        and *ydate*.  Note that `.plot` will work with `datetime` and\n",
      "        `numpy.datetime64` objects without resorting to this method.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like\n",
      "            The coordinates of the data points. If *xdate* or *ydate* is\n",
      "            *True*, the respective values *x* or *y* are interpreted as\n",
      "            :ref:`Matplotlib dates <date-format>`.\n",
      "        \n",
      "        fmt : str, optional\n",
      "            The plot format string. For details, see the corresponding\n",
      "            parameter in `.plot`.\n",
      "        \n",
      "        tz : timezone string or `datetime.tzinfo`, default: :rc:`timezone`\n",
      "            The time zone to use in labeling dates.\n",
      "        \n",
      "        xdate : bool, default: True\n",
      "            If *True*, the *x*-axis will be interpreted as Matplotlib dates.\n",
      "        \n",
      "        ydate : bool, default: False\n",
      "            If *True*, the *y*-axis will be interpreted as Matplotlib dates.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            Objects representing the plotted data.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*\n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.dates : Helper functions on dates.\n",
      "        matplotlib.dates.date2num : Convert dates to num.\n",
      "        matplotlib.dates.num2date : Convert num to dates.\n",
      "        matplotlib.dates.drange : Create an equally spaced sequence of dates.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        If you are using custom date tickers and formatters, it may be\n",
      "        necessary to set the formatters/locators after the call to\n",
      "        `.plot_date`. `.plot_date` will set the default tick locator to\n",
      "        `.AutoDateLocator` (if the tick locator is not already set to a\n",
      "        `.DateLocator` instance) and the default tick formatter to\n",
      "        `.AutoDateFormatter` (if the tick formatter is not already set to a\n",
      "        `.DateFormatter` instance).\n",
      "    \n",
      "    polar(*args, **kwargs)\n",
      "        Make a polar plot.\n",
      "        \n",
      "        call signature::\n",
      "        \n",
      "          polar(theta, r, **kwargs)\n",
      "        \n",
      "        Multiple *theta*, *r* arguments are supported, with format strings, as in\n",
      "        `plot`.\n",
      "    \n",
      "    prism()\n",
      "        Set the colormap to 'prism'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    psd(x, NFFT=None, Fs=None, Fc=None, detrend=None, window=None, noverlap=None, pad_to=None, sides=None, scale_by_freq=None, return_line=None, *, data=None, **kwargs)\n",
      "        Plot the power spectral density.\n",
      "        \n",
      "        The power spectral density :math:`P_{xx}` by Welch's average\n",
      "        periodogram method.  The vector *x* is divided into *NFFT* length\n",
      "        segments.  Each segment is detrended by function *detrend* and\n",
      "        windowed by function *window*.  *noverlap* gives the length of\n",
      "        the overlap between segments.  The :math:`|\\mathrm{fft}(i)|^2`\n",
      "        of each segment :math:`i` are averaged to compute :math:`P_{xx}`,\n",
      "        with a scaling to correct for power loss due to windowing.\n",
      "        \n",
      "        If len(*x*) < *NFFT*, it will be zero padded to *NFFT*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1-D array or sequence\n",
      "            Array or sequence containing the data\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  This can be different from *NFFT*, which specifies the number\n",
      "            of data points used.  While not increasing the actual resolution of the\n",
      "            spectrum (the minimum distance between resolvable peaks), this can give\n",
      "            more points in the plot, allowing for more detail. This corresponds to\n",
      "            the *n* parameter in the call to `~numpy.fft.fft`. The default is None,\n",
      "            which sets *pad_to* equal to *NFFT*\n",
      "        \n",
      "        NFFT : int, default: 256\n",
      "            The number of data points used in each block for the FFT.  A power 2 is\n",
      "            most efficient.  This should *NOT* be used to get zero padding, or the\n",
      "            scaling of the result will be incorrect; use *pad_to* for this instead.\n",
      "        \n",
      "        detrend : {'none', 'mean', 'linear'} or callable, default: 'none'\n",
      "            The function applied to each segment before fft-ing, designed to remove\n",
      "            the mean or linear trend.  Unlike in MATLAB, where the *detrend* parameter\n",
      "            is a vector, in Matplotlib it is a function.  The :mod:`~matplotlib.mlab`\n",
      "            module defines `.detrend_none`, `.detrend_mean`, and `.detrend_linear`,\n",
      "            but you can use a custom function as well.  You can also use a string to\n",
      "            choose one of the functions: 'none' calls `.detrend_none`. 'mean' calls\n",
      "            `.detrend_mean`. 'linear' calls `.detrend_linear`.\n",
      "        \n",
      "        scale_by_freq : bool, default: True\n",
      "            Whether the resulting density values should be scaled by the scaling\n",
      "            frequency, which gives density in units of 1/Hz.  This allows for\n",
      "            integration over the returned frequency values.  The default is True for\n",
      "            MATLAB compatibility.\n",
      "        \n",
      "        noverlap : int, default: 0 (no overlap)\n",
      "            The number of points of overlap between segments.\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        return_line : bool, default: False\n",
      "            Whether to include the line object plotted in the returned values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        Pxx : 1-D array\n",
      "            The values for the power spectrum :math:`P_{xx}` before scaling\n",
      "            (real valued).\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the elements in *Pxx*.\n",
      "        \n",
      "        line : `~matplotlib.lines.Line2D`\n",
      "            The line created by this function.\n",
      "            Only returned if *return_line* is True.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Keyword arguments control the `.Line2D` properties:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        specgram\n",
      "            Differs in the default overlap; in not returning the mean of the\n",
      "            segment periodograms; in returning the times of the segments; and\n",
      "            in plotting a colormap instead of a line.\n",
      "        magnitude_spectrum\n",
      "            Plots the magnitude spectrum.\n",
      "        csd\n",
      "            Plots the spectral density between two signals.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For plotting, the power is plotted as\n",
      "        :math:`10\\log_{10}(P_{xx})` for decibels, though *Pxx* itself\n",
      "        is returned.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Bendat & Piersol -- Random Data: Analysis and Measurement Procedures,\n",
      "        John Wiley & Sons (1986)\n",
      "    \n",
      "    quiver(*args, data=None, **kwargs)\n",
      "        Plot a 2D field of arrows.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "          quiver([X, Y], U, V, [C], **kwargs)\n",
      "        \n",
      "        *X*, *Y* define the arrow locations, *U*, *V* define the arrow directions, and\n",
      "        *C* optionally sets the color.\n",
      "        \n",
      "        **Arrow length**\n",
      "        \n",
      "        The default settings auto-scales the length of the arrows to a reasonable size.\n",
      "        To change this behavior see the *scale* and *scale_units* parameters.\n",
      "        \n",
      "        **Arrow shape**\n",
      "        \n",
      "        The arrow shape is determined by *width*, *headwidth*, *headlength* and\n",
      "        *headaxislength*. See the notes below.\n",
      "        \n",
      "        **Arrow styling**\n",
      "        \n",
      "        Each arrow is internally represented by a filled polygon with a default edge\n",
      "        linewidth of 0. As a result, an arrow is rather a filled area, not a line with\n",
      "        a head, and `.PolyCollection` properties like *linewidth*, *edgecolor*,\n",
      "        *facecolor*, etc. act accordingly.\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        X, Y : 1D or 2D array-like, optional\n",
      "            The x and y coordinates of the arrow locations.\n",
      "        \n",
      "            If not given, they will be generated as a uniform integer meshgrid based\n",
      "            on the dimensions of *U* and *V*.\n",
      "        \n",
      "            If *X* and *Y* are 1D but *U*, *V* are 2D, *X*, *Y* are expanded to 2D\n",
      "            using ``X, Y = np.meshgrid(X, Y)``. In this case ``len(X)`` and ``len(Y)``\n",
      "            must match the column and row dimensions of *U* and *V*.\n",
      "        \n",
      "        U, V : 1D or 2D array-like\n",
      "            The x and y direction components of the arrow vectors. The interpretation\n",
      "            of these components (in data or in screen space) depends on *angles*.\n",
      "        \n",
      "            *U* and *V* must have the same number of elements, matching the number of\n",
      "            arrow locations in  *X*, *Y*. *U* and *V* may be masked. Locations masked\n",
      "            in any of *U*, *V*, and *C* will not be drawn.\n",
      "        \n",
      "        C : 1D or 2D array-like, optional\n",
      "            Numeric data that defines the arrow colors by colormapping via *norm* and\n",
      "            *cmap*.\n",
      "        \n",
      "            This does not support explicit colors. If you want to set colors directly,\n",
      "            use *color* instead.  The size of *C* must match the number of arrow\n",
      "            locations.\n",
      "        \n",
      "        angles : {'uv', 'xy'} or array-like, default: 'uv'\n",
      "            Method for determining the angle of the arrows.\n",
      "        \n",
      "            - 'uv': Arrow direction in screen coordinates. Use this if the arrows\n",
      "              symbolize a quantity that is not based on *X*, *Y* data coordinates.\n",
      "        \n",
      "              If *U* == *V* the orientation of the arrow on the plot is 45 degrees\n",
      "              counter-clockwise from the  horizontal axis (positive to the right).\n",
      "        \n",
      "            - 'xy': Arrow direction in data coordinates, i.e. the arrows point from\n",
      "              (x, y) to (x+u, y+v). Use this e.g. for plotting a gradient field.\n",
      "        \n",
      "            - Arbitrary angles may be specified explicitly as an array of values\n",
      "              in degrees, counter-clockwise from the horizontal axis.\n",
      "        \n",
      "              In this case *U*, *V* is only used to determine the length of the\n",
      "              arrows.\n",
      "        \n",
      "            Note: inverting a data axis will correspondingly invert the\n",
      "            arrows only with ``angles='xy'``.\n",
      "        \n",
      "        pivot : {'tail', 'mid', 'middle', 'tip'}, default: 'tail'\n",
      "            The part of the arrow that is anchored to the *X*, *Y* grid. The arrow\n",
      "            rotates about this point.\n",
      "        \n",
      "            'mid' is a synonym for 'middle'.\n",
      "        \n",
      "        scale : float, optional\n",
      "            Scales the length of the arrow inversely.\n",
      "        \n",
      "            Number of data units per arrow length unit, e.g., m/s per plot width; a\n",
      "            smaller scale parameter makes the arrow longer. Default is *None*.\n",
      "        \n",
      "            If *None*, a simple autoscaling algorithm is used, based on the average\n",
      "            vector length and the number of vectors. The arrow length unit is given by\n",
      "            the *scale_units* parameter.\n",
      "        \n",
      "        scale_units : {'width', 'height', 'dots', 'inches', 'x', 'y', 'xy'}, optional\n",
      "            If the *scale* kwarg is *None*, the arrow length unit. Default is *None*.\n",
      "        \n",
      "            e.g. *scale_units* is 'inches', *scale* is 2.0, and ``(u, v) = (1, 0)``,\n",
      "            then the vector will be 0.5 inches long.\n",
      "        \n",
      "            If *scale_units* is 'width' or 'height', then the vector will be half the\n",
      "            width/height of the axes.\n",
      "        \n",
      "            If *scale_units* is 'x' then the vector will be 0.5 x-axis\n",
      "            units. To plot vectors in the x-y plane, with u and v having\n",
      "            the same units as x and y, use\n",
      "            ``angles='xy', scale_units='xy', scale=1``.\n",
      "        \n",
      "        units : {'width', 'height', 'dots', 'inches', 'x', 'y', 'xy'}, default: 'width'\n",
      "            Affects the arrow size (except for the length). In particular, the shaft\n",
      "            *width* is measured in multiples of this unit.\n",
      "        \n",
      "            Supported values are:\n",
      "        \n",
      "            - 'width', 'height': The width or height of the Axes.\n",
      "            - 'dots', 'inches': Pixels or inches based on the figure dpi.\n",
      "            - 'x', 'y', 'xy': *X*, *Y* or :math:`\\sqrt{X^2 + Y^2}` in data units.\n",
      "        \n",
      "            The following table summarizes how these values affect the visible arrow\n",
      "            size under zooming and figure size changes:\n",
      "        \n",
      "            =================  =================   ==================\n",
      "            units              zoom                figure size change\n",
      "            =================  =================   ==================\n",
      "            'x', 'y', 'xy'     arrow size scales   -\n",
      "            'width', 'height'  -                   arrow size scales\n",
      "            'dots', 'inches'   -                   -\n",
      "            =================  =================   ==================\n",
      "        \n",
      "        width : float, optional\n",
      "            Shaft width in arrow units. All head parameters are relative to *width*.\n",
      "        \n",
      "            The default depends on choice of *units* above, and number of vectors;\n",
      "            a typical starting value is about 0.005 times the width of the plot.\n",
      "        \n",
      "        headwidth : float, default: 3\n",
      "            Head width as multiple of shaft *width*. See the notes below.\n",
      "        \n",
      "        headlength : float, default: 5\n",
      "            Head length as multiple of shaft *width*. See the notes below.\n",
      "        \n",
      "        headaxislength : float, default: 4.5\n",
      "            Head length at shaft intersection as multiple of shaft *width*.\n",
      "            See the notes below.\n",
      "        \n",
      "        minshaft : float, default: 1\n",
      "            Length below which arrow scales, in units of head length. Do not\n",
      "            set this to less than 1, or small arrows will look terrible!\n",
      "        \n",
      "        minlength : float, default: 1\n",
      "            Minimum length as a multiple of shaft width; if an arrow length\n",
      "            is less than this, plot a dot (hexagon) of this diameter instead.\n",
      "        \n",
      "        color : color or color sequence, optional\n",
      "            Explicit color(s) for the arrows. If *C* has been set, *color* has no\n",
      "            effect.\n",
      "        \n",
      "            This is a synonym for the `.PolyCollection` *facecolor* parameter.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs : `~matplotlib.collections.PolyCollection` properties, optional\n",
      "            All other keyword arguments are passed on to `.PolyCollection`:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: array-like or scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa or antialiaseds: bool or list of bools\n",
      "            array: array-like or None\n",
      "            capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            clim: (vmin: float, vmax: float)\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            cmap: `.Colormap` or str or None\n",
      "            color: color or list of rgba tuples\n",
      "            edgecolor or ec or edgecolors: color or list of colors or 'face'\n",
      "            facecolor or facecolors or fc: color or list of colors\n",
      "            figure: `.Figure`\n",
      "            gid: str\n",
      "            hatch: {'/', '\\\\', '|', '-', '+', 'x', 'o', 'O', '.', '*'}\n",
      "            in_layout: bool\n",
      "            joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            label: object\n",
      "            linestyle or dashes or linestyles or ls: str or tuple or list thereof\n",
      "            linewidth or linewidths or lw: float or list of floats\n",
      "            mouseover: bool\n",
      "            norm: `.Normalize` or str or None\n",
      "            offset_transform or transOffset: unknown\n",
      "            offsets: (N, 2) or (2,) array-like\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            paths: list of array-like\n",
      "            picker: None or bool or float or callable\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sizes: ndarray or None\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            urls: list of str or None\n",
      "            verts: list of array-like\n",
      "            verts_and_codes: unknown\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.quiver.Quiver`\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Axes.quiverkey : Add a key to a quiver plot.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        **Arrow shape**\n",
      "        \n",
      "        The arrow is drawn as a polygon using the nodes as shown below. The values\n",
      "        *headwidth*, *headlength*, and *headaxislength* are in units of *width*.\n",
      "        \n",
      "        .. image:: /_static/quiver_sizes.svg\n",
      "           :width: 500px\n",
      "        \n",
      "        The defaults give a slightly swept-back arrow. Here are some guidelines how to\n",
      "        get other head shapes:\n",
      "        \n",
      "        - To make the head a triangle, make *headaxislength* the same as *headlength*.\n",
      "        - To make the arrow more pointed, reduce *headwidth* or increase *headlength*\n",
      "          and *headaxislength*.\n",
      "        - To make the head smaller relative to the shaft, scale down all the head\n",
      "          parameters proportionally.\n",
      "        - To remove the head completely, set all *head* parameters to 0.\n",
      "        - To get a diamond-shaped head, make *headaxislength* larger than *headlength*.\n",
      "        - Warning: For *headaxislength* < (*headlength* / *headwidth*), the \"headaxis\"\n",
      "          nodes (i.e. the ones connecting the head with the shaft) will protrude out\n",
      "          of the head in forward direction so that the arrow head looks broken.\n",
      "    \n",
      "    quiverkey(Q, X, Y, U, label, **kwargs)\n",
      "        Add a key to a quiver plot.\n",
      "        \n",
      "        The positioning of the key depends on *X*, *Y*, *coordinates*, and\n",
      "        *labelpos*.  If *labelpos* is 'N' or 'S', *X*, *Y* give the position of\n",
      "        the middle of the key arrow.  If *labelpos* is 'E', *X*, *Y* positions\n",
      "        the head, and if *labelpos* is 'W', *X*, *Y* positions the tail; in\n",
      "        either of these two cases, *X*, *Y* is somewhere in the middle of the\n",
      "        arrow+label key object.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        Q : `matplotlib.quiver.Quiver`\n",
      "            A `.Quiver` object as returned by a call to `~.Axes.quiver()`.\n",
      "        X, Y : float\n",
      "            The location of the key.\n",
      "        U : float\n",
      "            The length of the key.\n",
      "        label : str\n",
      "            The key label (e.g., length and units of the key).\n",
      "        angle : float, default: 0\n",
      "            The angle of the key arrow, in degrees anti-clockwise from the\n",
      "            x-axis.\n",
      "        coordinates : {'axes', 'figure', 'data', 'inches'}, default: 'axes'\n",
      "            Coordinate system and units for *X*, *Y*: 'axes' and 'figure' are\n",
      "            normalized coordinate systems with (0, 0) in the lower left and\n",
      "            (1, 1) in the upper right; 'data' are the axes data coordinates\n",
      "            (used for the locations of the vectors in the quiver plot itself);\n",
      "            'inches' is position in the figure in inches, with (0, 0) at the\n",
      "            lower left corner.\n",
      "        color : color\n",
      "            Overrides face and edge colors from *Q*.\n",
      "        labelpos : {'N', 'S', 'E', 'W'}\n",
      "            Position the label above, below, to the right, to the left of the\n",
      "            arrow, respectively.\n",
      "        labelsep : float, default: 0.1\n",
      "            Distance in inches between the arrow and the label.\n",
      "        labelcolor : color, default: :rc:`text.color`\n",
      "            Label color.\n",
      "        fontproperties : dict, optional\n",
      "            A dictionary with keyword arguments accepted by the\n",
      "            `~matplotlib.font_manager.FontProperties` initializer:\n",
      "            *family*, *style*, *variant*, *size*, *weight*.\n",
      "        **kwargs\n",
      "            Any additional keyword arguments are used to override vector\n",
      "            properties taken from *Q*.\n",
      "    \n",
      "    rc(group, **kwargs)\n",
      "        Set the current `.rcParams`.  *group* is the grouping for the rc, e.g.,\n",
      "        for ``lines.linewidth`` the group is ``lines``, for\n",
      "        ``axes.facecolor``, the group is ``axes``, and so on.  Group may\n",
      "        also be a list or tuple of group names, e.g., (*xtick*, *ytick*).\n",
      "        *kwargs* is a dictionary attribute name/value pairs, e.g.,::\n",
      "        \n",
      "          rc('lines', linewidth=2, color='r')\n",
      "        \n",
      "        sets the current `.rcParams` and is equivalent to::\n",
      "        \n",
      "          rcParams['lines.linewidth'] = 2\n",
      "          rcParams['lines.color'] = 'r'\n",
      "        \n",
      "        The following aliases are available to save typing for interactive users:\n",
      "        \n",
      "        =====   =================\n",
      "        Alias   Property\n",
      "        =====   =================\n",
      "        'lw'    'linewidth'\n",
      "        'ls'    'linestyle'\n",
      "        'c'     'color'\n",
      "        'fc'    'facecolor'\n",
      "        'ec'    'edgecolor'\n",
      "        'mew'   'markeredgewidth'\n",
      "        'aa'    'antialiased'\n",
      "        =====   =================\n",
      "        \n",
      "        Thus you could abbreviate the above call as::\n",
      "        \n",
      "              rc('lines', lw=2, c='r')\n",
      "        \n",
      "        Note you can use python's kwargs dictionary facility to store\n",
      "        dictionaries of default parameters.  e.g., you can customize the\n",
      "        font rc as follows::\n",
      "        \n",
      "          font = {'family' : 'monospace',\n",
      "                  'weight' : 'bold',\n",
      "                  'size'   : 'larger'}\n",
      "          rc('font', **font)  # pass in the font dict as kwargs\n",
      "        \n",
      "        This enables you to easily switch between several configurations.  Use\n",
      "        ``matplotlib.style.use('default')`` or :func:`~matplotlib.rcdefaults` to\n",
      "        restore the default `.rcParams` after changes.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Similar functionality is available by using the normal dict interface, i.e.\n",
      "        ``rcParams.update({\"lines.linewidth\": 2, ...})`` (but ``rcParams.update``\n",
      "        does not support abbreviations or grouping).\n",
      "    \n",
      "    rc_context(rc=None, fname=None)\n",
      "        Return a context manager for temporarily changing rcParams.\n",
      "        \n",
      "        The :rc:`backend` will not be reset by the context manager.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rc : dict\n",
      "            The rcParams to temporarily set.\n",
      "        fname : str or path-like\n",
      "            A file with Matplotlib rc settings. If both *fname* and *rc* are given,\n",
      "            settings from *rc* take precedence.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        :ref:`customizing-with-matplotlibrc-files`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Passing explicit values via a dict::\n",
      "        \n",
      "            with mpl.rc_context({'interactive': False}):\n",
      "                fig, ax = plt.subplots()\n",
      "                ax.plot(range(3), range(3))\n",
      "                fig.savefig('example.png')\n",
      "                plt.close(fig)\n",
      "        \n",
      "        Loading settings from a file::\n",
      "        \n",
      "             with mpl.rc_context(fname='print.rc'):\n",
      "                 plt.plot(x, y)  # uses 'print.rc'\n",
      "    \n",
      "    rcdefaults()\n",
      "        Restore the `.rcParams` from Matplotlib's internal default style.\n",
      "        \n",
      "        Style-blacklisted `.rcParams` (defined in\n",
      "        ``matplotlib.style.core.STYLE_BLACKLIST``) are not updated.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        matplotlib.rc_file_defaults\n",
      "            Restore the `.rcParams` from the rc file originally loaded by\n",
      "            Matplotlib.\n",
      "        matplotlib.style.use\n",
      "            Use a specific style file.  Call ``style.use('default')`` to restore\n",
      "            the default style.\n",
      "    \n",
      "    rgrids(radii=None, labels=None, angle=None, fmt=None, **kwargs)\n",
      "        Get or set the radial gridlines on the current polar plot.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "         lines, labels = rgrids()\n",
      "         lines, labels = rgrids(radii, labels=None, angle=22.5, fmt=None, **kwargs)\n",
      "        \n",
      "        When called with no arguments, `.rgrids` simply returns the tuple\n",
      "        (*lines*, *labels*). When called with arguments, the labels will\n",
      "        appear at the specified radial distances and angle.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        radii : tuple with floats\n",
      "            The radii for the radial gridlines\n",
      "        \n",
      "        labels : tuple with strings or None\n",
      "            The labels to use at each radial gridline. The\n",
      "            `matplotlib.ticker.ScalarFormatter` will be used if None.\n",
      "        \n",
      "        angle : float\n",
      "            The angular position of the radius labels in degrees.\n",
      "        \n",
      "        fmt : str or None\n",
      "            Format string used in `matplotlib.ticker.FormatStrFormatter`.\n",
      "            For example '%f'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lines : list of `.lines.Line2D`\n",
      "            The radial gridlines.\n",
      "        \n",
      "        labels : list of `.text.Text`\n",
      "            The tick labels.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            *kwargs* are optional `.Text` properties for the labels.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .pyplot.thetagrids\n",
      "        .projections.polar.PolarAxes.set_rgrids\n",
      "        .Axis.get_gridlines\n",
      "        .Axis.get_ticklabels\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "          # set the locations of the radial gridlines\n",
      "          lines, labels = rgrids( (0.25, 0.5, 1.0) )\n",
      "        \n",
      "          # set the locations and labels of the radial gridlines\n",
      "          lines, labels = rgrids( (0.25, 0.5, 1.0), ('Tom', 'Dick', 'Harry' ))\n",
      "    \n",
      "    savefig(*args, **kwargs)\n",
      "        Save the current figure.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "          savefig(fname, *, dpi='figure', format=None, metadata=None,\n",
      "                  bbox_inches=None, pad_inches=0.1,\n",
      "                  facecolor='auto', edgecolor='auto',\n",
      "                  backend=None, **kwargs\n",
      "                 )\n",
      "        \n",
      "        The available output formats depend on the backend being used.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fname : str or path-like or binary file-like\n",
      "            A path, or a Python file-like object, or\n",
      "            possibly some backend-dependent object such as\n",
      "            `matplotlib.backends.backend_pdf.PdfPages`.\n",
      "        \n",
      "            If *format* is set, it determines the output format, and the file\n",
      "            is saved as *fname*.  Note that *fname* is used verbatim, and there\n",
      "            is no attempt to make the extension, if any, of *fname* match\n",
      "            *format*, and no extension is appended.\n",
      "        \n",
      "            If *format* is not set, then the format is inferred from the\n",
      "            extension of *fname*, if there is one.  If *format* is not\n",
      "            set and *fname* has no extension, then the file is saved with\n",
      "            :rc:`savefig.format` and the appropriate extension is appended to\n",
      "            *fname*.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        dpi : float or 'figure', default: :rc:`savefig.dpi`\n",
      "            The resolution in dots per inch.  If 'figure', use the figure's\n",
      "            dpi value.\n",
      "        \n",
      "        format : str\n",
      "            The file format, e.g. 'png', 'pdf', 'svg', ... The behavior when\n",
      "            this is unset is documented under *fname*.\n",
      "        \n",
      "        metadata : dict, optional\n",
      "            Key/value pairs to store in the image metadata. The supported keys\n",
      "            and defaults depend on the image format and backend:\n",
      "        \n",
      "            - 'png' with Agg backend: See the parameter ``metadata`` of\n",
      "              `~.FigureCanvasAgg.print_png`.\n",
      "            - 'pdf' with pdf backend: See the parameter ``metadata`` of\n",
      "              `~.backend_pdf.PdfPages`.\n",
      "            - 'svg' with svg backend: See the parameter ``metadata`` of\n",
      "              `~.FigureCanvasSVG.print_svg`.\n",
      "            - 'eps' and 'ps' with PS backend: Only 'Creator' is supported.\n",
      "        \n",
      "        bbox_inches : str or `.Bbox`, default: :rc:`savefig.bbox`\n",
      "            Bounding box in inches: only the given portion of the figure is\n",
      "            saved.  If 'tight', try to figure out the tight bbox of the figure.\n",
      "        \n",
      "        pad_inches : float, default: :rc:`savefig.pad_inches`\n",
      "            Amount of padding around the figure when bbox_inches is 'tight'.\n",
      "        \n",
      "        facecolor : color or 'auto', default: :rc:`savefig.facecolor`\n",
      "            The facecolor of the figure.  If 'auto', use the current figure\n",
      "            facecolor.\n",
      "        \n",
      "        edgecolor : color or 'auto', default: :rc:`savefig.edgecolor`\n",
      "            The edgecolor of the figure.  If 'auto', use the current figure\n",
      "            edgecolor.\n",
      "        \n",
      "        backend : str, optional\n",
      "            Use a non-default backend to render the file, e.g. to render a\n",
      "            png file with the \"cairo\" backend rather than the default \"agg\",\n",
      "            or a pdf file with the \"pgf\" backend rather than the default\n",
      "            \"pdf\".  Note that the default backend is normally sufficient.  See\n",
      "            :ref:`the-builtin-backends` for a list of valid backends for each\n",
      "            file format.  Custom backends can be referenced as \"module://...\".\n",
      "        \n",
      "        orientation : {'landscape', 'portrait'}\n",
      "            Currently only supported by the postscript backend.\n",
      "        \n",
      "        papertype : str\n",
      "            One of 'letter', 'legal', 'executive', 'ledger', 'a0' through\n",
      "            'a10', 'b0' through 'b10'. Only supported for postscript\n",
      "            output.\n",
      "        \n",
      "        transparent : bool\n",
      "            If *True*, the Axes patches will all be transparent; the\n",
      "            Figure patch will also be transparent unless *facecolor*\n",
      "            and/or *edgecolor* are specified via kwargs.\n",
      "        \n",
      "            If *False* has no effect and the color of the Axes and\n",
      "            Figure patches are unchanged (unless the Figure patch\n",
      "            is specified via the *facecolor* and/or *edgecolor* keyword\n",
      "            arguments in which case those colors are used).\n",
      "        \n",
      "            The transparency of these patches will be restored to their\n",
      "            original values upon exit of this function.\n",
      "        \n",
      "            This is useful, for example, for displaying\n",
      "            a plot on top of a colored background on a web page.\n",
      "        \n",
      "        bbox_extra_artists : list of `~matplotlib.artist.Artist`, optional\n",
      "            A list of extra artists that will be considered when the\n",
      "            tight bbox is calculated.\n",
      "        \n",
      "        pil_kwargs : dict, optional\n",
      "            Additional keyword arguments that are passed to\n",
      "            `PIL.Image.Image.save` when saving the figure.\n",
      "    \n",
      "    sca(ax)\n",
      "        Set the current Axes to *ax* and the current Figure to the parent of *ax*.\n",
      "    \n",
      "    scatter(x, y, s=None, c=None, marker=None, cmap=None, norm=None, vmin=None, vmax=None, alpha=None, linewidths=None, *, edgecolors=None, plotnonfinite=False, data=None, **kwargs)\n",
      "        A scatter plot of *y* vs. *x* with varying marker size and/or color.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : float or array-like, shape (n, )\n",
      "            The data positions.\n",
      "        \n",
      "        s : float or array-like, shape (n, ), optional\n",
      "            The marker size in points**2 (typographic points are 1/72 in.).\n",
      "            Default is ``rcParams['lines.markersize'] ** 2``.\n",
      "        \n",
      "        c : array-like or list of colors or color, optional\n",
      "            The marker colors. Possible values:\n",
      "        \n",
      "            - A scalar or sequence of n numbers to be mapped to colors using\n",
      "              *cmap* and *norm*.\n",
      "            - A 2D array in which the rows are RGB or RGBA.\n",
      "            - A sequence of colors of length n.\n",
      "            - A single color format string.\n",
      "        \n",
      "            Note that *c* should not be a single numeric RGB or RGBA sequence\n",
      "            because that is indistinguishable from an array of values to be\n",
      "            colormapped. If you want to specify the same RGB or RGBA value for\n",
      "            all points, use a 2D array with a single row.  Otherwise, value-\n",
      "            matching will have precedence in case of a size matching with *x*\n",
      "            and *y*.\n",
      "        \n",
      "            If you wish to specify a single color for all points\n",
      "            prefer the *color* keyword argument.\n",
      "        \n",
      "            Defaults to `None`. In that case the marker color is determined\n",
      "            by the value of *color*, *facecolor* or *facecolors*. In case\n",
      "            those are not specified or `None`, the marker color is determined\n",
      "            by the next color of the ``Axes``' current \"shape and fill\" color\n",
      "            cycle. This cycle defaults to :rc:`axes.prop_cycle`.\n",
      "        \n",
      "        marker : `~.markers.MarkerStyle`, default: :rc:`scatter.marker`\n",
      "            The marker style. *marker* can be either an instance of the class\n",
      "            or the text shorthand for a particular marker.\n",
      "            See :mod:`matplotlib.markers` for more information about marker\n",
      "            styles.\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *c* is RGB(A).\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *c* is RGB(A).\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            This parameter is ignored if *c* is RGB(A).\n",
      "        \n",
      "        alpha : float, default: None\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        linewidths : float or array-like, default: :rc:`lines.linewidth`\n",
      "            The linewidth of the marker edges. Note: The default *edgecolors*\n",
      "            is 'face'. You may want to change this as well.\n",
      "        \n",
      "        edgecolors : {'face', 'none', *None*} or color or sequence of color, default: :rc:`scatter.edgecolors`\n",
      "            The edge color of the marker. Possible values:\n",
      "        \n",
      "            - 'face': The edge color will always be the same as the face color.\n",
      "            - 'none': No patch boundary will be drawn.\n",
      "            - A color or sequence of colors.\n",
      "        \n",
      "            For non-filled markers, *edgecolors* is ignored. Instead, the color\n",
      "            is determined like with 'face', i.e. from *c*, *colors*, or\n",
      "            *facecolors*.\n",
      "        \n",
      "        plotnonfinite : bool, default: False\n",
      "            Whether to plot points with nonfinite *c* (i.e. ``inf``, ``-inf``\n",
      "            or ``nan``). If ``True`` the points are drawn with the *bad*\n",
      "            colormap color (see `.Colormap.set_bad`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.collections.PathCollection`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*, *s*, *linewidths*, *edgecolors*, *c*, *facecolor*, *facecolors*, *color*\n",
      "        **kwargs : `~matplotlib.collections.Collection` properties\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        plot : To plot scatter plots when markers are identical in size and\n",
      "            color.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        * The `.plot` function will be faster for scatterplots where markers\n",
      "          don't vary in size or color.\n",
      "        \n",
      "        * Any or all of *x*, *y*, *s*, and *c* may be masked arrays, in which\n",
      "          case all masks will be combined and only unmasked points will be\n",
      "          plotted.\n",
      "        \n",
      "        * Fundamentally, scatter works with 1D arrays; *x*, *y*, *s*, and *c*\n",
      "          may be input as N-D arrays, but within scatter they will be\n",
      "          flattened. The exception is *c*, which will be flattened only if its\n",
      "          size matches the size of *x* and *y*.\n",
      "    \n",
      "    sci(im)\n",
      "        Set the current image.\n",
      "        \n",
      "        This image will be the target of colormap functions like\n",
      "        ``pyplot.viridis``, and other functions such as `~.pyplot.clim`.  The\n",
      "        current image is an attribute of the current Axes.\n",
      "    \n",
      "    semilogx(*args, **kwargs)\n",
      "        Make a plot with log scaling on the x axis.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            semilogx([x], y, [fmt], data=None, **kwargs)\n",
      "            semilogx([x], y, [fmt], [x2], y2, [fmt2], ..., **kwargs)\n",
      "        \n",
      "        This is just a thin wrapper around `.plot` which additionally changes\n",
      "        the x-axis to log scaling. All of the concepts and parameters of plot\n",
      "        can be used here as well.\n",
      "        \n",
      "        The additional parameters *base*, *subs*, and *nonpositive* control the\n",
      "        x-axis properties. They are just forwarded to `.Axes.set_xscale`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        base : float, default: 10\n",
      "            Base of the x logarithm.\n",
      "        \n",
      "        subs : array-like, optional\n",
      "            The location of the minor xticks. If *None*, reasonable locations\n",
      "            are automatically chosen depending on the number of decades in the\n",
      "            plot. See `.Axes.set_xscale` for details.\n",
      "        \n",
      "        nonpositive : {'mask', 'clip'}, default: 'mask'\n",
      "            Non-positive values in x can be masked as invalid, or clipped to a\n",
      "            very small positive number.\n",
      "        \n",
      "        **kwargs\n",
      "            All parameters supported by `.plot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            Objects representing the plotted data.\n",
      "    \n",
      "    semilogy(*args, **kwargs)\n",
      "        Make a plot with log scaling on the y axis.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            semilogy([x], y, [fmt], data=None, **kwargs)\n",
      "            semilogy([x], y, [fmt], [x2], y2, [fmt2], ..., **kwargs)\n",
      "        \n",
      "        This is just a thin wrapper around `.plot` which additionally changes\n",
      "        the y-axis to log scaling. All of the concepts and parameters of plot\n",
      "        can be used here as well.\n",
      "        \n",
      "        The additional parameters *base*, *subs*, and *nonpositive* control the\n",
      "        y-axis properties. They are just forwarded to `.Axes.set_yscale`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        base : float, default: 10\n",
      "            Base of the y logarithm.\n",
      "        \n",
      "        subs : array-like, optional\n",
      "            The location of the minor yticks. If *None*, reasonable locations\n",
      "            are automatically chosen depending on the number of decades in the\n",
      "            plot. See `.Axes.set_yscale` for details.\n",
      "        \n",
      "        nonpositive : {'mask', 'clip'}, default: 'mask'\n",
      "            Non-positive values in y can be masked as invalid, or clipped to a\n",
      "            very small positive number.\n",
      "        \n",
      "        **kwargs\n",
      "            All parameters supported by `.plot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            Objects representing the plotted data.\n",
      "    \n",
      "    set_cmap(cmap)\n",
      "        Set the default colormap, and applies it to the current image if any.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        cmap : `~matplotlib.colors.Colormap` or str\n",
      "            A colormap instance or the name of a registered colormap.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        colormaps\n",
      "        matplotlib.cm.register_cmap\n",
      "        matplotlib.cm.get_cmap\n",
      "    \n",
      "    set_loglevel(*args, **kwargs)\n",
      "        Set Matplotlib's root logger and root logger handler level, creating\n",
      "        the handler if it does not exist yet.\n",
      "        \n",
      "        Typically, one should call ``set_loglevel(\"info\")`` or\n",
      "        ``set_loglevel(\"debug\")`` to get additional debugging information.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        level : {\"notset\", \"debug\", \"info\", \"warning\", \"error\", \"critical\"}\n",
      "            The log level of the handler.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The first time this function is called, an additional handler is attached\n",
      "        to Matplotlib's root handler; this handler is reused every time and this\n",
      "        function simply manipulates the logger and handler's level.\n",
      "    \n",
      "    setp(obj, *args, **kwargs)\n",
      "        Set one or more properties on an `.Artist`, or list allowed values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        obj : `.Artist` or list of `.Artist`\n",
      "            The artist(s) whose properties are being set or queried.  When setting\n",
      "            properties, all artists are affected; when querying the allowed values,\n",
      "            only the first instance in the sequence is queried.\n",
      "        \n",
      "            For example, two lines can be made thicker and red with a single call:\n",
      "        \n",
      "            >>> x = arange(0, 1, 0.01)\n",
      "            >>> lines = plot(x, sin(2*pi*x), x, sin(4*pi*x))\n",
      "            >>> setp(lines, linewidth=2, color='r')\n",
      "        \n",
      "        file : file-like, default: `sys.stdout`\n",
      "            Where `setp` writes its output when asked to list allowed values.\n",
      "        \n",
      "            >>> with open('output.log') as file:\n",
      "            ...     setp(line, file=file)\n",
      "        \n",
      "            The default, ``None``, means `sys.stdout`.\n",
      "        \n",
      "        *args, **kwargs\n",
      "            The properties to set.  The following combinations are supported:\n",
      "        \n",
      "            - Set the linestyle of a line to be dashed:\n",
      "        \n",
      "              >>> line, = plot([1, 2, 3])\n",
      "              >>> setp(line, linestyle='--')\n",
      "        \n",
      "            - Set multiple properties at once:\n",
      "        \n",
      "              >>> setp(line, linewidth=2, color='r')\n",
      "        \n",
      "            - List allowed values for a line's linestyle:\n",
      "        \n",
      "              >>> setp(line, 'linestyle')\n",
      "              linestyle: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "        \n",
      "            - List all properties that can be set, and their allowed values:\n",
      "        \n",
      "              >>> setp(line)\n",
      "              agg_filter: a filter function, ...\n",
      "              [long output listing omitted]\n",
      "        \n",
      "            `setp` also supports MATLAB style string/value pairs.  For example, the\n",
      "            following are equivalent:\n",
      "        \n",
      "            >>> setp(lines, 'linewidth', 2, 'color', 'r')  # MATLAB style\n",
      "            >>> setp(lines, linewidth=2, color='r')        # Python style\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        getp\n",
      "    \n",
      "    show(*args, **kwargs)\n",
      "        Display all open figures.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        block : bool, optional\n",
      "            Whether to wait for all figures to be closed before returning.\n",
      "        \n",
      "            If `True` block and run the GUI main loop until all figure windows\n",
      "            are closed.\n",
      "        \n",
      "            If `False` ensure that all figure windows are displayed and return\n",
      "            immediately.  In this case, you are responsible for ensuring\n",
      "            that the event loop is running to have responsive figures.\n",
      "        \n",
      "            Defaults to True in non-interactive mode and to False in interactive\n",
      "            mode (see `.pyplot.isinteractive`).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ion : Enable interactive mode, which shows / updates the figure after\n",
      "              every plotting command, so that calling ``show()`` is not necessary.\n",
      "        ioff : Disable interactive mode.\n",
      "        savefig : Save the figure to an image file instead of showing it on screen.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        **Saving figures to file and showing a window at the same time**\n",
      "        \n",
      "        If you want an image file as well as a user interface window, use\n",
      "        `.pyplot.savefig` before `.pyplot.show`. At the end of (a blocking)\n",
      "        ``show()`` the figure is closed and thus unregistered from pyplot. Calling\n",
      "        `.pyplot.savefig` afterwards would save a new and thus empty figure. This\n",
      "        limitation of command order does not apply if the show is non-blocking or\n",
      "        if you keep a reference to the figure and use `.Figure.savefig`.\n",
      "        \n",
      "        **Auto-show in jupyter notebooks**\n",
      "        \n",
      "        The jupyter backends (activated via ``%matplotlib inline``,\n",
      "        ``%matplotlib notebook``, or ``%matplotlib widget``), call ``show()`` at\n",
      "        the end of every cell by default. Thus, you usually don't have to call it\n",
      "        explicitly there.\n",
      "    \n",
      "    specgram(x, NFFT=None, Fs=None, Fc=None, detrend=None, window=None, noverlap=None, cmap=None, xextent=None, pad_to=None, sides=None, scale_by_freq=None, mode=None, scale=None, vmin=None, vmax=None, *, data=None, **kwargs)\n",
      "        Plot a spectrogram.\n",
      "        \n",
      "        Compute and plot a spectrogram of data in *x*.  Data are split into\n",
      "        *NFFT* length segments and the spectrum of each section is\n",
      "        computed.  The windowing function *window* is applied to each\n",
      "        segment, and the amount of overlap of each segment is\n",
      "        specified with *noverlap*. The spectrogram is plotted as a colormap\n",
      "        (using imshow).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : 1-D array or sequence\n",
      "            Array or sequence containing the data.\n",
      "        \n",
      "        Fs : float, default: 2\n",
      "            The sampling frequency (samples per time unit).  It is used to calculate\n",
      "            the Fourier frequencies, *freqs*, in cycles per time unit.\n",
      "        \n",
      "        window : callable or ndarray, default: `.window_hanning`\n",
      "            A function or a vector of length *NFFT*.  To create window vectors see\n",
      "            `.window_hanning`, `.window_none`, `numpy.blackman`, `numpy.hamming`,\n",
      "            `numpy.bartlett`, `scipy.signal`, `scipy.signal.get_window`, etc.  If a\n",
      "            function is passed as the argument, it must take a data segment as an\n",
      "            argument and return the windowed version of the segment.\n",
      "        \n",
      "        sides : {'default', 'onesided', 'twosided'}, optional\n",
      "            Which sides of the spectrum to return. 'default' is one-sided for real\n",
      "            data and two-sided for complex data. 'onesided' forces the return of a\n",
      "            one-sided spectrum, while 'twosided' forces two-sided.\n",
      "        \n",
      "        pad_to : int, optional\n",
      "            The number of points to which the data segment is padded when performing\n",
      "            the FFT.  This can be different from *NFFT*, which specifies the number\n",
      "            of data points used.  While not increasing the actual resolution of the\n",
      "            spectrum (the minimum distance between resolvable peaks), this can give\n",
      "            more points in the plot, allowing for more detail. This corresponds to\n",
      "            the *n* parameter in the call to `~numpy.fft.fft`. The default is None,\n",
      "            which sets *pad_to* equal to *NFFT*\n",
      "        \n",
      "        NFFT : int, default: 256\n",
      "            The number of data points used in each block for the FFT.  A power 2 is\n",
      "            most efficient.  This should *NOT* be used to get zero padding, or the\n",
      "            scaling of the result will be incorrect; use *pad_to* for this instead.\n",
      "        \n",
      "        detrend : {'none', 'mean', 'linear'} or callable, default: 'none'\n",
      "            The function applied to each segment before fft-ing, designed to remove\n",
      "            the mean or linear trend.  Unlike in MATLAB, where the *detrend* parameter\n",
      "            is a vector, in Matplotlib it is a function.  The :mod:`~matplotlib.mlab`\n",
      "            module defines `.detrend_none`, `.detrend_mean`, and `.detrend_linear`,\n",
      "            but you can use a custom function as well.  You can also use a string to\n",
      "            choose one of the functions: 'none' calls `.detrend_none`. 'mean' calls\n",
      "            `.detrend_mean`. 'linear' calls `.detrend_linear`.\n",
      "        \n",
      "        scale_by_freq : bool, default: True\n",
      "            Whether the resulting density values should be scaled by the scaling\n",
      "            frequency, which gives density in units of 1/Hz.  This allows for\n",
      "            integration over the returned frequency values.  The default is True for\n",
      "            MATLAB compatibility.\n",
      "        \n",
      "        mode : {'default', 'psd', 'magnitude', 'angle', 'phase'}\n",
      "            What sort of spectrum to use.  Default is 'psd', which takes the\n",
      "            power spectral density.  'magnitude' returns the magnitude\n",
      "            spectrum.  'angle' returns the phase spectrum without unwrapping.\n",
      "            'phase' returns the phase spectrum with unwrapping.\n",
      "        \n",
      "        noverlap : int, default: 128\n",
      "            The number of points of overlap between blocks.\n",
      "        \n",
      "        scale : {'default', 'linear', 'dB'}\n",
      "            The scaling of the values in the *spec*.  'linear' is no scaling.\n",
      "            'dB' returns the values in dB scale.  When *mode* is 'psd',\n",
      "            this is dB power (10 * log10).  Otherwise this is dB amplitude\n",
      "            (20 * log10). 'default' is 'dB' if *mode* is 'psd' or\n",
      "            'magnitude' and 'linear' otherwise.  This must be 'linear'\n",
      "            if *mode* is 'angle' or 'phase'.\n",
      "        \n",
      "        Fc : int, default: 0\n",
      "            The center frequency of *x*, which offsets the x extents of the\n",
      "            plot to reflect the frequency range used when a signal is acquired\n",
      "            and then filtered and downsampled to baseband.\n",
      "        \n",
      "        cmap : `.Colormap`, default: :rc:`image.cmap`\n",
      "        \n",
      "        xextent : *None* or (xmin, xmax)\n",
      "            The image extent along the x-axis. The default sets *xmin* to the\n",
      "            left border of the first bin (*spectrum* column) and *xmax* to the\n",
      "            right border of the last bin. Note that for *noverlap>0* the width\n",
      "            of the bins is smaller than those of the segments.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*\n",
      "        \n",
      "        **kwargs\n",
      "            Additional keyword arguments are passed on to `~.axes.Axes.imshow`\n",
      "            which makes the specgram image. The origin keyword argument\n",
      "            is not supported.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        spectrum : 2D array\n",
      "            Columns are the periodograms of successive segments.\n",
      "        \n",
      "        freqs : 1-D array\n",
      "            The frequencies corresponding to the rows in *spectrum*.\n",
      "        \n",
      "        t : 1-D array\n",
      "            The times corresponding to midpoints of segments (i.e., the columns\n",
      "            in *spectrum*).\n",
      "        \n",
      "        im : `.AxesImage`\n",
      "            The image created by imshow containing the spectrogram.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        psd\n",
      "            Differs in the default overlap; in returning the mean of the\n",
      "            segment periodograms; in not returning times; and in generating a\n",
      "            line plot instead of colormap.\n",
      "        magnitude_spectrum\n",
      "            A single spectrum, similar to having a single segment when *mode*\n",
      "            is 'magnitude'. Plots a line instead of a colormap.\n",
      "        angle_spectrum\n",
      "            A single spectrum, similar to having a single segment when *mode*\n",
      "            is 'angle'. Plots a line instead of a colormap.\n",
      "        phase_spectrum\n",
      "            A single spectrum, similar to having a single segment when *mode*\n",
      "            is 'phase'. Plots a line instead of a colormap.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The parameters *detrend* and *scale_by_freq* do only apply when *mode*\n",
      "        is set to 'psd'.\n",
      "    \n",
      "    spring()\n",
      "        Set the colormap to 'spring'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    spy(Z, precision=0, marker=None, markersize=None, aspect='equal', origin='upper', **kwargs)\n",
      "        Plot the sparsity pattern of a 2D array.\n",
      "        \n",
      "        This visualizes the non-zero values of the array.\n",
      "        \n",
      "        Two plotting styles are available: image and marker. Both\n",
      "        are available for full arrays, but only the marker style\n",
      "        works for `scipy.sparse.spmatrix` instances.\n",
      "        \n",
      "        **Image style**\n",
      "        \n",
      "        If *marker* and *markersize* are *None*, `~.Axes.imshow` is used. Any\n",
      "        extra remaining keyword arguments are passed to this method.\n",
      "        \n",
      "        **Marker style**\n",
      "        \n",
      "        If *Z* is a `scipy.sparse.spmatrix` or *marker* or *markersize* are\n",
      "        *None*, a `.Line2D` object will be returned with the value of marker\n",
      "        determining the marker type, and any remaining keyword arguments\n",
      "        passed to `~.Axes.plot`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        Z : (M, N) array-like\n",
      "            The array to be plotted.\n",
      "        \n",
      "        precision : float or 'present', default: 0\n",
      "            If *precision* is 0, any non-zero value will be plotted. Otherwise,\n",
      "            values of :math:`|Z| > precision` will be plotted.\n",
      "        \n",
      "            For `scipy.sparse.spmatrix` instances, you can also\n",
      "            pass 'present'. In this case any value present in the array\n",
      "            will be plotted, even if it is identically zero.\n",
      "        \n",
      "        aspect : {'equal', 'auto', None} or float, default: 'equal'\n",
      "            The aspect ratio of the Axes.  This parameter is particularly\n",
      "            relevant for images since it determines whether data pixels are\n",
      "            square.\n",
      "        \n",
      "            This parameter is a shortcut for explicitly calling\n",
      "            `.Axes.set_aspect`. See there for further details.\n",
      "        \n",
      "            - 'equal': Ensures an aspect ratio of 1. Pixels will be square.\n",
      "            - 'auto': The Axes is kept fixed and the aspect is adjusted so\n",
      "              that the data fit in the Axes. In general, this will result in\n",
      "              non-square pixels.\n",
      "            - *None*: Use :rc:`image.aspect`.\n",
      "        \n",
      "        origin : {'upper', 'lower'}, default: :rc:`image.origin`\n",
      "            Place the [0, 0] index of the array in the upper left or lower left\n",
      "            corner of the Axes. The convention 'upper' is typically used for\n",
      "            matrices and images.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.image.AxesImage` or `.Line2D`\n",
      "            The return type depends on the plotting style (see above).\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            The supported additional parameters depend on the plotting style.\n",
      "        \n",
      "            For the image style, you can pass the following additional\n",
      "            parameters of `~.Axes.imshow`:\n",
      "        \n",
      "            - *cmap*\n",
      "            - *alpha*\n",
      "            - *url*\n",
      "            - any `.Artist` properties (passed on to the `.AxesImage`)\n",
      "        \n",
      "            For the marker style, you can pass any `.Line2D` property except\n",
      "            for *linestyle*:\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            antialiased or aa: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            color or c: color\n",
      "            dash_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            dash_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            dashes: sequence of floats (on/off ink in points) or (None, None)\n",
      "            data: (2, N) array or two 1D arrays\n",
      "            drawstyle or ds: {'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'}, default: 'default'\n",
      "            figure: `.Figure`\n",
      "            fillstyle: {'full', 'left', 'right', 'bottom', 'top', 'none'}\n",
      "            gapcolor: color or None\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linestyle or ls: {'-', '--', '-.', ':', '', (offset, on-off-seq), ...}\n",
      "            linewidth or lw: float\n",
      "            marker: marker style string, `~.path.Path` or `~.markers.MarkerStyle`\n",
      "            markeredgecolor or mec: color\n",
      "            markeredgewidth or mew: float\n",
      "            markerfacecolor or mfc: color\n",
      "            markerfacecoloralt or mfcalt: color\n",
      "            markersize or ms: float\n",
      "            markevery: None or int or (int, int) or slice or list[int] or float or (float, float) or list[bool]\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: float or callable[[Artist, Event], tuple[bool, dict]]\n",
      "            pickradius: unknown\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            solid_capstyle: `.CapStyle` or {'butt', 'projecting', 'round'}\n",
      "            solid_joinstyle: `.JoinStyle` or {'miter', 'round', 'bevel'}\n",
      "            transform: unknown\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xdata: 1D array\n",
      "            ydata: 1D array\n",
      "            zorder: float\n",
      "    \n",
      "    stackplot(x, *args, labels=(), colors=None, baseline='zero', data=None, **kwargs)\n",
      "        Draw a stacked area plot.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array-like\n",
      "        \n",
      "        y : (M, N) array-like\n",
      "            The data is assumed to be unstacked. Each of the following\n",
      "            calls is legal::\n",
      "        \n",
      "                stackplot(x, y)           # where y has shape (M, N)\n",
      "                stackplot(x, y1, y2, y3)  # where y1, y2, y3, y4 have length N\n",
      "        \n",
      "        baseline : {'zero', 'sym', 'wiggle', 'weighted_wiggle'}\n",
      "            Method used to calculate the baseline:\n",
      "        \n",
      "            - ``'zero'``: Constant zero baseline, i.e. a simple stacked plot.\n",
      "            - ``'sym'``:  Symmetric around zero and is sometimes called\n",
      "              'ThemeRiver'.\n",
      "            - ``'wiggle'``: Minimizes the sum of the squared slopes.\n",
      "            - ``'weighted_wiggle'``: Does the same but weights to account for\n",
      "              size of each layer. It is also called 'Streamgraph'-layout. More\n",
      "              details can be found at http://leebyron.com/streamgraph/.\n",
      "        \n",
      "        labels : list of str, optional\n",
      "            A sequence of labels to assign to each data series. If unspecified,\n",
      "            then no labels will be applied to artists.\n",
      "        \n",
      "        colors : list of color, optional\n",
      "            A sequence of colors to be cycled through and used to color the stacked\n",
      "            areas. The sequence need not be exactly the same length as the number\n",
      "            of provided *y*, in which case the colors will repeat from the\n",
      "            beginning.\n",
      "        \n",
      "            If not specified, the colors from the Axes property cycle will be used.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs\n",
      "            All other keyword arguments are passed to `.Axes.fill_between`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.PolyCollection`\n",
      "            A list of `.PolyCollection` instances, one for each element in the\n",
      "            stacked area plot.\n",
      "    \n",
      "    stairs(values, edges=None, *, orientation='vertical', baseline=0, fill=False, data=None, **kwargs)\n",
      "        A stepwise constant function as a line with bounding edges\n",
      "        or a filled plot.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        values : array-like\n",
      "            The step heights.\n",
      "        \n",
      "        edges : array-like\n",
      "            The edge positions, with ``len(edges) == len(vals) + 1``,\n",
      "            between which the curve takes on vals values.\n",
      "        \n",
      "        orientation : {'vertical', 'horizontal'}, default: 'vertical'\n",
      "            The direction of the steps. Vertical means that *values* are along\n",
      "            the y-axis, and edges are along the x-axis.\n",
      "        \n",
      "        baseline : float, array-like or None, default: 0\n",
      "            The bottom value of the bounding edges or when\n",
      "            ``fill=True``, position of lower edge. If *fill* is\n",
      "            True or an array is passed to *baseline*, a closed\n",
      "            path is drawn.\n",
      "        \n",
      "        fill : bool, default: False\n",
      "            Whether the area under the step curve should be filled.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        StepPatch : `matplotlib.patches.StepPatch`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        **kwargs\n",
      "            `~matplotlib.patches.StepPatch` properties\n",
      "    \n",
      "    stem(*args, linefmt=None, markerfmt=None, basefmt=None, bottom=0, label=None, use_line_collection=<deprecated parameter>, orientation='vertical', data=None)\n",
      "        Create a stem plot.\n",
      "        \n",
      "        A stem plot draws lines perpendicular to a baseline at each location\n",
      "        *locs* from the baseline to *heads*, and places a marker there. For\n",
      "        vertical stem plots (the default), the *locs* are *x* positions, and\n",
      "        the *heads* are *y* values. For horizontal stem plots, the *locs* are\n",
      "        *y* positions, and the *heads* are *x* values.\n",
      "        \n",
      "        Call signature::\n",
      "        \n",
      "          stem([locs,] heads, linefmt=None, markerfmt=None, basefmt=None)\n",
      "        \n",
      "        The *locs*-positions are optional. The formats may be provided either\n",
      "        as positional or as keyword-arguments.\n",
      "        Passing *markerfmt* and *basefmt* positionally is deprecated since\n",
      "        Matplotlib 3.5.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        locs : array-like, default: (0, 1, ..., len(heads) - 1)\n",
      "            For vertical stem plots, the x-positions of the stems.\n",
      "            For horizontal stem plots, the y-positions of the stems.\n",
      "        \n",
      "        heads : array-like\n",
      "            For vertical stem plots, the y-values of the stem heads.\n",
      "            For horizontal stem plots, the x-values of the stem heads.\n",
      "        \n",
      "        linefmt : str, optional\n",
      "            A string defining the color and/or linestyle of the vertical lines:\n",
      "        \n",
      "            =========  =============\n",
      "            Character  Line Style\n",
      "            =========  =============\n",
      "            ``'-'``    solid line\n",
      "            ``'--'``   dashed line\n",
      "            ``'-.'``   dash-dot line\n",
      "            ``':'``    dotted line\n",
      "            =========  =============\n",
      "        \n",
      "            Default: 'C0-', i.e. solid line with the first color of the color\n",
      "            cycle.\n",
      "        \n",
      "            Note: Markers specified through this parameter (e.g. 'x') will be\n",
      "            silently ignored (unless using ``use_line_collection=False``).\n",
      "            Instead, markers should be specified using *markerfmt*.\n",
      "        \n",
      "        markerfmt : str, optional\n",
      "            A string defining the color and/or shape of the markers at the stem\n",
      "            heads. If the marker is not given, use the marker 'o', i.e. filled\n",
      "            circles. If the color is not given, use the color from *linefmt*.\n",
      "        \n",
      "        basefmt : str, default: 'C3-' ('C2-' in classic mode)\n",
      "            A format string defining the properties of the baseline.\n",
      "        \n",
      "        orientation : str, default: 'vertical'\n",
      "            If 'vertical', will produce a plot with stems oriented vertically,\n",
      "            otherwise the stems will be oriented horizontally.\n",
      "        \n",
      "        bottom : float, default: 0\n",
      "            The y/x-position of the baseline (depending on orientation).\n",
      "        \n",
      "        label : str, default: None\n",
      "            The label to use for the stems in legends.\n",
      "        \n",
      "        use_line_collection : bool, default: True\n",
      "            *Deprecated since 3.6*\n",
      "        \n",
      "            If ``True``, store and plot the stem lines as a\n",
      "            `~.collections.LineCollection` instead of individual lines, which\n",
      "            significantly increases performance.  If ``False``, defaults to the\n",
      "            old behavior of using a list of `.Line2D` objects.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, all parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.StemContainer`\n",
      "            The container may be treated like a tuple\n",
      "            (*markerline*, *stemlines*, *baseline*)\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. seealso::\n",
      "            The MATLAB function\n",
      "            `stem <https://www.mathworks.com/help/matlab/ref/stem.html>`_\n",
      "            which inspired this method.\n",
      "    \n",
      "    step(x, y, *args, where='pre', data=None, **kwargs)\n",
      "        Make a step plot.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            step(x, y, [fmt], *, data=None, where='pre', **kwargs)\n",
      "            step(x, y, [fmt], x2, y2, [fmt2], ..., *, where='pre', **kwargs)\n",
      "        \n",
      "        This is just a thin wrapper around `.plot` which changes some\n",
      "        formatting options. Most of the concepts and parameters of plot can be\n",
      "        used here as well.\n",
      "        \n",
      "        .. note::\n",
      "        \n",
      "            This method uses a standard plot with a step drawstyle: The *x*\n",
      "            values are the reference positions and steps extend left/right/both\n",
      "            directions depending on *where*.\n",
      "        \n",
      "            For the common case where you know the values and edges of the\n",
      "            steps, use `~.Axes.stairs` instead.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array-like\n",
      "            1D sequence of x positions. It is assumed, but not checked, that\n",
      "            it is uniformly increasing.\n",
      "        \n",
      "        y : array-like\n",
      "            1D sequence of y levels.\n",
      "        \n",
      "        fmt : str, optional\n",
      "            A format string, e.g. 'g' for a green line. See `.plot` for a more\n",
      "            detailed description.\n",
      "        \n",
      "            Note: While full format strings are accepted, it is recommended to\n",
      "            only specify the color. Line styles are currently ignored (use\n",
      "            the keyword argument *linestyle* instead). Markers are accepted\n",
      "            and plotted on the given positions, however, this is a rarely\n",
      "            needed feature for step plots.\n",
      "        \n",
      "        where : {'pre', 'post', 'mid'}, default: 'pre'\n",
      "            Define where the steps should be placed:\n",
      "        \n",
      "            - 'pre': The y value is continued constantly to the left from\n",
      "              every *x* position, i.e. the interval ``(x[i-1], x[i]]`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'post': The y value is continued constantly to the right from\n",
      "              every *x* position, i.e. the interval ``[x[i], x[i+1])`` has the\n",
      "              value ``y[i]``.\n",
      "            - 'mid': Steps occur half-way between the *x* positions.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            An object with labelled data. If given, provide the label names to\n",
      "            plot in *x* and *y*.\n",
      "        \n",
      "        **kwargs\n",
      "            Additional parameters are the same as those for `.plot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        list of `.Line2D`\n",
      "            Objects representing the plotted data.\n",
      "    \n",
      "    streamplot(x, y, u, v, density=1, linewidth=None, color=None, cmap=None, norm=None, arrowsize=1, arrowstyle='-|>', minlength=0.1, transform=None, zorder=None, start_points=None, maxlength=4.0, integration_direction='both', broken_streamlines=True, *, data=None)\n",
      "        Draw streamlines of a vector flow.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : 1D/2D arrays\n",
      "            Evenly spaced strictly increasing arrays to make a grid.  If 2D, all\n",
      "            rows of *x* must be equal and all columns of *y* must be equal; i.e.,\n",
      "            they must be as if generated by ``np.meshgrid(x_1d, y_1d)``.\n",
      "        u, v : 2D arrays\n",
      "            *x* and *y*-velocities. The number of rows and columns must match\n",
      "            the length of *y* and *x*, respectively.\n",
      "        density : float or (float, float)\n",
      "            Controls the closeness of streamlines. When ``density = 1``, the domain\n",
      "            is divided into a 30x30 grid. *density* linearly scales this grid.\n",
      "            Each cell in the grid can have, at most, one traversing streamline.\n",
      "            For different densities in each direction, use a tuple\n",
      "            (density_x, density_y).\n",
      "        linewidth : float or 2D array\n",
      "            The width of the stream lines. With a 2D array the line width can be\n",
      "            varied across the grid. The array must have the same shape as *u*\n",
      "            and *v*.\n",
      "        color : color or 2D array\n",
      "            The streamline color. If given an array, its values are converted to\n",
      "            colors using *cmap* and *norm*.  The array must have the same shape\n",
      "            as *u* and *v*.\n",
      "        cmap, norm\n",
      "            Data normalization and colormapping parameters for *color*; only used\n",
      "            if *color* is an array of floats. See `~.Axes.imshow` for a detailed\n",
      "            description.\n",
      "        arrowsize : float\n",
      "            Scaling factor for the arrow size.\n",
      "        arrowstyle : str\n",
      "            Arrow style specification.\n",
      "            See `~matplotlib.patches.FancyArrowPatch`.\n",
      "        minlength : float\n",
      "            Minimum length of streamline in axes coordinates.\n",
      "        start_points : Nx2 array\n",
      "            Coordinates of starting points for the streamlines in data coordinates\n",
      "            (the same coordinates as the *x* and *y* arrays).\n",
      "        zorder : int\n",
      "            The zorder of the stream lines and arrows.\n",
      "            Artists with lower zorder values are drawn first.\n",
      "        maxlength : float\n",
      "            Maximum length of streamline in axes coordinates.\n",
      "        integration_direction : {'forward', 'backward', 'both'}, default: 'both'\n",
      "            Integrate the streamline in forward, backward or both directions.\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*, *u*, *v*, *start_points*\n",
      "        broken_streamlines : boolean, default: True\n",
      "            If False, forces streamlines to continue until they\n",
      "            leave the plot domain.  If True, they may be terminated if they\n",
      "            come too close to another streamline.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        StreamplotSet\n",
      "            Container object with attributes\n",
      "        \n",
      "            - ``lines``: `.LineCollection` of streamlines\n",
      "        \n",
      "            - ``arrows``: `.PatchCollection` containing `.FancyArrowPatch`\n",
      "              objects representing the arrows half-way along stream lines.\n",
      "        \n",
      "            This container will probably change in the future to allow changes\n",
      "            to the colormap, alpha, etc. for both lines and arrows, but these\n",
      "            changes should be backward compatible.\n",
      "    \n",
      "    subplot(*args, **kwargs)\n",
      "        Add an Axes to the current figure or retrieve an existing Axes.\n",
      "        \n",
      "        This is a wrapper of `.Figure.add_subplot` which provides additional\n",
      "        behavior when working with the implicit API (see the notes section).\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "           subplot(nrows, ncols, index, **kwargs)\n",
      "           subplot(pos, **kwargs)\n",
      "           subplot(**kwargs)\n",
      "           subplot(ax)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        *args : int, (int, int, *index*), or `.SubplotSpec`, default: (1, 1, 1)\n",
      "            The position of the subplot described by one of\n",
      "        \n",
      "            - Three integers (*nrows*, *ncols*, *index*). The subplot will take the\n",
      "              *index* position on a grid with *nrows* rows and *ncols* columns.\n",
      "              *index* starts at 1 in the upper left corner and increases to the\n",
      "              right. *index* can also be a two-tuple specifying the (*first*,\n",
      "              *last*) indices (1-based, and including *last*) of the subplot, e.g.,\n",
      "              ``fig.add_subplot(3, 1, (1, 2))`` makes a subplot that spans the\n",
      "              upper 2/3 of the figure.\n",
      "            - A 3-digit integer. The digits are interpreted as if given separately\n",
      "              as three single-digit integers, i.e. ``fig.add_subplot(235)`` is the\n",
      "              same as ``fig.add_subplot(2, 3, 5)``. Note that this can only be used\n",
      "              if there are no more than 9 subplots.\n",
      "            - A `.SubplotSpec`.\n",
      "        \n",
      "        projection : {None, 'aitoff', 'hammer', 'lambert', 'mollweide', 'polar', 'rectilinear', str}, optional\n",
      "            The projection type of the subplot (`~.axes.Axes`). *str* is the name\n",
      "            of a custom projection, see `~matplotlib.projections`. The default\n",
      "            None results in a 'rectilinear' projection.\n",
      "        \n",
      "        polar : bool, default: False\n",
      "            If True, equivalent to projection='polar'.\n",
      "        \n",
      "        sharex, sharey : `~.axes.Axes`, optional\n",
      "            Share the x or y `~matplotlib.axis` with sharex and/or sharey. The\n",
      "            axis will have the same limits, ticks, and scale as the axis of the\n",
      "            shared axes.\n",
      "        \n",
      "        label : str\n",
      "            A label for the returned axes.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.axes.SubplotBase`, or another subclass of `~.axes.Axes`\n",
      "        \n",
      "            The axes of the subplot. The returned axes base class depends on\n",
      "            the projection used. It is `~.axes.Axes` if rectilinear projection\n",
      "            is used and `.projections.polar.PolarAxes` if polar projection\n",
      "            is used. The returned axes is then a subplot subclass of the\n",
      "            base class.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            This method also takes the keyword arguments for the returned axes\n",
      "            base class; except for the *figure* argument. The keyword arguments\n",
      "            for the rectilinear base class `~.axes.Axes` can be found in\n",
      "            the following table but there might also be other keyword\n",
      "            arguments if another projection is used.\n",
      "        \n",
      "            Properties:\n",
      "            adjustable: {'box', 'datalim'}\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            anchor: (float, float) or {'C', 'SW', 'S', 'SE', 'E', 'NE', ...}\n",
      "            animated: bool\n",
      "            aspect: {'auto', 'equal'} or float\n",
      "            autoscale_on: bool\n",
      "            autoscalex_on: unknown\n",
      "            autoscaley_on: unknown\n",
      "            axes_locator: Callable[[Axes, Renderer], Bbox]\n",
      "            axisbelow: bool or 'line'\n",
      "            box_aspect: float or None\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            facecolor or fc: color\n",
      "            figure: `.Figure`\n",
      "            frame_on: bool\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            mouseover: bool\n",
      "            navigate: bool\n",
      "            navigate_mode: unknown\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            position: [left, bottom, width, height] or `~matplotlib.transforms.Bbox`\n",
      "            prop_cycle: unknown\n",
      "            rasterization_zorder: float or None\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            title: str\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            xbound: unknown\n",
      "            xlabel: str\n",
      "            xlim: (bottom: float, top: float)\n",
      "            xmargin: float greater than -0.5\n",
      "            xscale: unknown\n",
      "            xticklabels: unknown\n",
      "            xticks: unknown\n",
      "            ybound: unknown\n",
      "            ylabel: str\n",
      "            ylim: (bottom: float, top: float)\n",
      "            ymargin: float greater than -0.5\n",
      "            yscale: unknown\n",
      "            yticklabels: unknown\n",
      "            yticks: unknown\n",
      "            zorder: float\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Creating a new Axes will delete any preexisting Axes that\n",
      "        overlaps with it beyond sharing a boundary::\n",
      "        \n",
      "            import matplotlib.pyplot as plt\n",
      "            # plot a line, implicitly creating a subplot(111)\n",
      "            plt.plot([1, 2, 3])\n",
      "            # now create a subplot which represents the top plot of a grid\n",
      "            # with 2 rows and 1 column. Since this subplot will overlap the\n",
      "            # first, the plot (and its axes) previously created, will be removed\n",
      "            plt.subplot(211)\n",
      "        \n",
      "        If you do not want this behavior, use the `.Figure.add_subplot` method\n",
      "        or the `.pyplot.axes` function instead.\n",
      "        \n",
      "        If no *kwargs* are passed and there exists an Axes in the location\n",
      "        specified by *args* then that Axes will be returned rather than a new\n",
      "        Axes being created.\n",
      "        \n",
      "        If *kwargs* are passed and there exists an Axes in the location\n",
      "        specified by *args*, the projection type is the same, and the\n",
      "        *kwargs* match with the existing Axes, then the existing Axes is\n",
      "        returned.  Otherwise a new Axes is created with the specified\n",
      "        parameters.  We save a reference to the *kwargs* which we use\n",
      "        for this comparison.  If any of the values in *kwargs* are\n",
      "        mutable we will not detect the case where they are mutated.\n",
      "        In these cases we suggest using `.Figure.add_subplot` and the\n",
      "        explicit Axes API rather than the implicit pyplot API.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Figure.add_subplot\n",
      "        .pyplot.subplots\n",
      "        .pyplot.axes\n",
      "        .Figure.subplots\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            plt.subplot(221)\n",
      "        \n",
      "            # equivalent but more general\n",
      "            ax1 = plt.subplot(2, 2, 1)\n",
      "        \n",
      "            # add a subplot with no frame\n",
      "            ax2 = plt.subplot(222, frameon=False)\n",
      "        \n",
      "            # add a polar subplot\n",
      "            plt.subplot(223, projection='polar')\n",
      "        \n",
      "            # add a red subplot that shares the x-axis with ax1\n",
      "            plt.subplot(224, sharex=ax1, facecolor='red')\n",
      "        \n",
      "            # delete ax2 from the figure\n",
      "            plt.delaxes(ax2)\n",
      "        \n",
      "            # add ax2 to the figure again\n",
      "            plt.subplot(ax2)\n",
      "        \n",
      "            # make the first axes \"current\" again\n",
      "            plt.subplot(221)\n",
      "    \n",
      "    subplot2grid(shape, loc, rowspan=1, colspan=1, fig=None, **kwargs)\n",
      "        Create a subplot at a specific location inside a regular grid.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        shape : (int, int)\n",
      "            Number of rows and of columns of the grid in which to place axis.\n",
      "        loc : (int, int)\n",
      "            Row number and column number of the axis location within the grid.\n",
      "        rowspan : int, default: 1\n",
      "            Number of rows for the axis to span downwards.\n",
      "        colspan : int, default: 1\n",
      "            Number of columns for the axis to span to the right.\n",
      "        fig : `.Figure`, optional\n",
      "            Figure to place the subplot in. Defaults to the current figure.\n",
      "        **kwargs\n",
      "            Additional keyword arguments are handed to `~.Figure.add_subplot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.axes.SubplotBase`, or another subclass of `~.axes.Axes`\n",
      "        \n",
      "            The axes of the subplot.  The returned axes base class depends on the\n",
      "            projection used.  It is `~.axes.Axes` if rectilinear projection is used\n",
      "            and `.projections.polar.PolarAxes` if polar projection is used.  The\n",
      "            returned axes is then a subplot subclass of the base class.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The following call ::\n",
      "        \n",
      "            ax = subplot2grid((nrows, ncols), (row, col), rowspan, colspan)\n",
      "        \n",
      "        is identical to ::\n",
      "        \n",
      "            fig = gcf()\n",
      "            gs = fig.add_gridspec(nrows, ncols)\n",
      "            ax = fig.add_subplot(gs[row:row+rowspan, col:col+colspan])\n",
      "    \n",
      "    subplot_mosaic(mosaic, *, sharex=False, sharey=False, width_ratios=None, height_ratios=None, empty_sentinel='.', subplot_kw=None, gridspec_kw=None, **fig_kw)\n",
      "        Build a layout of Axes based on ASCII art or nested lists.\n",
      "        \n",
      "        This is a helper function to build complex GridSpec layouts visually.\n",
      "        \n",
      "        .. note::\n",
      "        \n",
      "           This API is provisional and may be revised in the future based on\n",
      "           early user feedback.\n",
      "        \n",
      "        See :doc:`/tutorials/provisional/mosaic`\n",
      "        for an example and full API documentation\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        mosaic : list of list of {hashable or nested} or str\n",
      "        \n",
      "            A visual layout of how you want your Axes to be arranged\n",
      "            labeled as strings.  For example ::\n",
      "        \n",
      "               x = [['A panel', 'A panel', 'edge'],\n",
      "                    ['C panel', '.',       'edge']]\n",
      "        \n",
      "            produces 4 axes:\n",
      "        \n",
      "            - 'A panel' which is 1 row high and spans the first two columns\n",
      "            - 'edge' which is 2 rows high and is on the right edge\n",
      "            - 'C panel' which in 1 row and 1 column wide in the bottom left\n",
      "            - a blank space 1 row and 1 column wide in the bottom center\n",
      "        \n",
      "            Any of the entries in the layout can be a list of lists\n",
      "            of the same form to create nested layouts.\n",
      "        \n",
      "            If input is a str, then it must be of the form ::\n",
      "        \n",
      "              '''\n",
      "              AAE\n",
      "              C.E\n",
      "              '''\n",
      "        \n",
      "            where each character is a column and each line is a row.\n",
      "            This only allows only single character Axes labels and does\n",
      "            not allow nesting but is very terse.\n",
      "        \n",
      "        sharex, sharey : bool, default: False\n",
      "            If True, the x-axis (*sharex*) or y-axis (*sharey*) will be shared\n",
      "            among all subplots.  In that case, tick label visibility and axis units\n",
      "            behave as for `subplots`.  If False, each subplot's x- or y-axis will\n",
      "            be independent.\n",
      "        \n",
      "        width_ratios : array-like of length *ncols*, optional\n",
      "            Defines the relative widths of the columns. Each column gets a\n",
      "            relative width of ``width_ratios[i] / sum(width_ratios)``.\n",
      "            If not given, all columns will have the same width.  Convenience\n",
      "            for ``gridspec_kw={'width_ratios': [...]}``.\n",
      "        \n",
      "        height_ratios : array-like of length *nrows*, optional\n",
      "            Defines the relative heights of the rows. Each row gets a\n",
      "            relative height of ``height_ratios[i] / sum(height_ratios)``.\n",
      "            If not given, all rows will have the same height. Convenience\n",
      "            for ``gridspec_kw={'height_ratios': [...]}``.\n",
      "        \n",
      "        empty_sentinel : object, optional\n",
      "            Entry in the layout to mean \"leave this space empty\".  Defaults\n",
      "            to ``'.'``. Note, if *layout* is a string, it is processed via\n",
      "            `inspect.cleandoc` to remove leading white space, which may\n",
      "            interfere with using white-space as the empty sentinel.\n",
      "        \n",
      "        subplot_kw : dict, optional\n",
      "            Dictionary with keywords passed to the `.Figure.add_subplot` call\n",
      "            used to create each subplot.\n",
      "        \n",
      "        gridspec_kw : dict, optional\n",
      "            Dictionary with keywords passed to the `.GridSpec` constructor used\n",
      "            to create the grid the subplots are placed on.\n",
      "        \n",
      "        **fig_kw\n",
      "            All additional keyword arguments are passed to the\n",
      "            `.pyplot.figure` call.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        fig : `.Figure`\n",
      "           The new figure\n",
      "        \n",
      "        dict[label, Axes]\n",
      "           A dictionary mapping the labels to the Axes objects.  The order of\n",
      "           the axes is left-to-right and top-to-bottom of their position in the\n",
      "           total layout.\n",
      "    \n",
      "    subplot_tool(targetfig=None)\n",
      "        Launch a subplot tool window for a figure.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `matplotlib.widgets.SubplotTool`\n",
      "    \n",
      "    subplots(nrows=1, ncols=1, *, sharex=False, sharey=False, squeeze=True, width_ratios=None, height_ratios=None, subplot_kw=None, gridspec_kw=None, **fig_kw)\n",
      "        Create a figure and a set of subplots.\n",
      "        \n",
      "        This utility wrapper makes it convenient to create common layouts of\n",
      "        subplots, including the enclosing figure object, in a single call.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        nrows, ncols : int, default: 1\n",
      "            Number of rows/columns of the subplot grid.\n",
      "        \n",
      "        sharex, sharey : bool or {'none', 'all', 'row', 'col'}, default: False\n",
      "            Controls sharing of properties among x (*sharex*) or y (*sharey*)\n",
      "            axes:\n",
      "        \n",
      "            - True or 'all': x- or y-axis will be shared among all subplots.\n",
      "            - False or 'none': each subplot x- or y-axis will be independent.\n",
      "            - 'row': each subplot row will share an x- or y-axis.\n",
      "            - 'col': each subplot column will share an x- or y-axis.\n",
      "        \n",
      "            When subplots have a shared x-axis along a column, only the x tick\n",
      "            labels of the bottom subplot are created. Similarly, when subplots\n",
      "            have a shared y-axis along a row, only the y tick labels of the first\n",
      "            column subplot are created. To later turn other subplots' ticklabels\n",
      "            on, use `~matplotlib.axes.Axes.tick_params`.\n",
      "        \n",
      "            When subplots have a shared axis that has units, calling\n",
      "            `~matplotlib.axis.Axis.set_units` will update each axis with the\n",
      "            new units.\n",
      "        \n",
      "        squeeze : bool, default: True\n",
      "            - If True, extra dimensions are squeezed out from the returned\n",
      "              array of `~matplotlib.axes.Axes`:\n",
      "        \n",
      "              - if only one subplot is constructed (nrows=ncols=1), the\n",
      "                resulting single Axes object is returned as a scalar.\n",
      "              - for Nx1 or 1xM subplots, the returned object is a 1D numpy\n",
      "                object array of Axes objects.\n",
      "              - for NxM, subplots with N>1 and M>1 are returned as a 2D array.\n",
      "        \n",
      "            - If False, no squeezing at all is done: the returned Axes object is\n",
      "              always a 2D array containing Axes instances, even if it ends up\n",
      "              being 1x1.\n",
      "        \n",
      "        width_ratios : array-like of length *ncols*, optional\n",
      "            Defines the relative widths of the columns. Each column gets a\n",
      "            relative width of ``width_ratios[i] / sum(width_ratios)``.\n",
      "            If not given, all columns will have the same width.  Equivalent\n",
      "            to ``gridspec_kw={'width_ratios': [...]}``.\n",
      "        \n",
      "        height_ratios : array-like of length *nrows*, optional\n",
      "            Defines the relative heights of the rows. Each row gets a\n",
      "            relative height of ``height_ratios[i] / sum(height_ratios)``.\n",
      "            If not given, all rows will have the same height. Convenience\n",
      "            for ``gridspec_kw={'height_ratios': [...]}``.\n",
      "        \n",
      "        subplot_kw : dict, optional\n",
      "            Dict with keywords passed to the\n",
      "            `~matplotlib.figure.Figure.add_subplot` call used to create each\n",
      "            subplot.\n",
      "        \n",
      "        gridspec_kw : dict, optional\n",
      "            Dict with keywords passed to the `~matplotlib.gridspec.GridSpec`\n",
      "            constructor used to create the grid the subplots are placed on.\n",
      "        \n",
      "        **fig_kw\n",
      "            All additional keyword arguments are passed to the\n",
      "            `.pyplot.figure` call.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        fig : `.Figure`\n",
      "        \n",
      "        ax : `~.axes.Axes` or array of Axes\n",
      "            *ax* can be either a single `~.axes.Axes` object, or an array of Axes\n",
      "            objects if more than one subplot was created.  The dimensions of the\n",
      "            resulting array can be controlled with the squeeze keyword, see above.\n",
      "        \n",
      "            Typical idioms for handling the return value are::\n",
      "        \n",
      "                # using the variable ax for single a Axes\n",
      "                fig, ax = plt.subplots()\n",
      "        \n",
      "                # using the variable axs for multiple Axes\n",
      "                fig, axs = plt.subplots(2, 2)\n",
      "        \n",
      "                # using tuple unpacking for multiple Axes\n",
      "                fig, (ax1, ax2) = plt.subplots(1, 2)\n",
      "                fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2)\n",
      "        \n",
      "            The names ``ax`` and pluralized ``axs`` are preferred over ``axes``\n",
      "            because for the latter it's not clear if it refers to a single\n",
      "            `~.axes.Axes` instance or a collection of these.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .pyplot.figure\n",
      "        .pyplot.subplot\n",
      "        .pyplot.axes\n",
      "        .Figure.subplots\n",
      "        .Figure.add_subplot\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            # First create some toy data:\n",
      "            x = np.linspace(0, 2*np.pi, 400)\n",
      "            y = np.sin(x**2)\n",
      "        \n",
      "            # Create just a figure and only one subplot\n",
      "            fig, ax = plt.subplots()\n",
      "            ax.plot(x, y)\n",
      "            ax.set_title('Simple plot')\n",
      "        \n",
      "            # Create two subplots and unpack the output array immediately\n",
      "            f, (ax1, ax2) = plt.subplots(1, 2, sharey=True)\n",
      "            ax1.plot(x, y)\n",
      "            ax1.set_title('Sharing Y axis')\n",
      "            ax2.scatter(x, y)\n",
      "        \n",
      "            # Create four polar axes and access them through the returned array\n",
      "            fig, axs = plt.subplots(2, 2, subplot_kw=dict(projection=\"polar\"))\n",
      "            axs[0, 0].plot(x, y)\n",
      "            axs[1, 1].scatter(x, y)\n",
      "        \n",
      "            # Share a X axis with each column of subplots\n",
      "            plt.subplots(2, 2, sharex='col')\n",
      "        \n",
      "            # Share a Y axis with each row of subplots\n",
      "            plt.subplots(2, 2, sharey='row')\n",
      "        \n",
      "            # Share both X and Y axes with all subplots\n",
      "            plt.subplots(2, 2, sharex='all', sharey='all')\n",
      "        \n",
      "            # Note that this is the same as\n",
      "            plt.subplots(2, 2, sharex=True, sharey=True)\n",
      "        \n",
      "            # Create figure number 10 with a single subplot\n",
      "            # and clears it if it already exists.\n",
      "            fig, ax = plt.subplots(num=10, clear=True)\n",
      "    \n",
      "    subplots_adjust(left=None, bottom=None, right=None, top=None, wspace=None, hspace=None)\n",
      "        Adjust the subplot layout parameters.\n",
      "        \n",
      "        Unset parameters are left unmodified; initial values are given by\n",
      "        :rc:`figure.subplot.[name]`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        left : float, optional\n",
      "            The position of the left edge of the subplots,\n",
      "            as a fraction of the figure width.\n",
      "        right : float, optional\n",
      "            The position of the right edge of the subplots,\n",
      "            as a fraction of the figure width.\n",
      "        bottom : float, optional\n",
      "            The position of the bottom edge of the subplots,\n",
      "            as a fraction of the figure height.\n",
      "        top : float, optional\n",
      "            The position of the top edge of the subplots,\n",
      "            as a fraction of the figure height.\n",
      "        wspace : float, optional\n",
      "            The width of the padding between subplots,\n",
      "            as a fraction of the average Axes width.\n",
      "        hspace : float, optional\n",
      "            The height of the padding between subplots,\n",
      "            as a fraction of the average Axes height.\n",
      "    \n",
      "    summer()\n",
      "        Set the colormap to 'summer'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    suptitle(t, **kwargs)\n",
      "        Add a centered suptitle to the figure.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        t : str\n",
      "            The suptitle text.\n",
      "        x : float, default: 0.5\n",
      "            The x location of the text in figure coordinates.\n",
      "        y : float, default: 0.98\n",
      "            The y location of the text in figure coordinates.\n",
      "        horizontalalignment, ha : {'center', 'left', 'right'}, default: center\n",
      "            The horizontal alignment of the text relative to (*x*, *y*).\n",
      "        verticalalignment, va : {'top', 'center', 'bottom', 'baseline'}, default: top\n",
      "            The vertical alignment of the text relative to (*x*, *y*).\n",
      "        fontsize, size : default: :rc:`figure.titlesize`\n",
      "            The font size of the text. See `.Text.set_size` for possible\n",
      "            values.\n",
      "        fontweight, weight : default: :rc:`figure.titleweight`\n",
      "            The font weight of the text. See `.Text.set_weight` for possible\n",
      "            values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        text\n",
      "            The `.Text` instance of the suptitle.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        fontproperties : None or dict, optional\n",
      "            A dict of font properties. If *fontproperties* is given the\n",
      "            default values for font size and weight are taken from the\n",
      "            `.FontProperties` defaults. :rc:`figure.titlesize` and\n",
      "            :rc:`figure.titleweight` are ignored in this case.\n",
      "        \n",
      "        **kwargs\n",
      "            Additional kwargs are `matplotlib.text.Text` properties.\n",
      "    \n",
      "    switch_backend(newbackend)\n",
      "        Close all open figures and set the Matplotlib backend.\n",
      "        \n",
      "        The argument is case-insensitive.  Switching to an interactive backend is\n",
      "        possible only if no event loop for another interactive backend has started.\n",
      "        Switching to and from non-interactive backends is always possible.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        newbackend : str\n",
      "            The name of the backend to use.\n",
      "    \n",
      "    table(cellText=None, cellColours=None, cellLoc='right', colWidths=None, rowLabels=None, rowColours=None, rowLoc='left', colLabels=None, colColours=None, colLoc='center', loc='bottom', bbox=None, edges='closed', **kwargs)\n",
      "        Add a table to an `~.axes.Axes`.\n",
      "        \n",
      "        At least one of *cellText* or *cellColours* must be specified. These\n",
      "        parameters must be 2D lists, in which the outer lists define the rows and\n",
      "        the inner list define the column values per row. Each row must have the\n",
      "        same number of elements.\n",
      "        \n",
      "        The table can optionally have row and column headers, which are configured\n",
      "        using *rowLabels*, *rowColours*, *rowLoc* and *colLabels*, *colColours*,\n",
      "        *colLoc* respectively.\n",
      "        \n",
      "        For finer grained control over tables, use the `.Table` class and add it to\n",
      "        the axes with `.Axes.add_table`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        cellText : 2D list of str, optional\n",
      "            The texts to place into the table cells.\n",
      "        \n",
      "            *Note*: Line breaks in the strings are currently not accounted for and\n",
      "            will result in the text exceeding the cell boundaries.\n",
      "        \n",
      "        cellColours : 2D list of colors, optional\n",
      "            The background colors of the cells.\n",
      "        \n",
      "        cellLoc : {'left', 'center', 'right'}, default: 'right'\n",
      "            The alignment of the text within the cells.\n",
      "        \n",
      "        colWidths : list of float, optional\n",
      "            The column widths in units of the axes. If not given, all columns will\n",
      "            have a width of *1 / ncols*.\n",
      "        \n",
      "        rowLabels : list of str, optional\n",
      "            The text of the row header cells.\n",
      "        \n",
      "        rowColours : list of colors, optional\n",
      "            The colors of the row header cells.\n",
      "        \n",
      "        rowLoc : {'left', 'center', 'right'}, default: 'left'\n",
      "            The text alignment of the row header cells.\n",
      "        \n",
      "        colLabels : list of str, optional\n",
      "            The text of the column header cells.\n",
      "        \n",
      "        colColours : list of colors, optional\n",
      "            The colors of the column header cells.\n",
      "        \n",
      "        colLoc : {'left', 'center', 'right'}, default: 'left'\n",
      "            The text alignment of the column header cells.\n",
      "        \n",
      "        loc : str, optional\n",
      "            The position of the cell with respect to *ax*. This must be one of\n",
      "            the `~.Table.codes`.\n",
      "        \n",
      "        bbox : `.Bbox`, optional\n",
      "            A bounding box to draw the table into. If this is not *None*, this\n",
      "            overrides *loc*.\n",
      "        \n",
      "        edges : substring of 'BRTL' or {'open', 'closed', 'horizontal', 'vertical'}\n",
      "            The cell edges to be drawn with a line. See also\n",
      "            `~.Cell.visible_edges`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.table.Table`\n",
      "            The created table.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            `.Table` properties.\n",
      "        \n",
      "        Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            clip_box: `.Bbox`\n",
      "            clip_on: bool\n",
      "            clip_path: Patch or (Path, Transform) or None\n",
      "            figure: `.Figure`\n",
      "            fontsize: float\n",
      "            gid: str\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            mouseover: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            rasterized: bool\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            transform: `.Transform`\n",
      "            url: str\n",
      "            visible: bool\n",
      "            zorder: float\n",
      "    \n",
      "    text(x, y, s, fontdict=None, **kwargs)\n",
      "        Add text to the Axes.\n",
      "        \n",
      "        Add the text *s* to the Axes at location *x*, *y* in data coordinates.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : float\n",
      "            The position to place the text. By default, this is in data\n",
      "            coordinates. The coordinate system can be changed using the\n",
      "            *transform* parameter.\n",
      "        \n",
      "        s : str\n",
      "            The text.\n",
      "        \n",
      "        fontdict : dict, default: None\n",
      "            A dictionary to override the default text properties. If fontdict\n",
      "            is None, the defaults are determined by `.rcParams`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.Text`\n",
      "            The created `.Text` instance.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `~matplotlib.text.Text` properties.\n",
      "            Other miscellaneous text parameters.\n",
      "        \n",
      "            Properties:\n",
      "            agg_filter: a filter function, which takes a (m, n, 3) float array and a dpi value, and returns a (m, n, 3) array and two offsets from the bottom left corner of the image\n",
      "            alpha: scalar or None\n",
      "            animated: bool\n",
      "            backgroundcolor: color\n",
      "            bbox: dict with properties for `.patches.FancyBboxPatch`\n",
      "            clip_box: unknown\n",
      "            clip_on: unknown\n",
      "            clip_path: unknown\n",
      "            color or c: color\n",
      "            figure: `.Figure`\n",
      "            fontfamily or family: {FONTNAME, 'serif', 'sans-serif', 'cursive', 'fantasy', 'monospace'}\n",
      "            fontproperties or font or font_properties: `.font_manager.FontProperties` or `str` or `pathlib.Path`\n",
      "            fontsize or size: float or {'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large'}\n",
      "            fontstretch or stretch: {a numeric value in range 0-1000, 'ultra-condensed', 'extra-condensed', 'condensed', 'semi-condensed', 'normal', 'semi-expanded', 'expanded', 'extra-expanded', 'ultra-expanded'}\n",
      "            fontstyle or style: {'normal', 'italic', 'oblique'}\n",
      "            fontvariant or variant: {'normal', 'small-caps'}\n",
      "            fontweight or weight: {a numeric value in range 0-1000, 'ultralight', 'light', 'normal', 'regular', 'book', 'medium', 'roman', 'semibold', 'demibold', 'demi', 'bold', 'heavy', 'extra bold', 'black'}\n",
      "            gid: str\n",
      "            horizontalalignment or ha: {'left', 'center', 'right'}\n",
      "            in_layout: bool\n",
      "            label: object\n",
      "            linespacing: float (multiple of font size)\n",
      "            math_fontfamily: str\n",
      "            mouseover: bool\n",
      "            multialignment or ma: {'left', 'right', 'center'}\n",
      "            parse_math: bool\n",
      "            path_effects: `.AbstractPathEffect`\n",
      "            picker: None or bool or float or callable\n",
      "            position: (float, float)\n",
      "            rasterized: bool\n",
      "            rotation: float or {'vertical', 'horizontal'}\n",
      "            rotation_mode: {None, 'default', 'anchor'}\n",
      "            sketch_params: (scale: float, length: float, randomness: float)\n",
      "            snap: bool or None\n",
      "            text: object\n",
      "            transform: `.Transform`\n",
      "            transform_rotates_text: bool\n",
      "            url: str\n",
      "            usetex: bool or None\n",
      "            verticalalignment or va: {'bottom', 'baseline', 'center', 'center_baseline', 'top'}\n",
      "            visible: bool\n",
      "            wrap: bool\n",
      "            x: float\n",
      "            y: float\n",
      "            zorder: float\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Individual keyword arguments can be used to override any given\n",
      "        parameter::\n",
      "        \n",
      "            >>> text(x, y, s, fontsize=12)\n",
      "        \n",
      "        The default transform specifies that text is in data coords,\n",
      "        alternatively, you can specify text in axis coords ((0, 0) is\n",
      "        lower-left and (1, 1) is upper-right).  The example below places\n",
      "        text in the center of the Axes::\n",
      "        \n",
      "            >>> text(0.5, 0.5, 'matplotlib', horizontalalignment='center',\n",
      "            ...      verticalalignment='center', transform=ax.transAxes)\n",
      "        \n",
      "        You can put a rectangular box around the text instance (e.g., to\n",
      "        set a background color) by using the keyword *bbox*.  *bbox* is\n",
      "        a dictionary of `~matplotlib.patches.Rectangle`\n",
      "        properties.  For example::\n",
      "        \n",
      "            >>> text(x, y, s, bbox=dict(facecolor='red', alpha=0.5))\n",
      "    \n",
      "    thetagrids(angles=None, labels=None, fmt=None, **kwargs)\n",
      "        Get or set the theta gridlines on the current polar plot.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "         lines, labels = thetagrids()\n",
      "         lines, labels = thetagrids(angles, labels=None, fmt=None, **kwargs)\n",
      "        \n",
      "        When called with no arguments, `.thetagrids` simply returns the tuple\n",
      "        (*lines*, *labels*). When called with arguments, the labels will\n",
      "        appear at the specified angles.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        angles : tuple with floats, degrees\n",
      "            The angles of the theta gridlines.\n",
      "        \n",
      "        labels : tuple with strings or None\n",
      "            The labels to use at each radial gridline. The\n",
      "            `.projections.polar.ThetaFormatter` will be used if None.\n",
      "        \n",
      "        fmt : str or None\n",
      "            Format string used in `matplotlib.ticker.FormatStrFormatter`.\n",
      "            For example '%f'. Note that the angle in radians will be used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lines : list of `.lines.Line2D`\n",
      "            The theta gridlines.\n",
      "        \n",
      "        labels : list of `.text.Text`\n",
      "            The tick labels.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs\n",
      "            *kwargs* are optional `.Text` properties for the labels.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .pyplot.rgrids\n",
      "        .projections.polar.PolarAxes.set_thetagrids\n",
      "        .Axis.get_gridlines\n",
      "        .Axis.get_ticklabels\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "          # set the locations of the angular gridlines\n",
      "          lines, labels = thetagrids(range(45, 360, 90))\n",
      "        \n",
      "          # set the locations and labels of the angular gridlines\n",
      "          lines, labels = thetagrids(range(45, 360, 90), ('NE', 'NW', 'SW', 'SE'))\n",
      "    \n",
      "    tick_params(axis='both', **kwargs)\n",
      "        Change the appearance of ticks, tick labels, and gridlines.\n",
      "        \n",
      "        Tick properties that are not explicitly set using the keyword\n",
      "        arguments remain unchanged unless *reset* is True.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        axis : {'x', 'y', 'both'}, default: 'both'\n",
      "            The axis to which the parameters are applied.\n",
      "        which : {'major', 'minor', 'both'}, default: 'major'\n",
      "            The group of ticks to which the parameters are applied.\n",
      "        reset : bool, default: False\n",
      "            Whether to reset the ticks to defaults before updating them.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        direction : {'in', 'out', 'inout'}\n",
      "            Puts ticks inside the Axes, outside the Axes, or both.\n",
      "        length : float\n",
      "            Tick length in points.\n",
      "        width : float\n",
      "            Tick width in points.\n",
      "        color : color\n",
      "            Tick color.\n",
      "        pad : float\n",
      "            Distance in points between tick and label.\n",
      "        labelsize : float or str\n",
      "            Tick label font size in points or as a string (e.g., 'large').\n",
      "        labelcolor : color\n",
      "            Tick label color.\n",
      "        colors : color\n",
      "            Tick color and label color.\n",
      "        zorder : float\n",
      "            Tick and label zorder.\n",
      "        bottom, top, left, right : bool\n",
      "            Whether to draw the respective ticks.\n",
      "        labelbottom, labeltop, labelleft, labelright : bool\n",
      "            Whether to draw the respective tick labels.\n",
      "        labelrotation : float\n",
      "            Tick label rotation\n",
      "        grid_color : color\n",
      "            Gridline color.\n",
      "        grid_alpha : float\n",
      "            Transparency of gridlines: 0 (transparent) to 1 (opaque).\n",
      "        grid_linewidth : float\n",
      "            Width of gridlines in points.\n",
      "        grid_linestyle : str\n",
      "            Any valid `.Line2D` line style spec.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        ::\n",
      "        \n",
      "            ax.tick_params(direction='out', length=6, width=2, colors='r',\n",
      "                           grid_color='r', grid_alpha=0.5)\n",
      "        \n",
      "        This will make all major ticks be red, pointing out of the box,\n",
      "        and with dimensions 6 points by 2 points.  Tick labels will\n",
      "        also be red.  Gridlines will be red and translucent.\n",
      "    \n",
      "    ticklabel_format(*, axis='both', style='', scilimits=None, useOffset=None, useLocale=None, useMathText=None)\n",
      "        Configure the `.ScalarFormatter` used by default for linear Axes.\n",
      "        \n",
      "        If a parameter is not set, the corresponding property of the formatter\n",
      "        is left unchanged.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        axis : {'x', 'y', 'both'}, default: 'both'\n",
      "            The axis to configure.  Only major ticks are affected.\n",
      "        \n",
      "        style : {'sci', 'scientific', 'plain'}\n",
      "            Whether to use scientific notation.\n",
      "            The formatter default is to use scientific notation.\n",
      "        \n",
      "        scilimits : pair of ints (m, n)\n",
      "            Scientific notation is used only for numbers outside the range\n",
      "            10\\ :sup:`m` to 10\\ :sup:`n` (and only if the formatter is\n",
      "            configured to use scientific notation at all).  Use (0, 0) to\n",
      "            include all numbers.  Use (m, m) where m != 0 to fix the order of\n",
      "            magnitude to 10\\ :sup:`m`.\n",
      "            The formatter default is :rc:`axes.formatter.limits`.\n",
      "        \n",
      "        useOffset : bool or float\n",
      "            If True, the offset is calculated as needed.\n",
      "            If False, no offset is used.\n",
      "            If a numeric value, it sets the offset.\n",
      "            The formatter default is :rc:`axes.formatter.useoffset`.\n",
      "        \n",
      "        useLocale : bool\n",
      "            Whether to format the number using the current locale or using the\n",
      "            C (English) locale.  This affects e.g. the decimal separator.  The\n",
      "            formatter default is :rc:`axes.formatter.use_locale`.\n",
      "        \n",
      "        useMathText : bool\n",
      "            Render the offset and scientific notation in mathtext.\n",
      "            The formatter default is :rc:`axes.formatter.use_mathtext`.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        AttributeError\n",
      "            If the current formatter is not a `.ScalarFormatter`.\n",
      "    \n",
      "    tight_layout(*, pad=1.08, h_pad=None, w_pad=None, rect=None)\n",
      "        Adjust the padding between and around subplots.\n",
      "        \n",
      "        To exclude an artist on the Axes from the bounding box calculation\n",
      "        that determines the subplot parameters (i.e. legend, or annotation),\n",
      "        set ``a.set_in_layout(False)`` for that artist.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pad : float, default: 1.08\n",
      "            Padding between the figure edge and the edges of subplots,\n",
      "            as a fraction of the font size.\n",
      "        h_pad, w_pad : float, default: *pad*\n",
      "            Padding (height/width) between edges of adjacent subplots,\n",
      "            as a fraction of the font size.\n",
      "        rect : tuple (left, bottom, right, top), default: (0, 0, 1, 1)\n",
      "            A rectangle in normalized figure coordinates into which the whole\n",
      "            subplots area (including labels) will fit.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        .Figure.set_layout_engine\n",
      "        .pyplot.tight_layout\n",
      "    \n",
      "    title(label, fontdict=None, loc=None, pad=None, *, y=None, **kwargs)\n",
      "        Set a title for the Axes.\n",
      "        \n",
      "        Set one of the three available Axes titles. The available titles\n",
      "        are positioned above the Axes in the center, flush with the left\n",
      "        edge, and flush with the right edge.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        label : str\n",
      "            Text to use for the title\n",
      "        \n",
      "        fontdict : dict\n",
      "            A dictionary controlling the appearance of the title text,\n",
      "            the default *fontdict* is::\n",
      "        \n",
      "               {'fontsize': rcParams['axes.titlesize'],\n",
      "                'fontweight': rcParams['axes.titleweight'],\n",
      "                'color': rcParams['axes.titlecolor'],\n",
      "                'verticalalignment': 'baseline',\n",
      "                'horizontalalignment': loc}\n",
      "        \n",
      "        loc : {'center', 'left', 'right'}, default: :rc:`axes.titlelocation`\n",
      "            Which title to set.\n",
      "        \n",
      "        y : float, default: :rc:`axes.titley`\n",
      "            Vertical Axes location for the title (1.0 is the top).  If\n",
      "            None (the default) and :rc:`axes.titley` is also None, y is\n",
      "            determined automatically to avoid decorators on the Axes.\n",
      "        \n",
      "        pad : float, default: :rc:`axes.titlepad`\n",
      "            The offset of the title from the top of the Axes, in points.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `.Text`\n",
      "            The matplotlib text instance representing the title\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `.Text` properties\n",
      "            Other keyword arguments are text properties, see `.Text` for a list\n",
      "            of valid text properties.\n",
      "    \n",
      "    tricontour(*args, **kwargs)\n",
      "        Draw contour lines on an unstructured triangular grid.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            tricontour(triangulation, z, [levels], ...)\n",
      "            tricontour(x, y, z, [levels], *, [triangles=triangles], [mask=mask], ...)\n",
      "        \n",
      "        The triangular grid can be specified either by passing a `.Triangulation`\n",
      "        object as the first parameter, or by passing the points *x*, *y* and\n",
      "        optionally the *triangles* and a *mask*. See `.Triangulation` for an\n",
      "        explanation of these parameters. If neither of *triangulation* or\n",
      "        *triangles* are given, the triangulation is calculated on the fly.\n",
      "        \n",
      "        It is possible to pass *triangles* positionally, i.e.\n",
      "        ``tricontour(x, y, triangles, z, ...)``. However, this is discouraged. For more\n",
      "        clarity, pass *triangles* via keyword argument.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        triangulation : `.Triangulation`, optional\n",
      "            An already created triangular grid.\n",
      "        \n",
      "        x, y, triangles, mask\n",
      "            Parameters defining the triangular grid. See `.Triangulation`.\n",
      "            This is mutually exclusive with specifying *triangulation*.\n",
      "        \n",
      "        z : array-like\n",
      "            The height values over which the contour is drawn.  Color-mapping is\n",
      "            controlled by *cmap*, *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        levels : int or array-like, optional\n",
      "            Determines the number and positions of the contour lines / regions.\n",
      "        \n",
      "            If an int *n*, use `~matplotlib.ticker.MaxNLocator`, which tries to\n",
      "            automatically choose no more than *n+1* \"nice\" contour levels between\n",
      "            between minimum and maximum numeric values of *Z*.\n",
      "        \n",
      "            If array-like, draw contour lines at the specified levels.  The values must\n",
      "            be in increasing order.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.tri.TriContourSet`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        colors : color string or sequence of colors, optional\n",
      "            The colors of the levels, i.e., the contour lines.\n",
      "        \n",
      "            The sequence is cycled for the levels in ascending order. If the sequence\n",
      "            is shorter than the number of levels, it is repeated.\n",
      "        \n",
      "            As a shortcut, single color strings may be used in place of one-element\n",
      "            lists, i.e. ``'red'`` instead of ``['red']`` to color all levels with the\n",
      "            same color. This shortcut does only work for color strings, not for other\n",
      "            ways of specifying colors.\n",
      "        \n",
      "            By default (value *None*), the colormap specified by *cmap* will be used.\n",
      "        \n",
      "        alpha : float, default: 1\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            If *vmin* or *vmax* are not given, the default color scaling is based on\n",
      "            *levels*.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        origin : {*None*, 'upper', 'lower', 'image'}, default: None\n",
      "            Determines the orientation and exact position of *z* by specifying the\n",
      "            position of ``z[0, 0]``.  This is only relevant, if *X*, *Y* are not given.\n",
      "        \n",
      "            - *None*: ``z[0, 0]`` is at X=0, Y=0 in the lower left corner.\n",
      "            - 'lower': ``z[0, 0]`` is at X=0.5, Y=0.5 in the lower left corner.\n",
      "            - 'upper': ``z[0, 0]`` is at X=N+0.5, Y=0.5 in the upper left corner.\n",
      "            - 'image': Use the value from :rc:`image.origin`.\n",
      "        \n",
      "        extent : (x0, x1, y0, y1), optional\n",
      "            If *origin* is not *None*, then *extent* is interpreted as in `.imshow`: it\n",
      "            gives the outer pixel boundaries. In this case, the position of z[0, 0] is\n",
      "            the center of the pixel, not a corner. If *origin* is *None*, then\n",
      "            (*x0*, *y0*) is the position of z[0, 0], and (*x1*, *y1*) is the position\n",
      "            of z[-1, -1].\n",
      "        \n",
      "            This argument is ignored if *X* and *Y* are specified in the call to\n",
      "            contour.\n",
      "        \n",
      "        locator : ticker.Locator subclass, optional\n",
      "            The locator is used to determine the contour levels if they are not given\n",
      "            explicitly via *levels*.\n",
      "            Defaults to `~.ticker.MaxNLocator`.\n",
      "        \n",
      "        extend : {'neither', 'both', 'min', 'max'}, default: 'neither'\n",
      "            Determines the ``tricontour``-coloring of values that are outside the\n",
      "            *levels* range.\n",
      "        \n",
      "            If 'neither', values outside the *levels* range are not colored.  If 'min',\n",
      "            'max' or 'both', color the values below, above or below and above the\n",
      "            *levels* range.\n",
      "        \n",
      "            Values below ``min(levels)`` and above ``max(levels)`` are mapped to the\n",
      "            under/over values of the `.Colormap`. Note that most colormaps do not have\n",
      "            dedicated colors for these by default, so that the over and under values\n",
      "            are the edge values of the colormap.  You may want to set these values\n",
      "            explicitly using `.Colormap.set_under` and `.Colormap.set_over`.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "                An existing `.TriContourSet` does not get notified if properties of its\n",
      "                colormap are changed. Therefore, an explicit call to\n",
      "                `.ContourSet.changed()` is needed after modifying the colormap. The\n",
      "                explicit call can be left out, if a colorbar is assigned to the\n",
      "                `.TriContourSet` because it internally calls `.ContourSet.changed()`.\n",
      "        \n",
      "        xunits, yunits : registered units, optional\n",
      "            Override axis units by specifying an instance of a\n",
      "            :class:`matplotlib.units.ConversionInterface`.\n",
      "        \n",
      "        antialiased : bool, optional\n",
      "            Enable antialiasing, overriding the defaults.  For\n",
      "            filled contours, the default is *True*.  For line contours,\n",
      "            it is taken from :rc:`lines.antialiased`.\n",
      "        \n",
      "        linewidths : float or array-like, default: :rc:`contour.linewidth`\n",
      "            The line width of the contour lines.\n",
      "        \n",
      "            If a number, all levels will be plotted with this linewidth.\n",
      "        \n",
      "            If a sequence, the levels in ascending order will be plotted with\n",
      "            the linewidths in the order specified.\n",
      "        \n",
      "            If None, this falls back to :rc:`lines.linewidth`.\n",
      "        \n",
      "        linestyles : {*None*, 'solid', 'dashed', 'dashdot', 'dotted'}, optional\n",
      "            If *linestyles* is *None*, the default is 'solid' unless the lines are\n",
      "            monochrome.  In that case, negative contours will take their linestyle\n",
      "            from :rc:`contour.negative_linestyle` setting.\n",
      "        \n",
      "            *linestyles* can also be an iterable of the above strings specifying a\n",
      "            set of linestyles to be used. If this iterable is shorter than the\n",
      "            number of contour levels it will be repeated as necessary.\n",
      "    \n",
      "    tricontourf(*args, **kwargs)\n",
      "        Draw contour regions on an unstructured triangular grid.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            tricontourf(triangulation, z, [levels], ...)\n",
      "            tricontourf(x, y, z, [levels], *, [triangles=triangles], [mask=mask], ...)\n",
      "        \n",
      "        The triangular grid can be specified either by passing a `.Triangulation`\n",
      "        object as the first parameter, or by passing the points *x*, *y* and\n",
      "        optionally the *triangles* and a *mask*. See `.Triangulation` for an\n",
      "        explanation of these parameters. If neither of *triangulation* or\n",
      "        *triangles* are given, the triangulation is calculated on the fly.\n",
      "        \n",
      "        It is possible to pass *triangles* positionally, i.e.\n",
      "        ``tricontourf(x, y, triangles, z, ...)``. However, this is discouraged. For more\n",
      "        clarity, pass *triangles* via keyword argument.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        triangulation : `.Triangulation`, optional\n",
      "            An already created triangular grid.\n",
      "        \n",
      "        x, y, triangles, mask\n",
      "            Parameters defining the triangular grid. See `.Triangulation`.\n",
      "            This is mutually exclusive with specifying *triangulation*.\n",
      "        \n",
      "        z : array-like\n",
      "            The height values over which the contour is drawn.  Color-mapping is\n",
      "            controlled by *cmap*, *norm*, *vmin*, and *vmax*.\n",
      "        \n",
      "        levels : int or array-like, optional\n",
      "            Determines the number and positions of the contour lines / regions.\n",
      "        \n",
      "            If an int *n*, use `~matplotlib.ticker.MaxNLocator`, which tries to\n",
      "            automatically choose no more than *n+1* \"nice\" contour levels between\n",
      "            between minimum and maximum numeric values of *Z*.\n",
      "        \n",
      "            If array-like, draw contour lines at the specified levels.  The values must\n",
      "            be in increasing order.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.tri.TriContourSet`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        colors : color string or sequence of colors, optional\n",
      "            The colors of the levels, i.e., the contour regions.\n",
      "        \n",
      "            The sequence is cycled for the levels in ascending order. If the sequence\n",
      "            is shorter than the number of levels, it is repeated.\n",
      "        \n",
      "            As a shortcut, single color strings may be used in place of one-element\n",
      "            lists, i.e. ``'red'`` instead of ``['red']`` to color all levels with the\n",
      "            same color. This shortcut does only work for color strings, not for other\n",
      "            ways of specifying colors.\n",
      "        \n",
      "            By default (value *None*), the colormap specified by *cmap* will be used.\n",
      "        \n",
      "        alpha : float, default: 1\n",
      "            The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
      "        \n",
      "        cmap : str or `~matplotlib.colors.Colormap`, default: :rc:`image.cmap`\n",
      "            The Colormap instance or registered colormap name used to map scalar data\n",
      "            to colors.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        norm : str or `~matplotlib.colors.Normalize`, optional\n",
      "            The normalization method used to scale scalar data to the [0, 1] range\n",
      "            before mapping to colors using *cmap*. By default, a linear scaling is\n",
      "            used, mapping the lowest value to 0 and the highest to 1.\n",
      "        \n",
      "            If given, this can be one of the following:\n",
      "        \n",
      "            - An instance of `.Normalize` or one of its subclasses\n",
      "              (see :doc:`/tutorials/colors/colormapnorms`).\n",
      "            - A scale name, i.e. one of \"linear\", \"log\", \"symlog\", \"logit\", etc.  For a\n",
      "              list of available scales, call `matplotlib.scale.get_scale_names()`.\n",
      "              In that case, a suitable `.Normalize` subclass is dynamically generated\n",
      "              and instantiated.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        vmin, vmax : float, optional\n",
      "            When using scalar data and no explicit *norm*, *vmin* and *vmax* define\n",
      "            the data range that the colormap covers. By default, the colormap covers\n",
      "            the complete value range of the supplied data. It is an error to use\n",
      "            *vmin*/*vmax* when a *norm* instance is given (but using a `str` *norm*\n",
      "            name together with *vmin*/*vmax* is acceptable).\n",
      "        \n",
      "            If *vmin* or *vmax* are not given, the default color scaling is based on\n",
      "            *levels*.\n",
      "        \n",
      "            This parameter is ignored if *colors* is set.\n",
      "        \n",
      "        origin : {*None*, 'upper', 'lower', 'image'}, default: None\n",
      "            Determines the orientation and exact position of *z* by specifying the\n",
      "            position of ``z[0, 0]``.  This is only relevant, if *X*, *Y* are not given.\n",
      "        \n",
      "            - *None*: ``z[0, 0]`` is at X=0, Y=0 in the lower left corner.\n",
      "            - 'lower': ``z[0, 0]`` is at X=0.5, Y=0.5 in the lower left corner.\n",
      "            - 'upper': ``z[0, 0]`` is at X=N+0.5, Y=0.5 in the upper left corner.\n",
      "            - 'image': Use the value from :rc:`image.origin`.\n",
      "        \n",
      "        extent : (x0, x1, y0, y1), optional\n",
      "            If *origin* is not *None*, then *extent* is interpreted as in `.imshow`: it\n",
      "            gives the outer pixel boundaries. In this case, the position of z[0, 0] is\n",
      "            the center of the pixel, not a corner. If *origin* is *None*, then\n",
      "            (*x0*, *y0*) is the position of z[0, 0], and (*x1*, *y1*) is the position\n",
      "            of z[-1, -1].\n",
      "        \n",
      "            This argument is ignored if *X* and *Y* are specified in the call to\n",
      "            contour.\n",
      "        \n",
      "        locator : ticker.Locator subclass, optional\n",
      "            The locator is used to determine the contour levels if they are not given\n",
      "            explicitly via *levels*.\n",
      "            Defaults to `~.ticker.MaxNLocator`.\n",
      "        \n",
      "        extend : {'neither', 'both', 'min', 'max'}, default: 'neither'\n",
      "            Determines the ``tricontourf``-coloring of values that are outside the\n",
      "            *levels* range.\n",
      "        \n",
      "            If 'neither', values outside the *levels* range are not colored.  If 'min',\n",
      "            'max' or 'both', color the values below, above or below and above the\n",
      "            *levels* range.\n",
      "        \n",
      "            Values below ``min(levels)`` and above ``max(levels)`` are mapped to the\n",
      "            under/over values of the `.Colormap`. Note that most colormaps do not have\n",
      "            dedicated colors for these by default, so that the over and under values\n",
      "            are the edge values of the colormap.  You may want to set these values\n",
      "            explicitly using `.Colormap.set_under` and `.Colormap.set_over`.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "                An existing `.TriContourSet` does not get notified if properties of its\n",
      "                colormap are changed. Therefore, an explicit call to\n",
      "                `.ContourSet.changed()` is needed after modifying the colormap. The\n",
      "                explicit call can be left out, if a colorbar is assigned to the\n",
      "                `.TriContourSet` because it internally calls `.ContourSet.changed()`.\n",
      "        \n",
      "        xunits, yunits : registered units, optional\n",
      "            Override axis units by specifying an instance of a\n",
      "            :class:`matplotlib.units.ConversionInterface`.\n",
      "        \n",
      "        antialiased : bool, optional\n",
      "            Enable antialiasing, overriding the defaults.  For\n",
      "            filled contours, the default is *True*.  For line contours,\n",
      "            it is taken from :rc:`lines.antialiased`.\n",
      "        \n",
      "        hatches : list[str], optional\n",
      "            A list of cross hatch patterns to use on the filled areas.\n",
      "            If None, no hatching will be added to the contour.\n",
      "            Hatching is supported in the PostScript, PDF, SVG and Agg\n",
      "            backends only.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `.tricontourf` fills intervals that are closed at the top; that is, for\n",
      "        boundaries *z1* and *z2*, the filled region is::\n",
      "        \n",
      "            z1 < Z <= z2\n",
      "        \n",
      "        except for the lowest interval, which is closed on both sides (i.e. it\n",
      "        includes the lowest value).\n",
      "    \n",
      "    tripcolor(*args, alpha=1.0, norm=None, cmap=None, vmin=None, vmax=None, shading='flat', facecolors=None, **kwargs)\n",
      "        Create a pseudocolor plot of an unstructured triangular grid.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "          tripcolor(triangulation, c, *, ...)\n",
      "          tripcolor(x, y, c, *, [triangles=triangles], [mask=mask], ...)\n",
      "        \n",
      "        The triangular grid can be specified either by passing a `.Triangulation`\n",
      "        object as the first parameter, or by passing the points *x*, *y* and\n",
      "        optionally the *triangles* and a *mask*. See `.Triangulation` for an\n",
      "        explanation of these parameters.\n",
      "        \n",
      "        It is possible to pass the triangles positionally, i.e.\n",
      "        ``tripcolor(x, y, triangles, c, ...)``. However, this is discouraged.\n",
      "        For more clarity, pass *triangles* via keyword argument.\n",
      "        \n",
      "        If neither of *triangulation* or *triangles* are given, the triangulation\n",
      "        is calculated on the fly. In this case, it does not make sense to provide\n",
      "        colors at the triangle faces via *c* or *facecolors* because there are\n",
      "        multiple possible triangulations for a group of points and you don't know\n",
      "        which triangles will be constructed.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        triangulation : `.Triangulation`\n",
      "            An already created triangular grid.\n",
      "        x, y, triangles, mask\n",
      "            Parameters defining the triangular grid. See `.Triangulation`.\n",
      "            This is mutually exclusive with specifying *triangulation*.\n",
      "        c : array-like\n",
      "            The color values, either for the points or for the triangles. Which one\n",
      "            is automatically inferred from the length of *c*, i.e. does it match\n",
      "            the number of points or the number of triangles. If there are the same\n",
      "            number of points and triangles in the triangulation it is assumed that\n",
      "            color values are defined at points; to force the use of color values at\n",
      "            triangles use the keyword argument ``facecolors=c`` instead of just\n",
      "            ``c``.\n",
      "            This parameter is position-only.\n",
      "        facecolors : array-like, optional\n",
      "            Can be used alternatively to *c* to specify colors at the triangle\n",
      "            faces. This parameter takes precedence over *c*.\n",
      "        shading : {'flat', 'gouraud'}, default: 'flat'\n",
      "            If  'flat' and the color values *c* are defined at points, the color\n",
      "            values used for each triangle are from the mean c of the triangle's\n",
      "            three points. If *shading* is 'gouraud' then color values must be\n",
      "            defined at points.\n",
      "        other_parameters\n",
      "            All other parameters are the same as for `~.Axes.pcolor`.\n",
      "    \n",
      "    triplot(*args, **kwargs)\n",
      "        Draw an unstructured triangular grid as lines and/or markers.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "          triplot(triangulation, ...)\n",
      "          triplot(x, y, [triangles], *, [mask=mask], ...)\n",
      "        \n",
      "        The triangular grid can be specified either by passing a `.Triangulation`\n",
      "        object as the first parameter, or by passing the points *x*, *y* and\n",
      "        optionally the *triangles* and a *mask*. If neither of *triangulation* or\n",
      "        *triangles* are given, the triangulation is calculated on the fly.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        triangulation : `.Triangulation`\n",
      "            An already created triangular grid.\n",
      "        x, y, triangles, mask\n",
      "            Parameters defining the triangular grid. See `.Triangulation`.\n",
      "            This is mutually exclusive with specifying *triangulation*.\n",
      "        other_parameters\n",
      "            All other args and kwargs are forwarded to `~.Axes.plot`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lines : `~matplotlib.lines.Line2D`\n",
      "            The drawn triangles edges.\n",
      "        markers : `~matplotlib.lines.Line2D`\n",
      "            The drawn marker nodes.\n",
      "    \n",
      "    twinx(ax=None)\n",
      "        Make and return a second axes that shares the *x*-axis.  The new axes will\n",
      "        overlay *ax* (or the current axes if *ax* is *None*), and its ticks will be\n",
      "        on the right.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        :doc:`/gallery/subplots_axes_and_figures/two_scales`\n",
      "    \n",
      "    twiny(ax=None)\n",
      "        Make and return a second axes that shares the *y*-axis.  The new axes will\n",
      "        overlay *ax* (or the current axes if *ax* is *None*), and its ticks will be\n",
      "        on the top.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        :doc:`/gallery/subplots_axes_and_figures/two_scales`\n",
      "    \n",
      "    uninstall_repl_displayhook()\n",
      "        Disconnect from the display hook of the current shell.\n",
      "    \n",
      "    violinplot(dataset, positions=None, vert=True, widths=0.5, showmeans=False, showextrema=True, showmedians=False, quantiles=None, points=100, bw_method=None, *, data=None)\n",
      "        Make a violin plot.\n",
      "        \n",
      "        Make a violin plot for each column of *dataset* or each vector in\n",
      "        sequence *dataset*.  Each filled area extends to represent the\n",
      "        entire data range, with optional lines at the mean, the median,\n",
      "        the minimum, the maximum, and user-specified quantiles.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        dataset : Array or a sequence of vectors.\n",
      "          The input data.\n",
      "        \n",
      "        positions : array-like, default: [1, 2, ..., n]\n",
      "          The positions of the violins. The ticks and limits are\n",
      "          automatically set to match the positions.\n",
      "        \n",
      "        vert : bool, default: True.\n",
      "          If true, creates a vertical violin plot.\n",
      "          Otherwise, creates a horizontal violin plot.\n",
      "        \n",
      "        widths : array-like, default: 0.5\n",
      "          Either a scalar or a vector that sets the maximal width of\n",
      "          each violin. The default is 0.5, which uses about half of the\n",
      "          available horizontal space.\n",
      "        \n",
      "        showmeans : bool, default: False\n",
      "          If `True`, will toggle rendering of the means.\n",
      "        \n",
      "        showextrema : bool, default: True\n",
      "          If `True`, will toggle rendering of the extrema.\n",
      "        \n",
      "        showmedians : bool, default: False\n",
      "          If `True`, will toggle rendering of the medians.\n",
      "        \n",
      "        quantiles : array-like, default: None\n",
      "          If not None, set a list of floats in interval [0, 1] for each violin,\n",
      "          which stands for the quantiles that will be rendered for that\n",
      "          violin.\n",
      "        \n",
      "        points : int, default: 100\n",
      "          Defines the number of points to evaluate each of the\n",
      "          gaussian kernel density estimations at.\n",
      "        \n",
      "        bw_method : str, scalar or callable, optional\n",
      "          The method used to calculate the estimator bandwidth.  This can be\n",
      "          'scott', 'silverman', a scalar constant or a callable.  If a\n",
      "          scalar, this will be used directly as `kde.factor`.  If a\n",
      "          callable, it should take a `matplotlib.mlab.GaussianKDE` instance as\n",
      "          its only parameter and return a scalar. If None (default), 'scott'\n",
      "          is used.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *dataset*\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        dict\n",
      "          A dictionary mapping each component of the violinplot to a\n",
      "          list of the corresponding collection instances created. The\n",
      "          dictionary has the following keys:\n",
      "        \n",
      "          - ``bodies``: A list of the `~.collections.PolyCollection`\n",
      "            instances containing the filled area of each violin.\n",
      "        \n",
      "          - ``cmeans``: A `~.collections.LineCollection` instance that marks\n",
      "            the mean values of each of the violin's distribution.\n",
      "        \n",
      "          - ``cmins``: A `~.collections.LineCollection` instance that marks\n",
      "            the bottom of each violin's distribution.\n",
      "        \n",
      "          - ``cmaxes``: A `~.collections.LineCollection` instance that marks\n",
      "            the top of each violin's distribution.\n",
      "        \n",
      "          - ``cbars``: A `~.collections.LineCollection` instance that marks\n",
      "            the centers of each violin's distribution.\n",
      "        \n",
      "          - ``cmedians``: A `~.collections.LineCollection` instance that\n",
      "            marks the median values of each of the violin's distribution.\n",
      "        \n",
      "          - ``cquantiles``: A `~.collections.LineCollection` instance created\n",
      "            to identify the quantile values of each of the violin's\n",
      "            distribution.\n",
      "    \n",
      "    viridis()\n",
      "        Set the colormap to 'viridis'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    vlines(x, ymin, ymax, colors=None, linestyles='solid', label='', *, data=None, **kwargs)\n",
      "        Plot vertical lines at each *x* from *ymin* to *ymax*.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : float or array-like\n",
      "            x-indexes where to plot the lines.\n",
      "        \n",
      "        ymin, ymax : float or array-like\n",
      "            Respective beginning and end of each line. If scalars are\n",
      "            provided, all lines will have the same length.\n",
      "        \n",
      "        colors : list of colors, default: :rc:`lines.color`\n",
      "        \n",
      "        linestyles : {'solid', 'dashed', 'dashdot', 'dotted'}, optional\n",
      "        \n",
      "        label : str, default: ''\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `~matplotlib.collections.LineCollection`\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *ymin*, *ymax*, *colors*\n",
      "        **kwargs : `~matplotlib.collections.LineCollection` properties.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        hlines : horizontal lines\n",
      "        axvline : vertical line across the Axes\n",
      "    \n",
      "    waitforbuttonpress(timeout=-1)\n",
      "        Blocking call to interact with the figure.\n",
      "        \n",
      "        Wait for user input and return True if a key was pressed, False if a\n",
      "        mouse button was pressed and None if no input was given within\n",
      "        *timeout* seconds.  Negative values deactivate *timeout*.\n",
      "    \n",
      "    winter()\n",
      "        Set the colormap to 'winter'.\n",
      "        \n",
      "        This changes the default colormap as well as the colormap of the current\n",
      "        image if there is one. See ``help(colormaps)`` for more information.\n",
      "    \n",
      "    xcorr(x, y, normed=True, detrend=<function detrend_none at 0x127a34dc0>, usevlines=True, maxlags=10, *, data=None, **kwargs)\n",
      "        Plot the cross correlation between *x* and *y*.\n",
      "        \n",
      "        The correlation with lag k is defined as\n",
      "        :math:`\\sum_n x[n+k] \\cdot y^*[n]`, where :math:`y^*` is the complex\n",
      "        conjugate of :math:`y`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like of length n\n",
      "        \n",
      "        detrend : callable, default: `.mlab.detrend_none` (no detrending)\n",
      "            A detrending function applied to *x* and *y*.  It must have the\n",
      "            signature ::\n",
      "        \n",
      "                detrend(x: np.ndarray) -> np.ndarray\n",
      "        \n",
      "        normed : bool, default: True\n",
      "            If ``True``, input vectors are normalised to unit length.\n",
      "        \n",
      "        usevlines : bool, default: True\n",
      "            Determines the plot style.\n",
      "        \n",
      "            If ``True``, vertical lines are plotted from 0 to the xcorr value\n",
      "            using `.Axes.vlines`. Additionally, a horizontal line is plotted\n",
      "            at y=0 using `.Axes.axhline`.\n",
      "        \n",
      "            If ``False``, markers are plotted at the xcorr values using\n",
      "            `.Axes.plot`.\n",
      "        \n",
      "        maxlags : int, default: 10\n",
      "            Number of lags to show. If None, will return all ``2 * len(x) - 1``\n",
      "            lags.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lags : array (length ``2*maxlags+1``)\n",
      "            The lag vector.\n",
      "        c : array  (length ``2*maxlags+1``)\n",
      "            The auto correlation vector.\n",
      "        line : `.LineCollection` or `.Line2D`\n",
      "            `.Artist` added to the Axes of the correlation:\n",
      "        \n",
      "            - `.LineCollection` if *usevlines* is True.\n",
      "            - `.Line2D` if *usevlines* is False.\n",
      "        b : `.Line2D` or None\n",
      "            Horizontal line at 0 if *usevlines* is True\n",
      "            None *usevlines* is False.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        linestyle : `.Line2D` property, optional\n",
      "            The linestyle for plotting the data points.\n",
      "            Only used if *usevlines* is ``False``.\n",
      "        \n",
      "        marker : str, default: 'o'\n",
      "            The marker for plotting the data points.\n",
      "            Only used if *usevlines* is ``False``.\n",
      "        \n",
      "        data : indexable object, optional\n",
      "            If given, the following parameters also accept a string ``s``, which is\n",
      "            interpreted as ``data[s]`` (unless this raises an exception):\n",
      "        \n",
      "            *x*, *y*\n",
      "        \n",
      "        **kwargs\n",
      "            Additional parameters are passed to `.Axes.vlines` and\n",
      "            `.Axes.axhline` if *usevlines* is ``True``; otherwise they are\n",
      "            passed to `.Axes.plot`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The cross correlation is performed with `numpy.correlate` with\n",
      "        ``mode = \"full\"``.\n",
      "    \n",
      "    xkcd(scale=1, length=100, randomness=2)\n",
      "        Turn on `xkcd <https://xkcd.com/>`_ sketch-style drawing mode.  This will\n",
      "        only have effect on things drawn after this function is called.\n",
      "        \n",
      "        For best results, the \"Humor Sans\" font should be installed: it is\n",
      "        not included with Matplotlib.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        scale : float, optional\n",
      "            The amplitude of the wiggle perpendicular to the source line.\n",
      "        length : float, optional\n",
      "            The length of the wiggle along the line.\n",
      "        randomness : float, optional\n",
      "            The scale factor by which the length is shrunken or expanded.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function works by a number of rcParams, so it will probably\n",
      "        override others you have set before.\n",
      "        \n",
      "        If you want the effects of this function to be temporary, it can\n",
      "        be used as a context manager, for example::\n",
      "        \n",
      "            with plt.xkcd():\n",
      "                # This figure will be in XKCD-style\n",
      "                fig1 = plt.figure()\n",
      "                # ...\n",
      "        \n",
      "            # This figure will be in regular style\n",
      "            fig2 = plt.figure()\n",
      "    \n",
      "    xlabel(xlabel, fontdict=None, labelpad=None, *, loc=None, **kwargs)\n",
      "        Set the label for the x-axis.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xlabel : str\n",
      "            The label text.\n",
      "        \n",
      "        labelpad : float, default: :rc:`axes.labelpad`\n",
      "            Spacing in points from the Axes bounding box including ticks\n",
      "            and tick labels.  If None, the previous value is left as is.\n",
      "        \n",
      "        loc : {'left', 'center', 'right'}, default: :rc:`xaxis.labellocation`\n",
      "            The label position. This is a high-level alternative for passing\n",
      "            parameters *x* and *horizontalalignment*.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `.Text` properties\n",
      "            `.Text` properties control the appearance of the label.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        text : Documents the properties supported by `.Text`.\n",
      "    \n",
      "    xlim(*args, **kwargs)\n",
      "        Get or set the x limits of the current axes.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            left, right = xlim()  # return the current xlim\n",
      "            xlim((left, right))   # set the xlim to left, right\n",
      "            xlim(left, right)     # set the xlim to left, right\n",
      "        \n",
      "        If you do not specify args, you can pass *left* or *right* as kwargs,\n",
      "        i.e.::\n",
      "        \n",
      "            xlim(right=3)  # adjust the right leaving left unchanged\n",
      "            xlim(left=1)  # adjust the left leaving right unchanged\n",
      "        \n",
      "        Setting limits turns autoscaling off for the x-axis.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        left, right\n",
      "            A tuple of the new x-axis limits.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Calling this function with no arguments (e.g. ``xlim()``) is the pyplot\n",
      "        equivalent of calling `~.Axes.get_xlim` on the current axes.\n",
      "        Calling this function with arguments is the pyplot equivalent of calling\n",
      "        `~.Axes.set_xlim` on the current axes. All arguments are passed though.\n",
      "    \n",
      "    xscale(value, **kwargs)\n",
      "        Set the xaxis' scale.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        value : {\"linear\", \"log\", \"symlog\", \"logit\", ...} or `.ScaleBase`\n",
      "            The axis scale type to apply.\n",
      "        \n",
      "        **kwargs\n",
      "            Different keyword arguments are accepted, depending on the scale.\n",
      "            See the respective class keyword arguments:\n",
      "        \n",
      "            - `matplotlib.scale.LinearScale`\n",
      "            - `matplotlib.scale.LogScale`\n",
      "            - `matplotlib.scale.SymmetricalLogScale`\n",
      "            - `matplotlib.scale.LogitScale`\n",
      "            - `matplotlib.scale.FuncScale`\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        By default, Matplotlib supports the above mentioned scales.\n",
      "        Additionally, custom scales may be registered using\n",
      "        `matplotlib.scale.register_scale`. These scales can then also\n",
      "        be used here.\n",
      "    \n",
      "    xticks(ticks=None, labels=None, *, minor=False, **kwargs)\n",
      "        Get or set the current tick locations and labels of the x-axis.\n",
      "        \n",
      "        Pass no arguments to return the current values without modifying them.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        ticks : array-like, optional\n",
      "            The list of xtick locations.  Passing an empty list removes all xticks.\n",
      "        labels : array-like, optional\n",
      "            The labels to place at the given *ticks* locations.  This argument can\n",
      "            only be passed if *ticks* is passed as well.\n",
      "        minor : bool, default: False\n",
      "            If ``False``, get/set the major ticks/labels; if ``True``, the minor\n",
      "            ticks/labels.\n",
      "        **kwargs\n",
      "            `.Text` properties can be used to control the appearance of the labels.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        locs\n",
      "            The list of xtick locations.\n",
      "        labels\n",
      "            The list of xlabel `.Text` objects.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Calling this function with no arguments (e.g. ``xticks()``) is the pyplot\n",
      "        equivalent of calling `~.Axes.get_xticks` and `~.Axes.get_xticklabels` on\n",
      "        the current axes.\n",
      "        Calling this function with arguments is the pyplot equivalent of calling\n",
      "        `~.Axes.set_xticks` and `~.Axes.set_xticklabels` on the current axes.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> locs, labels = xticks()  # Get the current locations and labels.\n",
      "        >>> xticks(np.arange(0, 1, step=0.2))  # Set label locations.\n",
      "        >>> xticks(np.arange(3), ['Tom', 'Dick', 'Sue'])  # Set text labels.\n",
      "        >>> xticks([0, 1, 2], ['January', 'February', 'March'],\n",
      "        ...        rotation=20)  # Set text labels and properties.\n",
      "        >>> xticks([])  # Disable xticks.\n",
      "    \n",
      "    ylabel(ylabel, fontdict=None, labelpad=None, *, loc=None, **kwargs)\n",
      "        Set the label for the y-axis.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        ylabel : str\n",
      "            The label text.\n",
      "        \n",
      "        labelpad : float, default: :rc:`axes.labelpad`\n",
      "            Spacing in points from the Axes bounding box including ticks\n",
      "            and tick labels.  If None, the previous value is left as is.\n",
      "        \n",
      "        loc : {'bottom', 'center', 'top'}, default: :rc:`yaxis.labellocation`\n",
      "            The label position. This is a high-level alternative for passing\n",
      "            parameters *y* and *horizontalalignment*.\n",
      "        \n",
      "        Other Parameters\n",
      "        ----------------\n",
      "        **kwargs : `.Text` properties\n",
      "            `.Text` properties control the appearance of the label.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        text : Documents the properties supported by `.Text`.\n",
      "    \n",
      "    ylim(*args, **kwargs)\n",
      "        Get or set the y-limits of the current axes.\n",
      "        \n",
      "        Call signatures::\n",
      "        \n",
      "            bottom, top = ylim()  # return the current ylim\n",
      "            ylim((bottom, top))   # set the ylim to bottom, top\n",
      "            ylim(bottom, top)     # set the ylim to bottom, top\n",
      "        \n",
      "        If you do not specify args, you can alternatively pass *bottom* or\n",
      "        *top* as kwargs, i.e.::\n",
      "        \n",
      "            ylim(top=3)  # adjust the top leaving bottom unchanged\n",
      "            ylim(bottom=1)  # adjust the bottom leaving top unchanged\n",
      "        \n",
      "        Setting limits turns autoscaling off for the y-axis.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        bottom, top\n",
      "            A tuple of the new y-axis limits.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Calling this function with no arguments (e.g. ``ylim()``) is the pyplot\n",
      "        equivalent of calling `~.Axes.get_ylim` on the current axes.\n",
      "        Calling this function with arguments is the pyplot equivalent of calling\n",
      "        `~.Axes.set_ylim` on the current axes. All arguments are passed though.\n",
      "    \n",
      "    yscale(value, **kwargs)\n",
      "        Set the yaxis' scale.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        value : {\"linear\", \"log\", \"symlog\", \"logit\", ...} or `.ScaleBase`\n",
      "            The axis scale type to apply.\n",
      "        \n",
      "        **kwargs\n",
      "            Different keyword arguments are accepted, depending on the scale.\n",
      "            See the respective class keyword arguments:\n",
      "        \n",
      "            - `matplotlib.scale.LinearScale`\n",
      "            - `matplotlib.scale.LogScale`\n",
      "            - `matplotlib.scale.SymmetricalLogScale`\n",
      "            - `matplotlib.scale.LogitScale`\n",
      "            - `matplotlib.scale.FuncScale`\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        By default, Matplotlib supports the above mentioned scales.\n",
      "        Additionally, custom scales may be registered using\n",
      "        `matplotlib.scale.register_scale`. These scales can then also\n",
      "        be used here.\n",
      "    \n",
      "    yticks(ticks=None, labels=None, *, minor=False, **kwargs)\n",
      "        Get or set the current tick locations and labels of the y-axis.\n",
      "        \n",
      "        Pass no arguments to return the current values without modifying them.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        ticks : array-like, optional\n",
      "            The list of ytick locations.  Passing an empty list removes all yticks.\n",
      "        labels : array-like, optional\n",
      "            The labels to place at the given *ticks* locations.  This argument can\n",
      "            only be passed if *ticks* is passed as well.\n",
      "        minor : bool, default: False\n",
      "            If ``False``, get/set the major ticks/labels; if ``True``, the minor\n",
      "            ticks/labels.\n",
      "        **kwargs\n",
      "            `.Text` properties can be used to control the appearance of the labels.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        locs\n",
      "            The list of ytick locations.\n",
      "        labels\n",
      "            The list of ylabel `.Text` objects.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Calling this function with no arguments (e.g. ``yticks()``) is the pyplot\n",
      "        equivalent of calling `~.Axes.get_yticks` and `~.Axes.get_yticklabels` on\n",
      "        the current axes.\n",
      "        Calling this function with arguments is the pyplot equivalent of calling\n",
      "        `~.Axes.set_yticks` and `~.Axes.set_yticklabels` on the current axes.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> locs, labels = yticks()  # Get the current locations and labels.\n",
      "        >>> yticks(np.arange(0, 1, step=0.2))  # Set label locations.\n",
      "        >>> yticks(np.arange(3), ['Tom', 'Dick', 'Sue'])  # Set text labels.\n",
      "        >>> yticks([0, 1, 2], ['January', 'February', 'March'],\n",
      "        ...        rotation=45)  # Set text labels and properties.\n",
      "        >>> yticks([])  # Disable yticks.\n",
      "\n",
      "DATA\n",
      "    color_sequences = <matplotlib.colors.ColorSequenceRegistry object>\n",
      "    colormaps = <matplotlib.cm.ColormapRegistry object>\n",
      "    rcParams = RcParams({'_internal.classic_mode': False,\n",
      "         ...nor.widt...\n",
      "    rcParamsDefault = RcParams({'_internal.classic_mode': False,\n",
      "         ...n...\n",
      "    rcParamsOrig = RcParams({'_internal.classic_mode': False,\n",
      "         ...nor....\n",
      "\n",
      "FILE\n",
      "    /opt/miniconda3/lib/python3.9/site-packages/matplotlib/pyplot.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(plt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diagramma dei codoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ottenere la lista dei codoni (asse x del diagramma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_codon_list = [t[0] for t in codon_frequencies]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gag',\n",
       " 'ctg',\n",
       " 'aac',\n",
       " 'gtg',\n",
       " 'aag',\n",
       " 'gac',\n",
       " 'atc',\n",
       " 'acc',\n",
       " 'ttc',\n",
       " 'ggc',\n",
       " 'ccc',\n",
       " 'ctc',\n",
       " 'cag',\n",
       " 'cac',\n",
       " 'atg',\n",
       " 'tgc',\n",
       " 'tac',\n",
       " 'gaa',\n",
       " 'ggg',\n",
       " 'gcc',\n",
       " 'gtc',\n",
       " 'tcc',\n",
       " 'gat',\n",
       " 'cgg',\n",
       " 'att',\n",
       " 'tct',\n",
       " 'aaa',\n",
       " 'tat',\n",
       " 'ttt',\n",
       " 'tgg',\n",
       " 'aat',\n",
       " 'agg',\n",
       " 'gga',\n",
       " 'tgt',\n",
       " 'ttg',\n",
       " 'agt',\n",
       " 'gcg',\n",
       " 'acg',\n",
       " 'cct',\n",
       " 'tca',\n",
       " 'cca',\n",
       " 'ccg',\n",
       " 'cga',\n",
       " 'cgc',\n",
       " 'agc',\n",
       " 'tcg',\n",
       " 'ggt',\n",
       " 'gct',\n",
       " 'act',\n",
       " 'gca',\n",
       " 'gtt',\n",
       " 'ctt',\n",
       " 'cat',\n",
       " 'aga',\n",
       " 'cta',\n",
       " 'cgt',\n",
       " 'caa',\n",
       " 'aca',\n",
       " 'tta',\n",
       " 'gta',\n",
       " 'ata',\n",
       " 'taa']"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_codon_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creare la corrispondente lista delle frequenze (asse y del diagramma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_freq_list = [t[1] for t in codon_frequencies]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[77,\n",
       " 67,\n",
       " 53,\n",
       " 49,\n",
       " 47,\n",
       " 44,\n",
       " 41,\n",
       " 39,\n",
       " 39,\n",
       " 37,\n",
       " 37,\n",
       " 35,\n",
       " 32,\n",
       " 31,\n",
       " 30,\n",
       " 30,\n",
       " 29,\n",
       " 29,\n",
       " 28,\n",
       " 28,\n",
       " 28,\n",
       " 27,\n",
       " 26,\n",
       " 25,\n",
       " 23,\n",
       " 22,\n",
       " 21,\n",
       " 21,\n",
       " 20,\n",
       " 20,\n",
       " 19,\n",
       " 18,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 15,\n",
       " 15,\n",
       " 14,\n",
       " 14,\n",
       " 13,\n",
       " 12,\n",
       " 12,\n",
       " 11,\n",
       " 11,\n",
       " 10,\n",
       " 10,\n",
       " 8,\n",
       " 8,\n",
       " 7,\n",
       " 7,\n",
       " 6,\n",
       " 5,\n",
       " 5,\n",
       " 5,\n",
       " 3,\n",
       " 2,\n",
       " 1,\n",
       " 1]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_freq_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Disegnare il diagramma."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABM0AAAIuCAYAAACsKv/uAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAABsPUlEQVR4nO3deZyN9f//8ecZzDHDzNhnTCZjGXuWshSVJXsfFS1Ki6VFH0qWIqmMYpTiQ5sSMZJKpajPpxBSSKaQkhBjqUwKzQzZxrx/f/jN+TrXzHDMeXOO8bjfbufGvM85r/O6rnNt53W9r+vtMsYYAQAAAAAAAPAICXQCAAAAAAAAQLChaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAACCyowZM+RyufJ8PPzww4FO74KQ8x1s3779jN+bmJgol8tlP6kC8Gc6fLV9+3a5XC7NmDHjrH0GAAAIjKKBTgAAACAv06dPV61atbzaYmNjA5QNfHXPPfeoY8eOgU7jnKlYsaK+/vprVatWLdCpAAAAyyiaAQCAoFSvXj01btzYp9ceO3ZMLpdLRYtyaBNolSpVUqVKlQKdxjnjdrt1+eWXBzoNAABwFnB5JgAAOK988cUXcrlcevPNNzVkyBBddNFFcrvd+uWXXyRJn3/+ua655hpFRkYqPDxcLVq00OLFi3PF+e9//6uGDRvK7XarSpUqev7553NdWniqS+9cLpcSExO92rZs2aIePXqoQoUKcrvdql27tl5++eU883/77bc1YsQIxcbGKjIyUm3bttWmTZtyvS6vR3x8vFfMd999V1dccYVKlCihkiVLqkOHDlq7dq1P83PVqlVq0aKFihcvrtjYWA0fPlzHjh3L87W+fM6ZXJ75zTffqEuXLipbtqyKFy+uatWqaeDAgV6vWb58ua655hpFREQoPDxczZs313//+98CT0d2drbGjRunWrVqye12q0KFCrrrrrv066+/er2uVatWqlevnlJSUnTVVVcpPDxcVatW1TPPPKPs7GzP67g8EwCAwouiGQAACErHjx9XVlaW1+Nkw4cP186dO/Xqq6/q448/VoUKFTRr1iy1b99ekZGRSk5O1pw5c1SmTBl16NDBq3C2ePFiXX/99YqIiNA777yj5557TnPmzNH06dMLnO9PP/2kJk2a6Mcff9T48eP1ySef6Nprr9WAAQM0atSoXK9/7LHHtGPHDk2dOlVTpkzRli1b1KVLFx0/flySdOmll+rrr7/2esycOVPFihVT3bp1PXGSkpJ02223qU6dOpozZ47efPNNZWZm6qqrrtJPP/102pyvueYa/f3335oxY4ZeffVVrV27VqNHj871Wn8+Jy8LFizQVVddpZ07d2rChAn69NNP9fjjj+uPP/7wvGbZsmVq06aN0tPTNW3aNL399tuKiIhQly5d9O677xZoOv79739r2LBhateunebPn6+nn35an332mZo3b66//vrL67VpaWm6/fbbdccdd2j+/Pnq1KmThg8frlmzZp3x9AIAgPOQAQAACCLTp083kvJ8HDt2zCxdutRIMldffbXX+w4ePGjKlCljunTp4tV+/Phx06BBA9O0aVNPW7NmzUxsbKw5dOiQpy0jI8OUKVPGnHx4lJqaaiSZ6dOn58pTkhk5cqTn7w4dOphKlSqZ9PR0r9c98MADpnjx4mbfvn3GGOPJv3Pnzl6vmzNnjpFkvv766zznyx9//GGqVq1q6tata/bv32+MMWbnzp2maNGi5sEHH/R6bWZmpomJiTG33HJLnrFydO/e3YSFhZm0tDRPW1ZWlqlVq5aRZFJTU8/4c0aOHGl8OcSsVq2aqVatmtd34HT55ZebChUqmMzMTK/86tWrZypVqmSys7PPaDo2btxoJJl+/fp5fc4333xjJJnHHnvM09ayZUsjyXzzzTder61Tp47p0KGD5+9TLSMAAOD8Rk8zAAAQlGbOnKmUlBSvx8n3LLvxxhu9Xr9y5Urt27dPPXv29Oqdlp2drY4dOyolJUUHDx7UwYMHlZKSom7duql48eKe9+f0YCqIw4cPa/HixeratavCw8O9Pr9z5846fPiwVq1a5fWe6667zuvv+vXrS5J27NiRK/7Bgwd17bXX6vDhw/r0009VqlQpSSd6a2VlZemuu+7y+szixYurZcuW+uKLL06Z99KlS3XNNdcoOjra01akSBF1797d63X+fo7T5s2btXXrVt19991e34Fzmr/55hvddNNNKlmypFd+d955p3799VfP5ay+TsfSpUslSb169fJqb9q0qWrXrp3rMt6YmBg1bdrUq61+/fp5fkcAAKDw4W65AAAgKNWuXfuUAwFUrFjR6++cy/puuummfN+zb98+uVwuZWdnKyYmJtfzebX5Yu/evcrKytKLL76oF198Mc/XOC/9K1u2rNffbrdbknTo0CGv9qysLN10003avHmzvvzyS8XFxXmey5nmJk2a5PmZISGnPj+6d+9en+aDv5/j9Oeff0rSKQcM2L9/v4wxub5n6f9GUd27d6/nX1+mI+f1+cV0FsOc35F04ntyfkcAAKBwomgGAADOS86bzZcrV06S9OKLL+Y7mmF0dLRnpM20tLRczzvbcnpBHTlyxKs9p/iSo3Tp0p4eUP3798/zs6tUqXKKqcnffffdp8WLF+t///ufGjRo4PVczjS///77qly58hnHLlu2rE/zwd/PcSpfvrwk5br5/slKly6tkJAQ7d69O9dzv//+u1devk5HThFs9+7duQp2v//+uyceAACARNEMAAAUEi1atFCpUqX0008/6YEHHsj3daGhoWratKnmzp2r5557zlMYy8zM1Mcff+z12ujoaBUvXlzr16/3ap83b57X3+Hh4WrdurXWrl2r+vXrKzQ01Mo0Pf7445o+fbqSk5PVtm3bXM936NBBRYsW1datW3NdruqL1q1ba/78+frjjz88lzYeP37c6yb7Nj7HqUaNGqpWrZreeOMNDR482NPL7mQlSpRQs2bNNHfuXD3//PMKCwuTdGL0y1mzZqlSpUqqUaPGGU1HmzZtJEmzZs3y6jWXkpKijRs3asSIEX5PGwAAKDwomgEAgEKhZMmSevHFF9WzZ0/t27dPN910kypUqKA///xT33//vf78809NnjxZkvT000+rY8eOateunYYMGaLjx4/r2WefVYkSJbRv3z5PTJfLpTvuuENvvPGGqlWrpgYNGmj16tWaPXt2rs+fNGmSrrzySl111VX697//rfj4eGVmZuqXX37Rxx9/rCVLlpzR9Lz33nsaM2aMbrrpJtWoUcPrnmhut1uNGjVSfHy8nnrqKY0YMULbtm1Tx44dVbp0af3xxx9avXq1SpQokefInTkef/xxzZ8/X23atNGTTz6p8PBwvfzyyzp48KDX6/z9nLy8/PLL6tKliy6//HINGjRIF198sXbu3KkFCxborbfekiSNHTtW7dq1U+vWrfXwww8rNDRUr7zyin788Ue9/fbbnt6Gvk5HzZo1dd999+nFF19USEiIOnXqpO3bt+uJJ55QXFycBg0adEbTAAAACjeKZgAAoNC44447dPHFF2vcuHHq27evMjMzVaFCBTVs2NDr5u/t2rXTRx99pMcff1zdu3dXTEyM+vXrp0OHDuUq/owfP16SNG7cOB04cEBt2rTRJ598ovj4eK/X1alTR2vWrNHTTz+txx9/XHv27FGpUqWUkJCgzp07n/G0bNiwQdKJSyLff/99r+cqV66s7du3S5KGDx+uOnXqaNKkSXr77bd15MgRxcTEqEmTJrr//vtP+Rn16tXT559/riFDhqhnz54qXbq07rzzTt1444267777vF7rz+fkpUOHDvryyy/11FNPacCAATp8+LAqVarkNUBCy5YttWTJEo0cOVK9evVSdna2GjRooPnz5+tf//pXgaZj8uTJqlatmqZNm6aXX35ZUVFR6tixo8aOHZvnPcwAAMCFy2WMMYFOAgAAIBgkJiZq1KhR4vAIAAAAZzbUEQAAAAAAAHABoGgGAAAAAAAAOHB5JgAAAAAAAOBATzMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHAoGugEzrbs7Gz9/vvvioiIkMvlCnQ6AAAAAAAACCBjjDIzMxUbG6uQkPz7kxX6otnvv/+uuLi4QKcBAAAAAACAILJr1y5VqlQp3+cDWjTLyspSYmKi3nrrLaWlpalixYrq1auXHn/8cU+lzxijUaNGacqUKdq/f7+aNWuml19+WXXr1vXpMyIiIiSdmBGRkZFnbVoAAAAAAAAQ/DIyMhQXF+epGeUnoEWzZ599Vq+++qqSk5NVt25dffvtt+rdu7eioqL00EMPSZLGjRunCRMmaMaMGapRo4ZGjx6tdu3aadOmTaedOEmeSzIjIyMpmgEAAAAAAECSTnsbr4AOBPD111/r+uuv17XXXqv4+HjddNNNat++vb799ltJJ3qZTZw4USNGjFC3bt1Ur149JScn659//tHs2bMDmToAAAAAAAAKsYAWza688kotXrxYmzdvliR9//33Wr58uTp37ixJSk1NVVpamtq3b+95j9vtVsuWLbVy5co8Yx45ckQZGRleDwAAAAAAAOBMBPTyzGHDhik9PV21atVSkSJFdPz4cY0ZM0a33XabJCktLU2SFB0d7fW+6Oho7dixI8+YY8eO1ahRo85u4gAAAAAAACjUAtrT7N1339WsWbM0e/ZsrVmzRsnJyXr++eeVnJzs9TrnNabGmHyvOx0+fLjS09M9j127dp21/AEAAAAAAFA4BbSn2SOPPKJHH31Ut956qyTpkksu0Y4dOzR27Fj17NlTMTExkuQZWTPHnj17cvU+y+F2u+V2u89+8gAAAAAAACi0AtrT7J9//lFIiHcKRYoUUXZ2tiSpSpUqiomJ0aJFizzPHz16VMuWLVPz5s3Paa4AAAAAAAC4cAS0p1mXLl00ZswYXXzxxapbt67Wrl2rCRMmqE+fPpJOXJY5cOBAJSUlKSEhQQkJCUpKSlJ4eLh69OgRyNQBAAAAAABQiAW0aPbiiy/qiSeeUL9+/bRnzx7Fxsaqb9++evLJJz2vGTp0qA4dOqR+/fpp//79atasmRYuXKiIiIgAZg4AAAAAAIDCzGWMMYFO4mzKyMhQVFSU0tPTFRkZGeh0AAAAAAAAEEC+1ooCek8zAAAAAAAAIBhRNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHIoGOgEUjGuUq0DvMyON5UwAAAAAAAAKH3qaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcKBoBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB4pmAAAAAAAAgENAi2bx8fFyuVy5Hv3795ckGWOUmJio2NhYhYWFqVWrVtqwYUMgUwYAAAAAAMAFIKBFs5SUFO3evdvzWLRokSTp5ptvliSNGzdOEyZM0EsvvaSUlBTFxMSoXbt2yszMDGTaAAAAAAAAKOQCWjQrX768YmJiPI9PPvlE1apVU8uWLWWM0cSJEzVixAh169ZN9erVU3Jysv755x/Nnj07kGkDAAAAAACgkAuae5odPXpUs2bNUp8+feRyuZSamqq0tDS1b9/e8xq3262WLVtq5cqV+cY5cuSIMjIyvB4AAAAAAADAmQiaotlHH32kv//+W7169ZIkpaWlSZKio6O9XhcdHe15Li9jx45VVFSU5xEXF3fWcgYAAAAAAEDhFDRFs2nTpqlTp06KjY31ane5XF5/G2NytZ1s+PDhSk9P9zx27dp1VvIFAAAAAABA4VU00AlI0o4dO/T5559r7ty5nraYmBhJJ3qcVaxY0dO+Z8+eXL3PTuZ2u+V2u89esgAAAAAAACj0gqKn2fTp01WhQgVde+21nrYqVaooJibGM6KmdOK+Z8uWLVPz5s0DkSYAAAAAAAAuEAHvaZadna3p06erZ8+eKlr0/9JxuVwaOHCgkpKSlJCQoISEBCUlJSk8PFw9evQIYMYAAAAAAAAo7AJeNPv888+1c+dO9enTJ9dzQ4cO1aFDh9SvXz/t379fzZo108KFCxURERGATAEAAAAAAHChcBljTKCTOJsyMjIUFRWl9PR0RUZGBjoda1yj8h8M4VTMyEL9dQMAAAAAAJySr7WioLinGQAAAAAAABBMKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA5FA50AAsc1ylWg95mRxnImAAAAAAAAwYWeZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcKBoBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOAQ8KLZb7/9pjvuuENly5ZVeHi4GjZsqO+++87zvDFGiYmJio2NVVhYmFq1aqUNGzYEMGMAAAAAAAAUdgEtmu3fv18tWrRQsWLF9Omnn+qnn37S+PHjVapUKc9rxo0bpwkTJuill15SSkqKYmJi1K5dO2VmZgYucQAAAAAAABRqRQP54c8++6zi4uI0ffp0T1t8fLzn/8YYTZw4USNGjFC3bt0kScnJyYqOjtbs2bPVt2/fc50yAAAAAAAALgAB7Wk2f/58NW7cWDfffLMqVKigRo0a6fXXX/c8n5qaqrS0NLVv397T5na71bJlS61cuTLPmEeOHFFGRobXAwAAAAAAADgTAS2abdu2TZMnT1ZCQoIWLFig+++/XwMGDNDMmTMlSWlpaZKk6Ohor/dFR0d7nnMaO3asoqKiPI+4uLizOxEAAAAAAAAodAJaNMvOztall16qpKQkNWrUSH379tW9996ryZMne73O5XJ5/W2MydWWY/jw4UpPT/c8du3addbyBwAAAAAAQOEU0KJZxYoVVadOHa+22rVra+fOnZKkmJgYScrVq2zPnj25ep/lcLvdioyM9HoAAAAAAAAAZyKgRbMWLVpo06ZNXm2bN29W5cqVJUlVqlRRTEyMFi1a5Hn+6NGjWrZsmZo3b35OcwUAAAAAAMCFI6CjZw4aNEjNmzdXUlKSbrnlFq1evVpTpkzRlClTJJ24LHPgwIFKSkpSQkKCEhISlJSUpPDwcPXo0SOQqQMAAAAAAKAQC2jRrEmTJvrwww81fPhwPfXUU6pSpYomTpyo22+/3fOaoUOH6tChQ+rXr5/279+vZs2aaeHChYqIiAhg5gAAAAAAACjMXMYYE+gkzqaMjAxFRUUpPT29UN3fzDUq74EQTseM/L+v20YMAAAAAACA84mvtaKA3tMMAAAAAAAACEYUzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcKBoBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAORQOdAM5vrlGuAr3PjDSWMwEAAAAAALCHnmYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcAho0SwxMVEul8vrERMT43neGKPExETFxsYqLCxMrVq10oYNGwKYMQAAAAAAAC4EAe9pVrduXe3evdvz+OGHHzzPjRs3ThMmTNBLL72klJQUxcTEqF27dsrMzAxgxgAAAAAAACjsAl40K1q0qGJiYjyP8uXLSzrRy2zixIkaMWKEunXrpnr16ik5OVn//POPZs+eHeCsAQAAAAAAUJgFvGi2ZcsWxcbGqkqVKrr11lu1bds2SVJqaqrS0tLUvn17z2vdbrdatmyplStX5hvvyJEjysjI8HoAAAAAAAAAZyKgRbNmzZpp5syZWrBggV5//XWlpaWpefPm2rt3r9LS0iRJ0dHRXu+Jjo72PJeXsWPHKioqyvOIi4s7q9MAAAAAAACAwiegRbNOnTrpxhtv1CWXXKK2bdvqv//9ryQpOTnZ8xqXy+X1HmNMrraTDR8+XOnp6Z7Hrl27zk7yAAAAAAAAKLQCfnnmyUqUKKFLLrlEW7Zs8Yyi6exVtmfPnly9z07mdrsVGRnp9QAAAAAAAADORFAVzY4cOaKNGzeqYsWKqlKlimJiYrRo0SLP80ePHtWyZcvUvHnzAGYJAAAAAACAwq5oID/84YcfVpcuXXTxxRdrz549Gj16tDIyMtSzZ0+5XC4NHDhQSUlJSkhIUEJCgpKSkhQeHq4ePXoEMm0AAAAAAAAUcgEtmv3666+67bbb9Ndff6l8+fK6/PLLtWrVKlWuXFmSNHToUB06dEj9+vXT/v371axZMy1cuFARERGBTBsAAAAAAACFXECLZu+8884pn3e5XEpMTFRiYuK5SQgB4RqV/8AOp2JGGsuZAAAAAAAAnBBU9zQDAAAAAAAAggFFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcKBoBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4FA00AkANrhGuQr0PjPSWM4EAAAAAAAUBvQ0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAB7+LZocPH7aRBwAAAAAAABA0ClQ0y87O1tNPP62LLrpIJUuW1LZt2yRJTzzxhKZNm2Y1QQAAAAAAAOBcK1DRbPTo0ZoxY4bGjRun0NBQT/sll1yiqVOnWksOAAAAAAAACIQCFc1mzpypKVOm6Pbbb1eRIkU87fXr19fPP/9sLTkAAAAAAAAgEApUNPvtt99UvXr1XO3Z2dk6duyY30kBAAAAAAAAgVSgolndunX11Vdf5Wp/77331KhRI7+TAgAAAAAAAAKpaEHeNHLkSN1555367bfflJ2drblz52rTpk2aOXOmPvnkE9s5AgAAAAAAAOdUgXqadenSRe+++67+97//yeVy6cknn9TGjRv18ccfq127drZzBAAAAAAAAM6pAvU0k6QOHTqoQ4cONnMBAAAAAAAAgkKBepr16dNHycnJudozMjLUp08fv5MCAAAAAAAAAqlARbMZM2aoX79+GjBggLKzsz3thw4dyrOYBgAAAAAAAJxPClQ0k6T//ve/+vTTT9WhQwft37/fZk4AAAAAAABAQBW4aFanTh2tWrVKx44dU5MmTbRx40abeQEAAAAAAAABU6CBAFwulySpbNmy+vzzz3X//ffr8ssv1/PPP281OeBcco1yFeh9ZqSxnAkAAAAAAAi0AhXNjPm/IkHRokU1depU1alTR/369bOWGAAAAAAAABAoBSqaLV26VGXKlPFqGzx4sOrXr68VK1ZYSQwAAAAAAAAIlAIVzVq2bJlne9u2bdW2bVu/EgIAAAAAAAACrUBFs+PHj2vGjBlavHix9uzZo+zsbK/nlyxZYiU5AAAAAAAAIBAKVDR76KGHNGPGDF177bWqV6+eZ2AAAAAAAAAAoDAoUNHsnXfe0Zw5c9S5c2fb+QAAAAAAAAABF1KQN4WGhqp69eq2cwEAAAAAAACCQoGKZkOGDNGkSZNkjLGdDwAAAAAAABBwBbo8c/ny5Vq6dKk+/fRT1a1bV8WKFfN6fu7cuVaSAwAAAAAAAAKhQEWzUqVKqWvXrrZzAQAAAAAAAIJCgYpm06dPt50HAAAAAAAAEDQKdE8zScrKytLnn3+u1157TZmZmZKk33//XQcOHLCWHAAAAAAAABAIBepptmPHDnXs2FE7d+7UkSNH1K5dO0VERGjcuHE6fPiwXn31Vdt5AgAAAAAAAOdMgXqaPfTQQ2rcuLH279+vsLAwT3vXrl21ePFia8kBAAAAAAAAgVDg0TNXrFih0NBQr/bKlSvrt99+s5IYAAAAAAAAECgF6mmWnZ2t48eP52r/9ddfFRER4XdSAAAAAAAAQCAVqKdZu3btNHHiRE2ZMkWS5HK5dODAAY0cOVKdO3e2miBwPnGNchXofWaksZwJAAAAAADwR4F6mv3nP//RsmXLVKdOHR0+fFg9evRQfHy8fvvtNz377LMFSmTs2LFyuVwaOHCgp80Yo8TERMXGxiosLEytWrXShg0bChQfAAAAAAAA8FWBeprFxsZq3bp1evvtt7VmzRplZ2fr7rvv1u233+41MICvUlJSNGXKFNWvX9+rfdy4cZowYYJmzJihGjVqaPTo0WrXrp02bdrEZaAAAAAAAAA4awpUNJOksLAw9enTR3369PErgQMHDuj222/X66+/rtGjR3vajTGaOHGiRowYoW7dukmSkpOTFR0drdmzZ6tv375+fS4AAAAAAACQnwIVzWbOnHnK5++66y6fY/Xv31/XXnut2rZt61U0S01NVVpamtq3b+9pc7vdatmypVauXEnRDAAAAAAAAGdNgYpmDz30kNffx44d0z///KPQ0FCFh4f7XDR75513tGbNGqWkpOR6Li0tTZIUHR3t1R4dHa0dO3bkG/PIkSM6cuSI5++MjAyfcgEAAAAAAAByFGgggP3793s9Dhw4oE2bNunKK6/U22+/7VOMXbt26aGHHtKsWbNUvHjxfF/ncnmPRmiMydV2srFjxyoqKsrziIuL822iAAAAAAAAgP+vQEWzvCQkJOiZZ57J1QstP99995327Nmjyy67TEWLFlXRokW1bNkyvfDCCypatKinh1lOj7Mce/bsydX77GTDhw9Xenq657Fr166CTxQAAAAAAAAuSAUeCCAvRYoU0e+//+7Ta6+55hr98MMPXm29e/dWrVq1NGzYMFWtWlUxMTFatGiRGjVqJEk6evSoli1bpmeffTbfuG63W263u+ATAQAAAAAAgAtegYpm8+fP9/rbGKPdu3frpZdeUosWLXyKERERoXr16nm1lShRQmXLlvW0Dxw4UElJSUpISFBCQoKSkpIUHh6uHj16FCRtAAAAAAAAwCcFKprdcMMNXn+7XC6VL19ebdq00fjx423kJUkaOnSoDh06pH79+mn//v1q1qyZFi5cqIiICGufAQAAAAAAADgVqGiWnZ1tOw9J0hdffOH1t8vlUmJiohITE8/K5wHByjUq/8Eu8mNGGusxAAAAAAC4UFkbCAAAAAAAAAAoLArU02zw4ME+v3bChAkF+QgAAAAAAAAgYApUNFu7dq3WrFmjrKws1axZU5K0efNmFSlSRJdeeqnndS7XmV8eBgAAAAAAAARagYpmXbp0UUREhJKTk1W6dGlJ0v79+9W7d29dddVVGjJkiNUkAQAAAAAAgHOpQPc0Gz9+vMaOHespmElS6dKlNXr0aKujZwIAAAAAAACBUKCiWUZGhv74449c7Xv27FFmZqbfSQEAAAAAAACBVKCiWdeuXdW7d2+9//77+vXXX/Xrr7/q/fff1913361u3brZzhEAAAAAAAA4pwp0T7NXX31VDz/8sO644w4dO3bsRKCiRXX33Xfrueees5ogAAAAAAAAcK4VqGgWHh6uV155Rc8995y2bt0qY4yqV6+uEiVK2M4PAAAAAAAAOOcKdHlmjt27d2v37t2qUaOGSpQoIWOMrbwAAAAAAACAgClQ0Wzv3r265pprVKNGDXXu3Fm7d++WJN1zzz0aMmSI1QQBAAAAAACAc61Al2cOGjRIxYoV086dO1W7dm1Pe/fu3TVo0CCNHz/eWoIAAsc1ynXG7zEjvXucFiRGXnEAAAAAADiXClQ0W7hwoRYsWKBKlSp5tSckJGjHjh1WEgMAAAAAAAACpUCXZx48eFDh4eG52v/66y+53W6/kwIAAAAAAAACqUBFs6uvvlozZ870/O1yuZSdna3nnntOrVu3tpYcAAAAAAAAEAgFujzzueeeU6tWrfTtt9/q6NGjGjp0qDZs2KB9+/ZpxYoVtnMEAAAAAAAAzqkC9TSrU6eO1q9fr6ZNm6pdu3Y6ePCgunXrprVr16patWq2cwQAAAAAAADOqTPuaXbs2DG1b99er732mkaNGnU2cgIAAAAAAAAC6ox7mhUrVkw//vijXC7X2cgHAAAAAAAACLgCXZ551113adq0abZzAQAAAAAAAIJCgQYCOHr0qKZOnapFixapcePGKlGihNfzEyZMsJIcAAAAAAAAEAhnVDTbtm2b4uPj9eOPP+rSSy+VJG3evNnrNVy2CQAAAAAAgPPdGRXNEhIStHv3bi1dulSS1L17d73wwguKjo4+K8kBgCS5RhWsGG9GGsuZAAAAAAAuFGd0TzNjvH+Afvrppzp48KDVhAAAAAAAAIBAK9BAADmcRTQAAAAAAACgMDijopnL5cp1zzLuYQYAAAAAAIDC5ozuaWaMUa9eveR2uyVJhw8f1v33359r9My5c+fayxAAAAAAAAA4x86oaNazZ0+vv++44w6ryQAAAAAAAADB4IyKZtOnTz9beQAAAAAAAABBw6+BAAAAAAAAAIDCiKIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBQNNAJAMC54BrlKtD7zEhjORMAAAAAwPmAnmYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAODB6JgD4iBE4AQAAAODCEdCeZpMnT1b9+vUVGRmpyMhIXXHFFfr00089zxtjlJiYqNjYWIWFhalVq1basGFDADMGAAAAAADAhSCgRbNKlSrpmWee0bfffqtvv/1Wbdq00fXXX+8pjI0bN04TJkzQSy+9pJSUFMXExKhdu3bKzMwMZNoAAAAAAAAo5AJaNOvSpYs6d+6sGjVqqEaNGhozZoxKliypVatWyRijiRMnasSIEerWrZvq1aun5ORk/fPPP5o9e3Yg0wYAAAAAAEAhFzQDARw/flzvvPOODh48qCuuuEKpqalKS0tT+/btPa9xu91q2bKlVq5cGcBMAQAAAAAAUNgFfCCAH374QVdccYUOHz6skiVL6sMPP1SdOnU8hbHo6Giv10dHR2vHjh35xjty5IiOHDni+TsjI+PsJA4AAAAAAIBCK+BFs5o1a2rdunX6+++/9cEHH6hnz55atmyZ53mXy3u0OmNMrraTjR07VqNGjTpr+QKAP2yMwMkongAAAABw9gX88szQ0FBVr15djRs31tixY9WgQQNNmjRJMTExkqS0tDSv1+/ZsydX77OTDR8+XOnp6Z7Hrl27zmr+AAAAAAAAKHwCXjRzMsboyJEjqlKlimJiYrRo0SLPc0ePHtWyZcvUvHnzfN/vdrsVGRnp9QAAAAAAAADOREAvz3zsscfUqVMnxcXFKTMzU++8846++OILffbZZ3K5XBo4cKCSkpKUkJCghIQEJSUlKTw8XD169Ahk2gAAAAAAACjkAlo0++OPP3TnnXdq9+7dioqKUv369fXZZ5+pXbt2kqShQ4fq0KFD6tevn/bv369mzZpp4cKFioiICGTaAAAAAAAAKOQCWjSbNm3aKZ93uVxKTExUYmLiuUkIAC4QtgYTKEgcBiQAAAAAcD4IunuaAQAAAAAAAIFG0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOAR09EwAAGyNwBipGXnEAAAAAFA70NAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAgdEzAQCwhBE4AQAAgMKDnmYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAODB6JgAAQYQROAEAAIDgQE8zAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAODAQAAAAhQyDCQAAAAD+o6cZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOBQNdAIAACD4uEa5CvQ+M9JYzgQAAAAIDHqaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGBAAAAwFnBYAIAAAA4n9HTDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHRs8EAABBy9YInAWJwyieAAAAFzZ6mgEAAAAAAAAOAS2ajR07Vk2aNFFERIQqVKigG264QZs2bfJ6jTFGiYmJio2NVVhYmFq1aqUNGzYEKGMAAAAAAABcCAJaNFu2bJn69++vVatWadGiRcrKylL79u118OBBz2vGjRunCRMm6KWXXlJKSopiYmLUrl07ZWZmBjBzAAAAAAAAFGYBvafZZ5995vX39OnTVaFCBX333Xe6+uqrZYzRxIkTNWLECHXr1k2SlJycrOjoaM2ePVt9+/YNRNoAAAAAAAAo5ILqnmbp6emSpDJlykiSUlNTlZaWpvbt23te43a71bJlS61cuTLPGEeOHFFGRobXAwAAAAAAADgTQTN6pjFGgwcP1pVXXql69epJktLS0iRJ0dHRXq+Njo7Wjh078owzduxYjRo16uwmCwAALjg2RuBkFE8AAIDzR9D0NHvggQe0fv16vf3227mec7m8DzCNMbnacgwfPlzp6emex65du85KvgAAAAAAACi8gqKn2YMPPqj58+fryy+/VKVKlTztMTExkk70OKtYsaKnfc+ePbl6n+Vwu91yu91nN2EAAAAAAAAUagHtaWaM0QMPPKC5c+dqyZIlqlKlitfzVapUUUxMjBYtWuRpO3r0qJYtW6bmzZuf63QBAAAAAABwgQhoT7P+/ftr9uzZmjdvniIiIjz3MIuKilJYWJhcLpcGDhyopKQkJSQkKCEhQUlJSQoPD1ePHj0CmToAAAAAAAAKsYAWzSZPnixJatWqlVf79OnT1atXL0nS0KFDdejQIfXr10/79+9Xs2bNtHDhQkVERJzjbAEAAAAAAHChCGjRzJjTjwblcrmUmJioxMTEs58QAABAkCvICJwSo3ACAACcqaAZPRMAAAAAAAAIFhTNAAAAAAAAAAeKZgAAAAAAAIADRTMAAAAAAADAIaADAQAAAODcYzABAACA06OnGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomgEAAAAAAAAOjJ4JAACAM2ZjBM7CFAMAABQ+9DQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIHRMwEAAAA/2RqBsyBxzkYMAABATzMAAAAAAAAgF4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgwEAAAAAAALwEakCCvOIAABAo9DQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIHRMwEAAAAEJRsjcDKKJwCgoOhpBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADo2cCAAAAwCkwAicAXJjoaQYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA6NnAgAAAMBZZmMETkbxBIBzi55mAAAAAAAAgANFMwAAAAAAAMCBohkAAAAAAADgQNEMAAAAAAAAcGAgAAAAAAC4QNgaTKAgcRiQAMD5hp5mAAAAAAAAgENAi2ZffvmlunTpotjYWLlcLn300UdezxtjlJiYqNjYWIWFhalVq1basGFDYJIFAAAAAADABSOgRbODBw+qQYMGeumll/J8fty4cZowYYJeeuklpaSkKCYmRu3atVNmZuY5zhQAAAAAAAAXkoDe06xTp07q1KlTns8ZYzRx4kSNGDFC3bp1kyQlJycrOjpas2fPVt++fc9lqgAAAAAAALiABO09zVJTU5WWlqb27dt72txut1q2bKmVK1fm+74jR44oIyPD6wEAAAAAAACciaAdPTMtLU2SFB0d7dUeHR2tHTt25Pu+sWPHatSoUWc1NwAAAACAf2yMwMkongDOpqDtaZbD5fLeCBpjcrWdbPjw4UpPT/c8du3adbZTBAAAAAAAQCETtD3NYmJiJJ3ocVaxYkVP+549e3L1PjuZ2+2W2+0+6/kBAAAAAACg8AranmZVqlRRTEyMFi1a5Gk7evSoli1bpubNmwcwMwAAAAAAABR2Ae1pduDAAf3yyy+ev1NTU7Vu3TqVKVNGF198sQYOHKikpCQlJCQoISFBSUlJCg8PV48ePQKYNQAAAAAAAAq7gBbNvv32W7Vu3drz9+DBgyVJPXv21IwZMzR06FAdOnRI/fr10/79+9WsWTMtXLhQERERgUoZAAAAAAAAF4CAFs1atWolY/IfucTlcikxMVGJiYnnLikAAAAAAABc8IL2nmYAAAAAAABAoFA0AwAAAAAAABwomgEAAAAAAAAOFM0AAAAAAAAAh4AOBAAAAAAAQCC5RrkK9D4z8v8GtbMRA0DwoacZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA6MngkAAAAAQIAxiicQfOhpBgAAAAAAADhQNAMAAAAAAAAcKJoBAAAAAAAADhTNAAAAAAAAAAeKZgAAAAAAAIADo2cCAAAAAABJjMAJnIyeZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4MHomAAAAAACwxtYInAWJczZi4MJFTzMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBQNNAJAAAAAAAABCvXKNcZv8eMNGchE5xr9DQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIHRMwEAAAAAAM4iRuA8P9HTDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4MBAAAAAAAABDkCjKYgOQ9oICNGBcSepoBAAAAAAAADudF0eyVV15RlSpVVLx4cV122WX66quvAp0SAAAAAAAACrGgL5q9++67GjhwoEaMGKG1a9fqqquuUqdOnbRz585ApwYAAAAAAIBCKuiLZhMmTNDdd9+te+65R7Vr19bEiRMVFxenyZMnBzo1AAAAAAAAFFJBPRDA0aNH9d133+nRRx/1am/fvr1WrlyZ53uOHDmiI0eOeP5OT0+XJGVkZJy9RAPhcMHe5jUfiBGcMQoYpzDFyBWHGMEZo4BxClOMXHGIEZwxChinMMXIFYcYwRmjgHEKU4xccYgRnDEKGKcwxcgVhxgsZ0EaI1ccW9/NeS5neow59QAHLnO6VwTQ77//rosuukgrVqxQ8+bNPe1JSUlKTk7Wpk2bcr0nMTFRo0aNOpdpAgAAAAAA4Dyza9cuVapUKd/ng7qnWQ6Xy3tIVGNMrrYcw4cP1+DBgz1/Z2dna9++fSpbtmy+7ylMMjIyFBcXp127dikyMpIYxAjaXIhBjPMlF2IQ43zJhRjEOF9yIQYxzpdciEGM8yWXYIlxPjHGKDMzU7Gxsad8XVAXzcqVK6ciRYooLS3Nq33Pnj2Kjo7O8z1ut1tut9urrVSpUmcrxaAVGRnp94JOjMIbI5hyIQYxzpdciEGM8yUXYhDjfMmFGMQ4X3IhBjHOl1yCJcb5Iioq6rSvCeqBAEJDQ3XZZZdp0aJFXu2LFi3yulwTAAAAAAAAsCmoe5pJ0uDBg3XnnXeqcePGuuKKKzRlyhTt3LlT999/f6BTAwAAAAAAQCEV9EWz7t27a+/evXrqqae0e/du1atXT//73/9UuXLlQKcWlNxut0aOHJnrElViECPYciEGMc6XXIhBjPMlF2IQ43zJhRjEOF9yIQYxzpdcgiVGYRTUo2cCAAAAAAAAgRDU9zQDAAAAAAAAAoGiGQAAAAAAAOBA0QwAAAAAAABwoGgGAAAAAAAAOFA0AwAAAAAAABwomp3nMjIy8nxkZmbq6NGjgU7vvJWamqotW7bkat+yZYu2b99+7hMKsP/9739asGBBrvYFCxbo008/DUBGyJGenq59+/blat+3b58yMjICkFFg2Vh3g2me9unTR5mZmbnaDx48qD59+vgUw8Z+4ssvv1RWVlau9qysLH355ZfnLIYNNvLYuXOn8hp83BijnTt3+p3jmZg5c6aOHDmSq/3o0aOaOXOmTzFsH0scOnQoV6zTeeqpp/TPP//kGeupp5464xz8ESzrnQ02lg/kZmNfEyzHmlWrVtXevXtztf/999+qWrWqTzFsrDPBsqza2hbZ2NcEyzyxwcYyAgSKy+R11IfzRkhIiFwuV77PV6pUSb169dLIkSMVEpJ/jbRr1655xnG5XCpevLiqV6+uHj16qGbNmvnGuOmmm9S4cWM9+uijXu3PPfecVq9erffee++001O6dOnT5tGrVy/17t073xiNGjXyKUbr1q3zjdGyZUv16dNHPXv29GqfNWuWpk6dqi+++OK00zJgwABVr15dAwYM8Gp/6aWX9Msvv2jixImnjWFjno4dO1bR0dG5dkhvvPGG/vzzTw0bNuy0MerXr69nnnlGnTt39mr/7LPPNGzYMH3//ffnZFokO/PVxjJiY3psrHedOnVSly5d1K9fP6/2V199VfPnz9f//ve/0+ZhY55K0vvvv685c+Zo586duX4Urlmz5rTvnz9/fp7tJ8+PKlWqnDKGjXU3mOZpkSJFtHv3blWoUMGr/a+//lJMTEyeB+RONvYT+eWxd+9eVahQQcePHy/wtJxJDEn69ddfNX/+/DyXswkTJpyTPGzEsLG/s5WLjWXkn3/+0dChQzVnzpw8f4CfLg9by8fgwYPzbD95vl5//fUqU6bMGedyrtc7G/sqW/P14MGDWrZsWZ7rnXM7lx9/113JzjyR/N9f2djX2Ihh4zgiJCREaWlpuZaRP/74QxdffHGeRRsnG+uMrWXV3+/WVh7Bsq+R/J8nkv/bABvLiGRnO2IjhiT99NNPeca47rrrfHp/SkqK3nvvvTxjzJ071+c8/P1ubGzfCz2D81pycrKpVKmSefzxx838+fPNvHnzzOOPP27i4uLMa6+9ZkaPHm1KlSplxowZc8o4PXv2NFFRUaZy5cqmW7dupmvXriY+Pt6UKlXK3HLLLaZmzZrG7Xab5cuX5xujXLlyZv369bna169fbypUqODT9EyYMMGULVvW3HHHHeaFF14wkyZNMnfccYcpV66cGTNmjLnnnnuM2+02U6ZMyTfGo48+aqKiosyVV15pBg8ebAYNGmSuuuoqExUVZR566CHTrl07ExISYj766KN8Y0RERJgtW7bkat+yZYuJioryaVpiY2PNt99+m6v9u+++MxdddJFPMWzM08qVK5sVK1bkal+1apWJj4/3KUbx4sVNampqrvbU1FQTHh7uUwwb02KMnflqYxmxMT021rvSpUubn376KVf7xo0bTZkyZXzKw8Y8nTRpkilZsqTp37+/CQ0NNX379jVt27Y1UVFR5rHHHvMphsvlMiEhIcblcnk9ctpCQkLM1Vdfbfbt25dvDBvrbjDM0/T0dPP3338bl8tlfvnlF5Oenu557Nu3zyQnJ5uKFSv6lIuN/YTL5TJ79uzJ1b5p0yYTERHhUx42Ynz++ecmPDzc1K1b1xQtWtQ0bNjQlCpVykRFRZnWrVufszzyi7F9+3aft4k29nenymXdunWmdOnSPuViYxnp16+fqV27tnnvvfdMWFiYeeONN8zTTz9tKlWqZGbNmnXaHPKbjsWLF5ty5cr5NB3GGNOqVSsTGRlpSpQoYS699FLTqFEjU7JkSRMVFWWaNWtmSpUqZUqXLm02bNiQ673Btt7Z2FfZWD7WrFljYmJiTGRkpClSpIgpX768cblcpkSJEqZKlSo+xbCx7hpjZ57Y2F/Z2NfYiOHPccS8efPMvHnzjMvlMjNnzvT8PW/ePDN37lzTv39/U6NGjVN+vs11xsayautYxMa26Gzua871PPFnG2BzGbGxHbERY+vWraZ+/fq5jltDQkJMSEiITzHefvttU6xYMXPttdea0NBQ869//cvUrFnTREVFmV69evkUwxj/t882tu8XAopm57k2bdqYd999N1f7u+++a9q0aWOMMWbmzJmmZs2ap4wzbNgw8+9//9scP37c03b8+HHzwAMPmOHDh5vs7Gxz3333mRYtWuQbo3jx4ubnn3/O1b5x40ZTvHhxn6anW7duZvLkybnaX331VdOtWzdjjDEvvPCCqVevXr4x7rnnHvPUU0/lan/66afNPffcY4wx5sknnzSXXXZZvjEiIyPNmjVrcrV/++23pmTJkqedDmOMcbvd+R4Mud1un2LYmKdut9ts27YtV/vWrVt9ziM6OtosXrw4V/uiRYtM+fLlfYphY1qMsTNfbSwjNqbHxnoXHh6eb/EuLCzMpzxszNOaNWua2bNnG2OMKVmypNm6dasxxpgnnnjC9O/f36cYn3/+uWnWrJn5/PPPTUZGhsnIyDCff/65ufzyy81///tfs3z5clO3bl3Tp0+ffGPYWHeDYZ6efPCV16NIkSJm9OjRPuXiz36ia9eupmvXriYkJMR07tzZ83fXrl3NddddZ+Lj402HDh1O+fk2YuRo0qSJeeKJJ4wx/7ecZWZmmuuuu8688sorZz2PQYMGmUGDBpmQkBDTt29fz9+DBg0yAwYMMM2aNTPNmzf3aVr83d81bNjQNGrUyISEhJhLLrnENGrUyPOoX7++iYiIMDfffLNPudg4loiLizNLly41xngXA2bOnGk6deqU7/tyilghISGe/+c8IiMjTUhIiOnXr59P02GMMf/5z39Mt27dTHp6uqctPT3d3HTTTWbixInm4MGD5vrrrzft27fP9d5gWe9y+LOvsrl8tGzZ0tx7770mKyvLs97t3LnTXH311eaDDz7wKYY/6+7JbOy/beyvbOxrbMTw5zjCeWLq5EdoaKipUaOG+fjjj0/5+TbWGZvLqj/fra1tkY19TbDMkxz+bANsbldtbEdsxPjXv/5lrr/+erNnzx5TsmRJ89NPP5mvvvrKNG3a1Hz55Zc+xbjkkkvMSy+95JVHdna2uffee82TTz7pUwxj/N8+29i+Xwgomp3nwsLCzObNm3O1b9682fPjbtu2baf9oVeuXDmzadOmXO2bNm0yZcuWNcac+MF4qjNfjRs3NqNGjcrVPnLkSHPppZee8vNzlChRIt8fmiVKlDDGGPPLL7+c8kx+ZGRkvjEiIyONMScKG6c6ILn22mvNzTffbLKysjxtWVlZ5sYbbzQdO3b0aVrq1q1rXnzxxVztL7zwgqldu7ZPMWzM0+rVq5s333wzV/vMmTN9PoNw7733mksuucT88ssvnrYtW7aY+vXrm7vvvtunGDamxRg789XGMmJjemysdy1btjQPPPBArvZ+/fqZK6+80qc8bMzTsLAws337dmOMMeXLlzfr1q0zxpzYFvnaO6tu3bp59opcvny5qVOnjjHmRKE2Li4u3xg21t1gmKdffPGFWbp0qXG5XGbu3Lnmiy++8DxWrlxpfvvtN5/yMMa//USvXr1Mr169jMvlMt27d/f83atXL3PfffeZpKQk8+eff57y823EyFGyZEnPdqhUqVLmxx9/NMacOOteuXLls55Hq1atTKtWrYzL5TLNmzf3/N2qVSvTvn17c9999+U5r/Pi7/4uMTHRJCYmGpfLZR5++GHP34mJiSYpKcnMnj3bHDlyxKdcbBxLlChRwrMNuOiii8w333zjeV/O9ORlxowZZvr06cblcplJkyaZGTNmeB6zZ882K1eu9GkacsTGxubZi+zHH380sbGxxpgTPT5ztrEnC5b1Loc/+yqby0dUVJTnJFFUVJSnJ+6qVatOe1I2hz/r7sls7L9t7K9s7GtsxLBxHBEfH+/zNtjJxjpje1tW0O/W1rbIxr4mWOZJDn+2ATa3qza2IzZilC1b1nz//ffGmBPbpJx5s3jxYtOwYUOfYoSHh3uu4ilbtqznZO1PP/1kYmJifIphjP/bZxvb9wtB0UBfHgr/VKpUSdOmTdMzzzzj1T5t2jTFxcVJOnHde+nSpU8ZJysrSz///LNq1Kjh1f7zzz97rpcvXrz4Ke/P8cQTT+jGG2/U1q1b1aZNG0nS4sWL9fbbb/t8v6oyZcro448/1qBBg7zaP/74Y8/9Rw4ePKiIiIh8YxQvXlwrV65U9erVvdpXrlyp4sWLS5Kys7PldrvzjTFu3DhdffXVqlmzpq666ipJ0ldffaWMjAwtWbLEp2kZPHiwHnjgAf35559e82P8+PE+38/Ixjy95557NHDgQB07dswrxtChQzVkyBCfYjz33HPq2LGjatWqpUqVKkk6cT+Aq666Ss8///w5mxbJzny1sYzYmB4b692YMWPUtm1bff/997rmmms8eaSkpGjhwoU+5WFjnsbExGjv3r2qXLmyKleurFWrVqlBgwZKTU3N84bpedm6dasiIyNztUdGRmrbtm2SpISEBP3111/5xrCx7gbDPG3ZsqWkEzeKjouLO+U9KU/Hn/3E9OnTJUnx8fF65JFHFB4efsafbyNGjhIlSnjurxMbG6utW7eqbt26knTK5cKZx8MPP6wSJUqc8ecvXbpUktS7d29NmjQpz+XVV/7u70aOHCnpxPR0797ds+0qCBvHElWrVtX27dtVuXJl1alTR3PmzFHTpk318ccfq1SpUvm+L+d+TlWqVFHz5s1VrFixAk+HdGIgjz179qhOnTpe7X/++adnQIJSpUrleTP+k9e7iy+++JTb3tOxMU/92VedvHzceuutp9yfnU6xYsU88yI6Olo7d+5U7dq1FRUV5fPAF/6suyezsf+2sb+ysa+xEcPGcURqaqpPn5UXG+uMzWXVn+/25G1RixYtVLRowX4m29jn2dy+21je/dkG2Nyu2tiO2Ihx/PhxlSxZUpJUrlw5/f7776pZs6YqV66sTZs2+RSjTJkynoERLrroIv3444+65JJL9Pfff+c5EEV+/N0+29i+XxACW7ODv+bNm2dCQ0M9PX7uuece06BBA+N2uz1dql955RUzaNCgU8Z58MEHTbly5cyECRPMV199ZZYvX24mTJhgypUrZwYMGGCMMeb1118/5WVixhjzySefmObNm5vw8HBTtmxZ07p1a/PFF1/4PD1TpkwxRYoUMV26dDFPP/20GT16tLnuuutM0aJFzdSpU40xxjz//PPmlltuyTfG008/bcLCwsyAAQPMm2++aWbNmmUGDBhgwsPDPd1/J0yYYNq2bXvKXH777TczfPhw07lzZ3PjjTeaUaNGmb179/o8LcacmPcXXXSRp7t7lSpVTHJy8hnF8HeeZmdnm6FDh5rixYt7ukKHh4ebUaNGmezs7DOKs2DBAjNu3Djz4osvmmXLlp3RdNiYlhz+zldby4i/02NrvVu7dq3p0aOHqVOnjrnssstM7969fe7tksPfeXr33XebxMREY4wxkydPNmFhYaZt27amVKlSp7yc8mQtWrQwHTt29Lp/x549e0zHjh3NVVddZYw50dMsISHhlHFsrLvBME9PdvDgQbNx40bz/fffez18YWM/0bp1a7N///5c7enp6T7fA2Tbtm359rzJ656Jebn++us99/h65JFHTPXq1c3o0aPNpZdeaq655hqfYtjQu3dvk5GRkav9wIEDpnfv3j7FsLG/s8XGMjJhwgQzadIkY4wxS5YsMWFhYSY0NNSEhISYiRMnnjaHk+9zc/IjIyPD5x4VxhjTo0cPU6VKFTN37lyza9cu8+uvv5q5c+eaqlWrmjvuuMMYc+JeMqe6fM+5nuU81q9fbzZv3mwOHz582jxszFMb+6rVq1ebVatW5WpftWqVSUlJOe10GGNMu3btzFtvvWWMMaZv376madOmZtasWaZDhw6madOmPsWwte7amCc29lfG2NnX+BvDxnHEpEmT8ny88MILZsqUKWbJkiVeveHy8sYbb5g5c+bkap8zZ46ZMWOGT9NiY1m18d2GhISYP/74I1f7X3/95fP9qoyxs9+0wcY8sbENsLFdtbEdsRHjyiuvNB9++KExxpjbbrvNdOzY0Sxfvtzcddddpm7duj7FuO2228z48eONMcaMHj3alC9f3txzzz2mcuXKpmvXrj7FMMb/78bGd3shoGhWCGzfvt08+uijpmvXruaGG24wjz76qM8/QHJkZWWZ0aNHm5iYGM8PvJiYGDNmzBjPjnLHjh1m165dZ2EKvC1fvtzceuutplGjRqZhw4bm1ltvzfOSrVOZNWuWufzyyz33Irj88ss9GwRjjPnnn3/MoUOHbKeerz179pjMzMxz9nl5yczMNKtXrzY//PCDTzum84E/8zUYlpFgWu9yFHSeHj9+3Bw7dszz97vvvmsefPBBM2nSJJ9/8P7888+mZs2aJjQ01FSrVs1Ur17dhIaGmlq1ankuP/nwww/NzJkzzzi/QPJnOd2zZ4+59tpr870XiK9SU1PNsGHDCryfyO9HxB9//GGKFi3qU4yrr746zx9Pb775pmnZsqVPMbZu3eopFh48eND8+9//Npdcconp2rWr5/ITX7z33nvm5ptvNs2aNfO6V0yjRo18en9+8+PPP/80RYoU8TkPG/u7rKws89xzz5kmTZqY6Ohor/vw+HqjaGP8X0acduzYYT744APPpUCnc7r73lx88cXmySef9Lp3U14yMzPNPffc4ynYhYSEmNDQUHPvvfd61sO1a9eatWvXFjgXt9tt7rrrrtPuI1JTU/0+PvN3X9WkSRPz3nvv5Wr/4IMPfP5BlJKSYpYsWWKMObFN6tSpk4mIiDCNGjU65Xw8ma111xj/54mN/VWwsHEcER8fb0qUKGFcLpcpU6aMKV26tOdG4NHR0cblcplq1aqZnTt35hujRo0anmXkZF988cVpBxPIYWNZtfHdulyuPLfvv/322xndg9fGftPG9t3GPLGxDbCxXbWxHbER47PPPvPc72vr1q2mdu3axuVymXLlypnPP//cpxh79+71XJp6/Phx8+yzz5ouXbqYQYMGnXLAKyd/v5tTvd/X/feFwGWMj/0yccHIuXzBn8tOCot//vknz+F369evf9r3tmnTRnPnzs11WUpGRoZuuOEGn7rep6SkKDs7W82aNfNq/+abb1SkSBE1btz4tDH69OmjSZMm5brE5+DBg3rwwQf1xhtvnDbGCy+8kGf7yUOaX3311SpSpMhZnRbJznwNRgVd73Le5+RyueR2uxUaGup3bueSMUYLFizQ5s2bZYxRrVq11K5dO58vT5w+fbpKliypm2++2av9vffe0z///OO5/OJUbAzxnpqaqqysLCUkJHi1b9myRcWKFVN8fPzpJ0bS7bffru3bt2vixIlq3bq1PvzwQ/3xxx8aPXq0xo8fr2uvvdanOAW1fv16SVLDhg21ZMkSz2WD0onLEz777DO99tpr2r59+2ljRUZGas2aNbkuq/rll1/UuHFj/f333zZTz9cLL7ygESNGqGfPnnr99dfVu3dvbd26VSkpKerfv7/GjBmT73szMjJkjFHp0qW1ZcsWlS9f3vPc8ePH9fHHH+vRRx/V77//fi4mRZL05JNPaurUqRo8eLCeeOIJjRgxQtu3b9dHH32kJ5988rwZLn7mzJkaMWKEevXqpaZNm8oYo5SUFCUnJ+vxxx/Xn3/+qeeff16PPPKIHnvssdPGO3DggLZt2yZjjKpVq+a5lMYX8+bN07Bhw/TII4945TJ+/HiNHDlSWVlZevTRR9W9e3efb1EQKCVLltT69etVtWpVr/bU1FTVr1/fc3kQzkzOttEp57jo4osvPu1lhjZinKygxxFvv/22pkyZoqlTp6patWqSTmyX+/btq/vuu08tWrTQrbfeqpiYGL3//vt5xihevLh+/vnnXPu27du3q3bt2jp06NBp8wj0sppzrDto0CA9/fTTXtuM48eP68svv9T27du1du3aU8axud8sLNt3qXBtV/Oyb98+lS5d2q/LTxHEAlaug1X+XLpjjJ0uxDbOhtjqEu0vG7078jtTdSZnmGycdbPRI8LGWUgb02KMnflqg43l3cZ6Z6t3hr89b3L4uy3yl42z3TbOMtvoVWWMMTExMZ4bqkdERHh63M2bN++0l+3m8OeSiJOXL+foai6Xy4SHh5tp06b5lIeNkeJs8GckMZujgNm6HLFq1armk08+8UxPzg2OJ02aZG677TafYti4bMbfS7xsjQZu49LZJk2amM8++yxX+2effWaaNGlijDnR67Vq1ar5xgiWy9XKlCmT5w3MV6xYYUqVKuVTDBuXVtuYFltxbCzveW0bz7TXjI0YNr6bqlWr5tkjZc2aNZ4Bo1asWHHKm5PHxcWZefPm5Wr/6KOPzEUXXeRTHjaWVX/Wu/j4eBMfH29cLpeJi4vz/B0fH29q1Khh2rdvn+ey52Rzvxks23cbx6s2tqs22NiG2NjPFLbLgAs7imbnOVuX7tgoRDzxxBOmYsWK5rnnnjPFixc3Tz/9tLn77rtN2bJlPfc5KWgeZ/Jj1UYxo0ePHqZ58+Zm9erVpkSJEmbhwoXmzTffNDVr1vTsvPKTszNyuVxm6dKlXjuoNWvWmKSkJJ9HZylRooTnB93Jtm3bdtofmenp6ebvv/82LpfL/PLLL14/yvbt22eSk5NNxYoVfcpj9uzZplWrVrlGz2zTpo155513zK5du0yLFi3MjTfeeFamxRi789XGMnI2l/czWe+Sk5NNpUqVzOOPP27mz59v5s2bZx5//HETFxdnXnvtNTN69GhTqlQpM2bMmHxjTJo0yZQsWdL079/fhIaGmr59+5q2bduaqKgo89hjj/mUh41tkY37qrjd7jx/KKSmpp52G5LzeSEhIWbMmDFeOUyYMMHccMMNPo+KFBERke8Ib6caxSyvODnTU7lyZbN8+XJjjG+jIufw54fZ9u3bTWpqqnG5XCYlJcVs377d8/j9999P+32czMZIcTbWXX9GEjt5FLAPPvggz1HAduzY4VMetgre4eHhns+MiYkx3333nTHmxCUjOaMJnkkuBf3x7u/JFVujgds4UVS8eHGzcePGXO0bN270bEdSU1NPmUuwXK7WvXt307JlS/P333972vbv329atmxpbr75Zp9i2DgJYOvEmY04Ni4T++ijj0zNmjXN1KlTzfr16833339vpk6damrXrm3eeecdM2vWLFOpUiUzZMiQsxrDxncTFhaWZ7Fg9erVnmU8NTX1lKPgPvLII6Zy5cqe/XRWVpZZvHixqVy58inzP5mNZdXGeteqVas8ixC+Onm/+e233/q137S9fS/o8m7jeNXGdtXGMUCwdEiwdRmwje/G1onzwoyi2XnOn+KOMXYLEf6cDbH5Y9VGMcOf3h02zzD5c9bNZo8IG2ch/T2DaHO+2lhG/Fneba53Nnpn+NPzJoe/2yJj7PRo9Odst62zzMbY61XVuHFjz5nZ66+/3tx5553m119/NUOHDvX5bKyNH2Y5NmzYYD799FMzb948r4cvfvzxR1O2bFlTrVo106tXL9OrVy9TrVo1U758efPDDz/4FMPGululShXPD4/GjRubV1991RhjzIIFC3w+6Ha5XOb333/P1X4mZ4htFLyNOfEjMWe5vPLKK83YsWONMca88847pnz58j7lYmMZ8ffkSkJCghk2bFiu9mHDhnl+7KakpJjY2Ng832/zRFHDhg1Nz549vXr8HT161PTs2dNzLLJ8+XITHx+fbwx/Cvg5/D3ZZIwxv/76q6lataqJiooyrVq1Mq1atTKlSpUyNWvWPOW29GQ2TgLYmBZbcWws7zZ6zdiIYeO76dy5s7n00ku99llr1qwxl112mbn22muNMcbMnz/f1KtXL98YR44cMbfccotxuVymWLFiplixYiYkJMT07t3b5/vo2lhWbax3AwcONIMGDcr1GDx4sHnsscfMG2+84dNgDWPGjMnzuHTatGnmmWee8SmXQG/fbR6v2tiu2jgGCHSHBFu/eW19NzZOnF8IKJqd5/y9dMdmIcKfsyE2f6za6MrsT+8Omz0z/DnrZrNHhI2zkP6eQbR55s7GMuLP8m5zvbPRO8Ofnjc5bFxGaKNHo42z3f6eZTbGTq8qY07c8Hr69OnGmBM/YsqXL29cLpdxu93mnXfe8SmGjR9m27ZtMw0aNPAsu87eSL7YsWNHviPF+botsrHu2hhJzOVyeY3ymmP79u0mPDzcpxi2LkccNmyYp7D23nvvmaJFi3oG0cirCJUXG8uIvydX/B1t0uaJohUrVpiyZcua8uXLm2uuuca0bdvWVKhQwZQtW9Z8/fXXxpgT3824cePyjREsl6sZc+KSoddee83069fPDBkyxCQnJ5ujR4/6/H4bJwFsTYuNODaWdxu9ZmzEsPHd7N6927Rt29a4XC4TGhpqQkNDjcvlMu3atTO7d+82xpwYEXfBggWnjbV582YzZ84c8/HHH5/xAA/G+L+s2ljvWrVqZSIjI02JEiXMpZdeaho1amRKlixpoqKiTLNmzUypUqVM6dKlzYYNG04Zp3LlynkO7LJq1apTFoZOFujtu83jVRvbVRvHAIHukBBslwHbOHF+IaBodp7z99Idm4UIG2dDbPxYtdGV2Ubvjhz+9MywcdbNRo8IG2chbUyLMXbO3NlYRvxZ3m0WVv3tnWGMnZ43Ni4jtNGj0cbZbhtnmW30qsrLwYMHzXfffWf+/PNPn99j44fZv/71L3P99debPXv2mJIlS5oNGzaYr776yjRt2tR8+eWXPuVh4/4dNtbdbdu2eZ3pPnkksbwK0CfLWRZCQkJM3759vZaPAQMGmGbNmpnmzZv7lIetyxGdVq1aZcaPH+/zfsYYO8uIvydXduzYYbZu3ZprtMmtW7f6VFS1eaLImBOjcE6ePNkMGjTIDBw40Lz66qt53sMmP8FyuVpSUpLf+0wbJwFsTIutODaWdxu9ZmzEsHWCxpgTI1jPmzfPfPTRR+bnn38+o/fmtc8sSO8sY/w7brax3v3nP/8x3bp1M+np6Z629PR0c9NNN5mJEyeagwcPmuuvv960b9/+lHHcbrfZtm1brvatW7cat9vtUy5O53r7bvN3ojH+b1dtHAMES4cEm5cB+/NbwsaJ8wtB0UAPRAD/1KxZU5s2bVJ8fLwaNmyo1157TfHx8Xr11VdVsWLF076/cuXKkqTRo0fr+++/V58+fbyef+ONN/Tnn39q2LBhp43VtWtXLV68WM2aNdNDDz2k2267TdOmTdPOnTs1aNAgn6anYcOGeuqpp3K1nzxK4/XXX+81Eo1TpUqVtHv3bl188cWqXr26Fi5cqEsvvVQpKSk+j0I0cOBA7d69W5I0cuRIdejQQbNmzVJoaKiSk5N9ipGamqquXbtq/fr1crlcMv9/oNqcUVV8GYHvoosu0vr16/XWW2/p+++/V1hYmHr37q3bbrtNxYoV8ykPSSpaNPeqfuDAARUvXtyn90+bNk133nmnLrvsMs/nHjt2TG3bttXUqVMlnRj1aPz48Wd9WqZMmaLZs2fnaq9bt65uvfVWn5ZVG8uIP8t7znqXnZ0tSfrpp5/yHKX1uuuuO20ezz//vG6++WZ9+umnatKkiVwul1JSUvTzzz97RrlKSUlR9+7d843Rpk0bffzxx7r00kt19913a9CgQXr//ff17bffqlu3bqfNQfJ/WyRJu3fvVlZWVq72rKwspaWlSZJiY2NPOYJWaGio3n33XY0ePVrr1q1TWFiYLrnkEs8898W6deu0Zs0aHT9+XDVr1pQxRlu2bFGRIkVUq1YtvfLKKxoyZIiWL1+uOnXq5BkjIiJC69ev10svveRZ3u+66y498MADOnDggM+5DB48OM/2M9km1qpVS88884ymTJniGU312LFjeuaZZ1SrVi1J0m+//abo6Oh8Y3z99ddasmSJypcvr5CQEBUpUkRXXnmlxo4dqwEDBpx2NDFJnm2g05lsi2ysu9WrV/caHfWWW27RLbfc4tPoqDnTaYzRDz/84DU6bWhoqBo0aKCHH37Y52mZNm2annnmGa/2adOmKS4uTtKJEVtLly7tU7yTtyM5I5TOnz/fp+2IjWWkdevW6tu3r6ZOnapGjRpJOjG//v3vf6tNmzaSpB9++EFVqlTJ8/1VqlTR7t27NXbsWK92X0etbdmypef/V1xxRa7tzt69exUXF+fTvlc6sU+7+uqrFR8f79k2L126VJJv2+bRo0drx44duuaaazz74OPHj6tnz56nHKH1ZOPHj9fVV1+typUre+bpunXrFB0drTfffNOnGK+99prf+8xnn31WLVu2VM2aNXXVVVdJkr766itlZGT4PGK1jWmxFcfG8v7yyy/ruuuuU6VKlVS/fn25XC6tX79ex48f1yeffCJJ2rZtm/r163dWY9j4bvLazyxbtuyM9jNr1671e7+5bds2de3aVT/88IPnuNl10kiEvqy7Nta75557TosWLfIahTQyMlKJiYlq3769HnroIT355JNq3779KePExcVpxYoVubZ5K1asUGxsrE+5jB07VtHR0Z7fZ82aNVOzZs30xhtv6Nlnn/Vp/fVnebf5OzFnWu6///4Cx7BxDODPNsTmfsbf37y2fkvExMRo7969qly5sipXrqxVq1apQYMGSk1Nzfe47YIUwIIdLLBx6Y4xdroQ5/XeMz0bYqNLtI2uzE4F6d1ho2dGjoKcdbPZIyKHP2chbZztNsbOmbuzsYwUZHm3dcmbP70zcvIoaM+bHDa2RTZ6NNo4223jLLOtUZFsbBNtXBJRqlQpT5f9qlWrem60/Msvv5y2J5TNbZGNdTe/m+aeyaWVvXr18lo+CsLfyxFzbN261dSvX99rO3Kmg3DYWEb8vcTLxiWvtuLYmKc5An25mo19po1Lq21Mi604NpZ3Y/zvNWMjho3vxtf9zI8//phvDBv7TZvHzf6sdyVKlDBLly7N1b506VLPJa9bt241ERERp4zzzDPPmLJly5o33njD0wNo2rRppmzZsiYpKcmnXGz8PrOxvNvII1guVzXG/22Ijf2MrcuA/f0tYeOWFRcClzGUEAuTf/75Rz///LMuvvhilStXzuf3FS9eXBs3bsx1NmTbtm2qU6eODh8+7HMsf3rNTJw4UV999ZWmT5/uOcOTkZGhu+++W1deeaXuvfde9ejRQ4cOHdKCBQt8yuebb77RihUrVL16dZ9ykOz07ihXrpyWLFmi+vXrKyoqSqtXr1bNmjW1ZMkSDRkyxKeeGf6cdWvdurWkE2cLr7jiilw9IuLj4/Xwww8rISHhtHnYmB/x8fGaPXu2mjdv7tX+zTff6NZbb1Vqaupp85CkhIQEjRw5UnfccYdX+5tvvqmRI0dq27ZtPsVx5nCmy4jz7F+OMzlj1qVLFxUpUkSvv/66qlatqm+++Ub79u3TkCFD9Pzzz3vOGp9KkSJFvHrN5PC1d4atGE4F2RalpaXpzjvv1OLFi3P1aJw5c6ZiYmK0dOlSHTt2LN8zvK1btz7l2e5NmzbJ5XKd8mz3RRddpEWLFuV6fsOGDWrfvr1+++03rVmzRu3bt9dff/2VZ4yQkBClpaXlmqc7duxQnTp1dPDgQZ/mia1t4oEDBzRr1ixt3rxZxhjVqlVLPXr0UEREhE95XHXVVRoyZIhuuOEG9ejRQ/v379fjjz+uKVOm6LvvvtOPP/6Y73ttbouczmTdzdmOTZo0Sffee6/Cw8M9zx0/flzffPONihQpohUrVpxxHgWxc+dOZWVl6fXXX9emTZs838u9996rokWL6uKLL/Ypjo3tiOT/MpJj06ZNXtNTs2bNU77e1vdi8/u1MU9t7Ddz+HNcZWOfaWMfYWOfaTOOv8u7rTwk/75fG9+Njf2Mjf2mjeNmG+vd7bffrq+//lrjx4/39OBfvXq1Hn74YTVv3lxvvvmm3nnnHT3//PP69ttv841jjNGjjz6qF154wfPdFi9eXMOGDdOTTz552mnJeb2N32f+Lu828rD5WzNHQY7fcxRkvbO5n7F1fOfv/io1NVUXXXSR59hszpw5Wr58uapXr65OnToV6NisUApQsQ6W2LqHQPXq1c2bb76Zq33mzJmeewmdjo0zs7GxsXlW1H/88UfPPZm+++47U7Zs2dPG8ueeCDaq//70zMhh46ybjR4RNs5C2rq3g40zdzZ6vdk4Y1a2bFnz/fffG2NO3Mw3p/fe4sWLfR4t1sbZLhs9b2zez8SfHo02znb7c5bZdg9PG9tEG8v7Z599Zj744ANjzIlpr127tnG5XKZcuXJm8eLFPsWwsS3KUZDte869FF0ul2nevLnn71atWpn27dub++67z+eelTbY6o1oYzuSw5/9ZkG3Aba+F5vfr415auM4wsZxlY19po19hK0rGmzECZZjABvfr43vxsZ+xkbvLBvHzTbWu8zMTHPPPfeY0NBQz3cRGhpq7r33XnPgwAFjjDFr167N8x6s+cVbvXq1+eGHH3y+t2oOG7/PcvizfbeRh40YNtZdf9Y7m/sZW795/d1f2ToWKewomp3nbHXttHFQZaPAY2Ona+MgxMYP7yuvvNJ8+OGHxhhjbrvtNtOxY0ezfPlyc9ddd5m6dev6lIfNH0P+sDE/bO34s7OzzdChQ03x4sU932l4eLgZNWqUzzFsHOzaKAIG+pI3mwUeG9siG4U3GwchPXr0MFWqVDFz5841u3btMr/++quZO3euqVq1qrnjjjuMMca8/fbb5rLLLstzPtgszNjYJp6Ny++NMWbv3r0mOzu7wO8vCBvbd5vFO3/YuhzRxg9NG/PV35Mrtr4XG3FszNNguVzNn32mzX2ErRNnNuIEyzGAP9+vze/Gxn7Gn/1mDhvHzbZu4m/MiWLX999/b9atW2cyMzN9+nzbbPw+C5bie7BcrhosHRJsXQbs7/7KRuH9QkDR7Dxna+dgoxBho8BjY6drY2No44e3jZ4ZNg7cbbAxP2zsLE/mz5k7Gwe7NoqA/hwg2ijO2Czw2NgW2ejRaOMgxMZZZlsFABvbxLMxgleg2LznTaDY7o1o44emjflq88dqoNmYpzb2mzZPnBVkn2lzH2HrxJmNOMFyDODP92vzu7Gxn7Gx37Rx3GzzipVgYOP3WaCL7zZj2Fh3g6VDgo31zpiC76/Oxr2vCzOKZuc52zsHfwoRNgo8Nna6NjaGtqr/TmfaM8PGgbsNNuaHjZ2lLTYOdm0UAYPlkjcbMWxsi2z86LZ1EGJMcJxltrFNtHl5R6AFy8GuP2z3RrSxHbExXwvTj1Ub8zRYLlezwcY+wtaJMxtxguUYwMb3a+O7sXk5ou395pkeN5+t4/dA8+f3WaCL7zZj2Fh3g2W7amu9K+j+KthuWRHsKJqd54Jp52CzwOPPTtfGxtDmD29/2Dhwt8F2IcLfHa6/bBzsnq0iYCAuebPBxrbIxo9u2/ciCRb+bBNt9/IMpGA52LXhbF4meqbbERvzNZiOR86GM52nwXK5WrCwtc+0ESdYjgGC7fsNhhNF/gqW4/dgUpj2mzbW3QthvTuT/VWw3LIi2FE0O88F084hWAo8NjaGwfzDOxBFlWCeHwVhs+AVDEXAYGBjW2TzR3dhOPi3JZh6efor2A52Cwsb8zWYjkeCQbBcrhZsbO0z/YkTLMcAhfH7DbTCdrxqQ2Hab9pYd1nvUBAuY4wJ9AieKLgDBw5o0KBBmjlzprKysiRJRYsWVc+ePfWf//xHJUqU0Lp16yRJDRs2POf57du3T6VLl5bL5Tpnn7lgwQIdPHhQ3bp107Zt2/Svf/1LP//8s8qWLat3331Xbdq08TnWgQMHtG3bNhljVK1aNZUsWfIsZh78Ctv8OHDggDZu3KiwsDAlJCTI7XYHOqXzlo1tka1h3pG3wrC829y+4//YmK/BfjwSKLb3m4E4riqsgnGbyPdrR2E7XvVHYdxv2l53We9wOhTNCgl2DqfGxhA4N/zZFvGjGwXB9v3sKOh85XgEAIIb+03gzFA0AwAEFX50AwAAAAgGFM0AAAAAAAAAh5BAJwAAAAAAAAAEG4pmAAAAAAAAgANFMwAAAAAAAMCBohkAAMAFIjEx0froszNmzFCpUqWsxgQAAAgGFM0AAACCUFpamh588EFVrVpVbrdbcXFx6tKlixYvXhzo1Lx0795dmzdvDnQaAAAA1hUNdAIAAADwtn37drVo0UKlSpXSuHHjVL9+fR07dkwLFixQ//799fPPPwc6RY+wsDCFhYUFOg0AAADr6GkGAAAQZPr16yeXy6XVq1frpptuUo0aNVS3bl0NHjxYq1atkiTt3LlT119/vUqWLKnIyEjdcsst+uOPP7ziPPPMM4qOjlZERITuvvtuHT582Ov57OxsPfXUU6pUqZLcbrcaNmyozz77zPP89u3b5XK5NHfuXLVu3Vrh4eFq0KCBvv76a89ruDwTAAAUVhTNAAAAgsi+ffv02WefqX///ipRokSu50uVKiVjjG644Qbt27dPy5Yt06JFi7R161Z1797d87o5c+Zo5MiRGjNmjL799ltVrFhRr7zyilesSZMmafz48Xr++ee1fv16dejQQdddd522bNni9boRI0bo4Ycf1rp161SjRg3ddtttysrKOjszAAAAIEi4jDEm0EkAAADghNWrV6tZs2aaO3euunbtmudrFi1apE6dOik1NVVxcXGSpJ9++kl169bV6tWr1aRJEzVv3lwNGjTQ5MmTPe+7/PLLdfjwYa1bt06SdNFFF6l///567LHHPK9p2rSpmjRpopdfflnbt29XlSpVNHXqVN19991en7Nx40bVqlVLM2bM0MCBA/X333+fnRkCAAAQIPQ0AwAACCI55zNdLle+r9m4caPi4uI8BTNJqlOnjkqVKqWNGzd6XnPFFVd4ve/kvzMyMvT777+rRYsWXq9p0aKFJ0aO+vXre/5fsWJFSdKePXvOZLIAAADOOxTNAAAAgkhCQoJcLleuwtXJjDF5FtXyaz8V5+vzilGsWLFcr8/Ozj6jzwEAADjfUDQDAAAIImXKlFGHDh308ssv6+DBg7me//vvv1WnTh3t3LlTu3bt8rT/9NNPSk9PV+3atSVJtWvX9gwakOPkvyMjIxUbG6vly5d7vWblypWeGAAAABeyooFOAAAAAN5eeeUVNW/eXE2bNtVTTz2l+vXrKysrS4sWLdLkyZP1008/qX79+rr99ts1ceJEZWVlqV+/fmrZsqUaN24sSXrooYfUs2dPNW7cWFdeeaXeeustbdiwQVWrVvV8ziOPPKKRI0eqWrVqatiwoaZPn65169bprbfeCtSkAwAABA2KZgAAAEGmSpUqWrNmjcaMGaMhQ4Zo9+7dKl++vC677DJNnjxZLpdLH330kR588EFdffXVCgkJUceOHfXiiy96YnTv3l1bt27VsGHDdPjwYd14443697//rQULFnheM2DAAGVkZGjIkCHas2eP6tSpo/nz5yshISEQkw0AABBUGD0TAAAAAAAAcOCeZgAAAAAAAIADRTMAAAAAAADAgaIZAAAAAAAA4EDRDAAAAAAAAHCgaAYAAAAAAAA4UDQDAAAAAAAAHCiaAQAAAAAAAA4UzQAAAAAAAAAHimYAAAAAAACAA0UzAAAAAAAAwIGiGQAAAAAAAOBA0QwAAAAAAABw+H+vhTpIK0MQdgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1500x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize = (15, 6))\n",
    "\n",
    "plt.bar(sorted_codon_list, codon_freq_list, color = 'green')\n",
    "\n",
    "plt.title('Frequenze dei codoni')\n",
    "plt.xlabel('Codoni')\n",
    "plt.ylabel('Frequenze')\n",
    "plt.xticks(rotation = 90)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diagramma degli amminoacidi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creare la lista dei simboli di amminoacido (asse x del diagramma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "ammino_acid_list = [t[0] for t in ammino_frequencies]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['L',\n",
       " 'E',\n",
       " 'S',\n",
       " 'G',\n",
       " 'V',\n",
       " 'R',\n",
       " 'P',\n",
       " 'N',\n",
       " 'T',\n",
       " 'D',\n",
       " 'K',\n",
       " 'A',\n",
       " 'I',\n",
       " 'F',\n",
       " 'Y',\n",
       " 'C',\n",
       " 'H',\n",
       " 'Q',\n",
       " 'M',\n",
       " 'W']"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ammino_acid_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creare la corrispondente lista delle frequenze (asse y del diagramma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_freq_list = [t[1] for t in ammino_frequencies]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[136,\n",
       " 106,\n",
       " 105,\n",
       " 93,\n",
       " 87,\n",
       " 82,\n",
       " 82,\n",
       " 72,\n",
       " 70,\n",
       " 70,\n",
       " 68,\n",
       " 65,\n",
       " 65,\n",
       " 59,\n",
       " 50,\n",
       " 47,\n",
       " 38,\n",
       " 37,\n",
       " 30,\n",
       " 20]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_freq_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABNYAAAIhCAYAAACCMjckAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAABLcElEQVR4nO3dZ5iV5YE+8HtCGUABBVfGUVCMWBB7SywBpFlQY4lm7RETs6jRiDGiiUBMcMVVMWKJkWaMLVmsWbtgQ9ce1xLLilgJiSJNpJ7/h/yZdQQNvIyemeH3u67z4TznOefczxQ43Dzv+1aUSqVSAAAAAICV8rVyBwAAAACAhkixBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYArLBx48aloqJiubfTTz+93PFWC0u/B2+++eaX+j4VFRUZOnRozf1JkyaloqIikyZN+lLf96s0dOjQVFRUlDvGSunRo0d69OjxT+e9+eabqaioyLhx42rGGuJ6AaC+a1ruAABAwzN27Nhsvvnmtcaqq6vLlIavwvbbb5/HHnssXbt2LXeUOnP88cdnr732KneMlXL55ZcXfm5DXC8A1HeKNQBgpXXr1i077rjjCs1duHBhKioq0rSpjx0NWZs2bfKNb3yj3DHq1AYbbJANNtig3DFWyqoUmw1xvQBQ3zkUFACoM0sPF/zd736XQYMGZf31109lZWVef/31JMl9992XXr16pU2bNmnVqlV222233H///cu8zp/+9Kdsu+22qaysTOfOnfMf//EfyxzGtrxD3Zb67GGMSfLaa6/l8MMPz7rrrpvKyspsscUWueyyy5ab//rrr8/ZZ5+d6urqtGnTJr17984rr7yyzLzl3TbaaKNar3njjTfmm9/8ZtZYY42sueaa6devX5599tkV+no+/vjj2W233dKiRYtUV1dn8ODBWbhw4XLnruj7/Pa3v82mm26aysrKdO3aNdddd12OPfbYZXJ/1ooeCvq3v/0tAwcOTNeuXbPmmmtm3XXXzZ577pmHH3641ryl378LLrgg559/fjbaaKO0bNkyPXr0yKuvvpqFCxfmzDPPTHV1ddq2bZsDDzww06dPr/UaG220Ufr375877rgj2223XVq2bJktttgid9xxR5J/HDa7xRZbZI011sjOO++cp556qtbzl3do5NLXvOuuu7L99tunZcuW2XzzzTNmzJhl1vrCCy/kgAMOyNprr50WLVpk2223zfjx42vN+eSTTzJo0KBsu+22adu2bdq1a5dvfvObufXWW5d5vSVLluTSSy/Ntttum5YtW2attdbKN77xjdx22201c5Z3KOh7772XQw89NK1bt07btm1z2GGHZdq0acu8vkNBAaDuKdYAgJW2ePHiLFq0qNbt0wYPHpy33norV155ZW6//fasu+66ufbaa9O3b9+0adMm48ePz0033ZR27dqlX79+tcq1+++/PwcccEBat26dG264IRdccEFuuummjB07tnDel156KTvttFNeeOGFXHjhhbnjjjuy77775kc/+lGGDRu2zPyzzjorU6dOzdVXX52rrroqr732Wvbbb78sXrw4yf8dFvnp2zXXXJNmzZplyy23rHmd4cOH51//9V/TtWvX3HTTTfnd736X2bNnZ4899shLL730TzP36tUrH330UcaNG5crr7wyzz77bH75y18uM3dF3+eqq67KD37wg2y99daZMGFCfvazn2XYsGF1et60Dz/8MEkyZMiQ/OlPf8rYsWOz8cYbp0ePHst9n8suuyyPPvpoLrvsslx99dX5y1/+kv322y8DBgzI3/72t4wZMyYjRozIfffdl+OPP36Z5//5z3/O4MGD89Of/jQTJkxI27Ztc9BBB2XIkCG5+uqrM3z48Pz+97/PzJkz079//8ybN++fruHPf/5zBg0alB//+Me59dZbs/XWW2fAgAF56KGHaua88sor2XXXXfPiiy/m17/+dSZMmJCuXbvm2GOPzYgRI2rmzZ8/Px9++GFOP/303HLLLbn++uuz++6756CDDso111xT632PPfbYnHLKKdlpp51y44035oYbbsj+++//hefTmzdvXnr37p177rkn5513Xv7whz+kqqoqhx122D9dJwBQB0oAACto7NixpSTLvS1cuLA0ceLEUpLSt771rVrPmzt3bqldu3al/fbbr9b44sWLS9tss01p5513rhnbZZddStXV1aV58+bVjM2aNavUrl270qc/ukyZMqWUpDR27NhlciYpDRkypOZ+v379ShtssEFp5syZteaddNJJpRYtWpQ+/PDDUqlUqsm/zz771Jp30003lZKUHnvsseV+Xf7617+WNt5449KWW25ZmjFjRqlUKpXeeuutUtOmTUsnn3xyrbmzZ88uVVVVlQ499NDlvtZShx12WKlly5aladOm1YwtWrSotPnmm5eSlKZMmbJS77N48eJSVVVVaZdddqk1b+rUqaVmzZqVNtxww1rjn/0aLv3aTJw48Qtzf9aiRYtKCxcuLPXq1at04IEH1owv/f5ts802pcWLF9eMjxw5spSktP/++9d6nVNPPbWUpNb3cMMNNyy1bNmy9M4779SMPffcc6UkpfXWW680d+7cmvFbbrmllKR022231YwNGTKk9NmPwxtuuGGpRYsWpalTp9aMzZs3r9SuXbvSCSecUDP23e9+t1RZWVl66623aj1/7733LrVq1ar00UcffeHXY8CAAaXtttuuZvyhhx4qJSmdffbZy33eUt27dy9179695v4VV1xRSlK69dZba837/ve/v8zvx/LWCwCsGjvWAICVds011+TJJ5+sdfv0OdQOPvjgWvMnT56cDz/8MMccc0ytXW5LlizJXnvtlSeffDJz587N3Llz8+STT+aggw5KixYtap7funXr7LfffoWyfvLJJ7n//vtz4IEHplWrVrXef5999sknn3ySxx9/vNZz9t9//1r3t9566yTJ1KlTl3n9uXPnZt99980nn3ySO++8M2uttVaS5O67786iRYty9NFH13rPFi1apHv37v90l9jEiRPTq1evdOjQoWasSZMmy+xEWtH3eeWVVzJt2rQceuihtZ7fqVOn7Lbbbl+YZWVdeeWV2X777dOiRYs0bdo0zZo1y/3335+XX355mbn77LNPvva1//tIusUWWyRJ9t1331rzlo6/9dZbtca33XbbrL/++svM69GjR1q1arXM+PK+h5+17bbbplOnTjX3W7RokU033bTWcx944IH06tUrHTt2rPXcY489Nh9//HEee+yxmrE//OEP2W233bLmmmvWfD1Gjx5d6+tx5513JklOPPHEf5rv0yZOnJjWrVsv8zN7+OGHr9TrAADFOIswALDStthiiy+8eMF6661X6/5f//rXJMkhhxzyuc/58MMPU1FRkSVLlqSqqmqZx5c3tiI++OCDLFq0KJdeemkuvfTS5c75+9//Xut++/bta92vrKxMkmUOI1y0aFEOOeSQvPrqq3nooYdqlSxL17zTTjst9z0/XSZ9Xu4V+Tqs6Pt88MEHSVKrqFuqQ4cOmTJlyhfmWVEXXXRRBg0alB/+8Ic599xzs84666RJkyb5+c9/vtxirV27drXuN2/e/AvHP/nkkzp9/vJ89vuf/ONn4NPf/w8++GCZn/Pk/66Ou/TrPWHChBx66KH5zne+k5/85CepqqpK06ZNc8UVV9Q6b9vf/va3NGnSZKV/zj/44IPlfk+L/r4AACtHsQYA1LnPniB9nXXWSZJceumln3tlyQ4dOtRcQXR5J17/7NjSHW3z58+vNb600Fhq7bXXTpMmTXLUUUd97m6gzp07f8FqPt8PfvCD3H///fmv//qvbLPNNrUeW7rmP/7xj9lwww1X+rXbt2+/Ql+HFX2fpWXR0iLui15zVVx77bXp0aNHrrjiilrjs2fPrrP3qA/at2+f999/f5nx9957L8n/fV+uvfbadO7cOTfeeGOt34vP/tz+y7/8SxYvXpxp06Ytt7D7ohxPPPHEMuN1+T0FAD6fYg0A+NLttttuWWuttfLSSy/lpJNO+tx5zZs3z84775wJEybkggsuqCnPZs+endtvv73W3A4dOqRFixZ5/vnna41/9mqLrVq1Ss+ePfPss89m6623rtm5tKp+9rOfZezYsRk/fnx69+69zOP9+vVL06ZN87//+7/LHBq7Inr27Jnbbrstf/3rX2t2JC1evDg33nhjoffZbLPNUlVVlZtuuimnnXZazfhbb72VyZMn1+y0WlUVFRU1O/yWev755/PYY48tc9hkQ9arV6/cfPPNee+992p97a655pq0atWqpkCuqKhI8+bNa5Vq06ZNW+bndO+99855552XK664Ir/4xS9WOEfPnj1z00035bbbbqt1OOh1111XdGkAwEpQrAEAX7o111wzl156aY455ph8+OGHOeSQQ7Luuuvmb3/7W/785z/nb3/7W80Op3PPPTd77bVX+vTpk0GDBmXx4sU5//zzs8Yaa9RccTL5R2Fx5JFHZsyYMfn617+ebbbZJk888cRyC4VLLrkku+++e/bYY4/827/9WzbaaKPMnj07r7/+em6//fY88MADK7WeP/zhD/nVr36VQw45JJtuummtc7RVVlZmu+22y0YbbZRf/OIXOfvss/PGG29kr732ytprr52//vWveeKJJ7LGGmss94qkS/3sZz/Lbbfdlj333DPnnHNOWrVqlcsuuyxz586tNW9F3+drX/tahg0blhNOOCGHHHJIjjvuuHz00UcZNmxY1ltvvX96aOqK6t+/f84999wMGTIk3bt3zyuvvJJf/OIX6dy58zJXj23IhgwZkjvuuCM9e/bMOeeck3bt2uX3v/99/vSnP2XEiBFp27Ztkn98PSZMmJCBAwfmkEMOydtvv51zzz036623Xl577bWa19tjjz1y1FFH5Ze//GX++te/pn///qmsrMyzzz6bVq1a5eSTT15ujqOPPjoXX3xxjj766PzqV79Kly5d8l//9V+5++67v5KvAwCs7hRrAMBX4sgjj0ynTp0yYsSInHDCCZk9e3bWXXfdbLvttjn22GNr5vXp0ye33HJLfvazn+Wwww5LVVVVBg4cmHnz5i1TRF144YVJkhEjRmTOnDnZc889c8cdd2SjjTaqNa9r16555plncu655+ZnP/tZpk+fnrXWWitdunTJPvvss9JrefHFF5P84/DLP/7xj7Ue23DDDfPmm28mSQYPHpyuXbvmkksuyfXXX5/58+enqqoqO+20U374wx9+4Xt069Yt9913XwYNGpRjjjkma6+9do466qgcfPDB+cEPflBr7oq+zw9+8INUVFRkxIgROfDAA7PRRhvlzDPPzK233rrMRQGKOvvss/Pxxx9n9OjRGTFiRLp27Zorr7wyN9988z+9YENDstlmm2Xy5Mk566yzcuKJJ2bevHnZYostMnbs2Fo/z9/73vcyffr0XHnllRkzZkw23njjnHnmmXnnnXeW+XkeN25ctt9++4wePTrjxo1Ly5Yt07Vr15x11lmfm6NVq1Z54IEHcsopp+TMM89MRUVF+vbtmxtuuCG77rrrl7V8AOD/qyiVSqVyhwAA+GeGDh2aYcOGxUeXuvXRRx9l0003zbe//e1cddVV5Y4DANCg2LEGALCamDZtWn71q1+lZ8+ead++faZOnZqLL744s2fPzimnnFLueAAADY5iDQBgNVFZWZk333wzAwcOzIcfflhzkv0rr7wyW265ZbnjAQA0OA4FBQAAAIAC6ubyTwAAAACwmlGsAQAAAEABijUAAAAAKMDFC5IsWbIk7733Xlq3bp2KiopyxwEAAACgTEqlUmbPnp3q6up87WtfvCdNsZbkvffeS8eOHcsdAwAAAIB64u23384GG2zwhXMUa0lat26d5B9fsDZt2pQ5DQAAAADlMmvWrHTs2LGmL/oiirWk5vDPNm3aKNYAAAAAWKHThbl4AQAAAAAUoFgDAAAAgALKWqw99NBD2W+//VJdXZ2Kiorccsstnzv3hBNOSEVFRUaOHFlrfP78+Tn55JOzzjrrZI011sj++++fd95558sNDgAAAMBqr6zF2ty5c7PNNttk1KhRXzjvlltuyX//93+nurp6mcdOPfXU3HzzzbnhhhvyyCOPZM6cOenfv38WL178ZcUGAAAAgPJevGDvvffO3nvv/YVz3n333Zx00km5++67s++++9Z6bObMmRk9enR+97vfpXfv3kmSa6+9Nh07dsx9992Xfv36fWnZAQAAAFi91etzrC1ZsiRHHXVUfvKTn2TLLbdc5vGnn346CxcuTN++fWvGqqur061bt0yePPlzX3f+/PmZNWtWrRsAAAAArIx6Xaydf/75adq0aX70ox8t9/Fp06alefPmWXvttWuNd+jQIdOmTfvc1z3vvPPStm3bmlvHjh3rNDcAAAAAjV+9LdaefvrpXHLJJRk3blwqKipW6rmlUukLnzN48ODMnDmz5vb222+valwAAAAAVjP1tlh7+OGHM3369HTq1ClNmzZN06ZNM3Xq1AwaNCgbbbRRkqSqqioLFizIjBkzaj13+vTp6dChw+e+dmVlZdq0aVPrBgAAAAAro94Wa0cddVSef/75PPfcczW36urq/OQnP8ndd9+dJNlhhx3SrFmz3HvvvTXPe//99/PCCy9k1113LVd0AAAAAFYDZb0q6Jw5c/L666/X3J8yZUqee+65tGvXLp06dUr79u1rzW/WrFmqqqqy2WabJUnatm2bAQMGZNCgQWnfvn3atWuX008/PVtttVXNVUIBAAAA4MtQ1mLtqaeeSs+ePWvun3baaUmSY445JuPGjVuh17j44ovTtGnTHHrooZk3b1569eqVcePGpUmTJl9GZAAAAABIklSUSqVSuUOU26xZs9K2bdvMnDnT+dYAAAAAVmMr0xPV23OsAQAAAEB9plgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAACigabkD8OUZVjGs3BFWyJDSkHJHAAAAAFhpdqwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUEBZi7WHHnoo++23X6qrq1NRUZFbbrml5rGFCxfmpz/9abbaaqusscYaqa6uztFHH5333nuv1mvMnz8/J598ctZZZ52sscYa2X///fPOO+98xSsBAAAAYHVT1mJt7ty52WabbTJq1KhlHvv444/zzDPP5Oc//3meeeaZTJgwIa+++mr233//WvNOPfXU3HzzzbnhhhvyyCOPZM6cOenfv38WL178VS0DAAAAgNVQ03K++d5775299957uY+1bds29957b62xSy+9NDvvvHPeeuutdOrUKTNnzszo0aPzu9/9Lr17906SXHvttenYsWPuu+++9OvXb7mvPX/+/MyfP7/m/qxZs+poRQAAAACsLhrUOdZmzpyZioqKrLXWWkmSp59+OgsXLkzfvn1r5lRXV6dbt26ZPHny577Oeeedl7Zt29bcOnbs+GVHBwAAAKCRaTDF2ieffJIzzzwzhx9+eNq0aZMkmTZtWpo3b56111671twOHTpk2rRpn/tagwcPzsyZM2tub7/99peaHQAAAIDGp6yHgq6ohQsX5rvf/W6WLFmSyy+//J/OL5VKqaio+NzHKysrU1lZWZcRAQAAAFjN1PsdawsXLsyhhx6aKVOm5N57763ZrZYkVVVVWbBgQWbMmFHrOdOnT0+HDh2+6qgAAAAArEbqdbG2tFR77bXXct9996V9+/a1Ht9hhx3SrFmzWhc5eP/99/PCCy9k1113/arjAgAAALAaKeuhoHPmzMnrr79ec3/KlCl57rnn0q5du1RXV+eQQw7JM888kzvuuCOLFy+uOW9au3bt0rx587Rt2zYDBgzIoEGD0r59+7Rr1y6nn356ttpqq5qrhAIAAADAl6GsxdpTTz2Vnj171tw/7bTTkiTHHHNMhg4dmttuuy1Jsu2229Z63sSJE9OjR48kycUXX5ymTZvm0EMPzbx589KrV6+MGzcuTZo0+UrWAAAAAMDqqaJUKpXKHaLcZs2albZt22bmzJm1zuHW0A2rGFbuCCtkSGlIuSMAAAAAJFm5nqhen2MNAAAAAOorxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABTcsdAFbGsIph5Y6wwoaUhpQ7AgAAAPAlsmMNAAAAAApQrAEAAABAAQ4FhTJzeCsAAAA0THasAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABZS1WHvooYey3377pbq6OhUVFbnllltqPV4qlTJ06NBUV1enZcuW6dGjR1588cVac+bPn5+TTz4566yzTtZYY43sv//+eeedd77CVQAAAACwOiprsTZ37txss802GTVq1HIfHzFiRC666KKMGjUqTz75ZKqqqtKnT5/Mnj27Zs6pp56am2++OTfccEMeeeSRzJkzJ/3798/ixYu/qmUAAAAAsBpqWs4333vvvbP33nsv97FSqZSRI0fm7LPPzkEHHZQkGT9+fDp06JDrrrsuJ5xwQmbOnJnRo0fnd7/7XXr37p0kufbaa9OxY8fcd9996dev31e2FgAAAABWL/X2HGtTpkzJtGnT0rdv35qxysrKdO/ePZMnT06SPP3001m4cGGtOdXV1enWrVvNnOWZP39+Zs2aVesGAAAAACuj3hZr06ZNS5J06NCh1niHDh1qHps2bVqaN2+etdde+3PnLM95552Xtm3b1tw6duxYx+kBAAAAaOzqbbG2VEVFRa37pVJpmbHP+mdzBg8enJkzZ9bc3n777TrJCgAAAMDqo94Wa1VVVUmyzM6z6dOn1+xiq6qqyoIFCzJjxozPnbM8lZWVadOmTa0bAAAAAKyMelusde7cOVVVVbn33ntrxhYsWJAHH3wwu+66a5Jkhx12SLNmzWrNef/99/PCCy/UzAEAAACAL0NZrwo6Z86cvP766zX3p0yZkueeey7t2rVLp06dcuqpp2b48OHp0qVLunTpkuHDh6dVq1Y5/PDDkyRt27bNgAEDMmjQoLRv3z7t2rXL6aefnq222qrmKqEAAAAA8GUoa7H21FNPpWfPnjX3TzvttCTJMccck3HjxuWMM87IvHnzMnDgwMyYMSO77LJL7rnnnrRu3brmORdffHGaNm2aQw89NPPmzUuvXr0ybty4NGnS5CtfDwAAAACrj7IWaz169EipVPrcxysqKjJ06NAMHTr0c+e0aNEil156aS699NIvISEAAAAALF+9PccaAAAAANRnijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKaFruAEDjNKxiWLkjrLAhpSHljgAAAEADZMcaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACmpY7AEBDMaxiWLkjrLAhpSHljgAAANDo2bEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAU3LHQCA8hpWMazcEVbIkNKQFZ7bUNaUrNy6AACA+sWONQAAAAAoYJWLtU8++aQucgAAAABAg1KoWFuyZEnOPffcrL/++llzzTXzxhtvJEl+/vOfZ/To0XUWbtGiRfnZz36Wzp07p2XLltl4443zi1/8IkuWLKmZUyqVMnTo0FRXV6dly5bp0aNHXnzxxTrLAAAAAADLU6hY++Uvf5lx48ZlxIgRad68ec34VlttlauvvrrOwp1//vm58sorM2rUqLz88ssZMWJELrjgglx66aU1c0aMGJGLLrooo0aNypNPPpmqqqr06dMns2fPrrMcAAAAAPBZhYq1a665JldddVWOOOKINGnSpGZ86623zl/+8pc6C/fYY4/lgAMOyL777puNNtoohxxySPr27ZunnnoqyT92q40cOTJnn312DjrooHTr1i3jx4/Pxx9/nOuuu67OcgAAAADAZxUq1t59991ssskmy4wvWbIkCxcuXOVQS+2+++65//778+qrryZJ/vznP+eRRx7JPvvskySZMmVKpk2blr59+9Y8p7KyMt27d8/kyZM/93Xnz5+fWbNm1boBAAAAwMpoWuRJW265ZR5++OFsuOGGtcb/8Ic/ZLvttquTYEny05/+NDNnzszmm2+eJk2aZPHixfnVr36Vf/3Xf02STJs2LUnSoUOHWs/r0KFDpk6d+rmve95552XYsGF1lhMAAACA1U+hYm3IkCE56qij8u6772bJkiWZMGFCXnnllVxzzTW544476izcjTfemGuvvTbXXXddttxyyzz33HM59dRTU11dnWOOOaZmXkVFRa3nlUqlZcY+bfDgwTnttNNq7s+aNSsdO3ass9wAAAAANH6FirX99tsvN954Y4YPH56Kioqcc8452X777XP77benT58+dRbuJz/5Sc4888x897vfTfKPiyNMnTo15513Xo455phUVVUl+cfOtfXWW6/medOnT19mF9unVVZWprKyss5yAgAAALD6KVSsJUm/fv3Sr1+/usyyjI8//jhf+1rt08A1adIkS5YsSZJ07tw5VVVVuffee2sOQV2wYEEefPDBnH/++V9qNgAAAABWb4UuXnDcccdl/Pjxy4zPmjUrxx133CqHWmq//fbLr371q/zpT3/Km2++mZtvvjkXXXRRDjzwwCT/OAT01FNPzfDhw3PzzTfnhRdeyLHHHptWrVrl8MMPr7McAAAAAPBZhXasjRs3LjfeeGOefvrpjBw5smZX2bx58zJ+/PiMGTOmTsJdeuml+fnPf56BAwdm+vTpqa6uzgknnJBzzjmnZs4ZZ5yRefPmZeDAgZkxY0Z22WWX3HPPPWndunWdZAAAAACA5Sl8KOif/vSnfP/738/LL7+cm266KWuvvXZd5kqStG7dOiNHjszIkSM/d05FRUWGDh2aoUOH1vn7AwAAAMDnKXQoaJJ07do1jz/+eBYuXJiddtopL7/8cl3mAgAAAIB6rVCxVlFRkSRp37597rvvvvTo0SPf+MY3ctttt9VpOAAAAACorwodCloqlf7vBZo2zdVXX52uXbtm4MCBdRYMAAAAAOqzQsXaxIkT065du1pjp512Wrbeeus8+uijdRIMAAAAAOqzQsVa9+7dlzveu3fv9O7de5UCAQAAAEBDUKhYW7x4ccaNG5f7778/06dPz5IlS2o9/sADD9RJOAAAAACorwoVa6ecckrGjRuXfffdN926dau5mAEAAAAArC4KFWs33HBDbrrppuyzzz51nQcAAAAAGoSvFXlS8+bNs8kmm9R1FgAAAABoMAoVa4MGDcoll1ySUqlU13kAAAAAoEEodCjoI488kokTJ+bOO+/MlltumWbNmtV6fMKECXUSDgAAAADqq0LF2lprrZUDDzywrrMAAAAAQINRqFgbO3ZsXecAAAAAgAal0DnWkmTRokW577778pvf/CazZ89Okrz33nuZM2dOnYUDAAAAgPqq0I61qVOnZq+99spbb72V+fPnp0+fPmndunVGjBiRTz75JFdeeWVd5wQAAACAeqXQjrVTTjklO+64Y2bMmJGWLVvWjB944IG5//776ywcAAAAANRXha8K+uijj6Z58+a1xjfccMO8++67dRIMAAAAAOqzQjvWlixZksWLFy8z/s4776R169arHAoAAAAA6rtCxVqfPn0ycuTImvsVFRWZM2dOhgwZkn322aeusgEAAABAvVXoUNCLL744PXv2TNeuXfPJJ5/k8MMPz2uvvZZ11lkn119/fV1nBIDV3rCKYeWOsMKGlIaUOwIAAHwlChVr1dXVee6553L99dfnmWeeyZIlSzJgwIAcccQRtS5mAAAAAACNVaFiLUlatmyZ4447Lscdd1xd5gEAAACABqFQsXbNNdd84eNHH310oTAAAAAA0FAUKtZOOeWUWvcXLlyYjz/+OM2bN0+rVq0UawAAAAA0eoWuCjpjxoxatzlz5uSVV17J7rvv7uIFAAAAAKwWCp9j7bO6dOmSf//3f8+RRx6Zv/zlL3X1sgBAI9ZQrna6Mlc6bShrSlzBFQBgVRXasfZ5mjRpkvfee68uXxIAAAAA6qVCO9Zuu+22WvdLpVLef//9jBo1KrvttludBAMAAACA+qxQsfbtb3+71v2Kior8y7/8S/bcc89ceOGFdZELAAAAAOq1QsXakiVL6joHAAAAADQodXqONQAAAABYXRTasXbaaaet8NyLLrqoyFsAAAAAQL1WqFh79tln88wzz2TRokXZbLPNkiSvvvpqmjRpku23375mXkVFRd2kBACgbIZVDCt3hBU2pDSk3BEAgNVIoWJtv/32S+vWrTN+/PisvfbaSZIZM2bke9/7XvbYY48MGjSoTkMCAAAAQH1T6BxrF154Yc4777yaUi1J1l577fzyl790VVAAAAAAVguFirVZs2blr3/96zLj06dPz+zZs1c5FAAAAADUd4WKtQMPPDDf+9738sc//jHvvPNO3nnnnfzxj3/MgAEDctBBB9V1RgAAAACodwqdY+3KK6/M6aefniOPPDILFy78xws1bZoBAwbkggsuqNOAAAAAAFAfFSrWWrVqlcsvvzwXXHBB/vd//zelUimbbLJJ1lhjjbrOBwAAAAD1UqFDQZd6//338/7772fTTTfNGmuskVKpVFe5AAAAAKBeK1SsffDBB+nVq1c23XTT7LPPPnn//feTJMcff3wGDRpUpwEBAAAAoD4qdCjoj3/84zRr1ixvvfVWtthii5rxww47LD/+8Y9z4YUX1llAAAD4MgyrGFbuCCtsSGnICs1rjGsCgPqsULF2zz335O67784GG2xQa7xLly6ZOnVqnQQDAAAAgPqs0KGgc+fOTatWrZYZ//vf/57KyspVDgUAAAAA9V2hYu1b3/pWrrnmmpr7FRUVWbJkSS644IL07NmzzsIBAAAAQH1V6FDQCy64ID169MhTTz2VBQsW5IwzzsiLL76YDz/8MI8++mhdZwQAAACAeqfQjrWuXbvm+eefz84775w+ffpk7ty5Oeigg/Lss8/m61//el1nBAAAAIB6Z6V3rC1cuDB9+/bNb37zmwwb1nCuOgQAAAAAdWmld6w1a9YsL7zwQioqKr6MPAAAAADQIBQ6FPToo4/O6NGj6zoLAAAAADQYhS5esGDBglx99dW59957s+OOO2aNNdao9fhFF11UJ+EAAAAAoL5aqWLtjTfeyEYbbZQXXngh22+/fZLk1VdfrTXHIaIAAAAArA5Wqljr0qVL3n///UycODFJcthhh+XXv/51OnTo8KWEAwAAAID6aqXOsVYqlWrdv/POOzN37tw6DQQAAAAADUGhixcs9dmi7cvw7rvv5sgjj0z79u3TqlWrbLvttnn66adrZRg6dGiqq6vTsmXL9OjRIy+++OKXngsAAACA1dtKFWsVFRXLnEPtyzyn2owZM7LbbrulWbNmufPOO/PSSy/lwgsvzFprrVUzZ8SIEbnooosyatSoPPnkk6mqqkqfPn0ye/bsLy0XAAAAAKzUOdZKpVKOPfbYVFZWJkk++eST/PCHP1zmqqATJkyok3Dnn39+OnbsmLFjx9aMbbTRRrXyjBw5MmeffXYOOuigJMn48ePToUOHXHfddTnhhBPqJAcAAAAAfNZK7Vg75phjsu6666Zt27Zp27ZtjjzyyFRXV9fcX3qrK7fddlt23HHHfOc738m6666b7bbbLr/97W9rHp8yZUqmTZuWvn371oxVVlame/fumTx58ue+7vz58zNr1qxaNwAAAABYGSu1Y+3TO8e+Cm+88UauuOKKnHbaaTnrrLPyxBNP5Ec/+lEqKytz9NFHZ9q0aUmyzFVJO3TokKlTp37u65533nkZNmzYl5odAAAAgMZtlS5e8GVbsmRJtt9++wwfPjzbbbddTjjhhHz/+9/PFVdcUWveZ8/zViqVvvDcb4MHD87MmTNrbm+//faXkh8AAACAxqteF2vrrbdeunbtWmtsiy22yFtvvZUkqaqqSpKanWtLTZ8+fZldbJ9WWVmZNm3a1LoBAAAAwMpYqUNBv2q77bZbXnnllVpjr776ajbccMMkSefOnVNVVZV777032223XZJkwYIFefDBB3P++ed/5XkBAIC6N6yiYZzGZUhpSLkjAPAVq9fF2o9//OPsuuuuGT58eA499NA88cQTueqqq3LVVVcl+cchoKeeemqGDx+eLl26pEuXLhk+fHhatWqVww8/vMzpAQAAAGjM6nWxttNOO+Xmm2/O4MGD84tf/CKdO3fOyJEjc8QRR9TMOeOMMzJv3rwMHDgwM2bMyC677JJ77rknrVu3LmNyAAAAABq7el2sJUn//v3Tv3//z328oqIiQ4cOzdChQ7+6UAAAAACs9ur1xQsAAAAAoL5SrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApoWu4AAAAAq5thFcPKHWGFDSkNKXcEgHrLjjUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAACigabkDAAAA0PANqxhW7ggrbEhpSLkjAI2EHWsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQQIMq1s4777xUVFTk1FNPrRkrlUoZOnRoqqur07Jly/To0SMvvvhi+UICAAAAsFpoMMXak08+mauuuipbb711rfERI0bkoosuyqhRo/Lkk0+mqqoqffr0yezZs8uUFAAAAIDVQYMo1ubMmZMjjjgiv/3tb7P22mvXjJdKpYwcOTJnn312DjrooHTr1i3jx4/Pxx9/nOuuu66MiQEAAABo7BpEsXbiiSdm3333Te/evWuNT5kyJdOmTUvfvn1rxiorK9O9e/dMnjz5c19v/vz5mTVrVq0bAAAAAKyMpuUO8M/ccMMNeeaZZ/Lkk08u89i0adOSJB06dKg13qFDh0ydOvVzX/O8887LsGHD6jYoAAAAAKuVer1j7e23384pp5ySa6+9Ni1atPjceRUVFbXul0qlZcY+bfDgwZk5c2bN7e23366zzAAAAACsHur1jrWnn34606dPzw477FAztnjx4jz00EMZNWpUXnnllST/2Lm23nrr1cyZPn36MrvYPq2ysjKVlZVfXnAAAAAAGr16vWOtV69e+Z//+Z8899xzNbcdd9wxRxxxRJ577rlsvPHGqaqqyr333lvznAULFuTBBx/MrrvuWsbkAAAAADR29XrHWuvWrdOtW7daY2ussUbat29fM37qqadm+PDh6dKlS7p06ZLhw4enVatWOfzww8sRGQAAAIDVRL0u1lbEGWeckXnz5mXgwIGZMWNGdtlll9xzzz1p3bp1uaMBAAAA0Ig1uGJt0qRJte5XVFRk6NChGTp0aFnyAAAAALB6anDFGgAAAHxVhlUMK3eEFTKkNKTcEWC1VK8vXgAAAAAA9ZViDQAAAAAKcCgoAAAArEYayuGtiUNcqf/sWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAU0LXcAAAAAgFUxrGJYuSOssCGlIeWOQB2yYw0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKKBeF2vnnXdedtppp7Ru3Trrrrtuvv3tb+eVV16pNadUKmXo0KGprq5Oy5Yt06NHj7z44otlSgwAAADA6qJeF2sPPvhgTjzxxDz++OO59957s2jRovTt2zdz586tmTNixIhcdNFFGTVqVJ588slUVVWlT58+mT17dhmTAwAAANDYNS13gC9y11131bo/duzYrLvuunn66afzrW99K6VSKSNHjszZZ5+dgw46KEkyfvz4dOjQIdddd11OOOGEcsQGAAAAYDVQr3esfdbMmTOTJO3atUuSTJkyJdOmTUvfvn1r5lRWVqZ79+6ZPHny577O/PnzM2vWrFo3AAAAAFgZ9XrH2qeVSqWcdtpp2X333dOtW7ckybRp05IkHTp0qDW3Q4cOmTp16ue+1nnnnZdhw4Z9eWEBAAAAVtGwiobTXQwpDSl3hLJoMDvWTjrppDz//PO5/vrrl3msoqKi1v1SqbTM2KcNHjw4M2fOrLm9/fbbdZ4XAAAAgMatQexYO/nkk3PbbbfloYceygYbbFAzXlVVleQfO9fWW2+9mvHp06cvs4vt0yorK1NZWfnlBQYAAACg0avXO9ZKpVJOOumkTJgwIQ888EA6d+5c6/HOnTunqqoq9957b83YggUL8uCDD2bXXXf9quMCAAAAsBqp1zvWTjzxxFx33XW59dZb07p165pzqrVt2zYtW7ZMRUVFTj311AwfPjxdunRJly5dMnz48LRq1SqHH354mdMDAAAA0JjV62LtiiuuSJL06NGj1vjYsWNz7LHHJknOOOOMzJs3LwMHDsyMGTOyyy675J577knr1q2/4rQAAAAArE7qdbFWKpX+6ZyKiooMHTo0Q4cO/fIDAQAAAMD/V6/PsQYAAAAA9ZViDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFgDAAAAgAIUawAAAABQgGINAAAAAApQrAEAAABAAYo1AAAAAChAsQYAAAAABSjWAAAAAKAAxRoAAAAAFKBYAwAAAIACFGsAAAAAUIBiDQAAAAAKUKwBAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACggEZTrF1++eXp3LlzWrRokR122CEPP/xwuSMBAAAA0Ig1imLtxhtvzKmnnpqzzz47zz77bPbYY4/svffeeeutt8odDQAAAIBGqlEUaxdddFEGDBiQ448/PltssUVGjhyZjh075oorrih3NAAAAAAaqablDrCqFixYkKeffjpnnnlmrfG+fftm8uTJy33O/PnzM3/+/Jr7M2fOTJLMmjXrywtaBp/kk3JHWCEr83VvKGtKVnxdjXFNSeNcV2NcU9Jw1tUY15T4+Wso62qMa0r8/DXGdTXGNSUNZ12NcU2Jn7+Gsq7GuKbEz19jXVd9t3QtpVLpn86tKK3IrHrsvffey/rrr59HH300u+66a8348OHDM378+LzyyivLPGfo0KEZNmzYVxkTAAAAgAbk7bffzgYbbPCFcxr8jrWlKioqat0vlUrLjC01ePDgnHbaaTX3lyxZkg8//DDt27f/3Ofwj8a2Y8eOefvtt9OmTZtyx6kTjXFNSeNcV2NcU9I412VNDUdjXJc1NRyNcV2NcU1J41yXNTUcjXFd1tRwNMZ1NcY1fRlKpVJmz56d6urqfzq3wRdr66yzTpo0aZJp06bVGp8+fXo6dOiw3OdUVlamsrKy1thaa631ZUVsdNq0adPofgEb45qSxrmuxrimpHGuy5oajsa4LmtqOBrjuhrjmpLGuS5rajga47qsqeFojOtqjGuqa23btl2heQ3+4gXNmzfPDjvskHvvvbfW+L333lvr0FAAAAAAqEsNfsdakpx22mk56qijsuOOO+ab3/xmrrrqqrz11lv54Q9/WO5oAAAAADRSjaJYO+yww/LBBx/kF7/4Rd5///1069Yt//Vf/5UNN9yw3NEalcrKygwZMmSZw2gbssa4pqRxrqsxrilpnOuypoajMa7LmhqOxriuxrimpHGuy5oajsa4LmtqOBrjuhrjmsqtwV8VFAAAAADKocGfYw0AAAAAykGxBgAAAAAFKNYAAAAAoADFGgAAAAAUoFhjhRx77LH59re/Xe4YdebYY49NRUXFMre99tqr3NFWyfTp03PCCSekU6dOqaysTFVVVfr165fHHnus3NFWybRp03LKKadkk002SYsWLdKhQ4fsvvvuufLKK/Pxxx+XO95K2W+//dK7d+/lPvbYY4+loqIizzzzzFecatV9+neqadOm6dSpU/7t3/4tM2bMKHe0VfLpdTVr1iwbb7xxTj/99MydO7fc0QpbuqZ///d/rzV+yy23pKKiokypVs3y/jz/9O3YY48td8SV9tmfvQ4dOqRPnz4ZM2ZMlixZUu54q2R5nyn++Mc/pkWLFhkxYkR5QtWRyZMnp0mTJg3+88RnNabPgZ/3GfD1118vd7RCSqVSevfunX79+i3z2OWXX562bdvmrbfeKkOyujFt2rScfPLJ2XjjjVNZWZmOHTtmv/32y/3331/uaCvt836PJk2alIqKinz00UdfeSaWtfTPiB/+8IfLPDZw4MAG97niyiuvTOvWrbNo0aKasTlz5qRZs2bZY489as19+OGHU1FRkVdfffWrjtngKdZYbe211155//33a92uv/76csdaJQcffHD+/Oc/Z/z48Xn11Vdz2223pUePHvnwww/LHa2wN954I9ttt13uueeeDB8+PM8++2zuu+++/PjHP87tt9+e++67r9wRV8qAAQPywAMPZOrUqcs8NmbMmGy77bbZfvvty5Bs1S39nXrzzTdz9dVX5/bbb8/AgQPLHWuVLV3XG2+8kV/+8pe5/PLLc/rpp5c71ipp0aJFzj///AZffC716T/HR44cmTZt2tQau+SSS8odsZBP/07deeed6dmzZ0455ZT079+/1gfkhu7qq6/OEUcckVGjRuWMM84od5xVMmbMmJx88sl55JFHGnSZ0dgt7zNg586dyx2rkIqKiowdOzb//d//nd/85jc141OmTMlPf/rTXHLJJenUqVMZExb35ptvZocddsgDDzyQESNG5H/+539y1113pWfPnjnxxBPLHY9PefvttzNgwIBUV1enefPm2XDDDXPKKafkgw8+KHe0Qjp27Jgbbrgh8+bNqxn75JNPcv311ze436eePXtmzpw5eeqpp2rGHn744VRVVeXJJ5+stUlh0qRJqa6uzqabblqOqA1a03IHgHJZuqOrsfjoo4/yyCOPZNKkSenevXuSZMMNN8zOO+9c5mSrZuDAgWnatGmeeuqprLHGGjXjW221VQ4++OCUSqUyplt5/fv3z7rrrptx48ZlyJAhNeMff/xxbrzxxgwfPryM6VbNp3+nNthggxx22GEZN25ceUPVgU+v6/DDD8/EiRNzyy235IorrihzsuJ69+6d119/Peedd16D3yGUpNaf5W3btk1FRUWj+PP90z9766+/frbffvt84xvfSK9evTJu3Lgcf/zxZU646kaMGJFzzjkn1113XQ4++OByx1klc+fOzU033ZQnn3wy06ZNy7hx43LOOeeUOxbL0dg+A3bs2DGXXHJJTjrppPTt2zcbbbRRBgwYkF69ejWonTWftXR30BNPPFHrM+CWW26Z4447rozJ+LQ33ngj3/zmN7Ppppvm+uuvT+fOnfPiiy/mJz/5Se688848/vjjadeuXbljrpTtt98+b7zxRiZMmJAjjjgiSTJhwoR07NgxG2+8cZnTrZzNNtss1dXVmTRpUr7xjW8k+UeBdsABB2TixImZPHlyzdE0kyZNSs+ePcsZt8GyYw0aiTXXXDNrrrlmbrnllsyfP7/ccerEBx98kHvuuScnnnhirQ9Un9bQDl1r2rRpjj766IwbN65WKfiHP/whCxYsqPnLu6F74403ctddd6VZs2bljlLnWrZsmYULF5Y7xipp0qRJhg8fnksvvTTvvPNOueOwEvbcc89ss802mTBhQrmjrLIzzzwz5557bu64444GX6olyY033pjNNtssm222WY488siMHTu2wf3nDw3XMccck169euV73/teRo0alRdeeCFXXXVVuWMV9uGHH+auu+763M+Aa6211lcfiuU68cQT07x589xzzz3p3r17OnXqlL333jv33Xdf3n333Zx99tnljljI9773vYwdO7bm/pgxYxpsodujR49MnDix5v7EiRPTo0ePdO/evWZ8wYIFeeyxxxRrBSnWWG3dcccdNWXU0tu5555b7liFNW3aNOPGjcv48eOz1lprZbfddstZZ52V559/vtzRCnv99ddTKpWy2Wab1RpfZ511ar5nP/3pT8uUrrjjjjsub775ZiZNmlQzNmbMmBx00EFZe+21yxdsFS39nWrZsmW+/vWv56WXXmqQ358v8sQTT+S6665Lr169yh1llR144IHZdttta+2cpGHYfPPN8+abb5Y7xiq58847c/755+fWW2/93PNONjSjR4/OkUcemeQfhxrOmTOnQZ4HanXw2c+A3/nOd8odqU5cddVVeemll3LqqafmN7/5TdZdd91yRyps6WfAzTffvNxR6tTy/v2x9957lztWYR9++GHuvvvuDBw4MC1btqz1WFVVVY444ojceOONDfI/GY466qg88sgjefPNNzN16tQ8+uijNX/GNzQ9evTIo48+mkWLFmX27Nl59tln861vfSvdu3ev+ffI448/nnnz5inWCnIoKKutnj17LnMoV0PbpvxZBx98cPbdd988/PDDeeyxx3LXXXdlxIgRufrqqxv0oQCf3ZX2xBNPZMmSJTniiCMa5O68zTffPLvuumvGjBmTnj175n//93/z8MMP55577il3tFWy9Hfq448/ztVXX51XX301J598crljrbKlH4IXLVqUhQsX5oADDsill15a7lh14vzzz8+ee+6ZQYMGlTsKK6FUKjW43bqftfXWW+fvf/97zjnnnOy0005p3bp1uSOtkldeeSVPPPFEzU7Cpk2b5rDDDsuYMWMaTXHYmHz2M+Dn7YpvaNZdd9384Ac/yC233JIDDzyw3HFWydIipqH/WfdZy/v3x3//93832MLmtddeS6lUyhZbbLHcx7fYYovMmDEjf/vb3xpc0bvOOutk3333zfjx41MqlbLvvvtmnXXWKXesQnr27Jm5c+fmySefzIwZM7Lppptm3XXXTffu3XPUUUdl7ty5mTRpUjp16tTgDnWtLxRrrLbWWGONbLLJJuWOUedatGiRPn36pE+fPjnnnHNy/PHHZ8iQIQ2yWNtkk01SUVGRv/zlL7XGl/6B/9n/GWtIBgwYkJNOOimXXXZZxo4dmw033LDB74L69O/Ur3/96/Ts2TPDhg1r0DtBk//7ENysWbNUV1c3qsNbv/Wtb6Vfv34566yzGuSfEaurl19+ucGeaH2p9ddfP//5n/+Znj17Zq+99spdd93VoMu10aNHZ9GiRVl//fVrxkqlUpo1a5YZM2Y06N3IjVFj/QyY/KPUbdq04f8Tr0uXLqmoqMjLL7/caK5Imyz/Z68xn5JhaUHavHnzMicp5rjjjstJJ52UJLnsssvKnKa4TTbZJBtssEEmTpyYGTNm1JyPu6qqKp07d86jjz6aiRMnZs899yxz0obLoaDQyHXt2jVz584td4xC2rdvnz59+mTUqFENdg2f59BDD02TJk1y3XXXZfz48fne977X6P5XdsiQIfmP//iPvPfee+WOskqWfgjecMMNG1WpttS///u/5/bbb8/kyZPLHYUV8MADD+R//ud/GsU5yTp16pQHH3ww06dPT9++fTNr1qxyRypk0aJFueaaa3LhhRfmueeeq7n9+c9/zoYbbpjf//735Y4IDU67du3Sr1+/XHbZZcv9DPjRRx999aFYxtL/BH/ppZeW+/hf/vKX/Mu//EuDPSfeXnvtlQULFmTBggXp169fueOskp49e2bSpEmZNGlSevToUTPevXv33H333Xn88ccdBroKFGussJkzZ9b6wPjcc8816EvJz58/P9OmTat1+/vf/17uWIV98MEH2XPPPXPttdfm+eefz5QpU/KHP/whI0aMyAEHHFDueIVdfvnlWbRoUXbcccfceOONefnll/PKK6/k2muvzV/+8pc0adKk3BELWXPNNXPYYYflrLPOynvvvdcodwv16NEjW265ZYO+0unqYKuttsoRRxzRaA5vbUyW/j317rvv5plnnsnw4cNzwAEHpH///jn66KPLHa9ObLDBBpk0aVI++OCD9O3bNzNnzix3pJV2xx13ZMaMGRkwYEC6detW63bIIYdk9OjR5Y4IDdLll1+exYsXZ+edd85//ud/5rXXXsvLL7+cX//61/nmN79Z7njk//4T/PLLL8+8efNqPTZt2rT8/ve/b9CfcZs0aZKXX345L7/8coP9N8dSPXv2zCOPPJLnnnuuZsda8o9i7be//W0++eQTxdoqUKyxwiZNmpTtttuu1q0hX0b+rrvuynrrrVfrtvvuu5c7VmFrrrlmdtlll1x88cX51re+lW7duuXnP/95vv/972fUqFHljlfY17/+9Tz77LPp3bt3Bg8enG222SY77rhjLr300px++ukN+jDDAQMGZMaMGendu3c6depU7jhfitNOOy2//e1v8/bbb5c7Cl/g3HPPbZAnFm7slv49tdFGG2WvvfbKxIkT8+tf/zq33nprg/+A/2nrr79+HnzwwXz00Ufp06dPg9uJMnr06PTu3Ttt27Zd5rGDDz44zz33XJ555pkyJIOGrXPnznnmmWfSs2fPDBo0KN26dUufPn1y//33L3OeMspn1KhRmT9/fvr165eHHnoob7/9du6666706dMnm266aYP+92KStGnTJm3atCl3jFXWs2fPzJs3L5tsskk6dOhQM969e/fMnj07X//619OxY8cyJmzYKko+SQMAAAAFvPnmmxk6dGjuuuuuTJ8+PaVSKQcddFB+97vfpVWrVuWOB186xRoAAABQJ4YMGZKLLroo99xzj8N2WS0o1gAAAIA6M3bs2MycOTM/+tGP8rWvOQMVjZtiDQAAAAAKUB0DAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAACrocmTJ6dJkybZa6+9PnfOddddlyZNmuSHP/zhV5gMAKDhqCiVSqVyhwAA4Kt1/PHHZ80118zVV1+dl156KZ06dVpmTu/evbPTTjvliiuuyHvvvZdWrVqVISkAQP1lxxoAwGpm7ty5uemmm/Jv//Zv6d+/f8aNG7fMnDfffDOTJ0/OmWeemc033zx//OMfv/qgAAD1nGINAGA1c+ONN2azzTbLZpttliOPPDJjx47NZw9iGDNmTPbdd9+0bds2Rx55ZEaPHl2mtAAA9ZdiDQBgNTN69OgceeSRSZK99torc+bMyf3331/z+JIlSzJu3LiaOd/97nfz2GOP5fXXXy9LXgCA+kqxBgCwGnnllVfyxBNP5Lvf/W6SpGnTpjnssMMyZsyYmjn33HNP5s6dm7333jtJss4666Rv37615gAAkDQtdwAAAL46o0ePzqJFi7L++uvXjJVKpTRr1iwzZszI2muvnTFjxuTDDz+sdbGCJUuW5Nlnn825556bJk2alCM6AEC946qgAACriUWLFmWDDTbIGWeckb59+9Z67OCDD87JJ5+cf/3Xf011dXWuueaabLnlljWPL1myJHvssUd+//vfp3///l91dACAesmONQCA1cQdd9yRGTNmZMCAAWnbtm2txw455JCa3Wzt27fPd77znXzta7XPGtK/f/+MHj1asQYA8P/ZsQYAsJrYb7/9smTJkvzpT39a5rFnnnkmO+ywQ1q3bp2jjjoql1122TJzJkyYkMMOOyzvvPNOOnTo8FVEBgCo1xRrAAAAAFCAq4ICAAAAQAGKNQAAAAAoQLEGAAAAAAUo1gAAAACgAMUaAAAAABSgWAMAAACAAhRrAAAAAFCAYg0AAAAAClCsAQAAAEABijUAAAAAKECxBgAAAAAF/D+hpbYnHfAAOAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1500x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize = (15, 6))\n",
    "\n",
    "plt.bar(ammino_acid_list, aa_freq_list, color = 'purple')\n",
    "\n",
    "plt.title('Frequenze degli amminoacidi')\n",
    "plt.xlabel('AA')\n",
    "plt.ylabel('Frequenze')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validare la sequenza della proteina letta dall'*entry*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Tradurre in proteina la sequenza della CDS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativa 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_translation = ''\n",
    "for i in range(0,len(cds_sequence)-3,3):\n",
    "    cds_translation = cds_translation + genetic_code_dict[cds_sequence[i:i+3]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativa 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_translation = ''.join(genetic_code_dict[codon] for codon in codon_list[:-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativa 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_translation = re.sub(r'.{3}', lambda x : genetic_code_dict[x.group()], cds_sequence[:-3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEYLSASDVFPCSVYVPDEWEVSREKITLLRELGQGSFGMVYEGNARDIIKGEAETRVAVKTVNESASLRERIEFLNEASVMKGFTCHHVVRLLGVVSKGQPTLVVMELMAHGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMAAEIADGMAYLNAKKFVHRDLAARNCMVAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPVRWMAPESLKDGVFTTSSDMWSFGVVLWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDNCPERVTDLMRMCWQFNPKMRPTFLEIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEMEFEDMENVPLDRSSHCQREEAGGRDGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSNPS'"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_translation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Verificare infine che la proteina letta dal file `EMBL` è uguale a quella ottenuta per traduzione della CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_translation == protein_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trovare una CDS \"sinonima\" della precedente ottenuta sostituendo il maggior numero di codoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Costruire il dizionario inverso del codice genetico:\n",
    "\n",
    "- **chiave**: simbolo di amminoacido\n",
    "- **valore**: lista dei codoni che corrispondono all'amminoacido"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "inverse_genetic_code_dict = dict((l[0], l[1:]) for l in split_genetic_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': ['gct', 'gcc', 'gca', 'gcg'],\n",
       " 'C': ['tgt', 'tgc'],\n",
       " 'D': ['gat', 'gac'],\n",
       " 'E': ['gaa', 'gag'],\n",
       " 'F': ['ttt', 'ttc'],\n",
       " 'G': ['ggt', 'ggc', 'gga', 'ggg'],\n",
       " 'H': ['cat', 'cac'],\n",
       " 'I': ['att', 'atc', 'ata'],\n",
       " 'K': ['aaa', 'aag'],\n",
       " 'L': ['tta', 'ttg', 'ctt', 'ctc', 'cta', 'ctg'],\n",
       " 'M': ['atg'],\n",
       " 'N': ['aat', 'aac'],\n",
       " 'P': ['cct', 'ccc', 'cca', 'ccg'],\n",
       " 'Q': ['caa', 'cag'],\n",
       " 'R': ['cgt', 'cgc', 'cga', 'cgg', 'aga', 'agg'],\n",
       " 'S': ['tct', 'tcc', 'tca', 'tcg', 'agt', 'agc'],\n",
       " 'T': ['act', 'acc', 'aca', 'acg'],\n",
       " 'V': ['gtt', 'gtc', 'gta', 'gtg'],\n",
       " 'W': ['tgg'],\n",
       " 'Y': ['tat', 'tac'],\n",
       " 's': ['tga', 'taa', 'tag']}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inverse_genetic_code_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Determinare una CDS sinonima sostituendo il maggior numero possibile di codoni."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "syn_cds_sequence = ''\n",
    "\n",
    "for codon in codon_list:\n",
    "    aa = genetic_code_dict[codon]\n",
    "    copy_list = list(inverse_genetic_code_dict[aa])\n",
    "    if len(copy_list) > 1:\n",
    "        copy_list.remove(codon)\n",
    "        \n",
    "    new_codon = copy_list.pop(0)\n",
    "    \n",
    "    syn_cds_sequence = syn_cds_sequence + new_codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'atgggtactggtggtcgtcgtggtgctgctgctgctcctttattagttgctgttgctgctttattattaggtgctgctggtcatttatatcctggtgaagtttgccctggtatggacattcgtaataatttaacccgtttacacgaattagaaaactgttccgttattgagggtcatttacaaattttattaatgtttaagactcgtcctgaggactttcgtgatttatcttttcctaagttaattatgattaccgactatttattattatttcgtgtttacggtttagaatctttaaaagatttatttcctaatttaactgttattcgtggttctcgtttatttttcaattatgctttagttatttttgaaatggtccatttaaaagagttaggtttatataatttaatgaatattactcgtggctccgttcgtattgaaaaaaataacgaattatgctatttagctaccattgattggtctcgcattttagactctgttgaagacaaccatattgttttaaataaggacgataatgaagaatgcggtgatatttgccctggcactgctaaaggtaaaactaattgtcctgctactgttattaatggtcaattcgttgagcgttgctggacccactctcattgtcaaaaggtctgtcctactatttgcaaatctcatggttgtactgctgagggtttatgctgtcattctgaatgtttaggtaattgctcccaacctgatgatcctactaaatgtgttgcttgtcgtaatttttatttagatggtcgttgcgttgaaacttgtcctcctccttattatcattttcaagattggcgttgcgttaatttttctttttgtcaagatttacatcataagtgtaaaaattctcgtcgtcaaggttgtcatcagtatgttatccataataataaatgtattcccgaatgcccttctggttatactatgaactcttctaatttattatgtactccttgtttaggcccttgccctaaagtttgtcatttattagagggtgaaaaaactattgattctgttacttccgctcaagaattacgtggttgtactgttattaatggttctttaattattaatatccgtggtggtaataacttagctgccgaattagaggctaatttaggtttaatcgaggagatctctggttacttaaagattcgtcgttcttatgccttagtttctttatctttttttcgtaaattgcgcttaatccgtggtgaaactttagagatcggtaattattctttttacgctttagataatcaaaatttacgtcaattatgggattggtctaagcataatttaactactacccaaggtaagttattttttcattacaatcctaagttatgtttatctgagattcataaaatggaggaggtctctggtactaaaggtcgtcaagaacgtaatgatatcgctttaaaaactaacggtgataaagcttcttgcgagaacgaattgttaaagttctcctatatccgtacttccttcgataaaattttattacgttgggaaccttattggcctcctgattttcgtgatttattaggttttatgttattttataaggaagctccctaccaaaacgttactgaatttgacggtcaagacgcttgcggctctaattcttggactgttgttgatatcgatcctcctttacgttctaatgatcctaagtctcaaaatcatcctggttggttaatgcgtggcttaaaaccttggactcaatacgctattttcgttaaaactttagttactttctctgacgagcgtcgtacttacggtgctaaatctgatattatctacgttcaaactgacgctactaatccttccgttcctttagaccctatttctgtttccaattcttcttctcaaatcatcttaaaatggaagcctccttctgatcctaacggtaatattactcattatttagtcttttgggaacgtcaagctgaggattctgaattatttgaattagactactgtttaaagggtttaaaattaccttctcgtacttggtcccctccttttgaatccgaggactcccaaaaacataatcaatctgaatacgaagactctgctggtgagtgttgttcttgccctaaaactgattcccaaattttaaaagaattagaagaatcttctttccgtaaaactttcgaagactatttacataatgttgtctttgttcctcgtaagacttcctctggtaccggcgctgaagatccccgtccttcccgtaagcgtcgttctttaggtgacgtcggtaacgttactgttgctgttcctactgttgctgcctttcctaatacctcttctacttctgttcctacttctcctgaagaacatcgtcccttcgaaaaagttgttaataaagaatctttagttatttctggtttacgtcattttactggttaccgtattgaattacaagcctgtaatcaagatactcccgaagagcgttgttctgttgctgcttatgtttctgctcgtactatgcccgaggctaaagccgacgatatcgtcggtcccgttactcacgagattttcgaaaataatgttgttcatttaatgtggcaagaacctaaagaacctaacggcttaattgttttatacgaggtttcttaccgtcgttacggcgacgaagaattacacttatgtgtttctcgtaaacattttgccttagagcgtggttgtcgtttacgcggtttatctcctggtaattattctgttcgtattcgtgctacttctttagctggtaatggttcctggactgagcctacttacttttatgttactgattacttggatgttccttctaacatcgctaagatcattattggtcctttaattttcgttttcttattttctgtcgttatcggttctatctacttatttttacgtaaacgtcaacctgacggtcctttaggtcctttatatgcctcctctaatcccgaatacttatctgcttctgacgttttcccttgttccgtttatgttcctgatgaatgggaagtttcccgtgaaaaaattactttattacgtgaattaggtcaaggttcttttggtatggtttacgaaggtaacgctcgtgatattattaaaggcgaagctgaaactcgtgttgctgttaaaactgttaatgaatctgcttctttacgtgaacgtatcgaatttttaaacgaagcttctgttatgaaaggttttacttgtcaccatgttgttcgtttattaggtgttgtttctaaaggtcaacctactttagttgttatggaattaatggcccatggtgatttaaaatcttatttacgctccttacgtcctgaagccgaaaacaaccccggtcgtccccctcccactttacaggaaatgatccaaatggctgctgaaatcgccgatggtatggcttatttaaatgctaaaaaattcgttcaccgtgatttagctgctcgtaattgtatggttgctcacgacttcaccgttaagatcggtgatttcggtatgactcgtgatatttacgagactgactattatcgtaagggtggtaaaggcttattacccgttcgttggatggctcctgaatctttaaaagacggtgtttttactacctcctccgatatgtggtctttcggtgttgttttatgggagattacttctttagctgagcaaccctatcagggtttatccaacgagcaagttttaaagttcgttatggacggtggttacttagaccagcctgataattgccctgaacgtgttaccgatttaatgcgtatgtgttggcagtttaatcctaaaatgcgtcctacttttttagaaatcgttaatttattaaaagatgatttacatccttctttccctgaagtttctttttttcattctgaagaaaataaagcccctgaatctgaagaattagaaatggaattcgaagatatggaaaacgttcctttagatcgctcttctcattgccaacgtgaagaagctggtggtcgtgacggtggttcttctttaggctttaaacgttcttatgaagagcatattccctatactcatatgaatggtggtaaaaagaatggtcgtatcttaactttaccccgttctaacccctcttga'"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "syn_cds_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) Verificare che fornisca la stessa proteina."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_translation == re.sub(r'.{3}', lambda x : genetic_code_dict[x.group()], syn_cds_sequence[:-3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "d) Misurare la differenza con la precedente CDS tramite distanza di Hamming, che è il numero di posizioni in cui le due sequenze hanno diverso carattere.\n",
    "\n",
    "**Esempio**: la distanza di Hamming di `ACGTG` e `GCTTG` è pari a 2, in quanto le basi diverse sono quelle in posizione 1 (`A` e `G`) e quelle in posizione 3 (`G` e `T`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "hamming_dist = [cds_sequence[i] != syn_cds_sequence[i] for i in range(len(cds_sequence))].count(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Percentuale di basi differenti rispetto alla lunghezza delle due CDS:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.36370209689081706"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hamming_dist / len(cds_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Numero di possibili CDS che esprimono la stessa proteina."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108882884128409166024746492002735855651428001792320127244126386647029980632289513784045044140581393841618827221748361594073355650045240363812106073841182979165897587490232446777783983313880563133730376849029385911803929480037903720952494698441685495017262385109095191569323097956789003231766319113433010022112570580209761930707000543314601534394699157328283463123684349487972283113201441194551015153866798709856868536638771549717606971376688439270202789826226688266419863942529413908709032123007393702560744391321515523540532123485323124014154147380386166265468444631304419682081513203797302729799078430988163220003146226946076604133339594066640842997415150550171343877731740457868722176"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count = 1\n",
    "for aa in cds_translation:\n",
    "    count = count * len(inverse_genetic_code_dict[aa])\n",
    "    \n",
    "count"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
